---
title: "The Pig protein Atlas"
Subtitle: "Normalization & classification"
author: "Max Jonatan Karlsson"
date: ""
output:
  pdf_document: default
editor_options:
  chunk_output_type: console
---

```{r start_up, warning=F, message=F, include = F}
require(tidyverse)
require(magrittr)
require(pcaMethods)
require(pheatmap)
require(dendextend)
require(ggdendro)
require(umap)
require(ggforce)
require(Hmisc)
require(NOISeq)
require(ggrepel)
require(gridExtra)
require(ape)
require(pbapply)
require(mgsub)
require(ggalluvial)
require(igraph)
require(viridis)
require(ggbeeswarm)
require(gtable)
require(grid)
require(ggridges)
require(broom)
require(patchwork)

setwd("~/Scilifelab/Projects/Pig Atlas/scripts/Normalization & classification/")
source("../theme.R")
source("../functions_normalization.R")
source("../functions_classification.R")
source("../functions_graphics.R")
source("../functions_utility.R")
```


```{r read_data, include = F}

# Load gene mapping tables
gene_mapping <- 
  read_delim("../../data/meta/ensembl_pig_gene.tab", delim = "\t") %>% 
  filter(biotype == "protein_coding")
gene_chromosome <- read_delim("../../data/meta/ensembl92_gene_chromosome.txt", delim = "\t")

human_gene_mapping <- read_delim("../../data/meta/human/geninfo_92.tsv", delim = "\t")
# Mapping to human genes

gene_orthologs <- read_delim("../../data/meta/ensembl_pig_ortholog.tab", delim = "\t")
gene_orthologs_all <- 
  read_delim("../../data/meta/Ensembl_orthologs_all.txt", delim = "\t") %>% 
  select(enssscg_id = 1,
         ensg_id = 3, 
         pig_gene_name = 2,
         human_gene_name = 4,
         ortholog_confidence = 10, 
         ortholog_type = 5,
         gene_order_conservation_score = 8, 
         whole_genome_alignment_coverage = 9, 
         sequence_identity = 6) %>%
  filter(!is.na(ortholog_confidence)) %>% 
  filter(ensg_id %in% human_gene_mapping$ensg_id) %>%
  filter(enssscg_id %in% gene_mapping$enssscg_id)

# Load sample mapping data
tissue_mapping <- read_delim("../../data/meta/tissue_mapping.tsv", delim = "\t")
## Tissue mapping
human_tissue_mapping <- read_delim("../../data/meta/human_tissue_mapping.tsv", delim = "\t")

sample_mapping <- 
  expand.grid(tissue = tissue_mapping$tissue, 
              individual = letters[1:4]) %>% 
  mutate(sample_ID = paste(tissue, individual, sep = "_")) %>%
  left_join(tissue_mapping, 
            by = "tissue") %>%
  as_tibble()

tissue_colors <- read_delim("../../data/meta/pig_tissue_colors.tsv", delim = "\t")

tissue_colors_palette <- setNames(tissue_colors$tissue_color, tissue_colors$tissue_name)

sample_meta <- 
  tibble(individual = letters[1:4],
         sex = c("female", "female", "male", "male"))

tissue_colors_palette_full <- 
  tissue_colors %$%
  tibble(color = c(tissue_color, region_tissue_color, consensus_tissue_color, organ_color),
         tissue = c(tissue_name, region_tissue_name, consensus_tissue_name, organ_name)) %>%
  unique() %$%
  setNames(color, tissue) 



tissue_colors_palette_full_humanpig <-
  tissue_colors %>%
  left_join(tissue_mapping) %$%
  tibble(color = c(tissue_color, region_tissue_color, consensus_tissue_color, organ_color, 
                   tissue_color, region_tissue_color, consensus_tissue_color),
         tissue = c(tissue_name, region_tissue_name, consensus_tissue_name, organ_name, 
                    human_tissue_name, human_region_name, human_consensus_tissue_name)) %>%
  unique() %$%
  setNames(color, tissue) 

tissue_colors_palette_full_humanpig_extension <- 
  set_names(c(rep(tissue_colors_palette_full_humanpig, 2), 
              "black", "black"),
            c(paste(names(tissue_colors_palette_full_humanpig), "human"), 
              paste(names(tissue_colors_palette_full_humanpig), "pig"),
              "not enriched in human human", "not enriched in pig pig"))

tissue_colors_palette_organs <- 
  tissue_colors %$%
  tibble(color = c(organ_color),
         tissue = c(organ_name)) %>%
  unique() %$%
  setNames(color, tissue) 

regions_organ_order <- 
  bind_rows(tissue_mapping %>% 
              select(region_tissue_name, organ_name),
            human_tissue_mapping %>% 
              select(region_tissue_name, organ_name)) %>% 
  unique() %>% 
  arrange(organ_name, region_tissue_name)

region_levels <- 
  unique(regions_organ_order$region_tissue_name)

region_palette <- 
  regions_organ_order %>%
  left_join(tissue_colors %>% 
              select(organ_name, organ_color)) %$%
  set_names(organ_color, region_tissue_name)
  

comparison_organ_order <- 
  bind_rows(tissue_mapping %>% 
              select(comparison_tissue, organ_name),
            human_tissue_mapping %>% 
              select(comparison_tissue, organ_name)) %>% 
  unique() %>% 
  arrange(organ_name, comparison_tissue)

comparison_levels <- 
  unique(comparison_organ_order$comparison_tissue)


consensus_organ_order <- 
  bind_rows(tissue_mapping %>% 
              select(consensus_tissue_name, organ_name),
            human_tissue_mapping %>% 
              select(consensus_tissue_name, organ_name)) %>% 
  unique() %>% 
  arrange(organ_name, consensus_tissue_name)

consensus_levels <- 
  unique(consensus_organ_order$consensus_tissue_name)
  
# Load data
pig_atlas_sample_temp <- 
  read_delim("../../data/processed/pig_filtered_sample_data.tab", delim = "\t")

# Load human data

## Tissue data
human_atlas_tissue <-
  bind_rows(read_delim("../../data/human data/lims/consensus_hpa_92.tsv", delim = "\t") %>%
              mutate(source = "hpa"),
            read_delim("../../data/human data/lims/consensus_gtex_92.tsv", delim = "\t") %>%
              mutate(source = "gtex"),
            read_delim("../../data/human data/lims/consensus_fantom_92.tsv", delim = "\t") %>%
              mutate(source = "fantom")) %>% 
  rename(tissue_name = celltype) %>%
  group_by(tissue_name, ensg_id) %>% 
  summarise(tpm = mean(tpm, na.rm = T),
            ptpm = mean(ptpm, na.rm = T),
            tmm = mean(tmm, na.rm = T),
            nx = mean(nx, na.rm = T)) %>%
  ungroup()


## Consensus data
human_atlas_consensus <- 
  read_delim("../../data/human data/lims/consensus_max_of_all_groups_92.tsv", delim = "\t") %>% 
  rename(consensus_tissue_name = tissue)

# Load human gene classifications 
human_gene_class <- 
  read_delim("../../data/human data/lims/consensus_all_category_92.tsv", delim = "\t") %>% 
  select(ensg_id,
         dist_category = distribution_category,
         spec_category = specificity_category,
         enriched_tissues = enhanced_tissues, 
         spec_score = ts_score, enhanced_score) %>% 
  mutate(enriched_tissues = gsub(", ", "_", enriched_tissues),
         enriched_tissues = gsub(",", ";", enriched_tissues),
         enriched_tissues = gsub("_", ", ", enriched_tissues),
         enriched_tissues = trimws(enriched_tissues),
         dist_category = tolower(dist_category),
         spec_category = tolower(spec_category))
  

full_gene_list <- 
  bind_rows(human_atlas_tissue %>% 
              select(gene = ensg_id) %>% 
              unique() %>% 
              mutate(species = "human"),
            pig_atlas_tissue %>% 
              select(gene = enssscg_id) %>% 
              unique() %>% 
              mutate(species = "pig"))


# human_gene_class <- 
#   read_delim("../../data/meta/human/proteinatlas.tsv", delim = "\t") %>% 
#   select(Gene, ensg_id = Ensembl, 
#          dist_category = `RNA tissue distribution`,
#          spec_category = `RNA tissue specificity`, 
#          enriched_tissues = `RNA tissue specific NX`) %>%
#   mutate(enriched_tissues = gsub("\\d|\\.|\\:", "", enriched_tissues), 
#          enriched_tissues = gsub(" ;", ", ", enriched_tissues),
#          enriched_tissues = trimws(enriched_tissues), 
#          dist_category = tolower(dist_category),
#          spec_category = tolower(spec_category))
```

#Tissue mapping


```{r determine_pig_human_tissue_overlap, echo=F, warning = F, message = F, results=F}

tissue_mapping %>% 
  arrange(organ_name, consensus_tissue_name, region_tissue_name, tissue_name) %>%
  tissue_connection_plot(col1 = "tissue_name", 
                         col2 = "human_tissue_name", 
                         col1_name = "Pig", 
                         col2_name = "Human", 
                         title = "Tissue overlap", 
                         pal = tissue_colors_palette_full, 
                         na.rm = T)
ggsave(savepath("Tissue Overlap tissue level.pdf"), width = 7, height = 9)


tissue_mapping %>% 
  arrange(organ_name, consensus_tissue_name, region_tissue_name, tissue_name) %>%
  tissue_connection_plot(col1 = "region_tissue_name", 
                         col2 = "human_region_name", 
                         col1_name = "Pig", 
                         col2_name = "Human", 
                         title = "Region overlap", 
                         pal = tissue_colors_palette_full, 
                         na.rm = T)
ggsave(savepath("Tissue Overlap region level.pdf"), width = 7, height = 9)

tissue_mapping %>% 
  arrange(organ_name, consensus_tissue_name, region_tissue_name, tissue_name) %>%
  tissue_connection_plot(col1 = "consensus_tissue_name", 
                         col2 = "human_consensus_tissue_name", 
                         col1_name = "Pig", 
                         col2_name = "Human", 
                         title = "Consensus overlap", 
                         pal = tissue_colors_palette_full, 
                         na.rm = T)
ggsave(savepath("Tissue Overlap consensus level.pdf"), width = 7, height = 9)

tissue_mapping %>% 
  arrange(organ_name, comparison_tissue, region_tissue_name, tissue_name) %>% 
  select(col1 = region_tissue_name, 
         col2 = consensus_tissue_name, 
         col3 = comparison_tissue, 
         col4 = human_consensus_tissue_name, 
         col5 = human_region_name) %>% 
  tissue_connection_multi_plot(col_names = c("Pig Tissue", 
                                             "Pig Consensus", 
                                             "Comparison", 
                                             "Human Consensus", 
                                             "Human Tissue"),
                                 title = "Comparison overlap", 
                               pal = tissue_colors_palette_full_humanpig, 
                               na.rm = F, 
                               expand_x = 0.3)
ggsave(savepath("Tissue Overlap for comparison.pdf"), width = 15, height = 20)

tissue_overlap <- 
  tissue_mapping %>% 
  arrange(organ_name, consensus_tissue_name, region_tissue_name, tissue_name) %>%
  filter(!is.na(human_tissue_name)) %>% 
  select(tissue_name, human_tissue_name) %>% 
  distinct() %>%
  mutate(comparison_id = case_when(tissue_name == human_tissue_name ~ tissue_name,
                                   T ~ paste(tissue_name, human_tissue_name, sep = " - ")))

region_overlap <- 
  tissue_mapping %>% 
  arrange(organ_name, consensus_tissue_name, region_tissue_name, tissue_name) %>%
  filter(!is.na(human_region_name)) %>% 
  select(region_tissue_name, human_region_name) %>% 
  distinct() %>%
  mutate(comparison_id = case_when(region_tissue_name == human_region_name ~ region_tissue_name,
                                   T ~ paste(region_tissue_name, human_region_name, sep = " - ")))

consensus_tissue_overlap <- 
  tissue_mapping %>% 
  arrange(organ_name, consensus_tissue_name, region_tissue_name, tissue_name) %>%
  filter(!is.na(consensus_tissue_name)) %>% 
  filter(!is.na(human_consensus_tissue_name)) %>% 
  select(consensus_tissue_name, human_consensus_tissue_name) %>% 
  distinct() %>%
  mutate(comparison_id = case_when(consensus_tissue_name == human_consensus_tissue_name ~ consensus_tissue_name,
                                   T ~ paste(consensus_tissue_name, human_consensus_tissue_name, sep = " - ")))


consensus_tissue_overlap_all <- 
  tissue_mapping %>% 
  arrange(organ_name, consensus_tissue_name, region_tissue_name, tissue_name) %>%
  
  select(consensus_tissue_name, human_consensus_tissue_name) %>% 
  distinct() %>%
  mutate(comparison_id = case_when(consensus_tissue_name == human_consensus_tissue_name ~ consensus_tissue_name,
                                   is.na(human_consensus_tissue_name) ~ consensus_tissue_name,
                                   is.na(consensus_tissue_name) ~ human_consensus_tissue_name,
                                   T ~ paste(consensus_tissue_name, human_consensus_tissue_name, sep = " - ")))



```



#Normalization

```{r PTPM_normalization, echo=F, warning = F, message = F, results=F}

pig_atlas_sample <- 
  pig_atlas_sample_temp %>% 
  group_by(sample_ID) %>% 
  group_by(sample_ID) %>% 
  mutate(sum_tpm = sum(tpm), 
         ptpm = tpm * 1e6 / sum_tpm) %>% 
  select(-sum_tpm) %>% 
  ungroup()
  


```

```{r normalization, echo=F, warning = F, message = F, results=F}

# Create wide format sample data
pig_atlas_sample_wide_ptpm <- 
  pig_atlas_sample %>%
  select(1, 2, ptpm) %>%
  spread(key = sample_ID, value = ptpm) %>%
  column_to_rownames("enssscg_id")

# TMM normalize sample data
pig_atlas_sample_wide_tmm <- 
  pig_atlas_sample_wide_ptpm %>%
  tmm_norm_median_ref()
  
# Create long format normalized sample data
pig_atlas_sample_norm <- 
  pig_atlas_sample %>% 
  left_join(pig_atlas_sample_wide_tmm %>% 
              as_tibble(rownames = "enssscg_id") %>%
              gather("sample_ID", "tmm", -1),
            by = c("enssscg_id", "sample_ID"))  %>%
  group_by(enssscg_id) %>%
  mutate(nx = case_when(tpm == 0 ~ 0, 
                        T ~ tmm/sqrt(sd(tmm)))) %>% 
  ungroup()

# Create wide format nx sample data
pig_atlas_sample_wide_nx <- 
  pig_atlas_sample_norm %>%
  select(1, 2, nx) %>%
  spread(key = sample_ID, value = nx) %>%
  column_to_rownames("enssscg_id")



```

```{r create_data_tables, echo=F, warning = F, message = F, results=F}

# Aggregate to tissue data
pig_atlas_tissue <- 
  pig_atlas_sample_norm %>% 
  group_by(enssscg_id, tissue_ID) %>% 
  summarise(tpm = mean(tpm), 
            ptpm = mean(ptpm), 
            tmm = mean(tmm), 
            nx = mean(nx)) %>% 
  ungroup()
  
pig_atlas_tissue_wide_tmm <- 
  pig_atlas_tissue %>%
  select(1, 2, tmm) %>%
  spread(key = tissue_ID, value = tmm) %>%
  column_to_rownames("enssscg_id")

pig_atlas_tissue_wide_nx <- 
  pig_atlas_tissue %>%
  select(1, 2, nx) %>%
  spread(key = tissue_ID, value = nx) %>%
  column_to_rownames("enssscg_id")

pig_atlas_tissue_name_wide_nx <- 
  pig_atlas_tissue_wide_nx %>% 
  set_colnames(with(tissue_mapping, tissue_name[match(colnames(.), tissue)]))

# Aggregate to regional tissue data
pig_atlas_region <- 
  pig_atlas_tissue %>% 
  left_join(tissue_mapping %>% 
              select(tissue, region_tissue_name), 
            by = c("tissue_ID" = "tissue")) %>%
  group_by(enssscg_id, region_tissue_name) %>% 
  summarise(tpm = max(tpm), 
            ptpm = max(ptpm), 
            tmm = max(tmm),
            nx = max(nx)) %>% 
  ungroup()

pig_atlas_region_wide_nx <- 
  pig_atlas_region %>%
  select(1, 2, nx) %>%
  spread(key = region_tissue_name, value = nx) %>%
  column_to_rownames("enssscg_id")


# Aggregate to consensus tissue data
pig_atlas_consensus <- 
  pig_atlas_tissue %>% 
  left_join(tissue_mapping %>% 
              select(tissue, consensus_tissue_name), 
            by = c("tissue_ID" = "tissue")) %>%
  group_by(enssscg_id, consensus_tissue_name) %>% 
  summarise(tpm = max(tpm), 
            ptpm = max(ptpm), 
            tmm = max(tmm),
            nx = max(nx)) %>% 
  ungroup()
  
pig_atlas_consensus_wide_tmm <- 
  pig_atlas_consensus %>%
  select(1, 2, tmm) %>%
  spread(key = consensus_tissue_name, value = tmm) %>%
  column_to_rownames("enssscg_id")

pig_atlas_consensus_wide_nx <- 
  pig_atlas_consensus %>%
  select(1, 2, nx) %>%
  spread(key = consensus_tissue_name, value = nx) %>%
  column_to_rownames("enssscg_id")


# Human

human_atlas_region <- 
  human_atlas_tissue %>% 
  left_join(human_tissue_mapping, 
            by = "tissue_name") %>% 
    group_by(region_tissue_name, ensg_id) %>% 
    summarise(tpm = max(tpm, na.rm = T),
              ptpm = max(ptpm, na.rm = T),
              tmm = max(tmm, na.rm = T),
              nx = max(nx, na.rm = T)) %>% 
  ungroup()


# ----- Aggregate to Comparison tissue -----

pig_atlas_comparison <- 
  pig_atlas_tissue %>% 
  inner_join(tissue_mapping %>% 
               select(tissue, comparison_tissue), 
             by = c("tissue_ID" = "tissue")) %>%
  group_by(enssscg_id, comparison_tissue) %>% 
  summarise(tpm = max(tpm), 
            ptpm = max(ptpm), 
            tmm = max(tmm),
            nx = max(nx)) %>% 
  ungroup() %>% 
  filter(!is.na(comparison_tissue))

human_atlas_comparison <- 
  human_atlas_tissue %>% 
  inner_join(human_tissue_mapping, 
             by = "tissue_name") %>% 
    group_by(comparison_tissue, ensg_id) %>% 
    summarise(tpm = max(tpm, na.rm = T),
              ptpm = max(ptpm, na.rm = T),
              tmm = max(tmm, na.rm = T),
              nx = max(nx, na.rm = T)) %>% 
  ungroup() %>% 
  filter(!is.na(comparison_tissue)) %>%
  filter(comparison_tissue %in% pig_atlas_comparison$comparison_tissue)

```

```{r create_human_pig_data_tables, echo=F, warning = F, message = F, results=F}

orth_region <- 
  pig_atlas_region %>% 
  # Filter to only tissues with overlap
  left_join(region_overlap,
            by = c("region_tissue_name")) %>%
  filter(!is.na(human_region_name)) %>% 
  
  # Filter to only genes with orthologs
  left_join(gene_orthologs,
            by = "enssscg_id") %>% 
  filter(!is.na(ensg_id)) %>% 
  
  # Join with human data 
  left_join(human_atlas_region,
            by = c("ensg_id", "human_region_name" = "region_tissue_name"), 
            suffix = c("_pig", "_human")) %>% 
  
  # Remove genes for tissues where we don't have full human gene coverage (FANTOM & GTEx)
  filter(!is.na(nx_human)) 

orth_all_region <- 
  pig_atlas_region %>% 
  # Filter to only tissues with overlap
  left_join(region_overlap,
            by = c("region_tissue_name")) %>%
  filter(!is.na(human_region_name)) %>% 
  
  # Filter to only genes with orthologs
  left_join(gene_orthologs_all,
            by = "enssscg_id") %>% 
  filter(!is.na(ensg_id)) %>% 
  
  # Join with human data 
  left_join(human_atlas_region,
            by = c("ensg_id", "human_region_name" = "region_tissue_name"), 
            suffix = c("_pig", "_human")) %>% 
  
  # Remove genes for tissues where we don't have full human gene coverage (FANTOM & GTEx)
  filter(!is.na(nx_human)) 

orth_region_wide <- 
  orth_region %>% 
  select(enssscg_id, ensg_id, ortholog_type, region_tissue_name, human_region_name, nx_pig, nx_human) %>%
  gather("species", "nx", nx_pig, nx_human) %>%
  mutate(species = gsub("nx_", "", species),
         species_region_tissue_name = paste(species, 
                                               case_when(species == "human" ~ human_region_name,
                                                         species == "pig" ~ region_tissue_name), 
                                               sep = "_")) %>% 
  select(-human_region_name, -region_tissue_name, -species) %>%
  
  # Remove duplicate of intestine
  distinct() %>%
  spread(species_region_tissue_name, nx)

# All orths
orth_all_region <- 
  pig_atlas_region %>% 
  # Filter to only tissues with overlap
  left_join(region_overlap,
            by = c("region_tissue_name")) %>%
  filter(!is.na(human_region_name)) %>% 
  
  # Filter to only genes with orthologs
  left_join(gene_orthologs_all,
            by = "enssscg_id") %>% 
  filter(!is.na(ensg_id)) %>% 
  
  # Join with human data 
  left_join(human_atlas_region,
            by = c("ensg_id", "human_region_name" = "region_tissue_name"), 
            suffix = c("_pig", "_human")) %>% 
  
  # Remove genes for tissues where we don't have full human gene coverage (FANTOM & GTEx)
  filter(!is.na(nx_human)) 


# ----- Consensus -----

orth_consensus <- 
  pig_atlas_consensus %>% 
  # Filter to only tissues with overlap
  left_join(consensus_tissue_overlap,
            by = c("consensus_tissue_name")) %>%
  filter(!is.na(human_consensus_tissue_name)) %>% 
  
  # Filter to only genes with orthologs
  left_join(gene_orthologs,
            by = "enssscg_id") %>% 
  filter(!is.na(ensg_id)) %>% 
  
  # Join with human data 
  left_join(human_atlas_consensus,
            by = c("ensg_id", "human_consensus_tissue_name" = "consensus_tissue_name"), 
            suffix = c("_pig", "_human")) %>% 
  
  # Remove genes for tissues where we don't have full human gene coverage (FANTOM & GTEx)
  filter(!is.na(nx_human)) %>%
  
  # Remove unnecessary column
  select(-num_samples) 

orth_consensus_wide <- 
  orth_consensus %>% 
  select(enssscg_id, ensg_id, ortholog_type, consensus_tissue_name, human_consensus_tissue_name, nx_pig, nx_human) %>%
  gather("species", "nx", nx_pig, nx_human) %>%
  mutate(species = gsub("nx_", "", species),
         species_consensus_tissue_name = paste(species, 
                                               case_when(species == "human" ~ human_consensus_tissue_name,
                                                         species == "pig" ~ consensus_tissue_name), 
                                               sep = "_")) %>% 
  select(-human_consensus_tissue_name, -consensus_tissue_name, -species) %>%
  
  # Remove duplicate of intestine
  distinct() %>%
  spread(species_consensus_tissue_name, nx)

```

```{r create_pca, echo=F, warning = F, message = F, results=F}

# Make PCAs
sample_nx_pca <- 
  pig_atlas_sample_wide_nx %>% 
  {log10(. + 1)} %>%
  t() %>%
  pca_calc(npcs = 150)

tissue_nx_pca <- 
  pig_atlas_tissue_name_wide_nx %>% 
  {log10(. + 1)} %>%
  t() %>%
  pca_calc(npcs = 50)

consensus_nx_pca <- 
  pig_atlas_consensus_wide_nx %>% 
  {log10(. + 1)} %>%
  t() %>%
  pca_calc(npcs = 50)


pca_R2_plot(sample_nx_pca$stats, mark_pc = sample_nx_pca$pc_95) + 
  ggtitle("Sample PCA")

pca_score_plot(sample_nx_pca$scores, 
               group_mapping = sample_mapping, 
               group_col = "tissue_name", 
               mapping_col = "sample_ID", 
               pal = tissue_colors_palette_full) + 
  ggtitle("Sample PCA")

pca_score_plot_simple(sample_nx_pca$scores) + 
  ggtitle("Sample PCA")

pca_R2_plot(tissue_nx_pca$stats, mark_pc = tissue_nx_pca$pc_95) + 
  ggtitle("Tissue PCA")

pca_score_plot(tissue_nx_pca$scores, 
               group_mapping = tissue_mapping, 
               group_col = "consensus_tissue_name", 
               mapping_col = "tissue_name", 
               pal = tissue_colors_palette_full) + 
  ggtitle("Tissue PCA")

pca_R2_plot(consensus_nx_pca$stats, mark_pc = consensus_nx_pca$pc_95) + 
  ggtitle("Consensus PCA")

pca_score_plot(consensus_nx_pca$scores, 
               group_mapping = tissue_mapping, 
               group_col = "consensus_tissue_name", 
               mapping_col = "consensus_tissue_name", 
               pal = tissue_colors_palette_full) + 
  ggtitle("Consensus PCA")



# ----- UMAP -----

sample_nx_umap <- 
  pig_atlas_sample_wide_nx %>% 
  {log10(. + 1)} %>%
  t() %>%
  umap_calc(npcs = 2)



umap_score_plot(sample_nx_umap$layout, 
               group_mapping = sample_mapping, 
               group_col = "tissue_name", 
               mapping_col = "sample_ID", 
               pal = tissue_colors_palette_full, repel = T) + 
  ggtitle("Sample UMAP")
ggsave(savepath("Pig Sample UMAP.pdf"), width = 10, height = 10, useDingbats = F)


region_nx_umap <- 
  pig_atlas_region_wide_nx %>% 
  {log10(. + 1)} %>%
  t() %>%
  umap_calc(npcs = 2)



umap_score_plot(region_nx_umap$layout, 
               group_mapping = tissue_mapping, 
               group_col = "tissue_name", 
               mapping_col = "region_tissue_name", 
               pal = tissue_colors_palette_full, repel = T) + 
  ggtitle("Sample UMAP")
ggsave(savepath("Pig Region UMAP.pdf"), width = 10, height = 10, useDingbats = F)

# ----- Plots to save -----
plot_lims_y <- 
  c(min(consensus_nx_pca$scores[,2]), max(consensus_nx_pca$scores[,2]))
plot_lims_x <- 
  c(min(consensus_nx_pca$scores[,1]), max(consensus_nx_pca$scores[,1]))

consensus_nx_pca$scores %>%
  as_tibble(rownames = "consensus_tissue_name") %>%
  
  left_join(tissue_mapping,
            by = c("consensus_tissue_name")) %>%
  mutate(group_col = case_when(organ_name == "Brain" ~ region_tissue_name,
                               T ~ consensus_tissue_name)) %>%
  group_by(group_col) %>%
  mutate(mean_x = mean(PC1),
         mean_y = mean(PC2)) %>%
  ungroup() %>%
  
  {ggplot(., aes(PC1, PC2, color = organ_name, fill = organ_name)) +
      
      scale_color_manual(values = tissue_colors_palette_full, name = "") +
      scale_fill_manual(values = tissue_colors_palette_full, name = "") +
      
      scale_x_continuous(limits = c(plot_lims_x)) + 
      scale_y_continuous(limits = c(plot_lims_y)) +
      
      simple_theme +
      coord_fixed((plot_lims_x[1] - plot_lims_x[2]) /
                    (plot_lims_y[1] - plot_lims_y[2])) +
      
      geom_point(show.legend = T, 
                 shape = 21, color = "gray25", 
                 size = 4) +
      xlab("PC1") +
      ylab("PC2") + 
      ggtitle("PCA") + 
      
      theme(plot.title = element_text(hjust = 0.5))}

ggsave(savepath("Pig consensus PCA.pdf"), width = 7, height = 7, useDingbats = F)



sample_nx_umap$layout %>%
  as_tibble(rownames = "sample_ID") %>%
  separate(sample_ID, into = c("mapping_col", "individual"), sep = "_") %>%
  left_join(tissue_mapping,
            by = c("mapping_col" = "tissue")) %>%
  mutate(group_col = case_when(organ_name == "Brain" ~ region_tissue_name,
                               T ~ consensus_tissue_name)) %>%
  group_by(group_col) %>%
  mutate(mean_x = mean(V1),
         mean_y = mean(V2)) %>%
  ungroup() %>%
  
  {ggplot(., aes(V1, V2, color = organ_name, fill = organ_name)) +
      
      scale_color_manual(values = tissue_colors_palette_full, name = "") +
      scale_fill_manual(values = tissue_colors_palette_full, name = "") +
      simple_theme +
      geom_segment(aes(xend = mean_x, yend = mean_y),
                   show.legend = F) +
      # geom_text_repel(data = select(., group_col, mean_x, mean_y) %>%
      #             unique(),
      #           aes(x = mean_x, y = mean_y, label = group_col),
      #           show.legend = F, 
      #           box.padding = 2) +
      # geom_label_repel(data = select(., group_col, mean_x, mean_y) %>%
      #             unique(),
      #           aes(x = mean_x, y = mean_y, label = group_col),
      #           show.legend = F, 
      #           color = "black", 
      #           box.padding = 2, 
      #           label.padding = 0.1) +
      
      geom_point(show.legend = T, 
                 shape = 21, color = "gray25", 
                 size = 2) +
      xlab("V1") +
      ylab("V2") + 
      ggtitle("UMAP") + 
      theme(plot.title = element_text(hjust = 0.5))}

ggsave(savepath("Pig sample UMAP.pdf"), width = 7, height = 7, useDingbats = F)

region_nx_umap$layout %>%
  as_tibble(rownames = "region_tissue_name") %>%
  left_join(tissue_mapping,
            by = c("region_tissue_name")) %>%
  mutate(group_col = case_when(organ_name == "Brain" ~ region_tissue_name,
                               T ~ consensus_tissue_name)) %>%
  group_by(group_col) %>%
  mutate(mean_x = mean(V1),
         mean_y = mean(V2)) %>%
  ungroup() %>%
  
  {ggplot(., aes(V1, V2, color = organ_name, fill = organ_name)) +
      
      scale_color_manual(values = tissue_colors_palette_full, name = "") +
      scale_fill_manual(values = tissue_colors_palette_full, name = "") +
      simple_theme +
      geom_segment(aes(xend = mean_x, yend = mean_y),
                   show.legend = F) +
      # geom_text_repel(data = select(., group_col, mean_x, mean_y) %>%
      #             unique(),
      #           aes(x = mean_x, y = mean_y, label = group_col),
      #           show.legend = F,
      #           box.padding = 2, 
      #           inherit.aes = F) +
      geom_label_repel(data = select(., group_col, organ_name, mean_x, mean_y) %>%
                  unique(),
                aes(x = mean_x, y = mean_y, label = group_col),
                show.legend = F,
                color = "black",
                box.padding = 0.25,
                label.padding = 0.25, 
                size = 4) +
      
      geom_point(show.legend = T, 
                 shape = 21, color = "gray25", 
                 size = 2) +
      xlab("V1") +
      ylab("V2") + 
      ggtitle("UMAP") + 
      theme(plot.title = element_text(hjust = 0.5))}

ggsave(savepath("Pig region UMAP.pdf"), width = 7, height = 7, useDingbats = F)
```

##Joined Atlases

```{r human_pig_normalization, echo=F, warning = F, message = F, results=F}
one2one_orth <- 
  gene_orthologs %>% 
  filter(ortholog_type == "ortholog_one2one")

joined_species_nx <- 
  left_join(pig_atlas_comparison %>%
            mutate(comparison_tissue = paste0("pig_", comparison_tissue)) %>% 
            inner_join(one2one_orth) %>% 
            unite(mutual_id, enssscg_id, ensg_id, sep = "_") %>% 
            select(mutual_id, comparison_tissue, nx) %>%
            mutate(nx = log10(nx + 1)) %>% 
            spread(comparison_tissue, nx),
          human_atlas_comparison %>%
            mutate(comparison_tissue = paste0("human_", comparison_tissue)) %>% 
            inner_join(one2one_orth) %>% 
            unite(mutual_id, enssscg_id, ensg_id, sep = "_") %>% 
            select(mutual_id, comparison_tissue, nx) %>%
            mutate(nx = log10(nx + 1)) %>% 
            spread(comparison_tissue, nx)) %>%
  column_to_rownames("mutual_id")
  
joined_species_names <- 
  names(joined_species_nx) %>% 
  enframe() %>% 
  separate(value, into = c("species", "tissue"), sep = "_")

joined_species_nx_norm <-
  limma::removeBatchEffect(joined_species_nx, 
                           batch = joined_species_names$species)

joined_species_atlas_comparison <-
  left_join(joined_species_nx_norm %>%
              as_tibble(rownames = "mutual_id") %>% 
              gather(comparison_tissue, species_norm_nx, -1) %>% 
              separate(comparison_tissue, into = c("species", "comparison_tissue"), sep = "_") %>% 
              spread(species, species_norm_nx),
            joined_species_nx %>%
              as_tibble(rownames = "mutual_id") %>% 
              gather(comparison_tissue, species_nx, -1) %>% 
              separate(comparison_tissue, into = c("species", "comparison_tissue"), sep = "_") %>% 
              spread(species, species_nx), 
            by = c("mutual_id", "comparison_tissue"),
            suffix = c("_norm_nx", "_nx")) %>%
  mutate(human_norm_nx = 10^human_norm_nx - 1,
         pig_norm_nx = 10^pig_norm_nx - 1,
         human_nx = 10^human_nx - 1,
         pig_nx = 10^pig_nx - 1)


joined_species_cor <- 
  joined_species_atlas_comparison %>% 
  group_by(comparison_tissue) %>%
  summarise(cor_nx = cor(human_nx, pig_nx, use = "pairwise.complete.obs", method = "spearman"), 
            cor_nx_norm = cor(human_norm_nx, pig_norm_nx, use = "pairwise.complete.obs", method = "spearman"))


species_atlas <- 
  joined_species_nx %>%
  as_tibble(rownames = "mutual_id") %>% 
  gather(comparison_tissue, nx, -1) %>% 
  separate(comparison_tissue, into = c("species", "comparison_tissue"), sep = "_") %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_", remove = F) %>%
  mutate(species_gene_id = case_when(species == "pig" ~ enssscg_id,
                                     species == "human" ~ ensg_id)) %>% 
  select(1:3, species_gene_id, everything()) %>% 
  mutate(comparison_tissue = factor(comparison_tissue, levels = comparison_levels)) %>%
  left_join(bind_rows(gene_mapping %>% 
                        select(species_gene_id = "enssscg_id",
                               gene_name = display_name),
                      human_gene_mapping %>% 
                        select(species_gene_id = ensg_id,
                               gene_name)) %>% 
              unique()) %>%
  mutate(nx = 10^nx - 1) 


####### ---- Consensus

joined_species_consensus_nx <- 
  left_join(pig_atlas_consensus %>%
            mutate(consensus_tissue_name = paste0("pig_", consensus_tissue_name)) %>% 
            inner_join(one2one_orth) %>% 
            unite(mutual_id, enssscg_id, ensg_id, sep = "_") %>% 
            select(mutual_id, consensus_tissue_name, nx) %>%
            mutate(nx = log10(nx + 1)) %>% 
            spread(consensus_tissue_name, nx),
          human_atlas_consensus %>%
            mutate(consensus_tissue_name = paste0("human_", consensus_tissue_name)) %>% 
            inner_join(one2one_orth) %>% 
            unite(mutual_id, enssscg_id, ensg_id, sep = "_") %>% 
            select(mutual_id, consensus_tissue_name, nx) %>%
            mutate(nx = log10(nx + 1)) %>% 
            spread(consensus_tissue_name, nx)) %>%
  column_to_rownames("mutual_id")

species_atlas_consensus <- 
  joined_species_consensus_nx %>%
  as_tibble(rownames = "mutual_id") %>% 
  gather(consensus_tissue_name, nx, -1) %>% 
  separate(consensus_tissue_name, into = c("species", "consensus_tissue_name"), sep = "_") %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_", remove = F) %>%
  mutate(species_gene_id = case_when(species == "pig" ~ enssscg_id,
                                     species == "human" ~ ensg_id)) %>% 
  select(1:3, species_gene_id, everything()) %>% 
  mutate(consensus_tissue_name = factor(consensus_tissue_name, levels = consensus_levels)) %>%
  left_join(bind_rows(gene_mapping %>% 
                        select(species_gene_id = "enssscg_id",
                               gene_name = display_name),
                      human_gene_mapping %>% 
                        select(species_gene_id = ensg_id,
                               gene_name)) %>% 
              unique()) %>%
  mutate(nx = 10^nx - 1) 

# All orthologs

joined_all_orth_species_consensus_nx <- 
  left_join(pig_atlas_consensus %>%
            mutate(consensus_tissue_name = paste0("pig_", consensus_tissue_name)) %>% 
            inner_join(gene_orthologs_all %>%
                         select(1, 2, ortholog_type)) %>% 
            unite(mutual_id, enssscg_id, ensg_id, sep = "_") %>% 
            select(mutual_id, consensus_tissue_name, nx) %>%
            mutate(nx = log10(nx + 1)) %>% 
            spread(consensus_tissue_name, nx),
          human_atlas_consensus %>%
            mutate(consensus_tissue_name = paste0("human_", consensus_tissue_name)) %>% 
            inner_join(gene_orthologs_all %>%
                         select(1, 2, ortholog_type)) %>% 
            unite(mutual_id, enssscg_id, ensg_id, sep = "_") %>% 
            select(mutual_id, consensus_tissue_name, nx) %>%
            mutate(nx = log10(nx + 1)) %>% 
            spread(consensus_tissue_name, nx)) %>%
  column_to_rownames("mutual_id")

species_all_orth_atlas_consensus <- 
  joined_all_orth_species_consensus_nx %>%
  as_tibble(rownames = "mutual_id") %>% 
  gather(consensus_tissue_name, nx, -1) %>% 
  separate(consensus_tissue_name, into = c("species", "consensus_tissue_name"), sep = "_") %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_", remove = F) %>%
  mutate(species_gene_id = case_when(species == "pig" ~ enssscg_id,
                                     species == "human" ~ ensg_id)) %>% 
  select(1:3, species_gene_id, everything()) %>% 
  mutate(consensus_tissue_name = factor(consensus_tissue_name, levels = consensus_levels)) %>%
  left_join(bind_rows(gene_mapping %>% 
                        select(species_gene_id = "enssscg_id",
                               gene_name = display_name),
                      human_gene_mapping %>% 
                        select(species_gene_id = ensg_id,
                               gene_name)) %>% 
              unique()) %>%
  mutate(nx = 10^nx - 1) 

####### ---- Region level

joined_region_species_nx <- 
  left_join(pig_atlas_region %>%
              mutate(region_tissue_name = paste0("pig_", region_tissue_name)) %>% 
              inner_join(one2one_orth) %>% 
              unite(mutual_id, enssscg_id, ensg_id, sep = "_") %>% 
              select(mutual_id, region_tissue_name, nx) %>%
              mutate(nx = log10(nx + 1)) %>% 
              spread(region_tissue_name, nx),
            
            human_atlas_region %>%
              mutate(region_tissue_name = paste0("human_", region_tissue_name)) %>% 
              inner_join(one2one_orth) %>% 
              unite(mutual_id, enssscg_id, ensg_id, sep = "_") %>% 
              select(mutual_id, region_tissue_name, nx) %>%
              mutate(nx = log10(nx + 1)) %>% 
              spread(region_tissue_name, nx)) %>%
  column_to_rownames("mutual_id")
  
joined_region_species_names <- 
  names(joined_region_species_nx) %>% 
  enframe() %>% 
  separate(value, into = c("species", "tissue"), sep = "_")

joined_region_species_nx_norm <-
  limma::removeBatchEffect(joined_region_species_nx, 
                           batch = joined_region_species_names$species)

joined_region_species_atlas <-
  left_join(region_overlap %>%
              filter(complete.cases(.)) %>% 
              left_join(joined_region_species_nx_norm %>%
                          as_tibble(rownames = "mutual_id") %>% 
                          gather(region_tissue_name, norm_nx, -1) %>% 
                          separate(region_tissue_name, into = c("species", "region_tissue_name"), sep = "_") %>% 
                          filter(species == "pig"), 
                        by = "region_tissue_name") %>%
              left_join(joined_region_species_nx_norm %>%
                          as_tibble(rownames = "mutual_id") %>% 
                          gather(region_tissue_name, norm_nx, -1) %>% 
                          separate(region_tissue_name, into = c("species", "region_tissue_name"), sep = "_") %>% 
                          filter(species == "human"), 
                        by = c("mutual_id", "human_region_name" = "region_tissue_name"), 
                        suffix = c("_pig", "_human")) %>% 
              select(-species_pig, -species_human) %>% 
              left_join(tissue_mapping %>% 
                          select(region_tissue_name, comparison_tissue) %>% 
                          unique(),
                        by = "region_tissue_name") %>%
              select(comparison_tissue, everything()),
            
            region_overlap %>%
              filter(complete.cases(.)) %>% 
              left_join(joined_region_species_nx %>%
                          as_tibble(rownames = "mutual_id") %>% 
                          gather(region_tissue_name, nx, -1) %>% 
                          separate(region_tissue_name, into = c("species", "region_tissue_name"), sep = "_") %>% 
                          filter(species == "pig"), 
                        by = "region_tissue_name") %>%
              left_join(joined_region_species_nx %>%
                          as_tibble(rownames = "mutual_id") %>% 
                          gather(region_tissue_name, nx, -1) %>% 
                          separate(region_tissue_name, into = c("species", "region_tissue_name"), sep = "_") %>% 
                          filter(species == "human"), 
                        by = c("mutual_id", "human_region_name" = "region_tissue_name"), 
                        suffix = c("_pig", "_human")) %>% 
              select(-species_pig, -species_human) %>% 
              left_join(tissue_mapping %>% 
                          select(region_tissue_name, comparison_tissue) %>% 
                          unique(),
                        by = "region_tissue_name") %>%
              select(comparison_tissue, everything()),
            by = c("comparison_tissue", "region_tissue_name",
                   "human_region_name", "comparison_id", "mutual_id")) %>%
  mutate(norm_nx_human = 10^norm_nx_human - 1,
         nx_human = 10^nx_human - 1, 
         norm_nx_pig = 10^norm_nx_pig - 1,
         nx_pig = 10^nx_pig - 1, 
         
         norm_nx_human = ifelse(norm_nx_human < 0, 0, norm_nx_human),
         norm_nx_pig = ifelse(norm_nx_pig < 0, 0, norm_nx_pig))


joined_region_species_cor <- 
  joined_region_species_atlas %>%
  group_by(comparison_tissue, 
           region_tissue_name,
           human_region_name,
           comparison_id) %>%
  summarise(cor_nx = cor(nx_human, nx_pig, use = "pairwise.complete.obs", method = "spearman"), 
            cor_nx_norm = cor(norm_nx_human, norm_nx_pig, use = "pairwise.complete.obs", method = "spearman"))


species_region_atlas <-
  joined_region_species_atlas %>%
  
  gather(type, value, norm_nx_pig, norm_nx_human, nx_pig, nx_human)  %>% 
  mutate(species = gsub("^norm_nx_|^nx_", "", type), 
         type = gsub("_pig$|_human$", "", type)) %>% 
  spread(type, value) %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_", remove = F) %>%
  mutate(species_gene_id = case_when(species == "pig" ~ enssscg_id,
                                     species == "human" ~ ensg_id), 
         species_region_name = case_when(species == "pig" ~ region_tissue_name,
                                     species == "human" ~ human_region_name)) %>% 
  left_join(gene_mapping %>%
              select(enssscg_id, 
                     pig_gene_name = display_name),
            by = "enssscg_id") %>% 
  left_join(human_gene_mapping %>%
              select(ensg_id, 
                     human_gene_name =  gene_name),
            by = "ensg_id") %>% 
  mutate(gene_name = case_when(species == "pig" ~ pig_gene_name,
                               species == "human" ~ human_gene_name)) %>%
  select(1:4, species_region_name, 
         species_gene_id, 
         pig_gene_name, 
         human_gene_name, 
         gene_name, 
         everything()) %>% 
  mutate(region_tissue_name = factor(region_tissue_name, levels = region_levels), 
         human_region_name = factor(human_region_name, levels = region_levels),
         species_region_name = factor(species_region_name, levels = region_levels)) 
```

###All orthologs

```{r joined_atlas_all_orth, echo=FALSE, message=FALSE, warning=FALSE}



joined_region_atlas_allorth <- 
  gene_orthologs_all %>%
  unite(mutual_id, enssscg_id, ensg_id, sep = "_", remove = F) %>%
  filter(ensg_id %in% full_gene_list$gene) %>% 
  filter(enssscg_id %in% full_gene_list$gene) %>%
  gather(gene_type, gene_id, enssscg_id, ensg_id) %>%
  mutate(species = case_when(gene_type == "enssscg_id" ~ "pig", 
                             gene_type == "ensg_id" ~ "human")) %>% 
  left_join(bind_rows(pig_atlas_region %>% 
            select(gene_id = enssscg_id, 
                   region_tissue_name, 
                   tpm, 
                   tmm, 
                   nx) %>% 
            mutate(species = "pig") %>% 
            left_join(tissue_mapping %>% 
                        select(region_tissue_name, 
                               comparison_tissue, 
                               human_region_name), 
                      by = "region_tissue_name"),
          
          human_atlas_region %>% 
            select(gene_id = ensg_id, 
                   region_tissue_name, 
                   tpm, 
                   tmm, 
                   nx) %>% 
            mutate(species = "human") %>% 
            left_join(human_tissue_mapping %>% 
                        select(region_tissue_name, 
                               comparison_tissue), 
                      by = "region_tissue_name") %>% 
            mutate(human_region_name = region_tissue_name)), 
          by = c("gene_id", "species")) %>% 
  select(mutual_id, gene_id, ortholog_type, human_gene_name, species, 
         region_tissue_name, human_region_name, comparison_tissue, everything(), -gene_type) %>% 
  
  mutate(region_tissue_name = factor(region_tissue_name, levels = region_levels), 
         human_region_name = factor(human_region_name, levels = region_levels),
         comparison_tissue = factor(comparison_tissue, levels = comparison_levels)) 


joined_region_atlas_allorth %>% 
  sapply(function(x) length(which(is.na(x))))
          

# Comparison level



joined_comparison_atlas_allorth <- 
  gene_orthologs_all %>%
  unite(mutual_id, enssscg_id, ensg_id, sep = "_", remove = F) %>%
  filter(ensg_id %in% full_gene_list$gene) %>% 
  filter(enssscg_id %in% full_gene_list$gene) %>%
  gather(gene_type, gene_id, enssscg_id, ensg_id) %>%
  mutate(species = case_when(gene_type == "enssscg_id" ~ "pig", 
                             gene_type == "ensg_id" ~ "human")) %>% 
  left_join(bind_rows(pig_atlas_comparison %>% 
                        select(gene_id = enssscg_id, 
                               comparison_tissue, 
                               tpm, 
                               tmm, 
                               nx) %>% 
                        mutate(species = "pig"),
                      
                      human_atlas_comparison %>% 
                        select(gene_id = ensg_id, 
                               comparison_tissue, 
                               tpm, 
                               tmm, 
                               nx) %>% 
                        mutate(species = "human")), 
            by = c("gene_id", "species")) %>% 
  select(mutual_id, gene_id, ortholog_type, human_gene_name, species, 
         comparison_tissue, everything(), -gene_type) %>% 
  
  mutate(comparison_tissue = factor(comparison_tissue, levels = comparison_levels)) 
  

```


#Classification

```{r classification, echo=F, warning = F, message = F, results=F}



pig_gene_class <- 
  hpa_gene_classification(data = pig_atlas_consensus,
                          expression_col = "nx",
                          tissue_col = "consensus_tissue_name",
                          gene_col = "enssscg_id",
                          enr_fold = 4,
                          max_group_n = 5,
                          det_lim = 1)
write_delim(pig_gene_class, savepath("Pig gene classification.txt"), delim = "\t")

# Detailed 
gene_chromosome %>% 
  select(enssscg_id = 1,
         description = `Gene description`,
         chromosome = `Chromosome/scaffold name`,
         gene_name = `Gene name`) %>% 
  unique() %>% 
  left_join(gene_orthologs, 
            by = "enssscg_id") %>% 
  left_join(human_gene_mapping %>% 
              select(1, 
                     human_gene_name = 2, 
                     ncbi_human_gene_summary = ncbi_gene_summary),
            by = "ensg_id") %>% 
  group_by(enssscg_id, ortholog_type, description, chromosome, gene_name) %>% 
  summarise(n_orthologs = n(),
            orth_ensg_id = paste(ensg_id, collapse = ";"),
            orth_gene_names = paste(human_gene_name, collapse = ";"), 
            orth_summary = paste(ncbi_human_gene_summary, collapse = ";")) %>% 
  full_join(pig_gene_class, 
            by = c("enssscg_id" = "gene")) %>% 
  filter(!is.na(spec_category)) %>%
  write_delim(savepath("Pig gene classification detailed table.txt"), delim = "\t")


human_gene_class_calculated <- 
  hpa_gene_classification(data = human_atlas_consensus,
                          expression_col = "nx",
                          tissue_col = "consensus_tissue_name",
                          gene_col = "ensg_id",
                          enr_fold = 4,
                          max_group_n = 5,
                          det_lim = 1)




left_join(human_gene_class,
          human_gene_class_calculated %>% 
            select(ensg_id = gene, 
                   dist_category,
                   spec_category,
                   enriched_tissues,
                   spec_score),
          by = "ensg_id", 
          suffix = c("", "_calc")) %T>% 
          {filter(., dist_category != dist_category_calc) %>% print()}
          # {filter(., spec_category != spec_category_calc) %>% print()}
          # {filter(., enriched_tissues != enriched_tissues_calc | 
          #           (is.na(enriched_tissues) & !is.na(enriched_tissues_calc)) |
          #           (!is.na(enriched_tissues) & is.na(enriched_tissues_calc))) %>% print()}
          # {filter(., spec_score != round(spec_score_calc) & enhanced_score != round(spec_score_calc)) %>% View()}


# ----- Comparison classification -----

pig_gene_class_comparison <- 
  hpa_gene_classification(data = pig_atlas_comparison,
                          expression_col = "nx",
                          tissue_col = "comparison_tissue",
                          gene_col = "enssscg_id",
                          enr_fold = 4,
                          max_group_n = 5,
                          det_lim = 1)
write_delim(pig_gene_class_comparison, savepath("Pig gene comparison classification.txt"), delim = "\t")


human_gene_class_comparison <- 
  hpa_gene_classification(data = human_atlas_comparison,
                          expression_col = "nx",
                          tissue_col = "comparison_tissue",
                          gene_col = "ensg_id",
                          enr_fold = 4,
                          max_group_n = 5,
                          det_lim = 1)
write_delim(human_gene_class_comparison, savepath("Human gene comparison classification.txt"), delim = "\t")

```

```{r classification_plots, echo=F, warning = F, message = F, results=F}

# n_genes barplot & pie
spec_dist_barplots <- class_spec_dist_barplot(pig_gene_class)
spec_dist_barplots

ggsave(savepath("class n_genes per dist barplot.pdf"), spec_dist_barplots[[1]], width = 5, height = 4)
ggsave(savepath("class n_genes per spec barplot.pdf"), spec_dist_barplots[[2]], width = 5, height = 4)

spec_dist_piecharts <- class_spec_dist_piechart(pig_gene_class)
spec_dist_piecharts

ggsave(savepath("class n_genes spec piechart.pdf"), spec_dist_piecharts[[1]], width = 6, height = 6)
ggsave(savepath("class n_genes dist piechart.pdf"), spec_dist_piecharts[[2]], width = 6, height = 6)


# n_genes chord
pdf(savepath("class comb n_genes chord.pdf"), width = 6, height = 6, useDingbats = F)
class_chord_plot(pig_gene_class)
dev.off()
class_chord_plot(pig_gene_class)

# n_genes tissue
n_genes_tissue_spec_plot <- 
  class_tissue_n_enriched_barplot(pig_gene_class) 
ggsave(savepath("class n_genes enriched per tissue barplot.pdf"), n_genes_tissue_spec_plot, width = 5, height = 6)

n_genes_tissue_dist_plot <-
  class_tissue_n_expressed_barplot(pig_gene_class)
ggsave(savepath("class n_genes detected per tissue barplot.pdf"), n_genes_tissue_dist_plot, width = 5, height = 6)

# ----- Comparison classifications -----

all_class_plots(pig_gene_class_comparison, "Pig comparison")
all_class_plots(human_gene_class_comparison, "Human comparison")
```



## Classification Figures

```{r Pig_class_figs}

pdf(savepath("N enr genes per tissue pig + dendro.pdf"), width = 6, height = 7)
class_tissue_n_enriched_barplot_dendro(class_table = pig_gene_class,
                                       dendro = pig_atlas_consensus_wide_nx %>%
                                         cor(method = "spearman") %>%
                                         {1 - .} %>%
                                         as.dist() %>% 
                                         hclust(method = "average"),
                                       pal = tissue_colors_palette_full, 
                                       width = 0.05)
dev.off()
```

```{r Pig_class_network}

pig_net_data <- 
  pig_gene_class %>% 
  select(1, 2, enriched_tissues) %>% 
  filter(spec_category %in% c("tissue enriched", "group enriched")) %>% 
  separate_rows(enriched_tissues, sep = ";") %>% 
  full_join(., 
            .,
            by = c("gene", "spec_category"), 
            suffix = c("1", "2")) %>%
  group_by(enriched_tissues1, enriched_tissues2, spec_category) %>% 
  summarise(n_genes = n()) %>% 
  filter(n_genes >= 8) %>% 
  filter(!(spec_category == "group enriched" & enriched_tissues1 == enriched_tissues2)) %>%
  filter(enriched_tissues1 >= enriched_tissues2) %>%
  ungroup() %>% 
  select(1,2, 4, 3) %>% 
  mutate(edge_id = paste("enrichment: ", enriched_tissues1, enriched_tissues2, spec_category))
  
pig_net_edges <- 
  pig_net_data %$% 
  bind_rows(tibble(node1 = enriched_tissues1, node2 = edge_id, n = n_genes, spec_category), 
            tibble(node1 = enriched_tissues2, node2 = edge_id, n = n_genes, spec_category)) %>% 
  unique()

g <-
  pig_net_edges %>%
  graph_from_data_frame(directed = FALSE) %>%
  ggraph(layout = "kk") 

link_map <- 
  pig_net_edges %>% 
  gather(node, id, -(3:4)) %>% 
  mutate(tissue_node = !grepl("^enrichment: ", id), 
         color_id = ifelse(tissue_node, id, spec_category), 
         label = ifelse(tissue_node, color_id, n)) %>%
  select(n, node, id, tissue_node, color_id, label) %>%
  unique()


edge_data <- get_edges()(g$data)
node_data <- 
  get_nodes()(g$data) %>% 
  as_tibble() %>%
  left_join(link_map, 
            by = c("name" = "id")) 
  

g + 
  geom_edge_arc(aes(width = n), 
                    color = "gray", 
                strength = 1, 
                show.legend = F) + 
    scale_edge_alpha_continuous(range = c(0.3, 1)) +
    scale_edge_width_continuous(range = c(1, 3)) +
  
  geom_node_point(data = node_data  %>%
                    filter(!tissue_node),
                  aes(size = log(n), 
                      fill = color_id), 
                  stroke = 1,
                  # size = 10,
                  shape = 21,
                  show.legend = F)+
  geom_node_point(data = node_data %>%
                    filter(tissue_node),
                  aes(fill = color_id), 
                  stroke = 1,
                  size = 20,
                  shape = 21,
                  show.legend = F)+
  geom_node_text(data = node_data,
                 aes(label = label),
                 size = 4) +
  scale_size_continuous(range = c(5, 10)) +
  scale_fill_manual(values = c(tissue_colors_palette_full_humanpig, gene_category_pal)) +
  
  theme_void()

ggsave(savepath("Network pig enrichment.pdf"), width = 16, height = 12, useDingbats = F)
ggsave(savepath("Network pig enrichment.png"), width = 16, height = 12)


```

```{r Pig_class_network_group_circular}

pig_net_data <- 
  pig_gene_class %>% 
  select(1, 2, enriched_tissues) %>% 
  filter(spec_category %in% c("tissue enriched", "group enriched")) %>% 
  separate_rows(enriched_tissues, sep = ";") %>% 
  full_join(., 
            .,
            by = c("gene", "spec_category"), 
            suffix = c("1", "2")) %>%
  group_by(enriched_tissues1, enriched_tissues2, spec_category) %>% 
  summarise(n_genes = n()) %>% 
  filter(!(spec_category == "group enriched" & enriched_tissues1 == enriched_tissues2)) %>%
  filter(enriched_tissues1 >= enriched_tissues2) %>%
  ungroup() %>% 
  select(1,2, 4, 3) 
 

g <-
  pig_net_data %>%
  graph_from_data_frame(directed = FALSE) %>%
  ggraph(layout = "linear", circular = T) 

# link_map <- 
#   pig_net_data %>% 
#   
#   mutate(tissue_node = !grepl("^enrichment: ", id), 
#          color_id = ifelse(tissue_node, id, spec_category), 
#          label = ifelse(tissue_node, color_id, n)) %>%
#   select(n, node, id, tissue_node, color_id, label) %>%
#   unique()


edge_data <- get_edges()(g$data)
node_data <- 
  get_nodes()(g$data) 
  

g + 
  geom_edge_arc(aes(width = n_genes), 
                    color = "gray", 
                # alpha = 0.1,
                strength = 1, 
                show.legend = F) + 
    scale_edge_alpha_continuous(range = c(0.3, 1)) +
    scale_edge_width_continuous(range = c(1, 3)) +
  g$data %>%
      mutate(degree = case_when(x >= 0 ~ asin(y) * 180 / pi,
                                x < 0 ~ 360 - asin(y) * 180 / pi)) %>%
      # left_join(color_mapping %>%
      #             select(label = label_col,
      #                    color = color_col),
      #           by = "label") %>%
                {geom_node_text(data = .,
                                aes(label = name),
                                angle = .$degree,
                                hjust = ifelse(.$x < 0, 1, 0),
                                vjust = 0.5,
                                size = 3)} +
  
  scale_size_continuous(range = c(1, 10)) +
  scale_fill_manual(values = c(tissue_colors_palette_full_humanpig, gene_category_pal)) +
  scale_x_continuous(expand = expand_scale(0.3)) +
  scale_y_continuous(expand = expand_scale(0.3)) +
  theme_void()


```



##Classification based on gene variance



# Clustering plots

```{r clustering_plots, echo=F, warning = F, message = F, results=F}
# tissue

tissue_wardclust <- 
  tissue_nx_pca$scores[, 1:tissue_nx_pca$pc_95] %>% 
  dist() %>% 
  hclust(method = "ward.D2")

tissue_spearclust <-
  pig_atlas_tissue_name_wide_nx %>%
  cor(method = "spearman") %>%
  {1 - .} %>%
  as.dist() %>% 
  hclust(method = "average")

circular_dendrogram_retinastyle(clust = tissue_wardclust, 
                                color_mapping = tissue_colors, 
                                label_col = "tissue_name", 
                                color_col = "tissue_color", 
                                mean_color = F)
ggsave(savepath("retinagram PC tissue wardclust tissue color.pdf"), width = 10, height = 10)

plot_dendrogram(tissue_wardclust, 
                color_mapping = tissue_colors, 
                  label_col = "tissue_name",
                  pal = tissue_colors_palette_full, 
                  do_label = F)

plot_dendrogram(tissue_spearclust, 
                color_mapping = tissue_colors, 
                  label_col = "tissue_name",
                  pal = tissue_colors_palette_full, 
                  do_label = F)


# consensus

consensus_wardclust <- 
  consensus_nx_pca$scores[, 1:consensus_nx_pca$pc_95] %>% 
  dist() %>% 
  hclust(method = "ward.D2")

consensus_spearclust <-
  pig_atlas_consensus_wide_nx %>%
  cor(method = "spearman") %>%
  {1 - .} %>%
  as.dist() %>% 
  hclust(method = "average")

circular_dendrogram_retinastyle(clust = consensus_wardclust, 
                                color_mapping = tissue_colors, 
                                label_col = "consensus_tissue_name", 
                                color_col = "consensus_tissue_color", 
                                mean_color = F)
ggsave(savepath("retinagram PC tissue wardclust consensus color.pdf"), width = 10, height = 10)


plot_dendrogram(consensus_wardclust, 
                color_mapping = tissue_colors, 
                  label_col = "consensus_tissue_name",
                  pal = tissue_colors_palette_full, 
                  do_label = F)

plot_dendrogram(consensus_spearclust, 
                color_mapping = tissue_colors, 
                  label_col = "consensus_tissue_name",
                  pal = tissue_colors_palette_full, 
                  do_label = F)
#

evolutionary_tree_plot(tissue_nx_pca$scores[, 1:tissue_nx_pca$pc_95],
                       dist_fun = dist,
                       tree_fun = nj, 
                       color_mapping = tissue_colors, 
                       mapping_col = "tissue", 
                       color_col = "tissue_color")

```


##Intraspecies correlation

```{r intraspec_cor, echo=F, warning = F, message = F, results=F}

pig_region_cor <- 
  pig_atlas_region_wide_nx %>%
  cor(method = "spearman")

pig_region_cor %>%
  pheatmap(color = heatmap_palette,
           border_color = NA, 
           annotation_row = tissue_mapping %>%
             select(region_tissue_name, 
                    Organ = organ_name) %>%
             unique() %>%
             filter(!is.na(region_tissue_name)) %>%
             column_to_rownames("region_tissue_name"),
           annotation_colors = list(Organ = tissue_colors %$% 
                                      set_names(organ_color, organ_name)),
           annotation_legend = F,
           show_colnames = F, 
           annotation_names_row = F, 
           clustering_method = "ward.D2", 
           filename = savepath("Spearman correlation pig heatmap.pdf"), height = 9, width = 6, useDingbats= F)

pig_region_cor %>%
  pheatmap(color = heatmap_palette,
           border_color = NA, 
           annotation_row = tissue_mapping %>%
             select(region_tissue_name, 
                    Organ = organ_name) %>%
             unique() %>%
             filter(!is.na(region_tissue_name)) %>%
             column_to_rownames("region_tissue_name"),
           annotation_colors = list(Organ = tissue_colors %$% 
                                      set_names(organ_color, organ_name)),
           annotation_legend = F,
           show_colnames = F, 
           annotation_names_row = F, 
           clustering_method = "ward.D2", 
           filename = savepath("Spearman correlation pig aspect 1 heatmap.pdf"), height = 9, width = 11, useDingbats= F)

```


# Correlation Human - Pig

```{r create_correlation_tables, echo=F, warning = F, message = F, results=F}

# Gene correlation

cor_s_gene_all_orth_consensus <- 
  orth_consensus %>% 
  calc_gene_correlations(var1 = "enssscg_id", 
                         var2 = "ensg_id", 
                         val1 = "nx_pig", 
                         val2 = "nx_human", 
                         cor_method = "spearman", 
                         p_adjust_method = "BH", 
                         alternative = "greater")

cor_p_gene_all_orth_consensus <- 
  orth_consensus %>% 
  mutate(log_nx_pig = log10(nx_pig + 1),
         log_nx_human = log10(nx_human + 1)) %>%
  calc_gene_correlations(var1 = "enssscg_id", 
                         var2 = "ensg_id", 
                         val1 = "log_nx_pig", 
                         val2 = "log_nx_human", 
                         cor_method = "pearson", 
                         p_adjust_method = "BH",
                         alternative = "greater")

cor_s_gene_all_orth_region <-
orth_region %>% 
  calc_gene_correlations(var1 = "enssscg_id", 
                         var2 = "ensg_id", 
                         val1 = "nx_pig", 
                         val2 = "nx_human", 
                         cor_method = "spearman", 
                         p_adjust_method = "BH", 
                         alternative = "greater")

cor_s_gene_all_orth_all_region <-
orth_all_region %>% 
  calc_gene_correlations(var1 = "enssscg_id", 
                         var2 = "ensg_id", 
                         val1 = "nx_pig", 
                         val2 = "nx_human", 
                         cor_method = "spearman", 
                         p_adjust_method = "BH", 
                         alternative = "greater")

cor_p_gene_all_orth_region <- 
  orth_region %>% 
  mutate(log_nx_pig = log10(nx_pig + 1),
         log_nx_human = log10(nx_human + 1)) %>%
  calc_gene_correlations(var1 = "enssscg_id", 
                         var2 = "ensg_id", 
                         val1 = "log_nx_pig", 
                         val2 = "log_nx_human", 
                         cor_method = "pearson", 
                         p_adjust_method = "BH",
                         alternative = "greater")
# Tissue correlation


cor_s_tissue_all_orth_consensus <- 
  orth_consensus %>% 
  calc_gene_correlations(var1 = "consensus_tissue_name", 
                         var2 = "human_consensus_tissue_name", 
                         val1 = "nx_pig", 
                         val2 = "nx_human", 
                         cor_method = "spearman", 
                         p_adjust_method = "BH", 
                         alternative = "greater")

cor_p_tissue_all_orth_consensus <- 
  orth_consensus %>% 
  mutate(log_nx_pig = log10(nx_pig + 1),
         log_nx_human = log10(nx_human + 1)) %>%
  calc_gene_correlations(var1 = "consensus_tissue_name", 
                         var2 = "human_consensus_tissue_name", 
                         val1 = "log_nx_pig", 
                         val2 = "log_nx_human", 
                         cor_method = "pearson", 
                         p_adjust_method = "BH",
                         alternative = "greater")

cor_s_tissue_all_orth_region <- 
  orth_region %>% 
  calc_gene_correlations(var1 = "region_tissue_name", 
                         var2 = "human_region_name", 
                         val1 = "nx_pig", 
                         val2 = "nx_human", 
                         cor_method = "spearman", 
                         p_adjust_method = "BH", 
                         alternative = "greater")

cor_p_tissue_all_orth_region <- 
  orth_region %>% 
  mutate(log_nx_pig = log10(nx_pig + 1),
         log_nx_human = log10(nx_human + 1)) %>%
  calc_gene_correlations(var1 = "region_tissue_name", 
                         var2 = "human_region_name", 
                         val1 = "log_nx_pig", 
                         val2 = "log_nx_human", 
                         cor_method = "pearson", 
                         p_adjust_method = "BH",
                         alternative = "greater")

# ----- wide tissue correlation ------

cor_all_orth_consensus_wide <- 
  orth_consensus_wide %>% 
  select(-(1:3)) %>% 
  cor(use = "pairwise.complete.obs", 
      method = "spearman")

cor_all_orth_region_wide <- 
  orth_region_wide %>% 
  select(-(1:3)) %>% 
  cor(use = "pairwise.complete.obs", 
      method = "spearman")

# ----- Tissue distance -----
dist_gene_all_orth_region <- 
  orth_region %>% 
  calc_gene_distance(var1 = "enssscg_id", 
                     var2 = "ensg_id", 
                     val1 = "nx_pig", 
                     val2 = "nx_human")

```

Here we see the 2-dimensional distribution of spearman correlation and pearson correlation for individual ortholog genes.

```{r compare_correlation_metrics, , echo=F, warning = F, message = F, results=F}

plot_data <- 
  left_join(cor_s_gene_all_orth_consensus, 
          cor_p_gene_all_orth_consensus, 
          by = c("enssscg_id", "ensg_id"),
          suffix = c("_spearman", "_pearson")) %>% 
  left_join(human_gene_class, 
            by = "ensg_id") %>% 
  left_join(pig_gene_class %>% 
              select(enssscg_id = gene,
                     spec_category,
                     dist_category, 
                     enriched_tissues), 
            by = "enssscg_id", 
            suffix = c("_human", "_pig")) %>% 
  mutate(spec_category_pig = factor(spec_category_pig, levels = spec_category_levels),
         spec_category_human = factor(spec_category_human, levels = spec_category_levels)) %>%
  filter(spec_category_human != "not detected" &
           spec_category_pig != "not detected")


all_orth_PvsS_cor_plot <- 
  plot_data %>% 
  ggplot(aes(cor_spearman, cor_pearson)) + 
  geom_hex(aes(fill = stat(log10(count))), 
           bins = 70) + 
  scale_fill_viridis_c() +
  # facet_wrap(significant_spearman~significant_pearson) +
  stripped_theme_facet + 
  ggtitle("Consensus level Gene Pearson correlation vs Spearman correlation")

plot_data <- 
  left_join(cor_s_gene_all_orth_region, 
          cor_p_gene_all_orth_region, 
          by = c("enssscg_id", "ensg_id"),
          suffix = c("_spearman", "_pearson")) %>% 
  left_join(human_gene_class, 
            by = "ensg_id") %>% 
  left_join(pig_gene_class %>% 
              select(enssscg_id = gene,
                     spec_category,
                     dist_category, 
                     enriched_tissues), 
            by = "enssscg_id", 
            suffix = c("_human", "_pig")) %>% 
  mutate(spec_category_pig = factor(spec_category_pig, levels = spec_category_levels),
         spec_category_human = factor(spec_category_human, levels = spec_category_levels)) %>%
  filter(spec_category_human != "not detected" &
           spec_category_pig != "not detected")


all_orthR_PvsS_cor_plot <- 
  plot_data %>% 
  ggplot(aes(cor_spearman, cor_pearson)) + 
  geom_hex(aes(fill = stat(log10(count))), 
           bins = 70) + 
  scale_fill_viridis_c() +
  # facet_wrap(significant_spearman~significant_pearson) +
  stripped_theme_facet + 
  ggtitle("Region level Gene Pearson correlation vs Spearman correlation")

all_orth_PvsS_cor_plot
all_orthR_PvsS_cor_plot

ggsave(savepath("Overall PvsS gene correlation PigHuman Consensus.pdf"), plot = all_orth_PvsS_cor_plot, width = 6, height = 6)
ggsave(savepath("Overall PvsS gene correlation PigHuman Region.pdf"), plot = all_orthR_PvsS_cor_plot, width = 6, height = 6)


plot_data %>%
  ggplot(aes(cor_spearman, cor_pearson)) + 
  geom_hex(aes(fill = stat(log10(count))), 
           bins = 30) + 
  scale_fill_viridis_c() +
  facet_grid(spec_category_human~spec_category_pig) +
  stripped_theme_facet




```

```{r human_pig_gene_correlation, echo=F, warning = F, message = F, results=F}



orth_consensus %>% 
  mutate(comparison_id = factor(comparison_id, levels = {cor_s_tissue_all_orth_consensus %>% 
      left_join(consensus_tissue_overlap) %$% 
      comparison_id[order(cor)]})) %>% 
  ggplot(aes(log10(nx_pig + 1), log10(nx_human + 1))) + 
  geom_hex(aes(fill = stat(log10(count))), 
           bins = 70,
           show.legend = F) + 
  geom_text(data = cor_s_tissue_all_orth_consensus %>% 
              left_join(consensus_tissue_overlap) %>% 
              mutate(comparison_id = factor(comparison_id, levels = {cor_s_tissue_all_orth_consensus %>% 
                  left_join(consensus_tissue_overlap) %$% 
                  comparison_id[order(cor)]})), 
            aes(label = round(cor, 2), 
                size = cor),
            x = 2.5, y = 0.5, 
            fontface = "bold", 
            show.legend = F) +
  scale_fill_viridis_c() +
  scale_size_continuous(range = c(7, 9))+
  stripped_theme_facet + 
  facet_wrap(~ comparison_id, nrow = 5) + 
  # facet_wrap(consensus_tissue_name ~ human_consensus_tissue_name, 
  #            labeller = label_unique) +
  ggtitle("Human - Pig spearman correlation of all orthologs")

ggsave(savepath("cor human pig consensus tis spearman scatter Consensus.pdf"), width = 15, height = 10, useDingbats = F)

orth_region %>% 
  mutate(comparison_id = factor(comparison_id, levels = {cor_s_tissue_all_orth_region %>% 
      left_join(region_overlap) %$% 
      comparison_id[order(cor)]})) %>% 
  ggplot(aes(log10(nx_pig + 1), log10(nx_human + 1))) + 
  geom_hex(aes(fill = stat(log10(count))), 
           bins = 70,
           show.legend = F) + 
  geom_text(data = cor_s_tissue_all_orth_region %>% 
              left_join(region_overlap) %>% 
              mutate(comparison_id = factor(comparison_id, levels = {cor_s_tissue_all_orth_region %>% 
                  left_join(region_overlap) %$% 
                  comparison_id[order(cor)]})), 
            aes(label = round(cor, 2), 
                size = cor),
            x = 2.5, y = 0.5, 
            fontface = "bold", 
            show.legend = F) +
  scale_fill_viridis_c() +
  scale_size_continuous(range = c(7, 9))+
  stripped_theme_facet + 
  facet_wrap(~ comparison_id, nrow = 5) + 
  # facet_wrap(consensus_tissue_name ~ human_consensus_tissue_name, 
  #            labeller = label_unique) +
  ggtitle("Human - Pig spearman correlation of all orthologs")

ggsave(savepath("cor human pig consensus tis spearman scatter Region.pdf"), width = 15, height = 10, useDingbats = F)

joined_species_atlas_comparison %>% 
  ggplot(aes(log10(pig_norm_nx + 1), log10(human_norm_nx + 1))) + 
  geom_hex(aes(fill = stat(log10(count))), 
           bins = 70,
           show.legend = F) + 
  geom_text(data = joined_species_cor, 
            aes(label = round(cor_nx_norm, 2), 
                size = cor_nx_norm),
            x = 2.5, y = 0.5, 
            fontface = "bold", 
            show.legend = F) +
  scale_fill_viridis_c() +
  scale_size_continuous(range = c(7, 9))+
  stripped_theme_facet + 
  facet_wrap(~ comparison_tissue, nrow = 5) + 
  # facet_wrap(consensus_tissue_name ~ human_consensus_tissue_name, 
  #            labeller = label_unique) +
  ggtitle("Human - Pig spearman correlation of 1-1 orthologs")

ggsave(savepath("cor human pig norm comparison scatter.pdf"), width = 15, height = 10, useDingbats = F)

joined_species_atlas_comparison %>% 
  ggplot(aes(log10(pig_nx + 1), log10(human_nx + 1))) + 
  geom_hex(aes(fill = stat(log10(count))), 
           bins = 70,
           show.legend = F) + 
  geom_text(data = joined_species_cor, 
            aes(label = round(cor_nx, 2), 
                size = cor_nx),
            x = 2.5, y = 0.5, 
            fontface = "bold", 
            show.legend = F) +
  scale_fill_viridis_c() +
  scale_size_continuous(range = c(7, 9))+
  stripped_theme_facet + 
  facet_wrap(~ comparison_tissue, nrow = 5) + 
  # facet_wrap(consensus_tissue_name ~ human_consensus_tissue_name, 
  #            labeller = label_unique) +
  ggtitle("Human - Pig spearman correlation of 1-1 orthologs")

ggsave(savepath("cor human pig nx comparison scatter.pdf"), width = 15, height = 10, useDingbats = F)


# ------ correlation heatmaps ------

cor_all_orth_consensus_wide %>%
  pheatmap(color = heatmap_palette, 
           border_color = NA, 
           clustering_method = "ward.D2")

cor_all_orth_region_wide %>%
  pheatmap(color = heatmap_palette, 
           border_color = NA, 
           clustering_method = "ward.D2")


cor_all_orth_consensus_wide %>%
{.[grep("^human", rownames(.)), grep("^pig", colnames(.))]} %>%

  pheatmap(color = heatmap_palette, 
           border_color = NA, 
           clustering_method = "ward.D2")

cor_all_orth_region_wide %>%
{.[grep("^human", rownames(.)), grep("^pig", colnames(.))]} %>%

  pheatmap(color = heatmap_palette, 
           border_color = NA, 
           clustering_method = "ward.D2")


# ----- interspecies correlation network -----


data_1 <- 
  cor_all_orth_region_wide %>%
  {.[grep("^human", rownames(.)), grep("^pig", colnames(.))]} %>%
  as_tibble(rownames = "human_tissue") %>% 
  gather("pig_tissue", "cor", -1) %>%
  mutate(human_label = gsub("^human_", "", human_tissue),
         pig_label = gsub("^pig_", "", pig_tissue)) 

bind_rows(data_1 %>%
            group_by(human_tissue) %>%
            top_n(2, wt = cor) %>%
            select(1:2) %>%
            ungroup(),
          data_1 %>%
            group_by(pig_tissue) %>%
            top_n(2, wt = cor) %>%
            select(1:2) %>%
            ungroup()) %>%
  unique()  %>%
  left_join(data_1,
            by = c("human_tissue", "pig_tissue")) %>%
  basic_network_plot(cor_col = "cor",
                     var1_col = "human_tissue",
                     var2_col = "pig_tissue",
                     var1_label_col = "human_label",
                     var2_label_col = "pig_label",
                     var1_fill_col = "human_label",
                     var2_fill_col = "pig_label",
                     fill_pal = tissue_colors_palette_full_humanpig,
                     var1_color = "darkblue",
                     var2_color = "pink",
                     scale_edge = T,
                     edge_color = "darkgray")
ggsave(savepath("network top2 human-pig cor region.pdf"), width = 12, height = 12)



```

# Compare classification

```{r tissue_name_replacement_test, echo = F, eval = F}

# Pig tissues

consensus_tissue_overlap_all %>% 
  filter(!is.na(consensus_tissue_name)) %$%
  tibble(O= consensus_tissue_name,
         A = comparison_id, 
         B = mgsub(consensus_tissue_name, 
                   mgsub(consensus_tissue_name, 
                         c("\\(", "\\)"), 
                         c("\\\\(", "\\\\)")),
                   comparison_id)) %>% 
  filter(A != B)
  
# Human tissues
consensus_tissue_overlap_all %>% 
  filter(!is.na(human_consensus_tissue_name)) %$%
  tibble(O= human_consensus_tissue_name,
         A = comparison_id, 
         B = mgsub(human_consensus_tissue_name, 
                   mgsub(human_consensus_tissue_name, 
                         c("\\(", "\\)"), 
                         c("\\\\(", "\\\\)")),
                   comparison_id)) %>% 
  filter(A != B)
```


```{r compare_classification_data_gen, warning=F, message=F, results=F}
# ----- comparison classification -----


human_pig_gene_class_comparison_long <- 
  full_join(gene_orthologs %>% 
              left_join(human_gene_class_comparison %>%
                          separate_rows(enriched_tissues, sep = ";") %>% 
                          select(ensg_id = gene,
                                 enriched_tissues), 
                        by = "ensg_id") %>% 
              mutate(comparison_id = enriched_tissues),
            
            
            gene_orthologs %>% 
              left_join(pig_gene_class_comparison %>%
                          separate_rows(enriched_tissues, sep = ";") %>% 
                          select(enssscg_id = gene,
                                 enriched_tissues), 
                        by = "enssscg_id") %>% 
              mutate(comparison_id = enriched_tissues),
            by = c("enssscg_id", "ensg_id", "ortholog_type", "comparison_id"), 
            suffix = c("_human", "_pig")) %>% 
  select(1, 2, 3, enriched_tissues_human, enriched_tissues_pig, comparison_id) 

         

         
         
human_pig_gene_class_comparison <- 
  human_pig_gene_class_comparison_long %>% 
  group_by(enssscg_id, ensg_id, ortholog_type) %>% 
  summarise(n_enriched_human = n_distinct(enriched_tissues_human, na.rm = T),
            n_enriched_pig = n_distinct(enriched_tissues_pig, na.rm = T),
            
            
            # Tissues count as overlapping if enriched in both human and pig, and the tissue is common in both datasets
            n_enriched_overlap = length(which((!is.na(enriched_tissues_human) & !is.na(enriched_tissues_pig)))),
            n_enriched_total = n_enriched_human + n_enriched_pig - n_enriched_overlap,
            
            enriched_tissues_overlap = paste(comparison_id[which(!is.na(enriched_tissues_human) & !is.na(enriched_tissues_pig))], 
                                             collapse = ";"),
            enriched_tissues_human = paste(enriched_tissues_human[which(!is.na(enriched_tissues_human))], 
                                           collapse = ";"),
            enriched_tissues_pig = paste(enriched_tissues_pig[which(!is.na(enriched_tissues_pig))], 
                                         collapse = ";"),
            
            enrichment_overlap = case_when(n_enriched_overlap == 0 ~ "no overlap", #n_enriched_total == 0 ~ "no enrichment", 
                                           n_enriched_overlap == n_enriched_total ~ "full overlap",
                                           n_enriched_overlap > 0 ~ "partial overlap")
            ) %>% 
  left_join(human_gene_class_comparison %>%
              select(ensg_id = gene,
                     dist_category,
                     spec_category),
            by = "ensg_id") %>% 
  left_join(pig_gene_class_comparison %>%
              select(enssscg_id = gene,
                     dist_category,
                     spec_category),
            by = "enssscg_id",
            suffix = c("_human", "_pig")) %>%
  mutate(dist_category_human_n = match(dist_category_human, dist_category_levels), 
         spec_category_human_n = match(spec_category_human, spec_category_levels), 
         dist_category_pig_n = match(dist_category_pig, dist_category_levels), 
         spec_category_pig_n = match(spec_category_pig, spec_category_levels), 
         dist_cat_diff = dist_category_human_n - dist_category_pig_n,
         spec_cat_diff = spec_category_human_n - spec_category_pig_n,
         shared_dist_category = case_when(abs(dist_cat_diff) == 0 ~ "shared",
                                          abs(dist_cat_diff) == 1 ~ "minor difference",
                                          abs(dist_cat_diff) == 2 ~ "medium difference",
                                          abs(dist_cat_diff) >= 3 ~ "major difference"),
         shared_spec_category = case_when(abs(spec_cat_diff) == 0 ~ "shared",
                                          abs(spec_cat_diff) == 1 ~ "minor difference",
                                          abs(spec_cat_diff) == 2 ~ "medium difference",
                                          abs(spec_cat_diff) >= 3 ~ "major difference")) %>% 
  ungroup()

# Dist

human_pig_gene_dist_class_comparison_long <- 
  full_join(gene_orthologs %>% 
              left_join(human_gene_class_comparison %>%
                          separate_rows(tissues_detected, sep = ";") %>% 
                          select(ensg_id = gene,
                                 tissues_detected), 
                        by = "ensg_id") %>% 
              mutate(comparison_id = tissues_detected),
            
            
            gene_orthologs %>% 
              left_join(pig_gene_class_comparison %>%
                          separate_rows(tissues_detected, sep = ";") %>% 
                          select(enssscg_id = gene,
                                 tissues_detected), 
                        by = "enssscg_id") %>% 
              mutate(comparison_id = tissues_detected),
            by = c("enssscg_id", "ensg_id", "ortholog_type", "comparison_id"), 
            suffix = c("_human", "_pig")) %>% 
  select(1, 2, 3, tissues_detected_human, tissues_detected_pig, comparison_id) 


    
human_pig_gene_dist_class_comparison <- 
  human_pig_gene_dist_class_comparison_long %>% 
  group_by(enssscg_id, ensg_id, ortholog_type) %>% 
  summarise(n_detected_human = n_distinct(tissues_detected_human, na.rm = T),
            n_detected_pig = n_distinct(tissues_detected_pig, na.rm = T),
            
            n_overlap_detected_human = n_distinct(tissues_detected_human[which(comparison_id %in%
                                                                                      consensus_tissue_overlap$comparison_id)], na.rm = T),
            n_overlap_detected_pig = n_distinct(tissues_detected_pig[which(comparison_id %in%
                                                                                  consensus_tissue_overlap$comparison_id)], na.rm = T),
            
            
            # Tissues count as overlapping if detected in both human and pig, and the tissue is common in both datasets
            n_detected_overlap = length(which((!is.na(tissues_detected_human) & !is.na(tissues_detected_pig)) &
                                                comparison_id %in% consensus_tissue_overlap$comparison_id)),
            n_detected_total = n_overlap_detected_human + n_overlap_detected_pig - n_detected_overlap,
            
            detected_tissues_overlap = paste(comparison_id[which(!is.na(tissues_detected_human) & !is.na(tissues_detected_pig))], 
                                             collapse = ";"),
            tissues_detected_human = paste(tissues_detected_human[which(!is.na(tissues_detected_human))], 
                                           collapse = ";"),
            tissues_detected_pig = paste(tissues_detected_pig[which(!is.na(tissues_detected_pig))], 
                                         collapse = ";"),
            
            detection_overlap_n = n_detected_overlap/n_detected_total
            ) 


# ----- Canon classification -----
human_pig_gene_class_long <- 
  full_join(gene_orthologs %>% 
              left_join(human_gene_class %>%
                          separate_rows(enriched_tissues, sep = ";") %>% 
                          select(ensg_id,
                                 enriched_tissues), 
                        by = "ensg_id") %>%
              left_join(consensus_tissue_overlap_all %>% 
                          filter(!is.na(human_consensus_tissue_name)) %>%
                          select(human_consensus_tissue_name, comparison_id), 
                        by = c("enriched_tissues" = "human_consensus_tissue_name")),
            
            
            gene_orthologs %>% 
              left_join(pig_gene_class %>%
                          separate_rows(enriched_tissues, sep = ";") %>% 
                          select(enssscg_id = gene,
                                 enriched_tissues), 
                        by = c("enssscg_id")) %>%
              left_join(consensus_tissue_overlap_all %>% 
                          filter(!is.na(consensus_tissue_name)) %>%
                          select(consensus_tissue_name, comparison_id), 
                        by = c("enriched_tissues" = "consensus_tissue_name")),
            by = c("enssscg_id", "ensg_id", "ortholog_type", "comparison_id"), 
            suffix = c("_human", "_pig")) %>% 
  select(1, 2, 3, enriched_tissues_human, enriched_tissues_pig, comparison_id) %>% 
  
  # Translate to the unique tissue comparison id
  mutate(enriched_tissues_comp_human = ifelse(is.na(enriched_tissues_human), NA, comparison_id),
         enriched_tissues_comp_pig = ifelse(is.na(enriched_tissues_pig), NA, comparison_id))

         
         
         
human_pig_gene_class <- 
  human_pig_gene_class_long %>% 
  group_by(enssscg_id, ensg_id, ortholog_type) %>% 
  summarise(n_enriched_human = n_distinct(enriched_tissues_comp_human, na.rm = T),
            n_enriched_pig = n_distinct(enriched_tissues_comp_pig, na.rm = T),
            
            n_overlap_enriched_human = n_distinct(enriched_tissues_comp_human[which(comparison_id %in%
                                                                                      consensus_tissue_overlap$comparison_id)], na.rm = T),
            n_overlap_enriched_pig = n_distinct(enriched_tissues_comp_pig[which(comparison_id %in%
                                                                                  consensus_tissue_overlap$comparison_id)], na.rm = T),
            
            
            # Tissues count as overlapping if enriched in both human and pig, and the tissue is common in both datasets
            n_enriched_overlap = length(which((!is.na(enriched_tissues_comp_human) & !is.na(enriched_tissues_comp_pig)) &
                                                comparison_id %in% consensus_tissue_overlap$comparison_id)),
            n_enriched_total = n_overlap_enriched_human + n_overlap_enriched_pig - n_enriched_overlap,
            
            enriched_tissues_overlap = paste(comparison_id[which(!is.na(enriched_tissues_comp_human) & !is.na(enriched_tissues_comp_pig))], 
                                             collapse = ";"),
            enriched_tissues_human = paste(enriched_tissues_comp_human[which(!is.na(enriched_tissues_comp_human))], 
                                           collapse = ";"),
            enriched_tissues_pig = paste(enriched_tissues_comp_pig[which(!is.na(enriched_tissues_comp_pig))], 
                                         collapse = ";"),
            
            enrichment_overlap = case_when(n_enriched_overlap == 0 ~ "no overlap", #n_enriched_total == 0 ~ "no enrichment", 
                                           n_enriched_overlap == n_enriched_total ~ "full overlap",
                                           n_enriched_overlap > 0 ~ "partial overlap")
            ) %>%
  left_join(human_gene_class %>%
              select(ensg_id,
                     dist_category,
                     spec_category),
            by = "ensg_id") %>% 
  left_join(pig_gene_class %>%
              select(enssscg_id = gene,
                     dist_category,
                     spec_category),
            by = "enssscg_id",
            suffix = c("_human", "_pig")) %>%
  mutate(dist_category_human_n = match(dist_category_human, dist_category_levels), 
         spec_category_human_n = match(spec_category_human, spec_category_levels), 
         dist_category_pig_n = match(dist_category_pig, dist_category_levels), 
         spec_category_pig_n = match(spec_category_pig, spec_category_levels), 
         dist_cat_diff = dist_category_human_n - dist_category_pig_n,
         spec_cat_diff = spec_category_human_n - spec_category_pig_n,
         shared_dist_category = case_when(abs(dist_cat_diff) == 0 ~ "shared",
                                          abs(dist_cat_diff) == 1 ~ "minor difference",
                                          abs(dist_cat_diff) == 2 ~ "medium difference",
                                          abs(dist_cat_diff) >= 3 ~ "major difference"),
         shared_spec_category = case_when(abs(spec_cat_diff) == 0 ~ "shared",
                                          abs(spec_cat_diff) == 1 ~ "minor difference",
                                          abs(spec_cat_diff) == 2 ~ "medium difference",
                                          abs(spec_cat_diff) >= 3 ~ "major difference"))
  


```

```{r compare_overlapping_enrichment, warning=F, message=F, results=F}

plot_data <- 
  human_pig_gene_class_comparison %>% 
  filter(ortholog_type == "ortholog_one2one") %>%
  separate_rows(enriched_tissues_human, sep = ";") %>%
  separate_rows(enriched_tissues_pig, sep = ";") %>%
  separate_rows(enriched_tissues_overlap, sep = ";") %>%
  mutate(is_overlap = enrichment_overlap %in% c("full overlap", "partial overlap"),
         is_elevated = spec_category_human %in% c("tissue enhanced", "group enriched", "tissue enriched") &
           spec_category_pig %in% c("tissue enhanced", "group enriched", "tissue enriched"), 
         same_spec = spec_category_human == spec_category_pig) %>%
  group_by(enriched_tissues_human, is_elevated, is_overlap, same_spec) %>% #enriched_tissues_pig
  summarise(n_genes = n_distinct(ensg_id)) %>%
  ungroup() %>%
  mutate(elevation_identity = case_when(is_elevated & is_overlap & same_spec ~ "identity",
                                        is_elevated & is_overlap ~ "overlapping tissues", 
                                        is_elevated ~ "different tissues", 
                                        T ~ "no enrichment"), 
         elevation_identity = factor(elevation_identity, levels = elevation_identity_levels), 
         enriched_tissues_human = case_when(enriched_tissues_human == "" ~ "not elevated", 
                                            T ~ enriched_tissues_human)) %>%
  group_by(enriched_tissues_human) %>%
  mutate(fract_genes = n_genes / sum(n_genes)) %>%
  ungroup() %>% 
  filter(enriched_tissues_human != "not elevated")

plot_order <- 
  plot_data %>% 
  select(enriched_tissues_human, elevation_identity, fract_genes) %>% 
  filter(!elevation_identity %in% c("no enrichment", "different tissues")) %>%
  group_by(enriched_tissues_human) %>% 
  summarise(fract_overlap = sum(fract_genes)) %>%
  arrange(fract_overlap) %>% 
  pull(enriched_tissues_human)
  

plot_data %>%
  mutate(enriched_tissues_human = factor(enriched_tissues_human, levels = plot_order)) %>%
  ggplot(aes(enriched_tissues_human, fract_genes, fill = elevation_identity, label = n_genes)) +
  geom_col(show.legend = F) + 
  geom_text(position = position_stack(vjust = 0.5), 
            fontface = "bold", 
            color = "white", 
            size = 2,
            show.legend = F) +
  # facet_wrap(~ enriched_tissues_human, nrow = 1) + 
  coord_flip() +
  scale_fill_manual(values = elevation_identity_pal) + #pals::ocean.curl(5)[-3]) +
  # scale_alpha_manual(values = set_names(c(1, 0.8, 0.5, 0.3), elevation_identity_levels)) +
  stripped_theme_facet + 
  scale_y_continuous(expand = expand_scale(0)) +
  scale_x_discrete(position = "top") +
  ylab("Fraction of genes") + 
  theme(axis.title.y = element_blank(), 
        axis.line = element_blank(),
        panel.border = element_blank(), 
        panel.spacing = unit(0, "cm"))
ggsave(savepath("Enrichment overlap fraction barplot.pdf"), width = 4, height = 4, useDingbats = F)


plot_data %>%
  mutate(enriched_tissues_human = factor(enriched_tissues_human, levels = plot_order)) %>%
  ggplot(aes(enriched_tissues_human, n_genes, fill = elevation_identity, label = n_genes)) +
  geom_col() + 
  # facet_wrap(~ enriched_tissues_human, nrow = 1) + 
  coord_flip() +
  scale_fill_manual(values = elevation_identity_pal) + 
  ylab("Number of genes") +
  theme(axis.text.y = element_blank(),
        axis.title.y =  element_blank(),
        axis.ticks = element_blank(), 
        panel.background = element_blank(), 
        panel.grid.major.x = element_line(color = "lightgray"))
  
ggsave(savepath("Enrichment overlap number barplot.pdf"), width = 4, height = 4, useDingbats = F)



plot_data <- 
  human_pig_gene_class_comparison %>% 
  filter(ortholog_type == "ortholog_one2one") %>%
  separate_rows(enriched_tissues_human, sep = ";") %>%
  separate_rows(enriched_tissues_pig, sep = ";") %>%
  separate_rows(enriched_tissues_overlap, sep = ";") %>%
  select(1, 2, spec_category_human, spec_category_pig, 
         enriched_tissues_human, enriched_tissues_pig,
         enrichment_overlap) %>%
  mutate(is_overlap = enriched_tissues_human == enriched_tissues_pig,
         is_elevated = spec_category_human %in% c("tissue enhanced", "group enriched", "tissue enriched") &
           spec_category_pig %in% c("tissue enhanced", "group enriched", "tissue enriched"), 
         same_spec = spec_category_human == spec_category_pig,
         
         elevation_identity = case_when(is_elevated & is_overlap & same_spec ~ "identity",
                                        is_elevated & is_overlap ~ "overlapping tissues", 
                                        is_elevated ~ "different tissues", 
                                        T ~ "no enrichment"), 
         elevation_identity = factor(elevation_identity, levels = elevation_identity_levels), 
         
         enriched_tissues_human = case_when(enriched_tissues_human == "" ~ "not elevated", 
                                            T ~ enriched_tissues_human),
         enriched_tissues_pig = case_when(enriched_tissues_pig == "" ~ "not elevated", 
                                          T ~ enriched_tissues_pig)) %>%
  select(1, 2, spec_category_human, elevation_identity, spec_category_pig, 
         enriched_tissues_human, enriched_tissues_pig,
         is_overlap, is_elevated, same_spec, elevation_identity)

pdf(savepath("tissue-wise enrichment overlap alluvial.pdf"), width = 10, height = 6, useDingbats = F)
pblapply(unique(c(plot_data$enriched_tissues_human, plot_data$enriched_tissues_pig)), 
         function(tissue) {
           plot_data %>% 
             filter(enriched_tissues_human == tissue | enriched_tissues_pig == tissue) %>% 
             group_by(enssscg_id, ensg_id, spec_category_human, spec_category_pig, same_spec, is_elevated) %>% 
             summarise(is_overlap = any(is_overlap), 
                       human = paste(enriched_tissues_human, collapse = ";"),
                       pig = paste(enriched_tissues_pig, collapse = ";")) %>% 
             ungroup() %>%
             mutate(elevation_identity = case_when(is_elevated & is_overlap & same_spec ~ "identity",
                                                   is_elevated & is_overlap ~ "overlapping tissues", 
                                                   is_elevated ~ "different tissues", 
                                                   T ~ "no enrichment"), 
                    elevation_identity = factor(elevation_identity, levels = elevation_identity_levels)) %>% 
             
             multi_alluvial_plot(vars = c("Human" = "spec_category_human", 
                                          "Enrichment overlap" = "elevation_identity", 
                                          "Pig" = "spec_category_pig"), 
                                 chunk_levels = c(spec_category_levels, elevation_identity_levels), 
                                 pal = c(gene_category_pal, elevation_identity_pal), 
                                 color_by = c(1, 3, 3)) +
             ggtitle(tissue) + 
             theme(plot.title = element_text(size = 20, hjust = 0.5, face = "bold"))
         })

dev.off()



plot_data <- 
  human_pig_gene_class_comparison %>% 
  filter(ortholog_type == "ortholog_one2one") %>%
  mutate(pig_elevated = spec_category_pig %in% c("tissue enhanced", "group enriched", "tissue enriched"), 
         human_elevated = spec_category_human %in% c("tissue enhanced", "group enriched", "tissue enriched"), 
         both_elevated = pig_elevated & human_elevated,
         is_overlap = enrichment_overlap != "no overlap")

plot_tissues <- 
  plot_data %$% 
  tibble(tissue = c(enriched_tissues_pig, enriched_tissues_human)) %>% 
  unique() %>% 
  separate_rows(tissue, sep = ";") %>% 
  unique() %>% 
  filter(tissue != "")

plot_data_2 <- 
  plot_tissues %>% 
  group_by(tissue) %>% 
  do({
    tissue <- .$tissue
    
    plot_data %>% 
      filter(grepl(paste0("(^|;)", tissue, "($|;)"), enriched_tissues_human) | 
               grepl(paste0("(^|;)", tissue, "($|;)"), enriched_tissues_pig)) %>%
      mutate(enriched_in = case_when(grepl(paste0("(^|;)", tissue, "($|;)"), enriched_tissues_human) & 
                                       grepl(paste0("(^|;)", tissue, "($|;)"), enriched_tissues_pig) ~ "Both",
                                     grepl(paste0("(^|;)", tissue, "($|;)"), enriched_tissues_human) ~ "Human",  
                                     grepl(paste0("(^|;)", tissue, "($|;)"), enriched_tissues_pig) ~ "Pig")) %>% 
      select(enssscg_id, ensg_id, 
             enriched_in,
             pig_elevated,
             human_elevated,
             both_elevated,
             is_overlap) %>% 
      mutate(tissue = tissue)
  })

plot_data_2 %>%
  # mutate(type = case_when(is_overlap & both_elevated ~ "Same enrichment in tissue",
  #                         !is_overlap & both_elevated ~ "Enriched in different tissues"
  #                         T ~ "Different enrichment")) %>% 
  group_by(tissue, enriched_in) %>%
  summarise(n_genes = n()) %>% 
  mutate(enriched_in = factor(enriched_in, levels = elevation_overlap_levels)) %>%
  ggplot(aes(tissue, n_genes, fill = enriched_in)) + 
  geom_col() + 
  
  scale_fill_manual(values = elevation_overlap_pal) +
  
  coord_flip() +
  stripped_theme_facet + 
  ylab("Number of genes") 
  

```

### All orthologs
```{r all_orth_comparison}

human_pig_gene_all_orths_class_comparison_long <- 
  full_join(gene_orthologs_all %>% 
              left_join(human_gene_class_comparison %>%
                          separate_rows(enriched_tissues, sep = ";") %>% 
                          select(ensg_id = gene,
                                 enriched_tissues), 
                        by = "ensg_id") %>% 
              mutate(comparison_id = enriched_tissues),
            
            
            gene_orthologs_all %>% 
              left_join(pig_gene_class_comparison %>%
                          separate_rows(enriched_tissues, sep = ";") %>% 
                          select(enssscg_id = gene,
                                 enriched_tissues), 
                        by = "enssscg_id") %>% 
              mutate(comparison_id = enriched_tissues),
            by = c("enssscg_id", "ensg_id", "ortholog_type", "comparison_id", 
                   "human_gene_name", "ortholog_confidence", 
                   "gene_order_conservation_score", 
                   "whole_genome_alignment_coverage", "sequence_identity"), 
            suffix = c("_human", "_pig")) 

         

         
         
human_pig_gene_all_orths_class_comparison <- 
  human_pig_gene_all_orths_class_comparison_long %>% 
  group_by(enssscg_id, ensg_id, ortholog_type, 
           human_gene_name, 
           ortholog_confidence,
           gene_order_conservation_score, 
           whole_genome_alignment_coverage, 
           sequence_identity) %>% 
  summarise(n_enriched_human = n_distinct(enriched_tissues_human, na.rm = T),
            n_enriched_pig = n_distinct(enriched_tissues_pig, na.rm = T),
            
            
            # Tissues count as overlapping if enriched in both human and pig, and the tissue is common in both datasets
            n_enriched_overlap = length(which((!is.na(enriched_tissues_human) & !is.na(enriched_tissues_pig)))),
            n_enriched_total = n_enriched_human + n_enriched_pig - n_enriched_overlap,
            
            enriched_tissues_overlap = paste(comparison_id[which(!is.na(enriched_tissues_human) & !is.na(enriched_tissues_pig))], 
                                             collapse = ";"),
            enriched_tissues_human = paste(enriched_tissues_human[which(!is.na(enriched_tissues_human))], 
                                           collapse = ";"),
            enriched_tissues_pig = paste(enriched_tissues_pig[which(!is.na(enriched_tissues_pig))], 
                                         collapse = ";"),
            
            enrichment_overlap = case_when(n_enriched_overlap == 0 ~ "no overlap", #n_enriched_total == 0 ~ "no enrichment", 
                                           n_enriched_overlap == n_enriched_total ~ "full overlap",
                                           n_enriched_overlap > 0 ~ "partial overlap")
            ) %>% 
  left_join(human_gene_class_comparison %>%
              select(ensg_id = gene,
                     dist_category,
                     spec_category),
            by = "ensg_id") %>% 
  left_join(pig_gene_class_comparison %>%
              select(enssscg_id = gene,
                     dist_category,
                     spec_category),
            by = "enssscg_id",
            suffix = c("_human", "_pig")) %>%
  mutate(dist_category_human_n = match(dist_category_human, dist_category_levels), 
         spec_category_human_n = match(spec_category_human, spec_category_levels), 
         dist_category_pig_n = match(dist_category_pig, dist_category_levels), 
         spec_category_pig_n = match(spec_category_pig, spec_category_levels), 
         dist_cat_diff = dist_category_human_n - dist_category_pig_n,
         spec_cat_diff = spec_category_human_n - spec_category_pig_n,
         shared_dist_category = case_when(abs(dist_cat_diff) == 0 ~ "shared",
                                          abs(dist_cat_diff) == 1 ~ "minor difference",
                                          abs(dist_cat_diff) == 2 ~ "medium difference",
                                          abs(dist_cat_diff) >= 3 ~ "major difference"),
         shared_spec_category = case_when(abs(spec_cat_diff) == 0 ~ "shared",
                                          abs(spec_cat_diff) == 1 ~ "minor difference",
                                          abs(spec_cat_diff) == 2 ~ "medium difference",
                                          abs(spec_cat_diff) >= 3 ~ "major difference")) %>% 
  ungroup()
```


```{r compare_classification, echo=F, warning = F, message = F, results=F}

human_pig_gene_class %>% 
  group_by(enrichment_overlap, shared_spec_category) %>%
  summarise(n_genes = n()) %>%
  arrange(shared_spec_category,enrichment_overlap) %$%
  chord_classification(from = enrichment_overlap,
                       to = shared_spec_category, 
                       sizes = n_genes, 
                       grid.col = set_names(c(viridis(3), inferno(4)),
                                            c("full overlap", "partial overlap", "no overlap",  
                                              "shared", "minor difference", "major difference", "medium difference")),
                       groups = c(rep(1, 3), 
                                  rep(2, 4)),
                       plot.order = c("full overlap", "partial overlap", "no overlap",  
                                      "shared", "minor difference", "medium difference", "major difference"),
                       size_labels = T)

human_pig_gene_class %>% 
  group_by(enrichment_overlap) %>%
  summarise(n_genes = n()) %>%
  mutate(enrichment_overlap = factor(enrichment_overlap, levels = c("full overlap", "partial overlap", "no overlap"))) %>%
  ggplot(aes(enrichment_overlap, n_genes, fill = enrichment_overlap)) + 
  geom_col() + 
  simple_theme + 
  scale_fill_viridis_d(direction = -1)

human_pig_gene_class %>% 
  group_by(shared_dist_category) %>%
  summarise(n_genes = n()) %>%
  # mutate(enrichment_overlap = factor(enrichment_overlap, levels = c("full overlap", "partial overlap", "no overlap", "no enrichment"))) %>%
  ggplot(aes(shared_dist_category, n_genes, fill = shared_dist_category)) + 
  geom_col() + 
  simple_theme + 
  scale_fill_viridis_d(direction = -1, option = "C")

human_pig_gene_class %>% 
  group_by(shared_spec_category) %>%
  summarise(n_genes = n()) %>%
  # mutate(enrichment_overlap = factor(enrichment_overlap, levels = c("full overlap", "partial overlap", "no overlap", "no enrichment"))) %>%
  ggplot(aes(shared_spec_category, n_genes, fill = shared_spec_category)) + 
  geom_col() + 
  simple_theme + 
  scale_fill_viridis_d(direction = -1, option = "C")
  


#####




#######


human_pig_gene_class_comparison_long %>% 
  
  left_join(human_gene_class_comparison %>%
              select(ensg_id = gene,
                     dist_category,
                     spec_category),
            by = "ensg_id") %>% 
  left_join(pig_gene_class_comparison %>%
              select(enssscg_id = gene,
                     dist_category,
                     spec_category),
            by = "enssscg_id",
            suffix = c("_human", "_pig")) %>% 
  group_by(enriched_tissues_human, enriched_tissues_pig) %>% 
  summarise(n_genes = n()) %>%
  ungroup() %>%
  mutate(enriched_tissues_human = ifelse(is.na(enriched_tissues_human), "not enriched in human", enriched_tissues_human),
         enriched_tissues_pig = ifelse(is.na(enriched_tissues_pig), "not enriched in pig", enriched_tissues_pig)) %$% 
  
  chord_with_title(from = paste(enriched_tissues_pig, "pig"), 
                   to = paste(enriched_tissues_human, "human"), 
                   sizes = n_genes, 
                   grid.col = tissue_colors_palette_full_humanpig_extension, 
                   groups = c(rep(1, n_distinct(enriched_tissues_pig)), 
                              rep(2, n_distinct(enriched_tissues_human))), 
                   plot.order = c(paste(unique(enriched_tissues_pig), "pig"), 
                                  paste(unique(enriched_tissues_human), "human")), 
                   titles = c("Pig", "Human"), 
                   from_labels = enriched_tissues_pig, 
                   to_labels = enriched_tissues_human, 
                   size_labels = F, 
                   label_style = "vertical")

human_pig_gene_class_comparison_long %>% 
  
  left_join(human_gene_class_comparison %>%
              select(ensg_id = gene,
                     dist_category,
                     spec_category),
            by = "ensg_id") %>% 
  left_join(pig_gene_class_comparison %>%
              select(enssscg_id = gene,
                     dist_category,
                     spec_category),
            by = "enssscg_id",
            suffix = c("_human", "_pig")) %>% 
  group_by(enriched_tissues_human, enriched_tissues_pig) %>% 
  summarise(n_genes = n()) %>%
  ungroup() %>%
  filter(!(is.na(enriched_tissues_human) & is.na(enriched_tissues_pig))) %>%
  mutate(enriched_tissues_human = ifelse(is.na(enriched_tissues_human), "not enriched in human", enriched_tissues_human),
         enriched_tissues_pig = ifelse(is.na(enriched_tissues_pig), "not enriched in pig", enriched_tissues_pig)) %$%  
  
  chord_with_title(from = paste(enriched_tissues_pig, "pig"), 
                   to = paste(enriched_tissues_human, "human"), 
                   sizes = n_genes, 
                   grid.col = tissue_colors_palette_full_humanpig_extension, 
                   groups = c(rep(1, n_distinct(enriched_tissues_pig)), 
                              rep(2, n_distinct(enriched_tissues_human))), 
                   plot.order = c(paste(unique(enriched_tissues_pig), "pig"), 
                                  paste(unique(enriched_tissues_human), "human")), 
                   titles = c("Pig", "Human"), 
                   from_labels = enriched_tissues_pig, 
                   to_labels = enriched_tissues_human, 
                   size_labels = F, 
                   label_style = "vertical")


#####

####




#######


######



```

```{r compare_spec_dist, echo=F, warning = F, message = F, results=F}

# ----- Canon -----
pdf(savepath("class n_genes dist orth overlap chord.pdf"), width = 10, height = 10)
human_pig_gene_class %>%
  group_by(dist_category_human, dist_category_pig) %>%
  summarise(n_genes = n()) %$%
  chord_with_title(from = paste(dist_category_pig, "pig"), 
                   to = paste(dist_category_human, "human"), 
                   sizes = n_genes, 
                   grid.col = gene_category_pal_human_pig, 
                   groups = c(rep(1, 5), 
                              rep(2, 5)), 
                   plot.order = c(paste(dist_category_levels, "pig"), 
                                  paste(dist_category_levels, "human")), 
                   titles = c("Pig", "Human"), 
                   from_labels = dist_category_pig, 
                   to_labels = dist_category_human, 
                   size_labels = T)
dev.off()


pdf(savepath("class n_genes spec orth overlap chord.pdf"), width = 10, height = 10)
human_pig_gene_class %>%
  group_by(spec_category_human, spec_category_pig) %>%
  summarise(n_genes = n()) %$%
  chord_with_title(from = paste(spec_category_pig, "pig"), 
                   to = paste(spec_category_human, "human"), 
                   sizes = n_genes, 
                   grid.col = gene_category_pal_human_pig, 
                   groups = c(rep(1, 5), 
                              rep(2, 5)), 
                   plot.order = c(paste(spec_category_levels, "pig"), 
                                  paste(spec_category_levels, "human")), 
                   titles = c("Pig", "Human"), 
                   from_labels = spec_category_pig, 
                   to_labels = spec_category_human, 
                   size_labels = T)

dev.off()

# ----- Comparison -----

pdf(savepath("class comparison n_genes dist orth overlap chord.pdf"), width = 10, height = 10)
human_pig_gene_class_comparison %>%
  group_by(dist_category_human, dist_category_pig) %>%
  summarise(n_genes = n()) %$%
  chord_with_title(from = paste(dist_category_pig, "pig"), 
                   to = paste(dist_category_human, "human"), 
                   sizes = n_genes, 
                   grid.col = gene_category_pal_human_pig, 
                   groups = c(rep(1, 5), 
                              rep(2, 5)), 
                   plot.order = c(paste(dist_category_levels, "pig"), 
                                  paste(dist_category_levels, "human")), 
                   titles = c("Pig", "Human"), 
                   from_labels = dist_category_pig, 
                   to_labels = dist_category_human, 
                   size_labels = T)
dev.off()


pdf(savepath("class comparison n_genes spec orth overlap chord.pdf"), width = 10, height = 10)
human_pig_gene_class_comparison %>%
  group_by(spec_category_human, spec_category_pig) %>%
  summarise(n_genes = n()) %$%
  chord_with_title(from = paste(spec_category_pig, "pig"), 
                   to = paste(spec_category_human, "human"), 
                   sizes = n_genes, 
                   grid.col = gene_category_pal_human_pig, 
                   groups = c(rep(1, 5), 
                              rep(2, 5)), 
                   plot.order = c(paste(spec_category_levels, "pig"), 
                                  paste(spec_category_levels, "human")), 
                   titles = c("Pig", "Human"), 
                   from_labels = spec_category_pig, 
                   to_labels = spec_category_human, 
                   size_labels = T)

dev.off()

# ----- compare canon with comparison ----

# Human
pdf(savepath("class comparison-canon dist Human chord.pdf"), width = 10, height = 10)
left_join(human_gene_class, 
          human_gene_class_comparison, 
          by = c("ensg_id" = "gene"), 
          suffix = c("_canon", "_comparison")) %>%
  group_by(dist_category_canon, dist_category_comparison) %>%
  summarise(n_genes = n()) %$%
  chord_with_title(from = paste(dist_category_canon, "canon"), 
                   to = paste(dist_category_comparison, "comparison"), 
                   sizes = n_genes, 
                   grid.col = gene_category_pal_comparison, 
                   groups = c(rep(1, 5), 
                              rep(2, 5)), 
                   plot.order = c(paste(dist_category_levels, "canon"), 
                                  paste(dist_category_levels, "comparison")), 
                   titles = c("canon", "comparison"), 
                   from_labels = dist_category_canon, 
                   to_labels = dist_category_comparison, 
                   size_labels = T)
dev.off()


pdf(savepath("class comparison-canon spec Human chord.pdf"), width = 10, height = 10)
left_join(human_gene_class, 
          human_gene_class_comparison, 
          by = c("ensg_id" = "gene"), 
          suffix = c("_canon", "_comparison")) %>%
  group_by(spec_category_canon, spec_category_comparison) %>%
  summarise(n_genes = n()) %$%
  chord_with_title(from = paste(spec_category_canon, "canon"), 
                   to = paste(spec_category_comparison, "comparison"), 
                   sizes = n_genes, 
                   grid.col = gene_category_pal_comparison, 
                   groups = c(rep(1, 5), 
                              rep(2, 5)), 
                   plot.order = c(paste(spec_category_levels, "canon"), 
                                  paste(spec_category_levels, "comparison")), 
                   titles = c("canon", "comparison"), 
                   from_labels = spec_category_canon, 
                   to_labels = spec_category_comparison, 
                   size_labels = T)

dev.off()

#Pig
# Human
pdf(savepath("class comparison-canon dist Pig chord.pdf"), width = 10, height = 10)
left_join(pig_gene_class, 
          pig_gene_class_comparison, 
          by = c("gene"), 
          suffix = c("_canon", "_comparison")) %>%
  group_by(dist_category_canon, dist_category_comparison) %>%
  summarise(n_genes = n()) %$%
  chord_with_title(from = paste(dist_category_canon, "canon"), 
                   to = paste(dist_category_comparison, "comparison"), 
                   sizes = n_genes, 
                   grid.col = gene_category_pal_comparison, 
                   groups = c(rep(1, 5), 
                              rep(2, 5)), 
                   plot.order = c(paste(dist_category_levels, "canon"), 
                                  paste(dist_category_levels, "comparison")), 
                   titles = c("canon", "comparison"), 
                   from_labels = dist_category_canon, 
                   to_labels = dist_category_comparison, 
                   size_labels = T)
dev.off()


pdf(savepath("class comparison-canon spec Pig chord.pdf"), width = 10, height = 10)
left_join(pig_gene_class, 
          pig_gene_class_comparison, 
          by = c("gene"), 
          suffix = c("_canon", "_comparison")) %>%
  group_by(spec_category_canon, spec_category_comparison) %>%
  summarise(n_genes = n()) %$%
  chord_with_title(from = paste(spec_category_canon, "canon"), 
                   to = paste(spec_category_comparison, "comparison"), 
                   sizes = n_genes, 
                   grid.col = gene_category_pal_comparison, 
                   groups = c(rep(1, 5), 
                              rep(2, 5)), 
                   plot.order = c(paste(spec_category_levels, "canon"), 
                                  paste(spec_category_levels, "comparison")), 
                   titles = c("canon", "comparison"), 
                   from_labels = spec_category_canon, 
                   to_labels = spec_category_comparison, 
                   size_labels = T)

dev.off()

```

```{r compare_spec_over_tissuechord, echo=F, warning = F, message = F, results=F}
plot_data <-
  human_pig_gene_class_comparison %>% 
  ungroup() %>%
  select(enssscg_id, ensg_id, 
         enriched_tissues_human, 
         enriched_tissues_pig) %>%
  separate_rows(enriched_tissues_pig, sep = ";") %>% 
  separate_rows(enriched_tissues_human, sep = ";") %>% 
  group_by(enriched_tissues_human, enriched_tissues_pig) %>%
  summarise(n_genes = n()) %>%
  ungroup() %>%
  arrange(-n_genes) %>% 
  mutate(enriched_tissues_human = ifelse(enriched_tissues_human == "", "not enriched in human", enriched_tissues_human),
         enriched_tissues_pig = ifelse(enriched_tissues_pig == "", "not enriched in pig", enriched_tissues_pig))

pdf(savepath("overlap enr pigtis - enr humantis.pdf"), width = 10, height = 10)
for(pig_tissue in sort(unique(plot_data$enriched_tissues_pig))) {
  plot_data %>% 
  filter(enriched_tissues_pig == pig_tissue) %$% 
  
  chord_with_title(from = paste(enriched_tissues_human, "human"), 
                   to = paste(enriched_tissues_pig, "pig"), 
                   sizes = n_genes, 
                   grid.col = tissue_colors_palette_full_humanpig_extension, 
                   groups = c(rep(1, n_distinct(enriched_tissues_human)), 
                              rep(2, n_distinct(enriched_tissues_pig))), 
                   plot.order = c(paste(unique(enriched_tissues_human), "human"), 
                                  paste(unique(enriched_tissues_pig), "pig")), 
                   titles = c("Human", "Pig"), 
                   from_labels = enriched_tissues_human, 
                   to_labels = enriched_tissues_pig, 
                   size_labels = F, 
                   label_style = "vertical", gap_adj = 0.2)
}
dev.off()

pdf(savepath("overlap enr humantis - enr pigtis.pdf"), width = 10, height = 10)
for(human_tissue in sort(unique(plot_data$enriched_tissues_human))) {
  plot_data %>% 
  filter(enriched_tissues_human == human_tissue) %$% 
  
  chord_with_title(from = paste(enriched_tissues_pig, "pig"), 
                   to = paste(enriched_tissues_human, "human"), 
                   sizes = n_genes, 
                   grid.col = tissue_colors_palette_full_humanpig_extension, 
                   groups = c(rep(1, n_distinct(enriched_tissues_pig)), 
                              rep(2, n_distinct(enriched_tissues_human))), 
                   plot.order = c(paste(unique(enriched_tissues_pig), "pig"), 
                                  paste(unique(enriched_tissues_human), "human")), 
                   titles = c("Pig", "Human"), 
                   from_labels = enriched_tissues_pig, 
                   to_labels = enriched_tissues_human, 
                   size_labels = F, 
                   label_style = "vertical", gap_adj = 0.2)
}
dev.off()

bind_rows(plot_data %>%
            filter(enriched_tissues_human != "not enriched in human") %>% 
            filter(enriched_tissues_pig != "not enriched in pig") %>%
            group_by(enriched_tissues_human) %>% 
            top_n(4, n_genes) %>% 
            ungroup(),
          
          plot_data %>% 
            filter(enriched_tissues_human != "not enriched in human") %>% 
            filter(enriched_tissues_pig != "not enriched in pig") %>%
            group_by(enriched_tissues_pig) %>% 
            top_n(4, n_genes) %>% 
            ungroup()) %>% 
  unique() %$%
  chord_with_title(from = paste(enriched_tissues_pig, "pig"), 
                   to = paste(enriched_tissues_human, "human"), 
                   sizes = n_genes, 
                   grid.col = tissue_colors_palette_full_humanpig_extension, 
                   groups = c(rep(1, n_distinct(enriched_tissues_pig)), 
                              rep(2, n_distinct(enriched_tissues_human))), 
                   plot.order = c(paste(unique(enriched_tissues_pig), "pig"), 
                                  rev(paste(unique(enriched_tissues_human), "human"))), 
                   titles = c("Pig", "Human"), 
                   from_labels = enriched_tissues_pig, 
                   to_labels = enriched_tissues_human, 
                   size_labels = F, 
                   label_style = "vertical", gap_adj = 0.2)


plot_data %>% 
  mutate(n_genes = log10(n_genes + 1)) %>%
  spread(enriched_tissues_human, n_genes, fill = 0) %>% 
  column_to_rownames("enriched_tissues_pig") %>% 
  pheatmap(color = heatmap_palette, clustering_method = "ward.D2", 
           cutree_rows = 10, 
           cutree_cols = 10)

# Plots showing common group enrichments

plot_data <- 
  pig_gene_class %>% 
  # filter(spec_category == "tissue enhanced") %>%
  select(enssscg_id = gene, 
         enriched_tissues) %>% 
  separate_rows(enriched_tissues, sep = ";") %>%
  full_join(., .,
            by = "enssscg_id", 
            suffix = c("_1", "_2")) %>% 
  group_by(enriched_tissues_1, enriched_tissues_2) %>% 
  summarise(n_genes = n()) %>% 
  ungroup() 

plot_data %>% 
  mutate(n_genes = log10(n_genes + 1)) %>%
  filter(!is.na(enriched_tissues_1)) %>%
  spread(enriched_tissues_2, n_genes, fill = 0) %>% 
  column_to_rownames("enriched_tissues_1") %>%
  pheatmap(color = heatmap_palette, 
           cutree_rows = 8, 
           cutree_cols = 8)

```

```{r compare_spec_over_tissuealluvial, echo=F, warning = F, message = F, results=F}
plot_data <-
  human_pig_gene_class_comparison %>% 
  ungroup() %>%
  select(enssscg_id, ensg_id, 
         enriched_tissues_human, 
         enriched_tissues_pig) %>%
  separate_rows(enriched_tissues_pig, sep = ";") %>% 
  separate_rows(enriched_tissues_human, sep = ";") %>% 
  mutate(enriched_tissues_human = ifelse(enriched_tissues_human == "", "not enriched in human", enriched_tissues_human),
         enriched_tissues_pig = ifelse(enriched_tissues_pig == "", "not enriched in pig", enriched_tissues_pig))




compare_tissue <- sample(unique(plot_data$enriched_tissues_human), 1)

plot_data %>%
  filter(enriched_tissues_human == compare_tissue | enriched_tissues_pig == compare_tissue) %>%
  mutate(row_n = row_number()) %>%
  gather(cat_type, cat, -enssscg_id, -ensg_id, -row_n) %>%
  mutate(cat = factor(cat, 
                      levels = {
                        filter(., cat_type == "enriched_tissues_pig") %>% 
                          filter(cat != "not enriched in pig") %>%
                          group_by(cat) %>% 
                          summarise(n_genes = n()) %>% 
                          ungroup() %$%
                          c(cat[order(n_genes)], "not enriched in pig", "not enriched in human")
                      }),
         cat_type = case_when(cat_type == "enriched_tissues_pig" ~ "Pig",
                              cat_type == "enriched_tissues_human" ~ "Human"),
         cat_type = factor(cat_type, 
                           levels = c("Pig", "Human"))) %>%

ggplot(aes(x = cat_type, stratum = cat, alluvium = row_n,
             y = 1,
             fill = cat)) +
  scale_x_discrete(expand = c(.1, .1), position = "top") +
  geom_flow(show.legend = F) +
  geom_stratum(show.legend = F, color = NA) + 
  scale_fill_manual(values = tissue_colors_palette_full_humanpig) + 
  geom_text(stat = "stratum", 
            label.strata = TRUE) +
  theme_void() + 
  theme(axis.text.x = element_text(size = 18, face = "bold"), 
        title = element_text(hjust = 0.5)) + 
  ggtitle(compare_tissue)

```

```{r compare_spec_over_alluvial, echo=F, warning = F, message = F, results=F}

pairwise_alluvial_plot(data = human_pig_gene_class_comparison, 
                       var1 = "enssscg_id", 
                       var2 = "ensg_id", 
                       cat1 = "spec_category_pig", 
                       cat2 = "spec_category_human", 
                       cat_levels = spec_category_levels, 
                       cat_names = c("Pig", "Human"), 
                       pal = gene_category_pal)
ggsave(savepath("Class comparison spec alluvial.pdf"), width = 6, height = 10, useDingbats = F)
ggsave(savepath("Class comparison spec alluvial.png"), width = 6, height = 10)

pairwise_subcat_alluvial_plot(data = human_pig_gene_class_comparison, 
                              var1 = "enssscg_id", 
                              var2 = "ensg_id", 
                              cat1 = "spec_category_pig", 
                              cat2 = "spec_category_human", 
                              cat_levels = spec_category_overlap_levels,
                              subcat = "enrichment_overlap",
                              cat_names = c("Pig", "Human"), 
                              pal = spec_category_overlap_pal, 
                              cat_pal = gene_category_pal)
ggsave(savepath("Class comparison spec_over alluvial.pdf"), width = 6, height = 10, useDingbats = F)
ggsave(savepath("Class comparison spec_over alluvial.png"), width = 6, height = 10)

# Dist

pairwise_alluvial_plot(data = human_pig_gene_class_comparison, 
                       var1 = "enssscg_id", 
                       var2 = "ensg_id", 
                       cat1 = "dist_category_pig", 
                       cat2 = "dist_category_human", 
                       cat_levels = dist_category_levels, 
                       cat_names = c("Pig", "Human"), 
                       pal = gene_category_pal)
ggsave(savepath("Class comparison dist alluvial.pdf"), width = 6, height = 10, useDingbats = F)
ggsave(savepath("Class comparison dist alluvial.png"), width = 6, height = 10)

# Regular classification - comparison
pig_gene_class %>% 
  left_join(pig_gene_class_comparison, 
            by = "gene", 
            suffix = c("", "_comparison")) %>%
  mutate(gene_comparison = gene) %>%
pairwise_alluvial_plot(var1 = "gene", 
                       var2 = "gene_comparison", 
                       cat1 = "spec_category", 
                       cat2 = "spec_category_comparison", 
                       cat_levels = spec_category_levels, 
                       cat_names = c("Original", "Comparison"), 
                       pal = gene_category_pal)
ggsave(savepath("Class Pig regular-comp alluvial.pdf"), width = 6, height = 10, useDingbats = F)
ggsave(savepath("Class Pig regular-comp alluvial.png"), width = 6, height = 10)

# 
# alluv_human_pig_class_1 <- 
#   human_pig_gene_class %>% 
#   select(enssscg_id, ensg_id, 
#          spec_category_pig, 
#          spec_category_human) %>% 
#   ungroup() %>%
#   mutate(row_n = row_number()) %>%
#   # group_by(spec_category_pig, spec_category_human) %>% 
#   # mutate(n_genes = n()) %>% 
#   gather(cat_type, cat, -enssscg_id, -ensg_id, -row_n) %>%
#   mutate(cat = factor(cat, levels = c(spec_category_levels, enrichment_overlap_levels)),
#          cat_type = case_when(cat_type == "spec_category_pig" ~ "Pig",
#                               cat_type == "spec_category_human" ~ "Human"),
#          cat_type = factor(cat_type, 
#                            levels = c("Pig", "Human"))) %>%
#   
#   ggplot(aes(x = cat_type, stratum = cat, alluvium = row_n,
#              y = 1,
#              fill = cat)) +
#   scale_x_discrete(expand = c(.1, .1), position = "top") +
#   geom_flow(show.legend = F) +
#   geom_stratum(show.legend = F, color = NA) + 
#   scale_fill_manual(values = c(gene_category_pal, enrichment_overlap_pal)) + 
#   # theme_void() +
#   theme(axis.text.x = element_text(size = 18, face = "bold"),
#         axis.text.y = element_blank(), 
#         axis.ticks = element_blank(), 
#         panel.background = element_blank(), 
#         axis.title = element_blank())
# 
# alluv_human_pig_class_1 <- 
#   alluv_human_pig_class_1 +
#     
#     # Flow label
#   geom_text(stat = "flow", 
#             aes(label = 1),
#             nudge_x = ifelse(ggplot_build(alluv_human_pig_class_1)$data[[1]]$x == 1, 
#                             1/6,
#                            -1/6), 
#             hjust = ifelse(ggplot_build(alluv_human_pig_class_1)$data[[1]]$x == 1, 
#                            0,
#                            1), 
#             size = 2) +
#     
#     # Stratum label
#   geom_text(stat = "stratum", 
#             aes(label = 1), 
#             size = 2,
#             nudge_x = ifelse(ggplot_build(alluv_human_pig_class_1)$data[[2]]$x == 1, 
#                              -1/6 - 1/100,
#                              1/6 + 1/100),
#             angle = ifelse(ggplot_build(alluv_human_pig_class_1)$data[[2]]$x == 1, 
#                              90,
#                              -90), 
#             vjust = 0) + 
#     
#     geom_text(stat = "stratum", 
#             aes(label = cat), 
#             size = 3)
#     # geom_label(stat = "stratum", 
#     #         aes(label = cat), 
#     #         color = "black",
#     #         label.size = 0,
#     #         size = 3, 
#     #         show.legend = F)
# 
# alluv_human_pig_class_1
# 
# ggsave(savepath("Class comparison spec alluvial.pdf"), width = 6, height = 10, useDingbats = F)
# 
# ###
# 
# alluv_human_pig_class_2 <- 
#   human_pig_gene_class %>% 
#   mutate(spec_category_pig_overlap = paste(spec_category_pig, enrichment_overlap),
#          spec_category_human_overlap = paste(spec_category_human, enrichment_overlap)) %>% 
#   select(enssscg_id, ensg_id, 
#          spec_category_pig_overlap, 
#          spec_category_human_overlap) %>% 
#   ungroup() %>%
#   mutate(row_n = row_number()) %>%
#   gather(cat_type, cat, -enssscg_id, -ensg_id, -row_n) %>%
#   mutate(cat = mgsub(cat, 
#                      c("full overlap", "partial overlap", "no overlap"), 
#                      c("FO", "PO", "NO")),
#          cat = factor(cat, levels = c(spec_category_overlap_levels_short)),
#          cat_type = case_when(cat_type == "spec_category_pig_overlap" ~ "Pig",
#                               cat_type == "spec_category_human_overlap" ~ "Human"),
#          cat_type = factor(cat_type, 
#                            levels = c("Pig", "Human"))) %>%
#   
#   ggplot(aes(x = cat_type, stratum = cat, alluvium = row_n,
#              y = 1,
#              fill = cat)) +
#   scale_x_discrete(expand = c(.1, .1), position = "top") +
#   geom_flow(show.legend = F) +
#   geom_stratum(show.legend = F, color = NA) + 
#   scale_fill_manual(values = spec_category_overlap_short_pal) + 
#   # theme_void() +
#   theme(axis.text.x = element_text(size = 18, face = "bold"),
#         axis.text.y = element_blank(), 
#         axis.ticks = element_blank(), 
#         panel.background = element_blank(), 
#         axis.title = element_blank())
# 
# flow_data <- 
#   ggplot_build(alluv_human_pig_class_2)$data[[1]]
# 
# stratum_data <- 
#   ggplot_build(alluv_human_pig_class_2)$data[[2]]
# 
# condensed_stratum_data <- 
#   stratum_data %>% 
#   mutate(label = gsub(" NO$| PO$| FO$", "", stratum)) %>% 
#   group_by(label, x) %>%
#   summarise(ymin = min(ymin), 
#             ymax = max(ymax), 
#             xmin = min(xmin), 
#             xmax = max(xmax),
#             y = (ymin + ymax) / 2) %>% 
#   left_join(gene_category_pal %>% 
#               enframe(), 
#             by = c("label" = "name"))
# 
# alluv_human_pig_class_2 <- 
#   alluv_human_pig_class_2 +
#   
#   
#   
#   # Stratum label
#   geom_text(stat = "stratum",
#             aes(label = 1), 
#             size = 2,
#             nudge_x = ifelse(stratum_data$x == 1, 
#                              -1/6 - 1/100,
#                              1/6 + 1/100),
#             angle = ifelse(stratum_data$x == 1, 
#                            90,
#                            -90), 
#             vjust = 0) + 
#   
#   geom_rect(data = condensed_stratum_data,
#             aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax),
#             inherit.aes = F, 
#             fill = NA, 
#             size = 0.1,
#             color = condensed_stratum_data$value) +
#   
#   geom_text(data = condensed_stratum_data,
#             aes(x = x, y = y, label = label), 
#             size = 3, 
#             inherit.aes = F) + 
#   
#   # Flow label
#   geom_text(stat = "flow", 
#             aes(label = 1),
#             nudge_x = ifelse(flow_data$x == 1, 
#                              1/6,
#                              -1/6), 
#             hjust = ifelse(flow_data$x == 1, 
#                            0,
#                            1), 
#             size = 2) 
#     
#     
# 
# alluv_human_pig_class_2
# 
# ggsave(savepath("Class comparison spec_over alluvial.pdf"), width = 6, height = 10, useDingbats = F)
# 
# 
# 
# 
# 
# # Legend
# enframe(spec_category_overlap_pal[1:9]) %>% 
#   mutate(spec = factor(rep(spec_category_levels[1:3], each = 3), levels = spec_category_levels),
#          overlap = factor(rep(enrichment_overlap_levels, 3), levels = rev(enrichment_overlap_levels))) %>% 
#   ggplot(aes(spec, overlap, fill = name)) + 
#   geom_tile(show.legend = F) + 
#   geom_text(data = tibble(label = c(spec_category_levels[1:3], rev(enrichment_overlap_levels)),
#                           x = c(1:3, rep(3.7, 3)),
#                           y = c(rep(3.8, 3), 1:3)), 
#             aes(x, y, label = label), 
#             inherit.aes = F, 
#             angle = c(45, 45, 45, 0, 0, 0), 
#             hjust = 0, 
#             vjust = 0.5) + 
#   theme_void() + 
#   scale_fill_manual(values = spec_category_overlap_pal) + 
#   scale_x_discrete(expand = expand_scale(3)) + 
#   scale_y_discrete(expand = expand_scale(3)) +
#   coord_fixed(ratio = 1) 
# 
# ggsave(savepath("spec_over legend.pdf"), width = 3, height = 3, useDingbats = F)

# ------ comparison ------

alluv_human_pig_class_1 <- 
  human_pig_gene_class_comparison %>% 
  select(enssscg_id, ensg_id, 
         spec_category_pig, 
         spec_category_human) %>% 
  ungroup() %>%
  mutate(row_n = row_number()) %>%
  # group_by(spec_category_pig, spec_category_human) %>% 
  # mutate(n_genes = n()) %>% 
  gather(cat_type, cat, -enssscg_id, -ensg_id, -row_n) %>%
  mutate(cat = factor(cat, levels = c(spec_category_levels, enrichment_overlap_levels)),
         cat_type = case_when(cat_type == "spec_category_pig" ~ "Pig",
                              cat_type == "spec_category_human" ~ "Human"),
         cat_type = factor(cat_type, 
                           levels = c("Pig", "Human"))) %>%
  
  ggplot(aes(x = cat_type, stratum = cat, alluvium = row_n,
             y = 1,
             fill = cat)) +
  scale_x_discrete(expand = c(.1, .1), position = "top") +
  geom_flow(show.legend = F) +
  geom_stratum(show.legend = F, color = NA) + 
  scale_fill_manual(values = c(gene_category_pal, enrichment_overlap_pal)) + 
  # theme_void() +
  theme(axis.text.x = element_text(size = 18, face = "bold"),
        axis.text.y = element_blank(), 
        axis.ticks = element_blank(), 
        panel.background = element_blank(), 
        axis.title = element_blank())

alluv_human_pig_class_1 <- 
  alluv_human_pig_class_1 +
    
    # Flow label
  geom_text(stat = "flow", 
            aes(label = 1),
            nudge_x = ifelse(ggplot_build(alluv_human_pig_class_1)$data[[1]]$x == 1, 
                            1/6,
                           -1/6), 
            hjust = ifelse(ggplot_build(alluv_human_pig_class_1)$data[[1]]$x == 1, 
                           0,
                           1), 
            size = 2) +
    
    # Stratum label
  geom_text(stat = "stratum", 
            aes(label = 1), 
            size = 2,
            nudge_x = ifelse(ggplot_build(alluv_human_pig_class_1)$data[[2]]$x == 1, 
                             -1/6 - 1/100,
                             1/6 + 1/100),
            angle = ifelse(ggplot_build(alluv_human_pig_class_1)$data[[2]]$x == 1, 
                             90,
                             -90), 
            vjust = 0) + 
    
    geom_text(stat = "stratum", 
            aes(label = cat), 
            size = 3)
    # geom_label(stat = "stratum", 
    #         aes(label = cat), 
    #         color = "black",
    #         label.size = 0,
    #         size = 3, 
    #         show.legend = F)

alluv_human_pig_class_1

ggsave(savepath("Class comparison spec alluvial.pdf"), width = 6, height = 10, useDingbats = F)

###

alluv_human_pig_class_2 <- 
  human_pig_gene_class_comparison %>% 
  mutate(spec_category_pig_overlap = paste(spec_category_pig, enrichment_overlap),
         spec_category_human_overlap = paste(spec_category_human, enrichment_overlap)) %>% 
  select(enssscg_id, ensg_id, 
         spec_category_pig_overlap, 
         spec_category_human_overlap) %>% 
  ungroup() %>%
  mutate(row_n = row_number()) %>%
  gather(cat_type, cat, -enssscg_id, -ensg_id, -row_n) %>%
  mutate(cat = mgsub(cat, 
                     c("full overlap", "partial overlap", "no overlap"), 
                     c("FO", "PO", "NO")),
         cat = factor(cat, levels = c(spec_category_overlap_levels_short)),
         cat_type = case_when(cat_type == "spec_category_pig_overlap" ~ "Pig",
                              cat_type == "spec_category_human_overlap" ~ "Human"),
         cat_type = factor(cat_type, 
                           levels = c("Pig", "Human"))) %>%
  
  ggplot(aes(x = cat_type, stratum = cat, alluvium = row_n,
             y = 1,
             fill = cat)) +
  scale_x_discrete(expand = c(.1, .1), position = "top") +
  geom_flow(show.legend = F) +
  geom_stratum(show.legend = F, color = NA) + 
  scale_fill_manual(values = spec_category_overlap_short_pal) + 
  # theme_void() +
  theme(axis.text.x = element_text(size = 18, face = "bold"),
        axis.text.y = element_blank(), 
        axis.ticks = element_blank(), 
        panel.background = element_blank(), 
        axis.title = element_blank())

flow_data <- 
  ggplot_build(alluv_human_pig_class_2)$data[[1]]

stratum_data <- 
  ggplot_build(alluv_human_pig_class_2)$data[[2]]

condensed_stratum_data <- 
  stratum_data %>% 
  mutate(label = gsub(" NO$| PO$| FO$", "", stratum)) %>% 
  group_by(label, x) %>%
  summarise(ymin = min(ymin), 
            ymax = max(ymax), 
            xmin = min(xmin), 
            xmax = max(xmax),
            y = (ymin + ymax) / 2) %>% 
  left_join(gene_category_pal %>% 
              enframe(), 
            by = c("label" = "name"))

alluv_human_pig_class_2 <- 
  alluv_human_pig_class_2 +
  
  
  
  # Stratum label
  geom_text(stat = "stratum",
            aes(label = 1), 
            size = 2,
            nudge_x = ifelse(stratum_data$x == 1, 
                             -1/6 - 1/100,
                             1/6 + 1/100),
            angle = ifelse(stratum_data$x == 1, 
                           90,
                           -90), 
            vjust = 0) + 
  
  geom_rect(data = condensed_stratum_data,
            aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax),
            inherit.aes = F, 
            fill = NA, 
            size = 0.1,
            color = condensed_stratum_data$value) +
  
  geom_text(data = condensed_stratum_data,
            aes(x = x, y = y, label = label), 
            size = 3, 
            inherit.aes = F) + 
  
  # Flow label
  geom_text(stat = "flow", 
            aes(label = 1),
            nudge_x = ifelse(flow_data$x == 1, 
                             1/6,
                             -1/6), 
            hjust = ifelse(flow_data$x == 1, 
                           0,
                           1), 
            size = 2) 
    
    

alluv_human_pig_class_2

ggsave(savepath("Class comparison spec_over alluvial.pdf"), width = 6, height = 10, useDingbats = F)





# Legend
enframe(spec_category_overlap_pal[1:9]) %>% 
  mutate(spec = factor(rep(spec_category_levels[1:3], each = 3), levels = spec_category_levels),
         overlap = factor(rep(enrichment_overlap_levels, 3), levels = rev(enrichment_overlap_levels))) %>% 
  ggplot(aes(spec, overlap, fill = name)) + 
  geom_tile(show.legend = F) + 
  geom_text(data = tibble(label = c(spec_category_levels[1:3], rev(enrichment_overlap_levels)),
                          x = c(1:3, rep(3.7, 3)),
                          y = c(rep(3.8, 3), 1:3)), 
            aes(x, y, label = label), 
            inherit.aes = F, 
            angle = c(45, 45, 45, 0, 0, 0), 
            hjust = 0, 
            vjust = 0.5) + 
  theme_void() + 
  scale_fill_manual(values = spec_category_overlap_pal) + 
  scale_x_discrete(expand = expand_scale(3)) + 
  scale_y_discrete(expand = expand_scale(3)) +
  coord_fixed(ratio = 1) 

ggsave(savepath("spec_over legend.pdf"), width = 3, height = 3, useDingbats = F)
```

#Project status visualisation

```{r project_status, include = F, message=F, warning=F}

tibble(Task = c("Data generation", 
                "Preprocessing", 
                "Pig Gene Classification", 
                "Human-Pig tissue mapping",
                "Compare classification", 
                "Human-Pig expression correlation", 
                "Make a mammalian RNA Atlas", 
                "Make figures for manuscript",
                "Write manuscript"),
       Progress = c(100, 
                    100, 
                    100,
                    100, 
                    50, 
                    60,
                    5,
                    25, 
                    1)) %>% 
  mutate(Task = factor(Task, levels = rev(Task))) %>%
  ggplot(aes(Task, Progress, fill = Progress)) + 
  geom_col(show.legend = F) + 
  coord_flip() + 
  stripped_theme + 
  scale_fill_gradient2(low = "darkred", 
                       mid = "orange", 
                       high = "green3", 
                       midpoint = 50)

ggsave(savepath("Project Progress.pdf"), width = 6, height = 3)

```




#Correlation and classification comparison

```{r cor_class_comp, include=F, message=F, warning=F}


# ----- Gene examples -----
complete_comparison_table <- 
  cor_p_gene_all_orth_region %>% 
  left_join(cor_s_gene_all_orth_region, 
            by = c("enssscg_id", "ensg_id"), 
            suffix = c("_pearson", "_spearman")) %>% 
  left_join(gene_orthologs) %>% 
  left_join(human_pig_gene_class_comparison) %>%
  left_join(human_pig_gene_dist_class_comparison) %>%
  left_join(dist_gene_all_orth_region)

comparison_table <- 
  complete_comparison_table %>% 
  select(1, 2, 
         ortholog_type,
         cor_pearson, cor_spearman, 
         enriched_tissues_overlap, enriched_tissues_human, enriched_tissues_pig, enrichment_overlap, 
         dist_category_human, spec_category_human, 
         dist_category_pig, spec_category_pig, 
         dist_cat_diff, spec_cat_diff, 
         shared_dist_category, shared_spec_category, 
         dist, mean_enssscg_id,
         mean_ensg_id, common_mean) %>%
  mutate(enrichment_overlap = factor(enrichment_overlap, levels = c("full overlap", "partial overlap", "no overlap")), 
         spec_category_pig = factor(spec_category_pig, levels = spec_category_levels),
         spec_category_human = factor(spec_category_human, levels = spec_category_levels))



pig_id <- "ENSSSCG00000022577"
human_id <- "ENSG00000237038"
     
pdf(savepath("Pig Human barplot examples.pdf"), width = 10, height = 6, useDingbats = F)
lapply(1:50, 
       function(x) {
         i <- sample(1:16445, 1)
         plot_ortholog_pair(pig_id = comparison_table$enssscg_id[i], 
                            human_id = comparison_table$ensg_id[i])
       })
dev.off()

# ----- Correlation distribution -----

comparison_table %>% 
  ggplot(aes(log2(common_mean + 1), log2(dist + 1))) + 
  geom_hex(aes(fill = stat(log10(count))), 
               bins = 60) + 
  # facet_grid(enrichment_overlap ~ spec_category_pig) +
  scale_fill_viridis() + 
  stripped_theme_facet

comparison_table %>% 
  ggplot(aes(log2(dist + 1), cor_spearman)) + 
  geom_hex(aes(fill = stat(log10(count))), 
               bins = 60) + 
  # facet_grid(enrichment_overlap ~ spec_category_pig) +
  scale_fill_viridis() + 
  stripped_theme_facet

comparison_table %>% 
  ggplot(aes(log2(dist/common_mean + 1), cor_spearman)) + 
  geom_hex(aes(fill = stat(log10(count))), 
               bins = 60) + 
  # facet_grid(enrichment_overlap ~ spec_category_pig) +
  scale_fill_viridis() + 
  stripped_theme_facet


comparison_table %>% 
  gather(cor_type, cor, cor_pearson, cor_spearman) %>% 
  ggplot(aes())



comparison_table %>% 
  ggplot(aes(cor_spearman, cor_pearson)) +  
  geom_hex(aes(fill = stat(log10(count))), 
               bins = 60) + 
  facet_grid(enrichment_overlap ~ spec_category_pig) +
  scale_fill_viridis() + 
  stripped_theme_facet

comparison_table %>% 
  ggplot(aes(cor_spearman, cor_pearson)) +  
  geom_hex(aes(fill = stat(log10(count))), 
               bins = 60) + 
  facet_grid(spec_category_human ~ spec_category_pig) +
  scale_fill_viridis() + 
  stripped_theme_facet


comparison_table %>% 
  ggplot(aes(cor_spearman, cor_pearson)) +  
  geom_hex(aes(fill = stat(log10(count))), 
               bins = 60) + 
  facet_grid(enrichment_overlap ~ shared_spec_category) +
  scale_fill_viridis() + 
  stripped_theme_facet




comparison_table %>% 
  ggplot(aes(cor_pearson, fill = ortholog_type)) + 
  # geom_density()+
  # geom_violin() +
  geom_density(alpha = 0.5, color = NA) +
  stripped_theme_facet

comparison_table %>% 
  ggplot(aes(cor_spearman, fill = ortholog_type)) + 
  # geom_density()+
  # geom_violin() +
  geom_density(alpha = 0.5, color = NA) +
  stripped_theme_facet

comparison_table %>% 
  ggplot(aes(cor_pearson, fill = enrichment_overlap)) + 
  # geom_density()+
  # geom_violin() +
  geom_density(alpha = 0.5, color = NA) +
  stripped_theme_facet

comparison_table %>% 
  ggplot(aes(cor_spearman, fill = enrichment_overlap)) + 
  # geom_density()+
  # geom_violin() +
  geom_density(alpha = 0.5, color = NA) +
  stripped_theme_facet


comparison_table %>% 
  ggplot(aes(cor_spearman, cor_pearson)) + 
  geom_abline(slope = 1) +
  geom_hex(aes(fill = stat(log10(count))), 
           bins = 70,
           show.legend = F) + 
  
  facet_wrap(~ enrichment_overlap) +
  scale_fill_viridis_c() +
  stripped_theme_facet

comparison_table %>% 
  ggplot(aes(cor_spearman, cor_pearson)) + 
  geom_abline(slope = 1) +
  geom_hex(aes(fill = stat(log10(count))), 
           bins = 70,
           show.legend = F) + 
  
  facet_wrap(~ shared_dist_category) +
  scale_fill_viridis_c() +
  stripped_theme_facet

comparison_table %>% 
  ggplot(aes(cor_spearman, cor_pearson)) + 
  geom_abline(slope = 1) +
  geom_hex(aes(fill = stat(log10(count))), 
           bins = 70,
           show.legend = F) + 
  
  facet_wrap(~ shared_spec_category) +
  scale_fill_viridis_c() +
  stripped_theme_facet

cor_class_pca <- 
  complete_comparison_table %>% 
  select(1, 2, 
         # cor_pearson, cor_spearman, 
         n_enriched_human, n_enriched_pig, n_enriched_overlap,
         n_detected_human, n_detected_pig, n_overlap_detected_human) %>% 
  unite(1,2, col = "ortholog") %>% 
  column_to_rownames("ortholog") %>% 
  apply(MARGIN = 2, scales::rescale, c(0,1)) %>%
  pca_calc(npcs = 3)



cor_class_pca$scores %>% 
  as_tibble(rownames = "names") %>%
  separate(names, into = c("enssscg_id", "ensg_id")) %>%
  left_join(comparison_table %>% 
              select(1:3, cor_pearson, cor_spearman)) %>%
  
  ggplot(aes(PC1, PC2, z = cor_pearson)) +
  
  # geom_point(aes(color = cor_spearman)) +
  # geom_hex(aes(fill = stat(log10(count)), weight = cor_pearson), 
  #          bins = 70,
  #          show.legend = F) + 
  stat_summary_hex(fun = mean, bins = 40) + 
  scale_fill_viridis_c() +
  scale_color_viridis_c() +
  stripped_theme_HPA + 
  geom_segment(data = cor_class_pca$loadings %>% 
                 as_tibble(rownames = "names"),
               aes(x = 0, xend = PC1, 
                   y = 0, yend = PC2), 
               arrow = arrow(),
               show.legend = F, 
               inherit.aes = F) + 
  geom_text(data = cor_class_pca$loadings %>% 
              as_tibble(rownames = "names"),
            aes(PC1, PC2, label = names), 
            inherit.aes = F)

  
  
  

cor_class_pca$scores %>% 
  as_tibble(rownames = "names") %>%
  ggplot(aes(PC1, PC2)) +
  geom_hex(aes(fill = stat(log10(count))), 
           bins = 70,
           show.legend = F) + 
  
  geom_segment(data = cor_class_pca$loadings  %>% 
                 as_tibble(rownames = "names") , 
               aes(x = 0, xend = PC1, 
                   y = 0, yend = PC2), 
               arrow = arrow(),
               
               show.legend = F) + 
  geom_text(data = cor_class_pca$loadings  %>% 
                 as_tibble(rownames = "names") , 
               aes(x = PC1, 
                   y = PC2, 
                   label = names)) + 
  
  scale_fill_viridis_c() +
  stripped_theme_HPA
  
cor_class_umap <- 
  complete_comparison_table %>% 
  select(1, 2, 
         # cor_pearson, cor_spearman, 
         n_enriched_human, n_enriched_pig, n_enriched_overlap,
         n_detected_human, n_detected_pig, n_overlap_detected_human) %>% 
  unite(1,2, col = "ortholog") %>% 
  column_to_rownames("ortholog") %>% 
  apply(MARGIN = 2, scales::rescale, c(0,1)) %>%
  umap()

cor_class_umap$layout %>% 
  as_tibble(rownames = "names") %>%
  separate(names, into = c("enssscg_id", "ensg_id")) %>%
  left_join(comparison_table %>% 
              select(1:3, cor_pearson, cor_spearman)) %>%
  
  ggplot(aes(V1, V2, z = cor_pearson)) +
  
  # geom_point(aes(color = cor_spearman)) +
  geom_hex(aes(fill = stat(log10(count)), weight = cor_pearson),
           bins = 70,
           show.legend = F) +
  # stat_summary_hex(fun = mean, bins = 40) + 
  scale_fill_viridis_c() +
  scale_color_viridis_c() +
  stripped_theme_HPA 



# ----- lm cor -----

lm_cor_res <-
  complete_comparison_table %>%
  gather(cor_type, cor, cor_pearson, cor_spearman) %>% 
  group_by(cor_type) %>%

  do({
    
    # model <-      lmer(cor ~ enrichment_overlap + ortholog_type + (1|enssscg_id) + (1|ensg_id), data=test.df)
    # model.red <-  lmer(v1 ~ (1|ID), data=test.df)
    # kenward<-KRmodcomp(model, model.red)
    # 
    # p.val<-kenward$stats$p.value
    # estimate<-coef(summary(model))[2,1]
    # 
    # paste(p.val, estimate)

    fit <-
      lm(cor ~
           n_enriched_overlap + #enrichment_overlap + 
           ortholog_type +
           n_detected_overlap,
         data = .)

    confs <-
      confint(fit) %>%
      as_tibble(rownames = "term") %>%
      rename(CIlow = 2,
             CIhigh = 3)
    fit %>%

      tidy() %>%
      left_join(confs,
                by = "term") %>%
      mutate(r_sqrt = summary(fit)$r.squared)
  }) %>%
  group_by(term) %>%
  mutate(adj_pval = p.adjust(p.value, method = "BH")) %>%
  ungroup() %>% 
  mutate(variable = str_extract(term, "enrichment_overlap|ortholog_type|shared_dist_category|shared_spec_category|Intercept"), 
         term = gsub("enrichment_overlap|ortholog_type|shared_dist_category|shared_spec_category", "", term), 
         cor_type = gsub("cor_", "", cor_type))

lm_cor_res %>% 
  arrange(adj_pval) %>% 
  filter(term != "(Intercept)") %>%
  mutate(logP = -log10(adj_pval)) %>%
  ggplot(aes(estimate, logP, label = term)) + 
  geom_text() + 
  facet_wrap(~ cor_type)

lm_cor_res %>% 
  ggplot(aes(term, estimate, fill = cor_type)) + 
  geom_col(position = "dodge") + 
  coord_flip() + 
  facet_grid(variable~., scales = "free_y") + 
  stripped_theme_facet
ggsave(savepath("Correlation lm estimates.pdf"), width = 4, height = 6)

```

```{r}
bind_rows(human_gene_class_comparison %>% 
            mutate(species = "human"), 
          pig_gene_class_comparison %>% 
            mutate(species = "pig")) %>% 
  separate_rows(enriched_tissues, sep = ";") %>%
  group_by(species, spec_category, enriched_tissues) %>% 
  summarise(n_genes = n()) %>% 
  spread(species, n_genes, fill = 0) %>%
  filter(!is.na(enriched_tissues)) %>%
  ggplot(aes(human, pig, color = spec_category)) + 
  geom_point(show.legend = F) + 
  scale_color_manual(values = gene_category_pal) +
  stripped_theme_facet +
  scale_x_log10() + 
  scale_y_log10() + 
  geom_abline(slope = 1) + 
  ggtitle("Number of genes in each enrichment category per tissue") +
  facet_wrap(~ enriched_tissues)
# 
# bind_rows(human_gene_class_comparison %>% 
#             mutate(species = "human"), 
#           pig_gene_class_comparison %>% 
#             mutate(species = "pig")) %>% 
#   separate_rows(tissues_detected, sep = ";") %>%
#   group_by(species, dist_category, tissues_detected) %>% 
#   summarise(n_genes = n()) %>% 
#   spread(species, n_genes, fill = 0) %>%
#   filter(!is.na(tissues_detected)) %>%
#   ggplot(aes(human, pig, color = dist_category)) + 
#   geom_point(show.legend = F) + 
#   scale_color_manual(values = gene_category_pal) +
#   stripped_theme_facet +
#   scale_x_log10() + 
#   scale_y_log10() + 
#   geom_abline(slope = 1) + 
#   ggtitle("Number of genes in each enrichment category per tissue") #+
#   facet_wrap(~ tissues_detected)


```

#ANOVA
##Sex specific genes

```{r sex_spec_genes}

# ANOVA

if(file.exists("../../data/processed/pig_anova.tab")) {
  anova_sex_res <- read_delim("../../data/processed/pig_anova.tab", delim = "\t")
} else {
  anova_sex_res <- 
  pig_atlas_sample_norm %>% 
  mutate(sex = case_when(individual %in% c("a", "b") ~ "female",
                         individual %in% c("c", "d") ~ "male")) %>%
  group_by(enssscg_id) %>% 
  mutate(sd_ = sd(tmm)) %>%
  filter(sd_ > 0) %>%
  select(-sd_) %>% 
  do({
    g_data <- .
    saved_data <<- g_data

    model <- aov(log2(tmm + 1) ~ sex + tissue_ID + individual,
                 data = g_data)

    
    res <-
      model %>%

      tidy() %>%
      left_join(omega_sq(model) %>% 
                  enframe("term", "sqrt_omega") %>%
                  mutate(term = trimws(term)), 
                by = "term")
    res
  }) %>%
  group_by(term) %>%
  mutate(adj_pval = p.adjust(p.value, method = "BH")) %>%
  ungroup() %>%
  mutate(significant = adj_pval < 0.05) 
  
  write_delim(anova_sex_res, "../../data/processed/pig_anova.tab", delim = "\t")
}

if(file.exists("../../data/processed/pig_anova_nosex_tissues.tab")) {
  anova_nosex_res <- read_delim("../../data/processed/pig_anova_nosex_tissues.tab", delim = "\t")
} else {
  anova_nosex_res <- 
    pig_atlas_sample_norm %>% 
    filter(!tissue_ID %in% c('bre', 'bul', 'cerv', 'end', 'epi', 'fal', 'mag', 'ova', 'pen', 'pro', 'sem', 'tes', 'vas')) %>% 
    mutate(sex = case_when(individual %in% c("a", "b") ~ "female",
                           individual %in% c("c", "d") ~ "male")) %>%
    group_by(enssscg_id) %>% 
    mutate(sd_ = sd(tmm)) %>%
    filter(sd_ > 0) %>%
    select(-sd_) %>% 
    do({
      g_data <- .
      saved_data <<- g_data
      
      model <- aov(log2(tmm + 1) ~ sex + tissue_ID + individual,
                   data = g_data)
      
    
    res <-
      model %>%

      tidy() %>%
      left_join(omega_sq(model) %>% 
                  enframe("term", "sqrt_omega") %>%
                  mutate(term = trimws(term)), 
                by = "term")
    res
  }) %>%
  group_by(term) %>%
  mutate(adj_pval = p.adjust(p.value, method = "BH")) %>%
  ungroup() %>%
  mutate(significant = adj_pval < 0.05) 
  
  write_delim(anova_nosex_res, "../../data/processed/pig_anova_nosex_tissues.tab", delim = "\t")
}

if(file.exists("../../data/processed/pig_anova_ward_cluster.Rdata")) {
  load("../../data/processed/pig_anova_ward_cluster.Rdata")
} else {
  pig_anova_clust <- 
    anova_sex_res %>% 
    select(1, 2, sqrt_omega) %>% 
    spread(term, sqrt_omega) %>% 
    column_to_rownames("enssscg_id") %>% 
    dist() %>% 
    hclust(method = "ward.D2")
  
  save(pig_anova_clust, 
       file = "../../data/processed/pig_anova_ward_cluster.Rdata")
}



anova_sex_res %>% 
  filter(term == "sex") %>%
  filter(adj_pval < 0.05) %>% 
  ggplot(aes(sqrt_omega)) + 
  geom_density()

anova_nosex_res %>% 
  filter(term == "sex") %>%
  filter(adj_pval < 0.05) %>% 
  ggplot(aes(sqrt_omega)) + 
  geom_density()

left_join(anova_nosex_res %>% 
            select(1, 2, sqrt_omega),
          anova_sex_res %>% 
            select(1, 2, sqrt_omega), 
          by = c("enssscg_id", "term")) %>% 
  filter(term == "sex") %>%
  filter(sqrt_omega.x > 0.05 | sqrt_omega.y > 0.05) %>%
  ggplot(aes(sqrt_omega.x, sqrt_omega.y)) +
  geom_point()
  


anova_sex_res_geneinfo <-
  anova_sex_res %>% 
  left_join(gene_chromosome %>% 
              select(enssscg_id = 1,
                     chromosome = 4, 
                     gene_name = 5) %>% 
              unique()) %>% 
  left_join(pig_gene_class, 
            by = c("enssscg_id" = "gene"))





# pairwise_alluvial_plot(data = anova_sex_res_cluster, 
#                        var1 = "enssscg_id", 
#                        var2 = "ensg_id", 
#                        cat1 = "spec_category_pig", 
#                        cat2 = "spec_category_human", 
#                        cat_levels = spec_category_levels, 
#                        cat_names = c("Pig", "Human"), 
#                        pal = gene_category_pal)

anova_sex_res_cluster_barplot <- 
  anova_sex_res %>% 
  arrange(term, sqrt_omega) %>%
  mutate(enssscg_id = factor(enssscg_id, unique(enssscg_id)), 
         term = factor(term, c("Residuals", "tissue_ID", "individual", "sex"))) %>%
  ggplot(aes(enssscg_id, sqrt_omega, fill = term)) + 
  geom_col() + 
  stripped_theme_facet + 
  scale_fill_manual(values = c("Residuals" = "gray", "tissue_ID" = "orangered", 
                               "individual" = "skyblue", "sex" = "darkblue")) + 
  
  theme(axis.text.x = element_blank())


ggsave(savepath("ANOVA Pig cluster barplot.pdf"), width = 10, height = 10)  
ggsave(savepath("ANOVA Pig cluster barplot2.pdf"), width = 10, height = 4)  

anova_sex_res_cluster_barplot <- 
  anova_sex_res %>% 
  mutate(enssscg_id = factor(enssscg_id, {filter(., term == "Residuals") %$% 
      enssscg_id[order(sqrt_omega)]}), 
         term = factor(term, c("Residuals", "tissue_ID", "individual", "sex"))) %>%
  ggplot(aes(enssscg_id, sqrt_omega, fill = term)) + 
  geom_col() + 
  stripped_theme_facet + 
  scale_fill_manual(values = c("Residuals" = "gray", "tissue_ID" = "orangered", 
                               "individual" = "skyblue", "sex" = "darkblue")) + 
  scale_y_continuous(expand = c(0,0)) +
  theme(axis.text.x = element_blank()) + 
  xlab("Gene") +
  ylab("% Variance explained") + 
  theme(axis.ticks.x = element_blank(), 
        panel.border = element_blank())
ggsave(savepath("ANOVA Pig cluster barplot3.pdf"), width = 10, height = 4)  

anova_sex_res_geneinfo %>% 
  group_by(spec_category, term) %>%
  summarise(sqrt_omega = mean(sqrt_omega)) %>% 
  ungroup() %>%
  mutate(spec_category = factor(spec_category, spec_category_levels)) %>%
  ggplot(aes(term, sqrt_omega, fill = term)) +
  geom_col() + 
  facet_grid(~spec_category) + 
  stripped_theme_facet + 
  scale_fill_manual(values = c("Residuals" = "gray", "tissue_ID" = "orangered", 
                               "individual" = "skyblue", "sex" = "darkblue")) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))

topsex_genes <- 
  anova_sex_res_geneinfo %>% 
  filter(term == "sex") %>% 
  filter(sqrt_omega >= 0.20) %>%
  arrange(-sqrt_omega) %>% 
  mutate(spec_category = factor(spec_category, levels = spec_category_levels), 
         chromosome = factor(chromosome, levels = c("X", "Y", 1:21))) 

topsex_genes %>%
  ggplot(aes(chromosome, 1, fill = spec_category, label = gene_name)) + 
  geom_col(color = "black") + 
  geom_text(position = "stack",
            vjust = 2, 
            size = 3,
            fontface = "bold",
            color = "white") +
  # coord_flip() + 
  stripped_theme_facet + 
  scale_fill_manual(values = gene_category_pal) + 
  ylab("Number of genes") + 
  xlab("Chromosome") + 
  ggtitle("Genes with >20% variance explained by sex")
  

plot_data <- 
  pig_atlas_sample_norm %>% 
  left_join(sample_meta) %>%
  left_join(gene_mapping,
            by = "enssscg_id") %>% 
  left_join(tissue_mapping %>% 
              select(tissue, region_tissue_name), 
            by = c("tissue_ID" = "tissue")) %>% 
  
  mutate(gene_name = case_when(display_name == "NULL" ~ enssscg_id, 
                               T ~ paste(display_name,  "-",  enssscg_id)), 
         region_tissue_name = factor(region_tissue_name, levels = region_levels))



  
anova_sex_res_geneinfo %>% 
  filter(enssscg_id == topsex_genes$enssscg_id[i]) %>%
  pie_plot(class_col = "term", 
           y_col = "sqrt_omega", 
           pal = anova_pal, 
           dodge = 0.3)
ggsave(savepath("Anova example pie.pdf"), width = 4, height = 4)


plot_data %>%
  filter(enssscg_id %in% topsex_genes$enssscg_id[i]) %>%
  ggplot(aes(region_tissue_name, tmm, group = individual, color = sex)) + 
  geom_point(show.legend = T) + 
  geom_line() +
  facet_wrap( ~ gene_name, scales = "free_y", ncol = 2) + 
  stripped_theme_facet + 
  ylab("Normalized TPM") +
  theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1), 
        strip.text.x = element_text(hjust = 0), 
        axis.title.x = element_blank()) + 
  scale_color_manual(values = sex_palette) + 
  scale_y_continuous(expand = expand_scale(c(0, 0.05)))


#####

anova_sex_res %>% 
  ggplot(aes(term, sqrt_omega)) + 
  geom_violin(draw_quantiles = 0.5, scale = "width")
  
sex_genes <- 
  anova_sex_res %>% 
  filter(sqrt_omega > 0.01 & term == "sex") %>% 
  arrange(-sqrt_omega)

anova_sex_res %>% 
  filter(enssscg_id %in% sex_genes$enssscg_id)  %>%
  mutate(enssscg_id = factor(enssscg_id, sex_genes$enssscg_id), 
         term = factor(term, c("Residuals", "tissue_ID", "individual", "sex"))) %>%
  ggplot(aes(enssscg_id, sqrt_omega, fill = term)) + 
  geom_col() + 
  stripped_theme_facet + 
  scale_fill_manual(values = c("Residuals" = "gray", "tissue_ID" = "orangered", 
                               "individual" = "skyblue", "sex" = "darkblue")) + 
  theme(axis.text.x = element_blank())


anova_sex_res %>% 
  select(1, 2, sqrt_omega) %>% 
  filter(enssscg_id %in% sex_genes$enssscg_id)  %>%
  spread(term, sqrt_omega) %>% 
  column_to_rownames("enssscg_id") %>% 
  pheatmap(clustering_method = "ward.D2")

```

##Species specific genes
```{r species_anova, eval = F}

if(file.exists("../../data/processed/pig_human_anova.tab")) {
  anova_pig_human_res <- read_delim("../../data/processed/pig_human_anova.tab", delim = "\t")
} else {
  anova_pig_human_res <- 
    joined_species_atlas_comparison %>%
    select(mutual_id, comparison_tissue, human_nx, pig_nx) %>%
    gather(species, nx, human_nx, pig_nx) %>%
    mutate(species = gsub("_nx", "", species)) %>% 
    group_by(mutual_id) %>% 
    mutate(sd_ = sd(nx)) %>%
    filter(sd_ > 0) %>%
    select(-sd_) %>% 
    do({
      g_data <- .
      saved_data <<- g_data
      
      model <- aov(log2(nx + 1) ~ comparison_tissue + species,
                   data = g_data)
      
    
    res <-
      model %>%

      tidy() %>%
      left_join(omega_sq(model) %>% 
                  enframe("term", "sqrt_omega") %>%
                  mutate(term = trimws(term)), 
                by = "term")
    res
  }) %>%
  group_by(term) %>%
  mutate(adj_pval = p.adjust(p.value, method = "BH")) %>%
  ungroup() %>%
  mutate(significant = adj_pval < 0.05) 
  
  write_delim(anova_pig_human_res, "../../data/processed/pig_human_anova.tab", delim = "\t")
}


anova_pig_human_res_geninfo <- 
  anova_pig_human_res %>% 
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>%
  left_join(human_pig_gene_class_comparison)


anova_pig_human_res_geninfo %>% 
  group_by(term) %>% 
  summarise(sqrt_omega = mean(sqrt_omega)) %>%
  mutate(term = factor(term, levels = anova_levels)) %>%
  pie_plot(class_col = "term", 
           y_col = "sqrt_omega", 
           pal = anova_pal, 
           dodge = 0.3)



anova_pig_human_res_geninfo %>% 
  group_by(term, spec_category_human) %>% 
  summarise(sqrt_omega = mean(sqrt_omega)) %>%
  ungroup() %>%
  mutate(spec_category_human = factor(spec_category_human, levels = spec_category_levels),
         term = factor(term, levels = anova_levels)) %>%
  ggplot(aes(spec_category_human, sqrt_omega, fill = term)) + 
  geom_col() +
  stripped_theme_facet + 
  scale_fill_manual(values = anova_pal) +
  theme(axis.text.x = element_text(angle = 37, hjust = 1))




anova_pig_human_res_geninfo %>% 
  # filter(significant) %>% 
  filter(spec_category_human == spec_category_pig) %>%
  group_by(term, spec_category_human) %>%
  summarise(sqrt_omega = sum(sqrt_omega)) %$%
  chord_with_title(from = spec_category_human, 
                   to = term, 
                   sizes = sqrt_omega, 
                   grid.col = c(gene_category_pal, anova_pal),
                   groups = c(rep(1, 4),
                              rep(2, 3)), 
                   plot.order = c("tissue enriched", "group enriched", "tissue enhanced", "low tissue specificity",
                                  "comparison_tissue", "Residuals", "species"), 
                   titles = c("Specificity", "Term"), 
                   from_labels = spec_category_human, 
                   to_labels = term) 

anova_pig_human_res_geninfo %>% 
  # filter(significant) %>% 
  filter(spec_category_human == spec_category_pig) %>%
  group_by(ensg_id, spec_category_human) %>%
  summarise(term = term[which(sqrt_omega == max(sqrt_omega))]) %>%
  group_by(term, spec_category_human) %>%
  summarise(n_genes = n()) %$%
  chord_with_title(from = spec_category_human, 
                   to = term, 
                   sizes = n_genes, 
                   grid.col = c(gene_category_pal, anova_pal),
                   groups = c(rep(1, 4),
                              rep(2, 3)), 
                   plot.order = c("tissue enriched", "group enriched", "tissue enhanced", "low tissue specificity",
                                  "comparison_tissue", "Residuals", "species"), 
                   titles = c("Specificity", "Term"), 
                   from_labels = spec_category_human, 
                   size_labels = T,
                   to_labels = term) 
                    
                   
anova_pig_human_res_geninfo %>% 
  # filter(significant) %>% 
  # filter(spec_category_human == spec_category_pig) %>%
  group_by(ensg_id, enssscg_id, spec_category_human, spec_category_pig, shared_spec_category, enrichment_overlap) %>%
  # mutate(res_omega = sqrt_omega[which(term == "Residuals")]) %>%
  # filter(res_omega < 0.5) %>% 
  summarise(max_omega = max(sqrt_omega), 
            max_term = term[which(sqrt_omega == max_omega)]) %>% 
  group_by(spec_category_human, spec_category_pig, shared_spec_category, enrichment_overlap, max_term) %>% 
  summarise(n_genes = n()) %>% 
  arrange(-n_genes) %>% 
  ungroup() %>%
  mutate(spec_category_human = factor(spec_category_human, levels = spec_category_levels),
         spec_category_pig = factor(spec_category_pig, levels = spec_category_levels),
         enrichment_overlap = factor(enrichment_overlap, levels = enrichment_overlap_levels)) %>%
  ggplot(aes(enrichment_overlap, n_genes, fill = max_term)) + 
  geom_col(position = "dodge") + 
  facet_grid(spec_category_human ~ spec_category_pig) +
  stripped_theme_facet + 
  scale_fill_manual(values = anova_pal)

```




###Species ANOVA plots

```{r species_anova_plots, echo=FALSE, message=FALSE, warning=FALSE}
plot_genes <- 
  anova_pig_human_res_geninfo$enssscg_id %>% 
  unique() 


# ANOVA ranked variance explained plot

plot_data <-
  anova_pig_human_res_geninfo %>%
  
  filter(enssscg_id %in% plot_genes) %>%
  
  
  mutate(term = case_when(term %in% c("Residuals", "species") ~ "Residuals and species", 
                          term == "comparison_tissue" ~ "Tissue")) %>%
  
  group_by(enssscg_id, 
           ensg_id, 
           term, 
           ortholog_type, 
           n_enriched_overlap, 
           enrichment_overlap, 
           dist_category_human, 
           spec_category_human, 
           dist_category_pig, 
           spec_category_pig, 
           enrichment_overlap,
           shared_dist_category, 
           shared_spec_category) %>% 
  summarise(var_exp = sum(sqrt_omega)) %>%
  ungroup() %>%
  left_join(complete_comparison_table %>% 
              select(1, 2, cor_pearson, cor_spearman), 
            by = c("ensg_id", "enssscg_id")) %>%
  
  mutate(enssscg_id = factor(enssscg_id, filter(., term == "Residuals and species") %>% 
                               arrange(var_exp) %>% 
                               pull(enssscg_id)), 
         term = factor(term, rev(anova_levels)), 
         rank = unclass(enssscg_id)) 

  
p1 <- 
  plot_data %>% 
  ggplot(aes(enssscg_id, 100 * var_exp, fill = term)) + 
  geom_col(position = "fill", 
           show.legend = F) + 
  stripped_theme_facet + 
  
  scale_fill_manual(values = anova_pal) + 
  scale_x_discrete(expand = c(0,0)) +
  scale_y_continuous(expand = c(0,0)) +
  
  ylab("Explained variance (%)") +
  xlab("Genes") +
  
  annotate("text", 
           x = c(0.25, 0.75) * max(plot_data$rank),
           y = c(0.75, 0.25),
           label = plot_data %>% 
             group_by(term) %>% 
             summarise(var_exp = 100 * mean(var_exp)) %$% 
             paste0(term, "\n", round(var_exp, 1), " %")) +
  
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(),
        
        panel.border = element_blank())


p2 <- 
  plot_data %>% 
  select(rank, 
         spec_category_human, 
         spec_category_pig,
         dist_category_human, 
         dist_category_pig) %>% 
  distinct() %>%
  gather(category_type, category, -rank) %>% 
  mutate(species = str_extract(category_type, "human$|pig$"),
         species = str_to_sentence(species),
         category_type = gsub("_human$|_pig$", "", category_type),
         category = str_to_sentence(category), 
         category = factor(category, unique(c(spec_category_levels, dist_category_levels)))) %>%
  ggplot(aes(x = rank, y = species, fill = category)) +
  ggridges::geom_density_ridges(show.legend = F, 
                                alpha = 0.6) +
  
  facet_wrap(~ category, ncol = 1, strip.position = "right") +
  
  scale_fill_manual(values = gene_category_pal) + 
  scale_x_continuous(expand = c(0,0), limits = c(0, max(plot_data$rank))) +
  
  stripped_theme + 
  theme(strip.text.y = element_text(angle = 0, hjust = 0, face = "bold"), 
        strip.background = element_blank(), 
        panel.spacing = unit(0, "cm"), 
        axis.title = element_blank(), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.line.x = element_blank(), 
        axis.line.y = element_blank())



p3 <- 
  plot_data %>% 
  select(rank, 
         spec_category_human, 
         spec_category_pig,
         n_enriched_overlap) %>% 
  distinct() %>%
  arrange(rank) %>%
  gather(species, category, -rank, -n_enriched_overlap) %>% 
  mutate(species = str_extract(species, "human$|pig$"), 
         species = str_to_sentence(species),
         category = factor(category, spec_category_levels), 
         category = str_to_sentence(category)) %>%
  arrange(rank) %>% 
  filter(!category %in% c("Low tissue specificity", "Not detected")) %>%
  ggplot(aes(x = rank, y = n_enriched_overlap, linetype = species,  color = category)) +
  # geom_line(show.legend = F) +
  geom_smooth(fill = NA, 
              show.legend = T) +
  facet_wrap(~ "Number of overlapping\nenriched tissues", ncol = 1, strip.position = "right") +
  
  scale_color_manual(values = gene_category_pal, guide= F) +
  scale_linetype_discrete(name = "") +
  scale_x_continuous(expand = c(0,0), limits = c(0, max(plot_data$rank))) +
  scale_y_continuous(expand = c(0,0), limits = c(0, 1)) +
  
  stripped_theme + 
  theme(strip.text.y = element_text(angle = 0, hjust = 0, face = "bold"), 
        strip.background = element_blank(), 
        panel.spacing = unit(0, "cm"), 
        axis.title = element_blank(), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.line.x = element_line(color = "gray"),
        axis.line.y = element_blank(),
        legend.position = c(0.8, 0.8))

p4 <- 
  plot_data %>% 
  select(rank, 
         # spec_category_human, 
         # spec_category_pig,
         cor_pearson, 
         cor_spearman) %>% 
  distinct() %>%
  # arrange(rank) %>% 
  gather(cor_type, cor, -rank) %>% 
  mutate(cor_type = gsub("^cor_", "", cor_type), 
         cor_type = str_to_sentence(cor_type)) %>%
  arrange(rank) %>% 
  ggplot(aes(x = rank, y =  cor)) +
  
  geom_abline(slope = 0,
              color = "gray") +
  
  # geom_hex(aes(fill = stat(log10(count))), 
  #          bins = 100, 
  #          show.legend = F) +
  geom_smooth(aes(color = cor_type),
              show.legend = T) +
  
  
  
  # facet_wrap(~ cor_type, ncol = 1, strip.position = "right") +
  
  
  
  scale_fill_viridis_c() +
  scale_color_discrete(name = "") +
  scale_x_continuous(expand = c(0,0), limits = c(0, max(plot_data$rank))) +
  scale_y_continuous(expand = c(0,0)) +
  
  stripped_theme + 
  theme(strip.text.y = element_text(angle = 0, hjust = 0, face = "bold"), 
        strip.background = element_blank(), 
        panel.spacing = unit(0, "cm"), 
        axis.title = element_blank(), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        legend.position = c(0.8, 0.8), 
        axis.line.x = element_blank(),
        axis.line.y = element_blank())

p5 <- 
  plot_data %>% 
  select(rank, 
         enrichment_overlap) %>% 
  distinct() %>% 
  mutate(enrichment_overlap = str_to_sentence(enrichment_overlap), 
         enrichment_overlap = factor(enrichment_overlap, enrichment_overlap_levels)) %>%
  arrange(rank) %>% 
  ggplot(aes(x = rank, fill = enrichment_overlap)) +
  geom_density(show.legend = T, 
                                alpha = 0.6) +
  
  
  
  scale_fill_manual(values = enrichment_overlap_pal, name = "Enrichment Overlap") +
  scale_x_continuous(expand = c(0,0), limits = c(0, max(plot_data$rank))) +
  
  stripped_theme + 
  theme(strip.text.y = element_text(angle = 0, hjust = 0, face = "bold"), 
        strip.background = element_blank(), 
        panel.spacing = unit(0, "cm"), 
        axis.title = element_blank(), 
        axis.text = element_blank(), 
        axis.ticks = element_blank(), 
        axis.line.x = element_blank(), 
        axis.line.y = element_blank(), 
        legend.position = c(0.8,0.8))

pdf(savepath("Species ANOVA summary plot.pdf"), 
    height = 10, width = 6, useDingbats = F)
p4 + p5 + p3 + p2 + p1 + 
  plot_layout(ncol = 1,
              heights = c(0.5, 0.5, 0.5, 1, 1))
dev.off()

marked_ranks <- 
  round(seq(1, max(plot_data$rank), length.out = 100))

example_genes <- 
  plot_data %>% 
  filter(rank %in% marked_ranks) %>%
  select(rank, 1, 2) %>% 
  left_join(human_gene_mapping %>% 
              select(1, 2), 
            by = "ensg_id") %>% 
  unique() %>% 
  arrange(rank)

pdf(savepath("Species ANOVA summary plot marked.pdf"), 
    height = 10, width = 6, useDingbats = F)
p4 + p5 + p3 + p2 + 
  (p1 + 
     geom_segment(data = example_genes, 
                  aes(x = rank, 
                      xend = rank, 
                      y = 0, 
                      yend = 0.1), 
                  inherit.aes = F)+ 
     geom_text(data = example_genes, 
               aes(x = rank, 
                   y = 0.1, 
                   label = gene_name), 
               hjust = 0,
               size = 1.5,
               angle = 90,
               inherit.aes = F)) + 
  plot_layout(ncol = 1,
              heights = c(0.5, 0.5, 0.5, 1, 1))
dev.off()


  

pdf(savepath("species ANOVA examples.pdf"), width = 10, height = 7, useDingbats = F)
lapply(1:nrow(example_genes), 
       function(i) {
         enssscg_id_ <- as.character(example_genes$enssscg_id[i])
         ensg_id_ <- as.character(example_genes$ensg_id[i])
         
         grid.arrange(anova_pig_human_res_geninfo %>% 
                        filter(enssscg_id == enssscg_id_ &
                                 ensg_id == ensg_id_) %>%
                        pie_plot(class_col = "term", 
                                 y_col = "sqrt_omega", 
                                 pal = anova_pal, 
                                 dodge = 0.3), 
                      
                      
                      human_pig_gene_class %>%
                        filter(enssscg_id == enssscg_id_ &
                                 ensg_id == ensg_id_) %$%
                        tibble(species = c("Human", "Pig"),
                               spec_cat = c(spec_category_human, spec_category_pig),
                               dist_cat = c(dist_category_human, dist_category_pig), 
                               enriched_tissues = c(enriched_tissues_human, enriched_tissues_pig)) %>% 
                        separate_rows(enriched_tissues, sep = ";") %>%
                        mutate(tissue_y = scales::rescale(unclass(factor(enriched_tissues)), c(3, 6)), 
                               species_x = case_when(species == "Pig" ~ 2,
                                                     T ~ 1)) %>%
                        ggplot(aes(x = species_x)) + 
                        
                        annotate("text", x = c(1,2), y = c(0, 0), 
                                 label = c("Human", "Pig"), 
                                 fontface = "bold") +
                        geom_label(aes(y = -1, label = spec_cat, fill = spec_cat), 
                                   show.legend = F) + 
                        geom_label(aes(y = -2, label = dist_cat, fill = dist_cat),
                                   show.legend = F) + 
                        
                        geom_label(aes(y = -tissue_y, label = enriched_tissues, fill = enriched_tissues),
                                   show.legend = F) + 
                        
                        annotate("text", x = 0.25, y = -c(1, 2, 3.5), hjust = 1, 
                                 label = c("Spec", "Dist", "Enriched\ntissues"), 
                                 fontface = "bold") +
                        
                        
                        geom_abline(slope = 0, intercept = -c(0.5, 1.5, 2.5), 
                                    color = "gray") +
                        
                        scale_fill_manual(values = c(gene_category_pal, tissue_colors_palette_full_humanpig)) +
                        scale_y_continuous(limits = c(-6, 0)) + 
                        scale_x_continuous(expand = expand_scale(0.4), 
                                           position = "top") + 
                        
                        
                        theme_void(),
                      
                      
                      orth_region %>% 
                        filter(enssscg_id == enssscg_id_ &
                                 ensg_id == ensg_id_) %>%
                        scatter_cor_plot(val1 = "nx_human", 
                                         val2 = "nx_pig", 
                                         lab1 = "ensg_id", 
                                         lab2 = "enssscg_id", 
                                         color_col = "region_tissue_name", 
                                         pal = tissue_colors_palette_full_humanpig),
                      
                      species_atlas_consensus %>% 
                        filter(species_gene_id %in% c(enssscg_id_,ensg_id_)) %>%
                        quick_atlas_barplot(tis_col = "consensus_tissue_name"), 
                      layout_matrix = rbind(c(1, 4),
                                            c(2, 4),
                                            c(3, 4),
                                            c(3, 4)))
         
         
         
       })
dev.off()

```



###Examples

```{r ANOVA_examples, echo=FALSE, message=FALSE, warning=FALSE}

# Examples

species_atlas %>% 
  filter(species_gene_id %in% c("ENSSSCG00000000059", "ENSG00000196419")) %>%
  quick_atlas_barplot

species_atlas %>% 
  filter(species_gene_id %in% c("ENSSSCG00000008209","ENSG00000172073")) %>%
  quick_atlas_barplot

species_atlas %>% 
  filter(species_gene_id %in% c("ENSSSCG00000013605","ENSG00000181786")) %>%
  quick_atlas_barplot


top_term_genes <- 
  anova_pig_human_res_geninfo %>% 
  arrange(-sqrt_omega) %>%
  {bind_rows(filter(., term == "comparison_tissue") %>% 
               select(enssscg_id, ensg_id) %>%
               head(40),
             filter(., term == "species") %>% 
               select(enssscg_id, ensg_id) %>%
               head(40),
             filter(., term == "Residuals") %>% 
               select(enssscg_id, ensg_id) %>%
               head(40),
             filter(., ensg_id %in% sample(unique(ensg_id), 50)) %>% 
               select(enssscg_id, ensg_id))}
  

pdf(savepath("ANOVA species examples.pdf"), width = 10, height = 7, useDingbats = F)
pblapply(1:nrow(top_term_genes), 
       function(i) {
         enssscg_id_ <- top_term_genes$enssscg_id[i]
         ensg_id_ <- top_term_genes$ensg_id[i]
         
         grid.arrange(anova_pig_human_res_geninfo %>% 
                        filter(enssscg_id == enssscg_id_ &
                                 ensg_id == ensg_id_) %>%
                        pie_plot(class_col = "term", 
                                 y_col = "sqrt_omega", 
                                 pal = anova_pal, 
                                 dodge = 0.3), 
                      human_pig_gene_class_comparison %>%
                        filter(enssscg_id == enssscg_id_ &
                                 ensg_id == ensg_id_) %$%
                        tibble(species = c("Human", "Pig"),
                               spec_cat = c(spec_category_human, spec_category_pig),
                               dist_cat = c(dist_category_human, dist_category_pig), 
                               enriched_tissues = c(enriched_tissues_human, enriched_tissues_pig)) %>% 
                        separate_rows(enriched_tissues, sep = ";") %>%
                        mutate(tissue_y = scales::rescale(unclass(factor(enriched_tissues)), c(3, 6)), 
                               species_x = case_when(species == "Pig" ~ 2,
                                                     T ~ 1)) %>%
                        ggplot(aes(x = species_x)) + 
                        
                        annotate("text", x = c(1,2), y = c(0, 0), 
                                 label = c("Human", "Pig"), 
                                 fontface = "bold") +
                        geom_label(aes(y = -1, label = spec_cat, fill = spec_cat), 
                                   show.legend = F) + 
                        geom_label(aes(y = -2, label = dist_cat, fill = dist_cat),
                                   show.legend = F) + 
                        
                        geom_label(aes(y = -tissue_y, label = enriched_tissues, fill = enriched_tissues),
                                   show.legend = F) + 
                        
                        annotate("text", x = 0.25, y = -c(1, 2, 3.5), hjust = 1, 
                                 label = c("Spec", "Dist", "Enriched\ntissues"), 
                                 fontface = "bold") +
                        
                        
                        geom_abline(slope = 0, intercept = -c(0.5, 1.5, 2.5), 
                                    color = "gray") +
                        
                        scale_fill_manual(values = c(gene_category_pal, tissue_colors_palette_full_humanpig)) +
                        scale_y_continuous(limits = c(-6, 0)) + 
                        scale_x_continuous(expand = expand_scale(0.4), 
                                           position = "top") + 
                        
                        
                        theme_void(),
                      orth_region %>% 
                        filter(enssscg_id == enssscg_id_ &
                                 ensg_id == ensg_id_) %>%
                        scatter_cor_plot(val1 = "nx_human", 
                                         val2 = "nx_pig", 
                                         lab1 = "ensg_id", 
                                         lab2 = "enssscg_id", 
                                         color_col = "region_tissue_name", 
                                         pal = tissue_colors_palette_full_humanpig),
                      species_atlas %>% 
                        filter(species_gene_id %in% c(enssscg_id_,ensg_id_)) %>%
                        quick_atlas_barplot(), 
                      layout_matrix = rbind(c(1, 4),
                                            c(2, 4),
                                            c(3, 4),
                                            c(3, 4)))
         
         
         
       })
dev.off()

pdf(savepath("ANOVA species examples all tissues.pdf"), width = 10, height = 7, useDingbats = F)
lapply(1:nrow(top_term_genes), 
       function(i) {
         enssscg_id_ <- top_term_genes$enssscg_id[i]
         ensg_id_ <- top_term_genes$ensg_id[i]
         
         grid.arrange(anova_pig_human_res_geninfo %>% 
                        filter(enssscg_id == enssscg_id_ &
                                 ensg_id == ensg_id_) %>%
                        pie_plot(class_col = "term", 
                                 y_col = "sqrt_omega", 
                                 pal = anova_pal, 
                                 dodge = 0.3), 
                      human_pig_gene_class %>%
                        filter(enssscg_id == enssscg_id_ &
                                 ensg_id == ensg_id_) %$%
                        tibble(species = c("Human", "Pig"),
                               spec_cat = c(spec_category_human, spec_category_pig),
                               dist_cat = c(dist_category_human, dist_category_pig), 
                               enriched_tissues = c(enriched_tissues_human, enriched_tissues_pig)) %>% 
                        separate_rows(enriched_tissues, sep = ";") %>%
                        mutate(tissue_y = scales::rescale(unclass(factor(enriched_tissues)), c(3, 6)), 
                               species_x = case_when(species == "Pig" ~ 2,
                                                     T ~ 1)) %>%
                        ggplot(aes(x = species_x)) + 
                        
                        annotate("text", x = c(1,2), y = c(0, 0), 
                                 label = c("Human", "Pig"), 
                                 fontface = "bold") +
                        geom_label(aes(y = -1, label = spec_cat, fill = spec_cat), 
                                   show.legend = F) + 
                        geom_label(aes(y = -2, label = dist_cat, fill = dist_cat),
                                   show.legend = F) + 
                        
                        geom_label(aes(y = -tissue_y, label = enriched_tissues, fill = enriched_tissues),
                                   show.legend = F) + 
                        
                        annotate("text", x = 0.25, y = -c(1, 2, 3.5), hjust = 1, 
                                 label = c("Spec", "Dist", "Enriched\ntissues"), 
                                 fontface = "bold") +
                        
                        
                        geom_abline(slope = 0, intercept = -c(0.5, 1.5, 2.5), 
                                    color = "gray") +
                        
                        scale_fill_manual(values = c(gene_category_pal, tissue_colors_palette_full_humanpig)) +
                        scale_y_continuous(limits = c(-6, 0)) + 
                        scale_x_continuous(expand = expand_scale(0.4), 
                                           position = "top") + 
                        
                        
                        theme_void(),
                      orth_region %>% 
                        filter(enssscg_id == enssscg_id_ &
                                 ensg_id == ensg_id_) %>%
                        scatter_cor_plot(val1 = "nx_human", 
                                         val2 = "nx_pig", 
                                         lab1 = "ensg_id", 
                                         lab2 = "enssscg_id", 
                                         color_col = "region_tissue_name", 
                                         pal = tissue_colors_palette_full_humanpig),
                      species_atlas_consensus %>% 
                        filter(species_gene_id %in% c(enssscg_id_,ensg_id_)) %>%
                        quick_atlas_barplot(tis_col = "consensus_tissue_name"), 
                      layout_matrix = rbind(c(1, 4),
                                            c(2, 4),
                                            c(3, 4),
                                            c(3, 4)))
         
         
         
       })
dev.off()
  

```

##Ortholog ANOVA

###Region level

```{r orth_assign_ANOVA, echo=FALSE, message=FALSE, warning=FALSE}

# left_join(anova_pig_human_res_geninfo %>% 
#             select(1:3, sqrt_omega) %>% 
#             mutate(term = ifelse(term == "comparison_tissue", "tissue", term)),
#           
# 
#   ortholog_anova_geninfo %>%
#     select(1:4, sqrt_omega) %>% 
#     mutate(term = ifelse(term == "human_region_name", "tissue", term)),
#   by = c("enssscg_id", "ensg_id", "term")) -> A
# 
# A %>% 
#   ggplot(aes(sqrt_omega.x, sqrt_omega.y)) + 
#   geom_hex(aes(fill = stat(log10(count))),
#            bins = 100) + 
#   scale_fill_viridis_c() + 
#   facet_wrap(~term)


if(file.exists("../../data/processed/ortholog_anova.tab")) {
  ortholog_anova <- read_delim("../../data/processed/ortholog_anova.tab", delim = "\t")
} else {
  ortholog_anova <- 
    joined_region_atlas_allorth %>%
    filter(!is.na(human_region_name)) %>% 
    group_by(mutual_id) %>% 
    mutate(sd_ = sd(nx)) %>%
    filter(sd_ > 0) %>%
    select(-sd_) %>% 
    do({
      g_data <- .
      saved_data <<- g_data
      
      model <- aov(log2(nx + 1) ~ human_region_name + species,
                   data = g_data)
      
    
    res <-
      model %>%

      tidy() %>%
      left_join(omega_sq(model) %>% 
                  enframe("term", "sqrt_omega") %>%
                  mutate(term = trimws(term)), 
                by = "term")
    res
  }) %>%
  group_by(term) %>%
  mutate(adj_pval = p.adjust(p.value, method = "BH")) %>%
  ungroup() %>%
  mutate(significant = adj_pval < 0.05) 
  
  write_delim(ortholog_anova, "../../data/processed/ortholog_anova.tab", delim = "\t")
}

ortholog_anova_geninfo_ <- 
  ortholog_anova %>% 
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_", remove = F) %>%
  left_join(human_pig_gene_all_orths_class_comparison) 


ortholog_anova_ortholog_assignment <- 
  ortholog_anova_geninfo_ %>% 
  filter(term == "human_region_name") %>% 
  select(mutual_id, 
         ensg_id, 
         enssscg_id, 
         ortholog_type,
         sqrt_omega) %>% 
  # gather(species, gene_id, human, pig) %>% 
  group_by(ensg_id) %>% 
  mutate(ortholog_class_human = case_when(sqrt_omega == max(sqrt_omega) ~ "primary",
                                          T ~ "secondary")) %>% 
  group_by(enssscg_id) %>% 
  mutate(ortholog_class_pig = case_when(sqrt_omega == max(sqrt_omega) ~ "primary",
                                        T ~ "secondary")) %>% 
  ungroup() %>% 
  mutate(mutual_primary = ortholog_class_human == "primary" & ortholog_class_pig == "primary")

ortholog_anova_geninfo <-
  ortholog_anova_geninfo_ %>%
  left_join(ortholog_anova_ortholog_assignment %>% 
              select(1, 
                     ortholog_class_pig, 
                     ortholog_class_human, 
                     mutual_primary), 
            by = "mutual_id") 






```

###Comparison level

```{r orth_assign_ANOVA_comparison, echo=FALSE, message=FALSE, warning=FALSE}


if(file.exists("../../data/processed/ortholog_comparison_anova.tab")) {
  ortholog_comparison_anova <- read_delim("../../data/processed/ortholog_comparison_anova.tab", delim = "\t")
} else {
  ortholog_comparison_anova <- 
    joined_comparison_atlas_allorth %>%
    group_by(mutual_id) %>% 
    mutate(sd_ = sd(nx)) %>%
    filter(sd_ > 0) %>%
    select(-sd_) %>% 
    do({
      g_data <- .
      saved_data <<- g_data
      
      model <- aov(log2(nx + 1) ~ comparison_tissue + species,
                   data = g_data)
      
    
    res <-
      model %>%

      tidy() %>%
      left_join(omega_sq(model) %>% 
                  enframe("term", "sqrt_omega") %>%
                  mutate(term = trimws(term)), 
                by = "term")
    res
  }) %>%
  group_by(term) %>%
  mutate(adj_pval = p.adjust(p.value, method = "BH")) %>%
  ungroup() %>%
  mutate(significant = adj_pval < 0.05) 
  
  write_delim(ortholog_comparison_anova, "../../data/processed/ortholog_comparison_anova.tab", delim = "\t")
}

ortholog_comparison_anova_geninfo_ <- 
  ortholog_comparison_anova %>% 
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_", remove = F) %>%
  left_join(human_pig_gene_all_orths_class_comparison) 


ortholog_comparison_anova_ortholog_assignment <- 
  ortholog_comparison_anova_geninfo_ %>% 
  filter(term == "comparison_tissue") %>% 
  select(mutual_id, 
         ensg_id, 
         enssscg_id, 
         ortholog_type,
         sqrt_omega) %>% 
  # gather(species, gene_id, human, pig) %>% 
  group_by(ensg_id) %>% 
  mutate(ortholog_class_human = case_when(sqrt_omega == max(sqrt_omega) ~ "primary",
                                          T ~ "secondary")) %>% 
  group_by(enssscg_id) %>% 
  mutate(ortholog_class_pig = case_when(sqrt_omega == max(sqrt_omega) ~ "primary",
                                        T ~ "secondary")) %>% 
  ungroup() %>% 
  mutate(mutual_primary = ortholog_class_human == "primary" & ortholog_class_pig == "primary")

ortholog_comparison_anova_high_conf_ortholog_assignment <- 
  ortholog_comparison_anova_geninfo_ %>% 
  filter(ortholog_confidence == 1) %>%
  filter(term == "comparison_tissue") %>% 
  select(mutual_id, 
         ensg_id, 
         enssscg_id, 
         ortholog_type,
         sqrt_omega) %>% 
  # gather(species, gene_id, human, pig) %>% 
  group_by(ensg_id) %>% 
  mutate(ortholog_class_human = case_when(sqrt_omega == max(sqrt_omega) ~ "primary",
                                          T ~ "secondary")) %>% 
  group_by(enssscg_id) %>% 
  mutate(ortholog_class_pig = case_when(sqrt_omega == max(sqrt_omega) ~ "primary",
                                        T ~ "secondary")) %>% 
  ungroup() %>% 
  mutate(mutual_primary = ortholog_class_human == "primary" & ortholog_class_pig == "primary")

ortholog_comparison_anova_geninfo <-
  ortholog_comparison_anova_geninfo_ %>%
  left_join(ortholog_comparison_anova_ortholog_assignment %>% 
              select(1, 
                     ortholog_class_pig, 
                     ortholog_class_human, 
                     mutual_primary), 
            by = "mutual_id") 






```



### Ortholog ANOVA plots

####All
```{r orth_assign_ANOVA_plots_all, echo=FALSE, message=FALSE, warning=FALSE}

ortholog_comparison_anova_geninfo %>% 
  filter(ortholog_type != "ortholog_one2one") %>%
  ggplot(aes(sqrt_omega, fill = ortholog_class_pig)) + 
  geom_density(alpha = 0.5) + 
  stripped_theme_facet



plot_genes <- 
  ortholog_comparison_anova_geninfo$enssscg_id %>% 
  unique() 


# ANOVA ranked variance explained plot

plot_data <-
  ortholog_comparison_anova_geninfo %>%
  
  filter(enssscg_id %in% plot_genes) %>%
  
  
  mutate(term = case_when(term %in% c("Residuals", "species") ~ "Residuals and species", 
                          term == "comparison_tissue" ~ "Tissue")) %>%
  
  group_by(mutual_id, 
           enssscg_id, 
           ensg_id, 
           term, 
           ortholog_type, 
           n_enriched_overlap, 
           enrichment_overlap, 
           dist_category_human, 
           spec_category_human, 
           dist_category_pig, 
           spec_category_pig, 
           enrichment_overlap,
           shared_dist_category, 
           shared_spec_category) %>% 
  summarise(var_exp = sum(sqrt_omega)) %>%
  ungroup() %>%
  left_join(complete_comparison_table %>% 
              select(1, 2, cor_pearson, cor_spearman), 
            by = c("ensg_id", "enssscg_id")) %>%
  
  mutate(mutual_id = factor(mutual_id, filter(., term == "Residuals and species") %>% 
                              arrange(var_exp) %>% 
                              pull(mutual_id)), 
         term = factor(term, rev(anova_levels)), 
         rank = unclass(mutual_id)) 

  
p1 <- 
  plot_data %>% 
  ggplot(aes(mutual_id, 100 * var_exp, fill = term)) + 
  geom_col(position = "fill", 
           show.legend = F) + 
  stripped_theme_facet + 
  
  scale_fill_manual(values = anova_pal) + 
  scale_x_discrete(expand = c(0,0)) +
  scale_y_continuous(expand = c(0,0)) +
  
  ylab("Explained variance (%)") +
  xlab("Genes") +
  
  annotate("text", 
           x = c(0.25, 0.75) * max(plot_data$rank),
           y = c(0.75, 0.25),
           label = plot_data %>% 
             group_by(term) %>% 
             summarise(var_exp = 100 * mean(var_exp)) %$% 
             paste0(term, "\n", round(var_exp, 1), " %")) +
  
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(),
        
        panel.border = element_blank())


p2 <- 
  plot_data %>% 
  select(rank, 
         spec_category_human, 
         spec_category_pig,
         dist_category_human, 
         dist_category_pig) %>% 
  distinct() %>%
  gather(category_type, category, -rank) %>% 
  mutate(species = str_extract(category_type, "human$|pig$"),
         species = str_to_sentence(species),
         category_type = gsub("_human$|_pig$", "", category_type),
         category = str_to_sentence(category), 
         category = factor(category, unique(c(spec_category_levels, dist_category_levels)))) %>%
  ggplot(aes(x = rank, y = species, fill = category)) +
  ggridges::geom_density_ridges(show.legend = F, 
                                alpha = 0.6) +
  
  facet_wrap(~ category, ncol = 1, strip.position = "right") +
  
  scale_fill_manual(values = gene_category_pal) + 
  scale_x_continuous(expand = c(0,0), limits = c(0, max(plot_data$rank))) +
  
  stripped_theme + 
  theme(strip.text.y = element_text(angle = 0, hjust = 0, face = "bold"), 
        strip.background = element_blank(), 
        panel.spacing = unit(0, "cm"), 
        axis.title = element_blank(), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.line.x = element_blank(), 
        axis.line.y = element_blank())



p3 <- 
  plot_data %>% 
  select(rank, 
         spec_category_human, 
         spec_category_pig,
         n_enriched_overlap) %>% 
  distinct() %>%
  arrange(rank) %>%
  gather(species, category, -rank, -n_enriched_overlap) %>% 
  mutate(species = str_extract(species, "human$|pig$"), 
         species = str_to_sentence(species),
         category = factor(category, spec_category_levels), 
         category = str_to_sentence(category)) %>%
  arrange(rank) %>% 
  filter(!category %in% c("Low tissue specificity", "Not detected")) %>%
  ggplot(aes(x = rank, y = n_enriched_overlap, linetype = species,  color = category)) +
  # geom_line(show.legend = F) +
  geom_smooth(fill = NA, 
              show.legend = T) +
  facet_wrap(~ "Number of overlapping\nenriched tissues", ncol = 1, strip.position = "right") +
  
  scale_color_manual(values = gene_category_pal, guide= F) +
  scale_linetype_discrete(name = "") +
  scale_x_continuous(expand = c(0,0), limits = c(0, max(plot_data$rank))) +
  scale_y_continuous(expand = c(0,0), limits = c(0, 1)) +
  
  stripped_theme + 
  theme(strip.text.y = element_text(angle = 0, hjust = 0, face = "bold"), 
        strip.background = element_blank(), 
        panel.spacing = unit(0, "cm"), 
        axis.title = element_blank(), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        axis.line.x = element_line(color = "gray"),
        axis.line.y = element_blank(),
        legend.position = c(0.8, 0.8))

p4 <- 
  plot_data %>% 
  select(rank, 
         # spec_category_human, 
         # spec_category_pig,
         cor_pearson, 
         cor_spearman) %>% 
  distinct() %>%
  # arrange(rank) %>% 
  gather(cor_type, cor, -rank) %>% 
  mutate(cor_type = gsub("^cor_", "", cor_type), 
         cor_type = str_to_sentence(cor_type)) %>%
  arrange(rank) %>% 
  ggplot(aes(x = rank, y =  cor)) +
  
  geom_abline(slope = 0,
              color = "gray") +
  
  # geom_hex(aes(fill = stat(log10(count))), 
  #          bins = 100, 
  #          show.legend = F) +
  geom_smooth(aes(color = cor_type),
              show.legend = T) +
  
  
  
  # facet_wrap(~ cor_type, ncol = 1, strip.position = "right") +
  
  
  
  scale_fill_viridis_c() +
  scale_color_discrete(name = "") +
  scale_x_continuous(expand = c(0,0), limits = c(0, max(plot_data$rank))) +
  scale_y_continuous(expand = c(0,0)) +
  
  stripped_theme + 
  theme(strip.text.y = element_text(angle = 0, hjust = 0, face = "bold"), 
        strip.background = element_blank(), 
        panel.spacing = unit(0, "cm"), 
        axis.title = element_blank(), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), 
        legend.position = c(0.8, 0.8), 
        axis.line.x = element_blank(),
        axis.line.y = element_blank())


p5 <- 
  plot_data %>% 
  select(rank, 
         enrichment_overlap) %>% 
  distinct() %>% 
  mutate(enrichment_overlap = str_to_sentence(enrichment_overlap), 
         enrichment_overlap = factor(enrichment_overlap, enrichment_overlap_levels)) %>%
  arrange(rank) %>% 
  ggplot(aes(x = rank, fill = enrichment_overlap)) +
  geom_density(show.legend = T, 
                                alpha = 0.6) +
  
  
  
  scale_fill_manual(values = enrichment_overlap_pal, name = "Enrichment Overlap") +
  scale_x_continuous(expand = c(0,0), limits = c(0, max(plot_data$rank))) +
  
  stripped_theme + 
  theme(strip.text.y = element_text(angle = 0, hjust = 0, face = "bold"), 
        strip.background = element_blank(), 
        panel.spacing = unit(0, "cm"), 
        axis.title = element_blank(), 
        axis.text = element_blank(), 
        axis.ticks = element_blank(), 
        axis.line.x = element_blank(), 
        axis.line.y = element_blank(), 
        legend.position = c(0.8,0.8))

pdf(savepath("Species ANOVA all orth summary plot.pdf"), 
    height = 10, width = 6, useDingbats = F)
p4 + p5 + p3 + p2 + p1 + 
  plot_layout(ncol = 1,
              heights = c(0.5, 0.5, 0.5, 1, 1))
dev.off()

##



marked_ranks <- 
  c(1:49, 
    round(seq(50, max(plot_data$rank), length.out = 30)))

example_genes <- 
  plot_data %>% 
  select(rank, 1, 2, 3) %>% 
  left_join(human_gene_mapping %>% 
              select(1, 2), 
            by = "ensg_id") %>% 
  unique() %>% 
  
  arrange(rank) %>% 
  filter(rank %in% marked_ranks) %>%
  pull(enssscg_id) %>%
  
  unique()

  

pdf(savepath("species ANOVA examples.pdf"), width = 18, height = 11, useDingbats = F)
lapply(example_genes, 
       function(enssscg_id_) {
         
         grid.arrange(ortholog_comparison_anova_geninfo %>% 
                        filter(enssscg_id == enssscg_id_) %>%
                        
                        mutate(term = case_when(term == "species" ~ "Species",
                                                term == "comparison_tissue" ~ "Tissue",
                                                term == "Residuals" ~ "Residuals")) %>%
                        mutate(ensg_id = factor(ensg_id, levels = {filter(., term == "Tissue") %$%
                                   ensg_id[order(sqrt_omega)]})) %>%
                        select(enssscg_id, ensg_id, term, sqrt_omega) %>%
                        mutate(term = factor(term, levels = c("Tissue", "Species", "Residuals"))) %>%
                        ggplot(aes(ensg_id, sqrt_omega, label = paste(round(sqrt_omega * 100), "%"), fill = term)) +
                        geom_col() + 
                        geom_text(position = "stack", 
                                  vjust = 1.1) + 
                        
                        scale_fill_manual(values = anova_pal) + 
                        stripped_theme_facet + 
                        theme(axis.line = element_blank(), 
                              axis.text.y = element_blank(), 
                              
                              axis.title = element_blank(),
                              axis.ticks = element_blank(), 
                              panel.background = element_blank(), 
                              panel.border = element_blank()), 
                      
                      
                      human_pig_gene_all_orths_class_comparison %>%
                        filter(enssscg_id == enssscg_id_) %>%
                        select(enssscg_id, 
                               ensg_id, 
                               spec_category_human, 
                               spec_category_pig,
                               dist_category_human, 
                               dist_category_pig, 
                               enriched_tissues_human, 
                               enriched_tissues_pig) %>% 
                        gather(gene_type, gene_id, 1:2) %>%
                        mutate(species = case_when(gene_type == "enssscg_id" ~ "Pig",
                                                   gene_type == "ensg_id" ~ "Human"),
                               spec_category = case_when(species == "Human" ~ spec_category_human,
                                                         species == "Pig" ~ spec_category_pig), 
                               dist_category = case_when(species == "Human" ~ dist_category_human,
                                                         species == "Pig" ~ dist_category_pig), 
                               enriched_tissues = case_when(species == "Human" ~ enriched_tissues_human,
                                                         species == "Pig" ~ enriched_tissues_pig)) %>% 
                        select(species, gene_id, 
                               spec_cat = spec_category, 
                               dist_cat = dist_category, 
                               enriched_tissues) %>% 
                        unique() %>%
                        separate_rows(enriched_tissues, sep = ";") %>%
                        mutate(tissue_y = scales::rescale(unclass(factor(enriched_tissues)), c(3, 6))) %>%
                        ggplot(aes(x = 1)) + 
                        
                        
                        geom_label(aes(y = -1, label = spec_cat, fill = spec_cat), 
                                   show.legend = F) + 
                        geom_label(aes(y = -2, label = dist_cat, fill = dist_cat),
                                   show.legend = F) + 
                        
                        geom_label(aes(y = -tissue_y, label = enriched_tissues, fill = enriched_tissues),
                                   show.legend = F) + 
                        # 
                        # annotate("text", x = 0.25, y = -c(1, 2, 3.5), hjust = 1, 
                        #          label = c("Spec", "Dist", "Enriched\ntissues"), 
                        #          fontface = "bold") +
                        
                        
                        geom_abline(slope = 0, intercept = -c(0.5, 1.5, 2.5), 
                                    color = "gray") +
                        
                        scale_fill_manual(values = c(gene_category_pal, tissue_colors_palette_full_humanpig)) +
                        scale_y_continuous(limits = c(-6, 0)) + 
                        scale_x_continuous(expand = expand_scale(0.4), 
                                           position = "top") + 
                        
                        facet_wrap(~species + gene_id) +
                        
                        theme_void(),
                      
                      species_all_orth_atlas_consensus %>% 
                        filter(enssscg_id == enssscg_id_) %>%
                        quick_atlas_barplot(tis_col = "consensus_tissue_name"), 
                      layout_matrix = rbind(c(1, 3),
                                            c(2, 3)))
         
         
         
       })
dev.off()
```

```{r joined_pca_umap, include = F, echo = F}



joined_species_nx_pca <- 
  joined_species_nx_norm %>% 
  t() %>%
  pca_calc(npcs = 50)


joined_species_nx_umap <- 
  joined_species_nx_norm %>% 
  {.[complete.cases(.),]} %>%
  t() %>%
  umap_calc(npcs = 2, n_neighbors = 6)

joined_species_not_norm_nx_umap <- 
  joined_species_nx %>% 
  {.[complete.cases(.),]} %>%
  t() %>%
  umap_calc(npcs = 2, n_neighbors = 6)

pca_R2_plot(joined_species_nx_pca$stats, 
            mark_pc = joined_species_nx_pca$pc_95) + 
  ggtitle("Sample PCA")


joined_species_nx_pca$scores %>% 
  as_tibble(rownames = "species_comparison_tissue") %>%
  separate(species_comparison_tissue, into = c("species", "comparison_tissue"), sep = "_", remove = F) %>%
    group_by(comparison_tissue) %>%
    mutate(mean_x = mean(PC3),
           mean_y = mean(PC2)) %>%
    ungroup() %>%

    {ggplot(., aes(PC3, PC2, fill = comparison_tissue,
                   shape = species)) +
        geom_segment(aes(xend = mean_x, yend = mean_y),
                     show.legend = F) +
        geom_point(show.legend = T, 
                   size = 3) +
        geom_text_repel(data = select(., comparison_tissue, mean_x, mean_y) %>%
                    unique(),
                  aes(x = mean_x, y = mean_y, label = comparison_tissue),
                  show.legend = F,
                  size = 3) +
        scale_fill_manual(values = tissue_colors_palette_full_humanpig) +
        scale_shape_manual(values = c("pig" = 21, "human" = 22)) +
        guides(fill = FALSE) +
        simple_theme + 
        theme(legend.position = c(0.8,0.8))}
ggsave(savepath("Human-Pig normalized PCA.pdf"), height = 8, width = 8, useDingbats = F)
ggsave(savepath("Human-Pig normalized PCA.png"), height = 8, width = 8)

joined_species_nx_umap$layout %>% 
  as_tibble(rownames = "species_comparison_tissue") %>%
  separate(species_comparison_tissue, into = c("species", "comparison_tissue"), sep = "_", remove = F) %>%
    group_by(comparison_tissue) %>%
    mutate(mean_x = mean(V1),
           mean_y = mean(V2)) %>%
    ungroup() %>%

    {ggplot(., aes(V1, V2, fill = comparison_tissue,
                   shape = species)) +
        geom_segment(aes(xend = mean_x, yend = mean_y),
                     show.legend = F) +
        geom_point(show.legend = T, 
                   size = 3) +
        geom_text_repel(data = select(., comparison_tissue, mean_x, mean_y) %>%
                    unique(),
                  aes(x = mean_x, y = mean_y, label = comparison_tissue),
                  show.legend = F, 
                  size = 3) +
        scale_fill_manual(values = tissue_colors_palette_full_humanpig) +
        scale_shape_manual(values = c("pig" = 21, "human" = 22)) +
        guides(fill = FALSE) +
        simple_theme + 
        theme(legend.position = c(0.8,0.8))}
ggsave(savepath("Human-Pig normalized UMAP.pdf"), height = 8, width = 8, useDingbats = F)
ggsave(savepath("Human-Pig normalized UMAP.png"), height = 8, width = 8)

joined_species_not_norm_nx_umap$layout %>% 
  as_tibble(rownames = "species_comparison_tissue") %>%
  separate(species_comparison_tissue, into = c("species", "comparison_tissue"), sep = "_", remove = F) %>%
    group_by(comparison_tissue) %>%
    mutate(mean_x = mean(V1),
           mean_y = mean(V2)) %>%
    ungroup() %>%

    {ggplot(., aes(V1, V2, fill = comparison_tissue,
                   shape = species)) +
        geom_segment(aes(xend = mean_x, yend = mean_y),
                     show.legend = F) +
        geom_point(show.legend = T, 
                   size = 3) +
        geom_text_repel(data = select(., comparison_tissue, mean_x, mean_y) %>%
                    unique(),
                  aes(x = mean_x, y = mean_y, label = comparison_tissue),
                  show.legend = F, 
                  size = 3) +
        scale_fill_manual(values = tissue_colors_palette_full_humanpig) +
        scale_shape_manual(values = c("pig" = 21, "human" = 22)) +
        guides(fill = FALSE) +
        simple_theme + 
        theme(legend.position = c(0.8,0.8))}
ggsave(savepath("Human-Pig not normalized UMAP.pdf"), height = 8, width = 8, useDingbats = F)
ggsave(savepath("Human-Pig not normalized UMAP.png"), height = 8, width = 8)

```


#Orthologs 
## Orthology network

```{r orth_net, include = F, message=F, warning=F}


gene_ortholog_network <- 
  make_ortholog_net(gene_orthologs_all)

gene_ortholog_high_conf_network <- 
  make_ortholog_net(gene_orthologs_all %>% 
                      filter(ortholog_confidence == 1))

ortholog_connection_data <- 
  gene_orthologs_all %>% 
  left_join(gene_ortholog_network$communities, 
            by = c("enssscg_id" = "node")) %>% 
  group_by(community) %>%
  mutate(edge_size = n()) %>%
  ungroup() %>%
  arrange(-community_size) %>%
  left_join(cor_s_gene_all_orth_all_region %>% 
              select(1, 2, 4), 
            by = c("enssscg_id", "ensg_id")) %>%
  left_join(ortholog_comparison_anova_ortholog_assignment %>% 
              select(enssscg_id, ensg_id, sqrt_omega, 
                     ortholog_class_human, ortholog_class_pig, 
                     mutual_primary)) %>%
  rename(expression_identity = sqrt_omega) %>% 
  mutate(ortholog_confidence = factor(ortholog_confidence))
  

ortholog_connection_data_high_conf <- 
  gene_orthologs_all %>% 
  filter(ortholog_confidence == 1) %>%
  left_join(gene_ortholog_high_conf_network$communities, 
            by = c("enssscg_id" = "node")) %>% 
  group_by(community) %>%
  mutate(edge_size = n()) %>%
  ungroup() %>%
  arrange(-community_size) %>%
  left_join(cor_s_gene_all_orth_all_region %>% 
              select(1, 2, 4), 
            by = c("enssscg_id", "ensg_id")) %>%
  left_join(ortholog_comparison_anova_high_conf_ortholog_assignment %>% 
              select(enssscg_id, ensg_id, sqrt_omega, 
                     ortholog_class_human, ortholog_class_pig, 
                     mutual_primary)) %>%
  rename(expression_identity = sqrt_omega) %>% 
  mutate(ortholog_confidence = factor(ortholog_confidence))


```

##Ortholog assignment omega

All genes are part of ortholog network communities. As one gene is retrieved, data on the whole network is retrieved. The primary ortholog is defined as the one with highest amount of variance explained by tissue rather than species differences or residuals.

```{r orth_assignment, echo=FALSE, warning=FALSE, paged.print=FALSE}

example_gene <- "ENSG00000279395"
example_genes <- 
  ortholog_connection_data_high_conf %>% 
  filter(community_size > 2) %$%
  sample(c(ensg_id,
           enssscg_id), 50)

pdf(savepath("ortholog network examples.pdf"), width = 8, height = 5, useDingbats = F)
lapply(example_genes, 
       function(x) {print(x)
         ortholog_connection_plot(ortholog_connection_data, x)})
dev.off()
  
pdf(savepath("ortholog network high conf examples.pdf"), width = 8, height = 5, useDingbats = F)
lapply(example_genes, 
       function(x) {print(x)
         ortholog_connection_plot(ortholog_connection_data_high_conf, x)})
dev.off()
  

```


##Orthology visualization

```{r orth_vis, eval = F, include = F, message=F, warning=F}


orth_gs <- c("ENSSSCG00000021258", "ENSSSCG00000037747", "ENSSSCG00000012238", "ENSSSCG00000032981", "ENSSSCG00000022577", "ENSSSCG00000013919", 
             "ENSSSCG00000040140", "ENSSSCG00000011590", "ENSSSCG00000028613")
i <- 1

ortholog_heatmap(ortholog_connection_data, 
                        gene = orth_gs[i], 
                        species = "pig")
i <- i + 1
g1 <- 
  ortholog_network_plot(ortholog_connection_data, 
                        gene = orth_gs[i], 
                        species = "pig")

g2 <- 
  ortholog_barplot(ortholog_connection_data, 
                   gene = orth_gs[i], 
                   species = "pig")

g3 <- 
  ortholog_scatterplot(ortholog_connection_data, 
                   gene = orth_gs[i], 
                   species = "pig")

g4 <- 
  ortholog_metrics(ortholog_connection_data, 
                   gene = orth_gs[i], 
                   species = "pig")



grid.arrange(g2[[orth_gs[i]]], g1, 
             g2[[2]], g3[[1]], g4[[1]],
             g2[[3]], g3[[2]], g4[[2]],
             layout_matrix = rbind(c(1, 1, 2),
                                   c(1, 1, 2),
                                   c(1, 1, 2),
                                   c(3, 3, 4),
                                   c(3, 3, 4),
                                   c(3, 3, 5),
                                   c(6, 6, 7),
                                   c(6, 6, 7),
                                   c(6, 6, 8)))
             # widths = c(3, 1))

pdf(savepath("Gene examples Pig Atlas.pdf"), width = 8, height = 4, useDingbats = F)             
for(i in 1:length(orth_gs)) {
   
  ortholog_network_plot(ortholog_connection_data, 
                        gene = orth_gs[i], 
                        species = "pig") %>% print


  ortholog_barplot(ortholog_connection_data, 
                   gene = orth_gs[i], 
                   species = "pig") %>% lapply(print)


  ortholog_scatterplot(ortholog_connection_data, 
                   gene = orth_gs[i], 
                   species = "pig") %>% lapply(print)
  
  ortholog_metrics(ortholog_connection_data, 
                   gene = orth_gs[i], 
                   species = "pig") %>% lapply(print)
}
dev.off()



###




g <- tibble(gene = rep(c("ENSSSCG000001284", "ENSG000005494",  "ENSG000005495", "ENSG000005502", "ENSG000005503"), each = 10),
       name = rep(c("ENGEN1", "ENGEN1",  "ENGEN2", "NOGEN1", "NOGEN2"), each = 10),
       sample_ID = rep(c(1:10), 5),
       species = factor(rep(c("PIG", "HUMAN", "HUMAN", "HUMAN", "HUMAN"), each = 10), levels = c("PIG", "HUMAN")),
       NX = c(rnorm(10, mean = 5, sd = 1), rnorm(10, mean = 7, sd = 1), rnorm(10, mean = 4, sd = 1), 
              rnorm(10, mean = 13, sd = 1), rnorm(10, mean = 4, sd = 1))
       ) %>% 
  ggplot(aes(name, NX)) + 
  geom_boxplot(outlier.colour = NA, fill = "#4682B4", color = "#454545", 
               width = 1) + 
  geom_beeswarm() +
  facet_wrap(~species, scales = "free_x") +
  scale_x_discrete(expand = expand_scale(add = 0))+
  stripped_theme_HPA + 
  theme(axis.text.x = element_text(angle = 60, hjust = 1), 
        panel.spacing = unit(0, "cm"), 
        axis.title.x = element_blank())
g

gt = ggplot_gtable(ggplot_build(g))

panel_i <- gt$layout$l[grep('panel-1-1', gt$layout$name)]
gt$widths[panel_i] = gt$widths[panel_i]*0.25
pdf(savepath("Detailed ortholog expression display.pdf"), width = 5, height = 4, useDingbats = F)
grid.draw(gt)

dev.off()

####

g <- 
  tibble(name = c("ENGEN1", "ENGEN2", "AGENE1", "AGENE2"), 
         `Ortholog score` = c(0.95, 0.70, 0.43, 0.20),
         `Ortholog confidence` = c(1.00, 1.00, 0.50, 0.50),
         `Sequene homology` = c(0.95, 0.90, 0.89, 0.69),
         `Expression correlation` = c(0.73, 0.57, 0.65, 0.04),
         `Specifity class score` = c(1.00, 0.75, 0.75, 0.00)) %>% 
  mutate(name = factor(name, levels = name)) %>%
  gather(Term, Score, -1) %>%
  mutate(is_orth_score = factor(case_when(Term == "Ortholog score" ~ "TOTAL", 
                                          T ~ "SCORE" ),
                                levels = c("TOTAL", "SCORE"))) %>% 
  
  ggplot(aes(name, Term, fill = Score, label = Score)) + 
  geom_tile(show.legend = F) + 
  geom_text() + 
  stripped_theme_HPA +
  scale_x_discrete(position = "top") +
  scale_y_discrete(expand = c(0,0)) +
  theme(axis.title = element_blank(), 
        axis.text.x = element_text(angle = 60, hjust = 0),
        panel.background = element_rect(fill = "#D9D9D9"),
        panel.spacing = unit(0.1, "lines")) + 
  scale_fill_gradient2(low = "darkred", mid = "#D9D9D9", high = "darkgreen", 
                       midpoint = 0.5) + 
  
  facet_wrap(is_orth_score ~ ., scales = "free_y", strip.position="right", ncol = 1)
g  
gt = ggplot_gtable(ggplot_build(g))
gtable_show_layout(gt)

panel_i <- gt$layout$t[grep('panel-1-1', gt$layout$name)]
gt$heights[panel_i] = gt$heights[panel_i]*0.25
pdf(savepath("Ortholog score display.pdf"), width = 5, height = 3, useDingbats = F)
grid.draw(gt)

dev.off()
```

##Ortholog examples

```{r orth_examples, echo=FALSE, message=FALSE, warning=FALSE, paged.print=TRUE}

Evelinas_gene_examples <- 
  read_delim("../../doc/Notes from Evelina/genes with ab staining in pig tissue.txt", delim = "\t") %>% 
  select(ensg_id = 1, 
         gene_name = 2, 
         rel_score = 3) %>% 
  left_join(gene_orthologs, 
            by = "ensg_id") %>% 
  mutate(mutual_id = paste(ensg_id, enssscg_id, sep = " - "))

plots <- 
  Evelinas_gene_examples %>% 
  select(ensg_id, gene_name, enssscg_id, mutual_id) %>%
  gather(type, gene_id, ensg_id, enssscg_id) %>% 
  
  {lapply(unique(.$mutual_id),
          function(g_id) {
            
            genes <- 
              filter(., mutual_id == g_id)
            
            species_atlas %>% 
              filter(species_gene_id %in% genes$gene_id) %>%
              quick_atlas_barplot()
          })}

pdf(savepath("Evelinas gene example consensus barplots.pdf"), 
    width = 6, 
    height = 4, 
    useDingbats = F)
plots
dev.off()

plots <- 
  Evelinas_gene_examples %>% 
  select(ensg_id, gene_name, enssscg_id, mutual_id) %>%
  gather(type, gene_id, ensg_id, enssscg_id) %>% 
  
  {lapply(unique(.$mutual_id),
          function(g_id) {
            
            genes <- 
              filter(., mutual_id == g_id)
            
            species_region_atlas %>% 
              filter(species_gene_id %in% genes$gene_id) %>% 
              quick_atlas_barplot(tis_col = "species_region_name")
          })}

pdf(savepath("Evelinas gene example region barplots.pdf"), 
    width = 10, 
    height = 5, 
    useDingbats = F)
plots
dev.off()
```



# Tests

```{r normalization_parameter_test, include = F, eval = F}

norm_settings_grid <- 
  expand.grid(logratioTrim = seq(0.2, 0.3, length.out = 5), # default = 0.3 
              sumTrim = seq(0.04, 0.06, length.out = 5) # default = 0.05
              ) %>%
  mutate(i = row_number())

norm_parameter_test_result <- 
  pblapply(norm_settings_grid$i, 
         FUN = function(i) {
           pig_atlas_sample_wide_ptpm %>%
           
           tmm_norm_median_ref(logratioTrim = norm_settings_grid$logratioTrim[i],
                               sumTrim = norm_settings_grid$sumTrim[i])
         })

# 
# A <- 
#   norm_parameter_test_result %>% 
#   set_names(norm_settings_grid$i) %>%
#   plyr::ldply(.id = "i") %>% 
#   left_join(norm_settings_grid,
#             by = "i")

norm_parameter_test_result

```

```{r plsda_test, include = F, eval = F}

require(caret)
require(ropls)

X <- 
  sample_nx_pca$scores[,1:sample_nx_pca$pc_95]

Y <- 
  rownames(X) %>%
  enframe("name", "sample_ID") %>% 
  left_join(sample_mapping, 
            by = "sample_ID") %$%
  consensus_tissue_name

pig_sample_plsda <- 
  plsda(X, factor(Y))

plot(pig_sample_plsda)


pig_sample_opls <- 
  opls(X, factor(Y))

plot(pig_sample_opls)

```

