---
title: "main"
author: "Max J. Karlsson"
date: "2020 M04 6"
output: pdf_document
editor_options: 
  chunk_output_type: console
---

#Setup

```{r start_up, warning=F, message=F, include = F}
require(tidyverse)
require(magrittr)
require(pcaMethods)
require(pheatmap)
require(dendextend)
require(ggdendro)
require(umap)
require(ggforce)
require(Hmisc)
require(NOISeq)
require(ggrepel)
require(gridExtra)
require(ape)
require(pbapply)
require(mgsub)
require(ggalluvial)
require(igraph)
require(ggraph)
require(viridis)
require(ggbeeswarm)
require(gtable)
require(grid)
require(ggridges)
require(broom)
require(patchwork)
require(propr)
require(ellipse)
require(tidygraph)
require(concaveman)
library(enrichR)
library(tm)
library(ggwordcloud)
library(sf)
rename <- dplyr::rename

setwd("~/Scilifelab/Projects/Pig Atlas")
source("scripts/theme.R")
source("scripts/functions_normalization.R")
source("scripts/functions_classification.R")
source("scripts/functions_graphics.R")
source("scripts/functions_utility.R")
```

## Designate orthologs

```{r orthologs}

# Load gene mapping tables
gene_mapping <- 
  read_delim("data/meta/ensembl_pig_gene.tab", delim = "\t") %>% 
  filter(biotype == "protein_coding")
gene_chromosome <- read_delim("data/meta/ensembl92_gene_chromosome.txt", delim = "\t")

human_gene_mapping <- read_delim("data/meta/human/geninfo_92.tsv", delim = "\t")

# gene_orthologs <- read_delim("data/meta/ensembl_pig_ortholog.tab", delim = "\t")
gene_orthologs_all <- 
  read_delim("data/meta/Ensembl_orthologs_all.txt", delim = "\t") %>% 
  select(enssscg_id = 1,
         ensg_id = 3, 
         pig_gene_name = 2,
         human_gene_name = 4,
         ortholog_confidence = 10, 
         ortholog_type = 5,
         gene_order_conservation_score = 8, 
         whole_genome_alignment_coverage = 9, 
         sequence_identity = 6) %>%
  filter(!is.na(ortholog_confidence)) %>% 
  filter(ensg_id %in% human_gene_mapping$ensg_id) %>%
  filter(enssscg_id %in% gene_mapping$enssscg_id)

gene_orthologs_data <- 
  gene_orthologs_all %>% 
  group_by(enssscg_id) %>% 
  mutate(pig_n_high_conf = length(which(ortholog_confidence == 1)), 
         pig_n_orthologs = n()) %>%
  group_by(ensg_id) %>% 
  mutate(human_n_high_conf = length(which(ortholog_confidence == 1)), 
         human_n_orthologs = n()) %>%
  ungroup()
  
gene_orthologs_data %>% 
  # filter(ortholog_type == "ortholog_one2many" & 
  #          n_high_conf == 1) %>%
  filter(pig_n_orthologs == 3 & human_n_orthologs == 7) %>% 
  mutate(pig_y = unclass(factor(enssscg_id)),
         human_y = unclass(factor(ensg_id))) %>%
  ggplot() +
  geom_text(aes(x = 1, y = pig_y, label = enssscg_id), 
            hjust = 1) +
  geom_text(aes(x = 2, y = human_y, label = ensg_id),
            hjust = 0) + 
  geom_segment(aes(x = 1, xend = 2, 
                   y = pig_y,  yend = human_y,
                   color = as.factor(ortholog_confidence),
                   size = pig_n_high_conf == 1 & human_n_high_conf == 1 & ortholog_confidence == 1),
               show.legend = F) + 
  theme_void() +
  scale_size_discrete(range = c(1, 4)) + 
  scale_x_continuous(expand = expansion(1.5))
ggsave(savepath("gene ortholog many2many example.pdf"), width = 5, height = 5)

# Orthologs are filtered by these rules to have a unique one to one relationship between human and pig:
# The gene must:
# 1) be high confidence and have only one high confidence relationship for both human and pig, or
# 2) be a one2one ortholog
gene_orthologs <- 
  gene_orthologs_data %>% 
  filter((pig_n_high_conf == 1 & human_n_high_conf == 1 & ortholog_confidence == 1 &
             ortholog_type == "ortholog_one2many") | 
           ortholog_type == "ortholog_one2one") 
  
if(!file.exists("data/processed/gene_orthologs.csv")) {
  write_csv(gene_orthologs,
            "data/processed/gene_orthologs.csv")
}
  
```


## Read data
```{r read_data, include = F}


# Load sample mapping data
tissue_mapping <- read_delim("data/meta/tissue_mapping.tsv", delim = "\t")
## Tissue mapping
human_tissue_mapping <- read_delim("data/meta/human_tissue_mapping.tsv", delim = "\t")

sample_mapping <- 
  expand.grid(tissue = tissue_mapping$tissue, 
              individual = letters[1:4]) %>% 
  mutate(sample_ID = paste(tissue, individual, sep = "_")) %>%
  left_join(tissue_mapping, 
            by = "tissue") %>%
  as_tibble()

tissue_colors <- read_delim("data/meta/pig_tissue_colors.tsv", delim = "\t")

tissue_colors_palette <- 
  c(setNames(tissue_colors$tissue_color, tissue_colors$tissue_name),
    setNames(tissue_colors$tissue_color, str_to_sentence(tissue_colors$tissue_name)))

sample_meta <- 
  tibble(individual = letters[1:4],
         sex = c("female", "female", "male", "male"))

tissue_colors_palette_full <- 
  tissue_colors %$%
  tibble(color = c(tissue_color, region_tissue_color, consensus_tissue_color, organ_color),
         tissue = c(tissue_name, region_tissue_name, consensus_tissue_name, organ_name)) %>%
  unique() %$%
  c(setNames(color, tissue),
    setNames(color, str_to_sentence(tissue)))


tissue_colors_palette_full_humanpig <-
  tissue_colors %>%
  left_join(tissue_mapping) %$%
  tibble(color = c(tissue_color, region_tissue_color, consensus_tissue_color, organ_color, 
                   tissue_color, region_tissue_color, consensus_tissue_color),
         tissue = c(tissue_name, region_tissue_name, consensus_tissue_name, organ_name, 
                    human_tissue_name, human_region_name, human_consensus_tissue_name)) %>%
  unique() %$%
  setNames(color, tissue) 

tissue_colors_palette_comparison <-
  tissue_mapping %>% 
  select(comparison_tissue, organ_name) %>% 
  filter(!is.na(comparison_tissue)) %>%
  distinct() %>% 
  left_join(tissue_colors %>% 
              select(organ_name, organ_color) %>%
              distinct()) %$%
  tibble(color = c(organ_color),
         tissue = c(comparison_tissue)) %$%
  c(setNames(color, tissue),
    setNames(color, str_to_sentence(tissue)))

tissue_colors_palette_full_humanpig_extension <- 
  set_names(c(rep(tissue_colors_palette_full_humanpig, 2), 
              "black", "black"),
            c(paste(names(tissue_colors_palette_full_humanpig), "human"), 
              paste(names(tissue_colors_palette_full_humanpig), "pig"),
              "not enriched in human human", "not enriched in pig pig"))

tissue_colors_palette_organs <- 
  tissue_colors %$%
  tibble(color = c(organ_color),
         tissue = c(organ_name)) %>%
  unique() %$%
  setNames(color, tissue) 

regions_organ_order <- 
  bind_rows(tissue_mapping %>% 
              select(region_tissue_name, organ_name),
            human_tissue_mapping %>% 
              select(region_tissue_name, organ_name)) %>% 
  unique() %>% 
  arrange(organ_name, region_tissue_name)

region_levels <- 
  unique(regions_organ_order$region_tissue_name)

region_palette <- 
  regions_organ_order %>%
  left_join(tissue_colors %>% 
              select(organ_name, organ_color)) %$%
  set_names(organ_color, region_tissue_name)
  

comparison_organ_order <- 
  bind_rows(tissue_mapping %>% 
              select(comparison_tissue, organ_name),
            human_tissue_mapping %>% 
              select(comparison_tissue, organ_name)) %>% 
  unique() %>% 
  arrange(organ_name, comparison_tissue)

comparison_levels <- 
  unique(comparison_organ_order$comparison_tissue)


consensus_organ_order <- 
  bind_rows(tissue_mapping %>% 
              select(consensus_tissue_name, organ_name),
            human_tissue_mapping %>% 
              select(consensus_tissue_name, organ_name)) %>% 
  unique() %>% 
  arrange(organ_name, consensus_tissue_name)

consensus_levels <- 
  unique(consensus_organ_order$consensus_tissue_name)
  
# Load data


## Unfiltered
pig_atlas_unfiltered_sample_temp <- 
  read_delim("data/processed/pig_unfiltered_sample_data.tab", delim = "\t")

pig_atlas_consensus <-
  read_delim("data/HPA/consensus_pig_tissues_92.tsv", delim = "\t") %>%
  rename(enssscg_id = ensg_id) %>%
  mutate(tissue = gsub(" 1$", "", tissue))



# pig_atlas_tissue_per <- 
#   read_delim("data/HPA/rna_pigtissue_pig_92.tsv", delim = "\t")
##classification

pig_gene_classification <- 
  read_delim("data/HPA/pig_consensus_category_piggenome_92.tsv", delim = "\t") %>%
  rename(enssscg_id = ensg_id) %>%
  mutate(enhanced_tissues = gsub(", ", "_", enhanced_tissues),
         enhanced_tissues = gsub(",", ";", enhanced_tissues),
         enhanced_tissues = gsub("_", ", ", enhanced_tissues),
         enhanced_tissues = trimws(enhanced_tissues))

pig_gene_comparison_classification <- 
  read_delim("data/HPA/pig_compare_category_piggenome_92.tsv", delim = "\t") %>%
  rename(enssscg_id = ensg_id) %>%
  mutate(enhanced_tissues = gsub(", ", "_", enhanced_tissues),
         enhanced_tissues = gsub(",", ";", enhanced_tissues),
         enhanced_tissues = gsub("_", ", ", enhanced_tissues),
         enhanced_tissues = trimws(enhanced_tissues))

human_gene_class <- 
  read_delim("data/human data/lims/consensus_all_category_92.tsv", delim = "\t") %>% 
  mutate(enhanced_tissues = gsub(", ", "_", enhanced_tissues),
         enhanced_tissues = gsub(",", ";", enhanced_tissues),
         enhanced_tissues = gsub("_", ", ", enhanced_tissues),
         enhanced_tissues = trimws(enhanced_tissues))

human_gene_comparison_classification <- 
  read_delim("data/HPA/tissue_compare_category_92.tsv", delim = "\t") %>%
  mutate(enhanced_tissues = gsub(", ", "_", enhanced_tissues),
         enhanced_tissues = gsub(",", ";", enhanced_tissues),
         enhanced_tissues = gsub("_", ", ", enhanced_tissues),
         enhanced_tissues = trimws(enhanced_tissues))

# Load human data

## Tissue data
human_atlas_tissue <-
  bind_rows(read_delim("data/human data/lims/consensus_hpa_92.tsv", delim = "\t") %>%
              mutate(source = "hpa"),
            read_delim("data/human data/lims/consensus_gtex_92.tsv", delim = "\t") %>%
              mutate(source = "gtex"),
            read_delim("data/human data/lims/consensus_fantom_92.tsv", delim = "\t") %>%
              mutate(source = "fantom")) %>% 
  rename(tissue_name = celltype) %>%
  group_by(tissue_name, ensg_id) %>% 
  summarise(tpm = mean(tpm, na.rm = T),
            ptpm = mean(ptpm, na.rm = T),
            tmm = mean(tmm, na.rm = T),
            nx = mean(nx, na.rm = T)) %>%
  ungroup()

## Blood data
human_gene_class_blood <-
  read_delim("data/meta/human/bloodcells_hpa_category_92.tsv", delim = "\t") 

human_gene_class_bloodlineage <-
  read_delim("data/meta/human/bloodcells_hpa_regional_category_92.tsv", delim = "\t") 
              
## All rna data 

HPA_rna <-
  read_delim("data/HPA/rna_tissue_consensus_HPA19.tsv", delim = "\t") %>%
  rename(ensg_id = 1, 
         gene_name = 2, 
         tissue = 3, 
         nx = 4)

## Consensus data
human_atlas_consensus <- 
  read_delim("data/human data/lims/consensus_max_of_all_groups_92.tsv", delim = "\t") %>% 
  rename(consensus_tissue_name = tissue)

  

gene_list_alun <- 
  read_delim("data/meta/Gene list from Alun.txt", delim = "\t")

###

SLA_genes <- 
  read_csv("data/meta/SLA genes.csv")


### disease genes

disease_genes <- 
  read_delim("data/HPA/proteinatlas_disease_involvement_92.tsv", delim = "\t") %>% 
  select(gene_name = Gene, 
         ensg_id = Ensembl,
         disease = `Disease involvement`) %>% 
  separate_rows(disease, sep = ", ")


xeno_ko_genes <- 
  c("GGTA1" = "ENSSSCG00000005518", 
    "CMAH" = "ENSSSCG00000001099", 
    "B4GALNT2" = "ENSSSCG00000040942",
    "VWF" = "ENSSSCG00000000712", 
    "ULBP1" = "ENSSSCG00000026042") %>% 
  enframe("gene_name", "enssscg_id") %>% 
  mutate(name_id = paste(gene_name, enssscg_id))


disease_genes_full <- 
  disease_genes %>% 
  filter(!is.na(disease)) %>%
  inner_join(gene_orthologs) %>%
  select(enssscg_id, ensg_id, 
         gene_name, disease) %>% 
  bind_rows(xeno_ko_genes %>% 
              select(enssscg_id, 
                     gene_name) %>% 
              mutate(disease = "Xenograft rejection")) %>%
  filter(!disease %in% c("Disease mutation",
                         "Cancer-related genes",
                         "FDA approved drug targets",
                         "Proto-oncogene"))

disease_genes_full %>% 
  pull(enssscg_id) %>% 
  n_distinct()

disease_genes_selected <-
  readxl::read_xlsx("data/meta/alun disease gene list.xlsx")  %>%
  left_join(gene_orthologs,
            by = c("gene" = "human_gene_name")) %>% 
  select(enssscg_id, ensg_id, 
         gene_name = gene, 
         disease = Disease,
         organ = Organ)

disease_genes_selected2 <-
  readxl::read_xlsx("data/meta/alun disease gene list 2.xlsx", sheet = 2)  %>%
  select(1:4, gene = 5) %>%
  separate_rows(gene, sep = ", ") %>%
  left_join(gene_orthologs,
            by = c("gene" = "human_gene_name")) %>% 
  select(enssscg_id, ensg_id, 
         gene_name = gene, 
         disease = Disease,
         organ = Organ)

# disease_genes %>%
#   filter(!is.na(disease)) %>%
#   write_csv(savepath("HPA_E92_disease_genes.csv"))
```


# Compare Pers data

```{r}

pig_gene_classification
pig_gene_classification_2 <- 
  pig_atlas_consensus %>%
  hpa_gene_classification(expression_col = "tmm", 
                          tissue_col = "tissue", 
                          gene_col = "ensg_id", 
                          enr_fold = 4, 
                          max_group_n = 5, 
                          det_lim = 1)


joined_classes <- 
  pig_gene_classification_2 %>% 
  select(1:4, enriched_tissues) %>%
  left_join(pig_gene_classification,
            by = c("gene" = "ensg_id")) 
  
joined_classes %>%
  group_by(spec_category, specificity_category) %>% 
  summarise(n = n())

joined_classes %>% 
  mutate(enhanced_tissues = gsub(",", ";", enhanced_tissues),
         enriched_tissues = gsub(" 1", "", enriched_tissues),
         same_enr = enhanced_tissues == enriched_tissues) %>%
  filter(!same_enr)
```

#Data generation

##Data tables for saving

```{r tables_for_saving, echo=FALSE, message=FALSE, warning=FALSE}

pig_gene_classification %>% 
  left_join(gene_orthologs) %>% 
  left_join(human_gene_class %>% 
              select(1:4),
            by = "ensg_id") %>% 
  left_join(human_gene_mapping) %>% 
  write_csv(savepath("Pig class ortholog table.csv"))

# pig_gene_classification %>% 
#   left_join(gene_orthologs) %>% 
#   left_join(human_gene_class %>% 
#               select(1:4),
#             by = "ensg_id") %>% 
#   left_join(human_gene_mapping) %>% 
#   select(enssscg_id, gene_name, 2, 3, 4, 5, 6) %>% 
#   distinct() %>%
#   group_by(enssscg_id, specificity_category, distribution_category, ts_score, enhanced_score, enhanced_tissues) %>%
#   summarise(gene_name = paste(sort(gene_name), collapse = ";")) %>%
#   ungroup() %>%
#   select(enssscg_id, gene_name, everything()) %>%
#   write_csv(savepath("Pig class table.csv"))

```

##Normalization
```{r}

if(file.exists("data/processed/normalized_res.Rdata")) {
  load("data/processed/normalized_res.Rdata")
  
} else {
  source("scripts/normalization.R", local = T)
  load("data/processed/normalized_res.Rdata")
}


```



##PCA

```{r}

if(file.exists("data/processed/pig_pca.Rdata")) {
  load("data/processed/pig_pca.Rdata")
} else {
  
  sample_pca <-
    pig_atlas_sample_norm %>% 
    group_by(enssscg_id) %>% 
    mutate(sd = sd(tmm)) %>%
    ungroup() %>% 
    filter(sd > 0) %>%
    select(enssscg_id, sample_ID, tmm) %>% 
    
    spread(enssscg_id, tmm) %>%
    column_to_rownames("sample_ID") %>%
    {. + 1} %>%
    compositions::clr() %>%
    as.data.frame() %>%
    pca_calc(npcs = 150)
  
  
  
  
  tissue_pca <-
    pig_atlas_tissue %>% 
    group_by(enssscg_id) %>% 
    mutate(sd = sd(tmm)) %>%
    ungroup() %>% 
    filter(sd > 0) %>%
    select(enssscg_id, tissue_name, tmm) %>% 
    
    spread(enssscg_id, tmm) %>%
    column_to_rownames("tissue_name") %>%
    {. + 1} %>%
    compositions::clr() %>%
    as.data.frame() %>%
    pca_calc(npcs = 50)
  
  save(list = c("sample_pca",
                "tissue_pca"),
       file = "data/processed/pig_pca.Rdata")
}

```

## Basic gene stats

```{r}
basic_gene_stats <- 
  pig_atlas_sample %>%
  select(1, 2, ptpm) %>% 
  group_by(enssscg_id) %>% 
  summarise(sd = sd(ptpm),
            mean = mean(ptpm),
            median = median(ptpm),
            max = max(ptpm),
            fract_0 = length(which(ptpm < 0.1)) / length(ptpm)) %>%
  ungroup() %>% 
  mutate(CV = 100 * sd / mean)

basic_gene_stats %>% 
  filter(max >= 1) 

basic_gene_stats %>% 
  ggplot(aes(fract_0)) +
  geom_histogram() +
  
  basic_gene_stats %>% 
  ggplot(aes(sd)) +
  geom_density() +
  scale_x_log10() +
  
  basic_gene_stats %>% 
  ggplot(aes(CV)) +
  geom_histogram()


```




##Classification

###Generate comparison classification

```{r comparison_class, echo=FALSE, message=FALSE, warning=FALSE}


# pig_gene_class_comparison <- 
#   hpa_gene_classification(data = pig_atlas_comparison,
#                           expression_col = "nx",
#                           tissue_col = "comparison_tissue",
#                           gene_col = "enssscg_id",
#                           enr_fold = 4,
#                           max_group_n = 5,
#                           det_lim = 1)
# write_delim(pig_gene_class_comparison, savepath("Pig gene comparison classification.txt"), delim = "\t")

# 
# A <- 
#   left_join(pig_gene_comparison_classification %>%
#             select(1,
#                    spec_category = specificity_category,
#                    dist_category = distribution_category, 
#                    enriched_tissues = enhanced_tissues) %>% 
#             mutate(spec_category = tolower(spec_category)),
#           pig_gene_class_comparison %>%
#             select(1,2, 3, enriched_tissues),
#           by = c("enssscg_id" = "gene"),
#           suffix = c("1", "2")) 
# 
# A %>% 
#   filter(dist_category1 == "detected in all" &dist_category2 != "detected in all")
# 
# pig_gene_comparison_classification %>%
#   filter(enssscg_id == "ENSSSCG00000001790") 
# 
# pig_atlas_comparison %>%
#   filter(enssscg_id == "ENSSSCG00000000006") %>%
#   ggplot(aes(nx, comparison_tissue)) +
#   geom_col() +
#   geom_vline(xintercept = 1)
# 
# pig_atlas_sample %>%
#   filter(enssscg_id == "ENSSSCG00000001790") %>%
#   ggplot(aes(ptpm, sample_ID)) +
#   geom_col()
# 
# human_gene_class_comparison <- 
#   hpa_gene_classification(data = human_atlas_comparison,
#                           expression_col = "nx",
#                           tissue_col = "comparison_tissue",
#                           gene_col = "ensg_id",
#                           enr_fold = 4,
#                           max_group_n = 5,
#                           det_lim = 1)
# write_delim(human_gene_class_comparison, savepath("Human gene comparison classification.txt"), delim = "\t")


```


##Intraconsensus classification

### Generate data
```{r intra_consensus, echo=FALSE, warning=FALSE}


consensus_tissues_with_multiple <- 
  tissue_mapping %>% 
  filter(tissue_name %in% pig_atlas_tissue$tissue_name) %>%
  group_by(consensus_tissue_name) %>% 
  summarise(n_tissues = n_distinct(tissue_name)) %>%
  ungroup() %>%
  filter(n_tissues > 1)


anova_data <- 
  pig_atlas_sample_norm %>% 
  # filter(enssscg_id %in% sample(unique(enssscg_id), 100)) %>%
  left_join(tissue_mapping,
            by = c("tissue_ID" = "tissue")) %>% 
  mutate(sex = case_when(individual %in% c("a", "b") ~ "female",
                         individual %in% c("c", "d") ~ "male")) %>%
  group_by(consensus_tissue_name) %>% 
  mutate(n_tissues = n_distinct(tissue_name)) %>%
  ungroup() %>%
  filter(n_tissues > 1) %>%
  select(enssscg_id, consensus_tissue_name, tissue_name, individual, sex, tmm) %>%
  group_by(enssscg_id) %>% 
    mutate(sd_ = sd(tmm),
           max_ = max(tmm),
           range_ = max(tmm) / min(tmm[which(tmm > 0)])) %>%
    filter(sd_ > 0 & max_ > 1 & range_ >= 4) %>%
    select(-sd_, -max_, -range_) %>%
  ungroup() #%>%
  # filter(enssscg_id %in% sample(unique(enssscg_id), 100)) 


save(anova_data,
     file = "data/processed/pig_intraconsensus_anova_data.Rdata")

if(file.exists("data/processed/pig_intraconsensus_anova2.Rdata")) {
  load("data/processed/pig_intraconsensus_anova2.Rdata")
} else {
  
  
  intraconsens_anova_diffs <- 
    anova_data %>% 
    rename(tissue1 = tissue_name) %>%
    left_join(., 
              select(.,enssscg_id, consensus_tissue_name, tissue2 = tissue1, tmm, individual), 
              by = c("consensus_tissue_name",
                     "individual",
                     "enssscg_id"),
              suffix = c("1", "2")) %>%
    filter(tissue1 != tissue2 & tissue1 < tissue2) %>%
    group_by(consensus_tissue_name, tissue1, tissue2, enssscg_id) %>%
    summarise(mean_tmm1 = mean(tmm1, na.rm = T),
              mean_tmm2 = mean(tmm2, na.rm = T),
              ratio = max(mean_tmm1, 1) / max(mean_tmm2, 1)) %>% 
    ungroup() 
  
  diff_data <- 
    anova_data %>% 
    mutate(log_tmm = log2(tmm + 1)) %>% 
    mutate(id = paste(tissue_name, individual)) %>% 
    select(1, consensus_tissue_name, id, log_tmm) %>% 
    spread(enssscg_id, log_tmm) %>% 
    arrange(consensus_tissue_name, id) 
  
  diff_data_meta <- 
    anova_data %>%
    mutate(id = paste(tissue_name, individual)) %>% 
    select(consensus_tissue_name, id, tissue_name, sex) %>%
    distinct() %>%
    arrange(consensus_tissue_name, id) %>%
    mutate_all(.funs = as.factor)
  
  ##
  # require(multidplyr)
  
  
  # cluster <- new_cluster(detectCores() - 1)
  
  # cluster_copy(cluster, names = c("%>%", "%$%", "diff_data_meta", "diff_data", "set_colnames", "set_names") )
  
  intraconsens_anova_temp <- 
    anova_data %>% 
    select(consensus_tissue_name, tissue_name) %>%
    distinct() %>% 
    arrange(consensus_tissue_name) %>%
    group_by(consensus_tissue_name) %>% 
    # partition(cluster) %>%
    do({
      g_data <- .
      
      tis_ <- 
        unique(g_data$consensus_tissue_name)
      
      contrasts_ <-
        gsub("\\(|\\)", "", gsub(" ", "_", g_data$tissue_name)) %>%
        tidyr::crossing(V1 = ., V2 = .) %>% 
        dplyr::filter(V1 != V2 & V1 < V2) %>% 
        dplyr::mutate(name = paste(V1, V2, sep = "__"),
                      formula = paste(V1, V2, sep = " - "))
      
      meta_ <- 
        diff_data_meta %>% 
        dplyr::filter(consensus_tissue_name == tis_) %>%
        dplyr::mutate(tissue_name = gsub("\\(|\\)", "", gsub(" ", "_", tissue_name))) %>%
        dplyr::mutate(tissue_name = factor(tissue_name, unique(tissue_name)))
      
      data_ <-
        diff_data %>% 
        dplyr::filter(consensus_tissue_name == tis_) %>%
        dplyr::select_if(.predicate = function(x) !is.na(x[1])) %>%
        dplyr::select(-1) %>% 
        tibble::column_to_rownames("id") %>% 
        as.matrix()
      
      # Create a design matrix that includes separate coefficients for each level
      # within the grouping factor so that the desired differences can be extracted as contrasts.
      
      design <-
        model.matrix(~ 0 + tissue_name,# + tissue_name:sex + sex, 
                     meta_,
                     contrasts.arg = list(tissue_name = contrasts(meta_$tissue_name, contrasts=F))) %>%
        set_colnames(gsub("tissue_name|sex", "", gsub(" |:", "_", colnames(.))))
      
      ### Fit model and extract contrasts of interest
      
      # corfit <-
      #   duplicateCorrelation(t(data_),
      #                        design)
      
      fit <- limma::lmFit(t(data_),
                          design)
      
      contrasts <- limma::makeContrasts(contrasts = contrasts_$formula,
                                        
                                        levels=design) %>%
        set_colnames(contrasts_$name)
      
      fit2 <- limma::contrasts.fit(fit, contrasts)
      fit3 <- limma::eBayes(fit2, trend = FALSE)
      lapply(colnames(fit3$contrasts),
             FUN = function(con) limma::topTable(fit3, coef=con, number=Inf, adjust.method="BH", p.value=1) %>%
               tibble::as_tibble(rownames = "variable")) %>%
        set_names(colnames(fit3$contrasts)) %>%
        dplyr::bind_rows(.id = "term") %>%
        dplyr::mutate(term = gsub("_", " ", term)) %>%
        tidyr::separate(term, into = c("tissue1", "tissue2"), sep = "  ") 
      
    }) %>%
    ungroup() #%>% 
    # collect()
  
  
  name_dictionary <- 
    anova_data %>% 
    select(tissue_name) %>% 
    distinct() %>% 
    mutate(new_name = gsub("_", " ", gsub("\\(|\\)", "", gsub(" ", "_", tissue_name)))) 
  
  
  intraconsens_anova_temp2 <-
    intraconsens_anova_temp %>%
    mutate(tissue1 = name_dictionary$tissue_name[match(tissue1, name_dictionary$new_name)],
           tissue2 = name_dictionary$tissue_name[match(tissue2, name_dictionary$new_name)]) 
  
  rm(intraconsens_anova_temp)
    
  
  intraconsens_anova_res <- 
    intraconsens_anova_temp2 %>%
    arrange(consensus_tissue_name, tissue1, tissue2, variable) %>% 
    {
      temp <- .
      
      temp_diffs <- 
        intraconsens_anova_diffs %>%
        arrange(consensus_tissue_name, tissue1, tissue2, enssscg_id)
      
      temp %>%
        mutate(mean_tmm1 = temp_diffs$mean_tmm1,
               mean_tmm2 = temp_diffs$mean_tmm2,
               ratio = temp_diffs$ratio)
    } %>%
    select(consensus_tissue_name, enssscg_id = variable, logFC, AveExpr, t, 
           pval = P.Value, 
           adj_pval = adj.P.Val,
           everything()) %>%
    mutate(adj_p_global = p.adjust(pval, method = "BH")) %>%
    mutate(sign = adj_p_global < 0.05,
           relevant = abs(ratio) >= 4 | abs(ratio) <= 0.25,
           show = sign & relevant)
  
  rm(intraconsens_anova_temp2)
  
  save(intraconsens_anova_res,
       file = "data/processed/pig_intraconsensus_anova.Rdata")
  
  
  
  intraconsens_anova_res2 <- 
    intraconsens_anova_res %>% 
    select(consensus_tissue_name,
           enssscg_id,
           tissue1,
           tissue2,
           mean_tmm1,
           mean_tmm2,
           ratio, 
           adj_p_global,
           show)
  
  
  save(intraconsens_anova_res2,
       file = "data/processed/pig_intraconsensus_anova2.Rdata")
  
  rm(intraconsens_anova_res)
  rm(intraconsens_anova_diffs)
  rm(diff_data_meta)
  rm(diff_data)
}

# To be relevant, at least one tissue has to be deviating 2sd - for brain only as it has many more subtissues
outlim_tissues <- 
  anova_data %>% 
  group_by(enssscg_id, consensus_tissue_name, tissue_name) %>% 
  summarise(tmm = mean(tmm)) %>%
  group_by(consensus_tissue_name, enssscg_id) %>% 
    mutate(tmm = log10(tmm + 1),
           mean = mean(tmm),
           sd = sd(tmm),
           lolim = mean - 2*sd,
           hilim = mean + 2*sd) %>% 
  ungroup() %>%
  filter(tmm < lolim | tmm > hilim | consensus_tissue_name != "brain") %>%
  select(enssscg_id, consensus_tissue_name) %>%
  distinct()
  

intraconsens_relevant_genes <- 
  intraconsens_anova_res2 %>%
  filter(show) %>%
  inner_join(outlim_tissues) %>%
  left_join(pig_gene_classification) 

intraconsens_relevant_genes_summary <- 
  intraconsens_relevant_genes %>%
  # filter(consensus_tissue_name == "lymphoid tissue") %>%
  mutate(direction = ifelse(ratio > 1, 1, -1),
         hi_tissue = ifelse(direction == 1, tissue1, tissue2),
         lo_tissue = ifelse(direction == 1, tissue2, tissue1),
         ratio = ifelse(direction == 1, ratio, 1/ratio)) %>%
  select(consensus_tissue_name, enssscg_id, direction, hi_tissue, lo_tissue, ratio) %>% 
  group_by(consensus_tissue_name, enssscg_id) %>%
  summarise(high_tissues = paste(sort(unique(hi_tissue[!hi_tissue %in% lo_tissue])), collapse = ";"),
            n_high = n_distinct(hi_tissue[!hi_tissue %in% lo_tissue]),
            low_tissues = paste(sort(unique(lo_tissue[!lo_tissue %in% hi_tissue])), collapse = ";"),
            n_low = n_distinct(lo_tissue[!lo_tissue %in% hi_tissue])) %>%
  ungroup() %>% 
  arrange(enssscg_id)

intraconsens_highlow_res <- 
  intraconsens_relevant_genes_summary %>% 
  {bind_rows(select(., 1:2, 
                    tissue = high_tissues,
                    n = n_high) %>% 
               mutate(type = "high"),
             select(., 1:2, 
                    tissue = low_tissues,
                    n = n_low) %>% 
               mutate(type = "low"))} %>%
  separate_rows(tissue, sep = ";") 

intraconsensus_cluster_settings <- 
  bind_rows(tibble(consensus_tissue_name = 'adipose tissue',
                   cutree_tissue = 1,
                   cutree_genes = 6),
            tibble(consensus_tissue_name = 'brain',
                   cutree_tissue = 7,
                   cutree_genes = 22),
            tibble(consensus_tissue_name = 'cartilage',
                   cutree_tissue = 1,
                   cutree_genes = 4),
            tibble(consensus_tissue_name = 'heart',
                   cutree_tissue = 1,
                   cutree_genes = 7),
            tibble(consensus_tissue_name = 'kidney',
                   cutree_tissue = 1,
                   cutree_genes = 5),
            tibble(consensus_tissue_name = 'large intestine',
                   cutree_tissue = 1,
                   cutree_genes = 10),
            tibble(consensus_tissue_name = 'lung',
                   cutree_tissue = 1,
                   cutree_genes = 2),
            tibble(consensus_tissue_name = 'lymphoid tissue',
                   cutree_tissue = 3,
                   cutree_genes = 9),
            tibble(consensus_tissue_name = 'male glands',
                   cutree_tissue = 3,
                   cutree_genes = 8),
            tibble(consensus_tissue_name = 'mesothelial tissue',
                   cutree_tissue = 1,
                   cutree_genes = 5),
            tibble(consensus_tissue_name = 'mouth',
                   cutree_tissue = 1,
                   cutree_genes = 8),
            tibble(consensus_tissue_name = 'skin',
                   cutree_tissue = 1,
                   cutree_genes = 4),
            tibble(consensus_tissue_name = 'small intestine',
                   cutree_tissue = 1,
                   cutree_genes = 6),
            tibble(consensus_tissue_name = 'stomach',
                   cutree_tissue = 1,
                   cutree_genes = 4),
            tibble(consensus_tissue_name = 'upper respiratory system',
                   cutree_tissue = 1,
                   cutree_genes = 7))


  

anova_data %>%
  filter(enssscg_id == "ENSSSCG00000000026") %>% 
  filter(consensus_tissue_name == "brain") %>%
  mutate(id = paste(tissue_name, individual)) %>%
  ggplot(aes(tmm, tissue_name)) +
  geom_boxplot() +
  geom_point()

pig_atlas_sample_norm %>%
  filter(enssscg_id == "ENSSSCG00000036923") %>% 
  left_join(tissue_mapping,
            by = c("tissue_ID" = "tissue")) %>%
  filter(consensus_tissue_name == "brain") %>%
  mutate(id = paste(tissue_name, individual)) %>%
  ggplot(aes(tpm, tissue_name)) +
  geom_boxplot() +
  geom_point()

pig_atlas_consensus %>% 
  filter(enssscg_id == "ENSSSCG00000036923") 

intraconsens_relevant_genes %>%
  filter(enssscg_id == "ENSSSCG00000000026") %>% 
  filter(consensus_tissue_name == "brain")  

######


intraconsens_heatmap_data <- 
  pig_atlas_sample_norm %>% 
  inner_join(tissue_mapping %>%
               filter(consensus_tissue_name %in% anova_data$consensus_tissue_name) ,
            by = c("tissue_ID" = "tissue")) %>% 
  
  
  inner_join(pig_gene_classification %>% 
               separate_rows(enhanced_tissues, sep = ",") %>%
               select(enssscg_id, enhanced_tissues) %>%
               bind_rows(intraconsens_relevant_genes %>%
                           select(enssscg_id, 
                                  enhanced_tissues = consensus_tissue_name)) %>%
               distinct(),
             by = c("enssscg_id", "consensus_tissue_name" = "enhanced_tissues")) %>%
  
  group_by(consensus_tissue_name, tissue_name, enssscg_id) %>% 
  summarise(tmm = mean(tmm)) %>%
  ungroup()  %>%
  group_by(enssscg_id, consensus_tissue_name) %>% 
  mutate(log_tmm = log10(tmm + 1),
         tmm = tmm / max(tmm, na.rm = T)) %>%
  ungroup() %>%
  filter(is.finite(tmm))


#####
intraconsens_heatmap_data %>% 
  filter(enssscg_id == "ENSSSCG00000040244")

pig_gene_classification %>% filter(enssscg_id == "ENSSSCG00000036923")
pig_gene_classification %>% filter(enssscg_id == "ENSSSCG00000040244")
pig_atlas_sample_norm %>%
  filter(enssscg_id == "ENSSSCG00000040244") %>% 
  left_join(tissue_mapping,
            by = c("tissue_ID" = "tissue")) %>%
  filter(consensus_tissue_name == "large intestine") %>%
  mutate(id = paste(tissue_name, individual)) %>%
  ggplot(aes(tpm, tissue_name)) +
  geom_boxplot() +
  geom_point()
pig_gene_classification %>% filter(enssscg_id == "ENSSSCG00000034997")
pig_atlas_sample_norm %>%
  filter(enssscg_id == "ENSSSCG00000022233") %>% 
  left_join(tissue_mapping,
            by = c("tissue_ID" = "tissue")) %>%
  filter(consensus_tissue_name == "brain") %>%
  mutate(id = paste(tissue_name, individual)) %>%
  ggplot(aes(tpm, tissue_name)) +
  geom_boxplot() +
  geom_point()

#####

intraconsens_clusters <- 
  lapply(sort(unique(intraconsens_heatmap_data$consensus_tissue_name)),
         function(orgn) {
           print(orgn)
           cluster_data_organ <- 
             intraconsens_heatmap_data %>%
             filter(consensus_tissue_name == orgn) %>% 
             select(-log_tmm) %>%
             spread(tissue_name, tmm) %>%
             select(-consensus_tissue_name) %>%
             column_to_rownames("enssscg_id") 
           
           cluster_settings <- 
             intraconsensus_cluster_settings %>%
             filter(consensus_tissue_name == orgn)
           
           gene_cluster <-
             cluster_data_organ %>%
             dist() %>%
             hclust(method = "ward.D2")
           
           tis_cluster <-
             cluster_data_organ %>%
             t() %>%
             dist() %>%
             hclust(method = "ward.D2")
           
           gene_cluster_temp <- 
             cutree(gene_cluster,
                    k = cluster_settings$cutree_genes) %>%
             enframe("enssscg_id", "cluster_i")
           
           gene_cluster_id <- 
             gene_cluster_temp %>%
             left_join(labels(gene_cluster) %>% 
                         enframe("row", "enssscg_id") %>% 
                         left_join(gene_cluster_temp) %>% 
                         select(cluster_i) %>%
                         distinct() %>%
                         mutate(cluster_id = LETTERS[row_number()]))
             
           
           tis_cluster_temp <- 
             cutree(tis_cluster,
                    k = cluster_settings$cutree_tissue) %>%
             enframe("tissue", "cluster_i")
           
           tis_cluster_id <- 
             tis_cluster_temp %>%
             left_join(labels(tis_cluster) %>% 
                         enframe("row", "tissue") %>% 
                         left_join(tis_cluster_temp) %>% 
                         select(cluster_i) %>%
                         distinct() %>%
                         mutate(cluster_id = LETTERS[row_number()]))
           
           list(gene_cluster = gene_cluster,
                tis_cluster = tis_cluster,
                gene_cluster_id = gene_cluster_id,
                tis_cluster_id = tis_cluster_id)
         }) %>%
  set_names(sort(unique(intraconsens_heatmap_data$consensus_tissue_name)))

#####

intraconsens_gene_clusters <- 
  intraconsens_clusters %>%
  map(.f = function(x) x$gene_cluster_id) %>%
  bind_rows(.id = "consensus_tissue_name") 

intraconsens_tissue_clusters <- 
  intraconsens_clusters %>%
  map(.f = function(x) x$tis_cluster_id) %>%
  bind_rows(.id = "consensus_tissue_name") 

intraconsens_tissue_cluster_dict <- 
  intraconsens_relevant_genes_summary %>% 
  select(high_tissues, low_tissues) %>% 
  distinct() %>% 
  gather(temp, tissue) %>%
  select(-1) %>%
  distinct() %>%
  mutate(all_tissues = tissue) %>%
  separate_rows(tissue, sep = ";") %>% 
  left_join(intraconsens_tissue_clusters,
            by = "tissue") %>%
  group_by(all_tissues) %>% 
  summarise(tissue_clusters = paste(unique(cluster_id), collapse = ","))

intraconsens_classification <-
  intraconsens_relevant_genes_summary %>%
  left_join(intraconsens_gene_clusters) %>% 
  left_join(intraconsens_tissue_cluster_dict,
            by = c("high_tissues" = "all_tissues")) %>%
  left_join(intraconsens_tissue_cluster_dict,
            by = c("low_tissues" = "all_tissues"),
            suffix = c("_high", "_low"))
  
######

intraconsens_gene_enrich_overlap <- 
  intraconsens_classification %>% 
  left_join(pig_gene_classification %>% 
              separate_rows(enhanced_tissues, sep = ";"),
            by = c("enssscg_id")) %>%
  group_by(consensus_tissue_name, enssscg_id, high_tissues, n_high, low_tissues, n_low, cluster_id) %>%
  summarise(overlap = case_when(any(is.na(enhanced_tissues)) ~ "variable - not elevated",
                                any(enhanced_tissues == consensus_tissue_name) ~ "variable and elevated",
                                T ~ "variable - elevated elsewhere")) %>%
  left_join(pig_gene_classification,
            by = c("enssscg_id")) %>%
  ungroup() %>%
  left_join(gene_mapping %>% 
              select(1:2)) %>%
  select(1, 2, gene_name = display_name, everything())




intraconsens_gene_enrich_overlap_all <- 
  pig_gene_classification %>% 
  # head %>%
  separate_rows(enhanced_tissues, sep = ";") %>%
  rename(consensus_tissue_name = enhanced_tissues) %>%
  filter(!is.na(consensus_tissue_name)) %>%
  filter(consensus_tissue_name %in% consensus_tissues_with_multiple$consensus_tissue_name) %>%
  filter(!(paste(consensus_tissue_name, 
                 enssscg_id) %in% 
             paste(intraconsens_gene_enrich_overlap$consensus_tissue_name,
                   intraconsens_gene_enrich_overlap$enssscg_id))) %>%
  left_join(gene_mapping) %>%
  mutate(high_tissues = NA,
         n_high = 0,
         low_tissues = NA, 
         n_low = 0,
         overlap = "not variable - elevated") %>%
  left_join(intraconsens_gene_clusters %>% 
              select(1, 2, 4)) %>% 
  select(consensus_tissue_name, 
         enssscg_id,
         gene_name = display_name,
         high_tissues,
         n_high,
         low_tissues, 
         n_low, cluster_id,
         overlap,
         specificity_category,
         distribution_category,
         ts_score, 
         enhanced_score) %>%
  bind_rows(intraconsens_gene_enrich_overlap)

intraconsens_gene_enrich_overlap %>%
  write_csv(savepath("intraconsensus results.csv"))

intraconsens_gene_enrich_overlap %>%
  select(1:9) %>%
  write_csv(savepath("intraconsensus results small.csv"))

intraconsens_gene_enrich_overlap %>%
  group_by(enssscg_id, gene_name) %>% 
  summarise(consensus_tissue_name = paste(sort(unique(consensus_tissue_name)),
                                          collapse = ";")) %>%
  write_csv(savepath("intraconsensus results summarised.csv"))

intraconsens_gene_enrich_overlap_all %>%
  arrange(consensus_tissue_name, enssscg_id) %>%
  write_csv(savepath("intraconsensus results including elevated.csv"))

######


```

###Figures

```{r intra_consensus2, echo=FALSE, warning=FALSE}



########
intraconsens_relevant_genes_summary %>% 
  mutate(n_var = ifelse(n_high == 1 | n_low == 1,
                        "single region",
                        "multiple regions")) %>%
  group_by(consensus_tissue_name, n_var) %>% 
  summarise(n_genes = n()) %>%
  group_by(consensus_tissue_name) %>%
  mutate(n_tot = sum(n_genes)) %>%
  ungroup() %>%
  arrange(n_tot) %>%
  mutate(consensus_tissue_name = factor(consensus_tissue_name, unique(consensus_tissue_name))) %>%
  ggplot(aes(n_genes, consensus_tissue_name, fill = n_var)) +
  geom_col() +
  stripped_theme_facet +
  xlab("Number of genes with local variability") +
  scale_fill_discrete(name = "Variable regions") +
  theme(strip.text.y = element_text(angle = 0),
        panel.spacing = unit(0, "mm"),
        axis.title.y = element_blank())

ggsave(savepath("intraconsensus n var genes barplot.pdf"), width = 5, height = 3)

intraconsens_gene_enrich_overlap_all %>% 
  group_by(consensus_tissue_name, overlap) %>% 
  summarise(n_genes = n()) %>%
  group_by(consensus_tissue_name) %>%
  mutate(n_tot = sum(n_genes)) %>%
  ungroup() %>%
  arrange(n_tot) %>%
  mutate(consensus_tissue_name = factor(consensus_tissue_name, unique(consensus_tissue_name)),
         overlap = factor(overlap, variability_levels)) %>%
  ggplot(aes(n_genes, consensus_tissue_name, fill = overlap)) +
  geom_col() +
  stripped_theme_facet +
  xlab("Number of genes") +
  
  theme(strip.text.y = element_text(angle = 0),
        panel.spacing = unit(0, "mm"),
        axis.title.y = element_blank()) +
  scale_fill_manual(values = variability_palette,
                    name = "Variable regions") +
  ggtitle("Number of genes elevated or with local variability")

ggsave(savepath("intraconsensus n var + ele barplot.pdf"), width = 5, height = 3)

intraconsens_highlow_res %>%
  # mutate(single = n == 1) %>%
  group_by(consensus_tissue_name, tissue, type) %>%
  summarise(n_genes = n()) %>%
  ungroup() %>%
  ggplot(aes(n_genes, tissue, fill = type)) +
  geom_col() +
  facet_grid(consensus_tissue_name ~ ., space = "free", scales = "free_y") +
  stripped_theme_facet +
  theme(strip.text.y = element_text(angle = 0),
        panel.spacing = unit(0, "mm"))

ggsave(savepath("intraconsensus summary barplot.pdf"), width = 8, height = 8)

intraconsens_relevant_genes_summary %>% 
  group_by(n_high) %>% 
  summarise(n = n()) %>%
  ggplot(aes(n_high, n)) +
  geom_col() + 
  xlab("Number of elevated tissues")  +
  stripped_theme +
  scale_y_continuous(expand = c(0,0))

ggsave(savepath("intraconsensus number of enriched.pdf"), width = 4, height = 4)

intraconsens_highlow_res %>%
  # mutate(single = n == 1) %>%
  filter(type == "high") %>%
  group_by(consensus_tissue_name, tissue) %>%
  summarise(n_genes = n()) %>%
  ungroup() %>%
  ggplot(aes(n_genes, tissue)) +
  geom_col() +
  facet_grid(consensus_tissue_name ~ ., space = "free", scales = "free_y") +
  stripped_theme_facet +
  theme(strip.text.y = element_text(angle = 0),
        panel.spacing = unit(0, "mm"))

ggsave(savepath("intraconsensus summary barplot only high.pdf"), width = 8, height = 8)



#####






```

####Heatmaps

```{r}



# 
# gene_cluster <- 
#   intraconsens_clusters[[orgn]]$gene_cluster
# 
# tis_cluster <- 
#   intraconsens_clusters[[orgn]]$tis_cluster
# 
# ggdendrogram(tis_cluster) +
#   theme(axis.text = element_blank()) +
# 
#   plot_spacer() +
#   
# intraconsens_heatmap_data %>%
#   filter(consensus_tissue_name == orgn) %>%
#   mutate(tissue_name = factor(tissue_name, labels(tis_cluster)),
#          enssscg_id = factor(enssscg_id, labels(gene_cluster))) %>% 
#   ggplot(aes(tissue_name, enssscg_id, fill = tmm)) +
#   geom_tile() +
#   scale_fill_viridis_c(option = "B", direction = -1) +
#   theme(axis.text.y = element_blank(),
#         axis.text.x = element_text(angle = 60, hjust = 1),
#         axis.ticks.y = element_blank(),
#         axis.title = element_blank()) +
#   
#   # ggdendrogram(gene_cluster, labels = F) +
#   # theme(axis.text = element_blank()) +
#   
#   plot_layout(nrow = 2, ncol = 2)
# dendro_data



pdf(savepath("Intraconsensus anova significant heatmaps.pdf"), width = 8, height = 7)
plots <- 
  lapply(unique(intraconsens_heatmap_data$consensus_tissue_name),
         function(orgn) {
           print(orgn)
           heatmap_data_organ <- 
             intraconsens_heatmap_data %>%
             filter(consensus_tissue_name == orgn) %>% 
             select(-log_tmm) %>%
             spread(tissue_name, tmm) %>%
             select(-consensus_tissue_name) %>%
             column_to_rownames("enssscg_id") 
           
           heatmap_meta_organ <- 
             rownames(heatmap_data_organ) %>% 
             enframe("name", "enssscg_id") %>%
             select(-1) %>%
             left_join(intraconsens_gene_enrich_overlap_all %>%
                         filter(consensus_tissue_name == orgn)) %>% 
             
             select(enssscg_id, overlap) %>% 
             mutate(overlap = ifelse(is.na(overlap), "elevated in tissue", overlap)) %>%
             left_join(intraconsens_clusters[[orgn]]$gene_cluster_id %>%
                         select(enssscg_id, 
                                `gene cluster` = cluster_id)) %>%
             
             select(1, 3, 2) %>%
             column_to_rownames("enssscg_id")
           
           
           heatmap_settings <- 
             intraconsensus_cluster_settings %>%
             filter(consensus_tissue_name == orgn)
           
           cluster_pal <- 
             n_distinct(heatmap_meta_organ$`gene cluster`) %>%
             {set_names(viridis(.), LETTERS[1:.])}
           
           g <- 
             heatmap_data_organ %>%
             pheatmap(clustering_method = "ward.D2", 
                      color = heatmap_palette,
                      border_color = NA, 
                      show_rownames = F, 
                      annotation_row = heatmap_meta_organ,
                      annotation_colors = list(overlap = variability_palette, 
                                               "gene cluster" =  cluster_pal),
                      main = orgn,
                      cutree_rows = heatmap_settings$cutree_genes,
                      cutree_cols = 1,
                      silent = T)
           
           plot(g$gtable)
         })

dev.off()


heatmap_data <- 
  pig_atlas_sample_norm  %>%
  inner_join(tissue_mapping %>%
               filter(consensus_tissue_name %in% anova_data$consensus_tissue_name) ,
            by = c("tissue_ID" = "tissue")) %>% 
  filter(enssscg_id %in% intraconsens_heatmap_data$enssscg_id) %>%
  inner_join(pig_gene_classification %>% 
               separate_rows(enhanced_tissues, sep = ",") %>%
               select(enssscg_id, enhanced_tissues) %>%
               bind_rows(intraconsens_relevant_genes %>%
                           select(enssscg_id, 
                                  enhanced_tissues = consensus_tissue_name)) %>%
               distinct(),
             by = c("enssscg_id", "consensus_tissue_name" = "enhanced_tissues")) %>%
  
  mutate(log_tmm = log10(tmm + 1),
         tmm = tmm / max(tmm, na.rm = T),
         sample_ID = paste(tissue_name, individual))


pdf(savepath("Intraconsensus anova significant heatmaps logtmm.pdf"), width = 8, height = 7)
plots <- 
  lapply(unique(heatmap_data$consensus_tissue_name),
         function(orgn) {
           print(orgn)
           heatmap_data_organ <- 
             heatmap_data %>%
             filter(consensus_tissue_name == orgn) %>% 
             select(enssscg_id, sample_ID, log_tmm) %>%
             spread(sample_ID, log_tmm) %>%
             column_to_rownames("enssscg_id") 
           
           heatmap_meta_organ <- 
             rownames(heatmap_data_organ) %>% 
             enframe("name", "enssscg_id") %>%
             select(-1) %>%
             left_join(intraconsens_gene_enrich_overlap_all %>%
                         filter(consensus_tissue_name == orgn)) %>% 
             
             select(enssscg_id, overlap) %>% 
             mutate(overlap = ifelse(is.na(overlap), "elevated in tissue", overlap)) %>%
             left_join(intraconsens_clusters[[orgn]]$gene_cluster_id %>%
                         select(enssscg_id, 
                                `gene cluster` = cluster_id)) %>%
             
             select(1, 3, 2) %>%
             column_to_rownames("enssscg_id")
           
           
           heatmap_settings <- 
             intraconsensus_cluster_settings %>%
             filter(consensus_tissue_name == orgn)
           
           
           cluster_pal <- 
             n_distinct(heatmap_meta_organ$`gene cluster`) %>%
             {set_names(viridis(.), LETTERS[1:.])}
           
           
           g <- 
             heatmap_data_organ %>%
             pheatmap(cluster_rows = intraconsens_clusters[[orgn]]$gene_cluster,
                      clustering_method = "ward.D2", 
                      color = heatmap_palette,
                      border_color = NA, 
                      show_rownames = F, 
                      annotation_row = heatmap_meta_organ,
                      annotation_colors = list(overlap = variability_palette,
                                               "gene cluster" =  cluster_pal) ,
                      main = orgn,
                      cutree_rows = heatmap_settings$cutree_genes,
                      cutree_cols = 1,
                      silent = T)
           
           plot(g$gtable)
         })

dev.off()

  
```


#### Network

```{r}
plots <- 
  intraconsens_relevant_genes_summary %>% 
  mutate(spec_category = ifelse(n_high == 1, 
                                "tissue enriched",
                                "group enriched")) %>%
  filter(n_high <= 10) %>%
  {lapply(sort(unique(.$consensus_tissue_name)),
          function(tis_) {
            filter(., consensus_tissue_name == tis_) %>%
              classification_network_plot_2(gene_col = "enssscg_id", 
                                            spec_col = "spec_category", 
                                            enriched_col = "high_tissues", 
                                            spec_filter = c("tissue enriched",
                                                            "group enriched"), 
                                            pal = c(gene_category_pal, tissue_colors_palette), 
                                            savename = paste("Intraconsensus", tis_), 
                                            enriched_sep = ";", 
                                            node_filter_rank = 2, 
                                            node_filter_show_cat = c("tissue enriched"), 
                                            node_filter_min = 2, 
                                            node_filter_n_show = 8, 
                                            scale_factor = 1) +
              ggtitle(tis_)
          })}

pdf(savepath("intraconsensus networks.pdf"), width = 12, height = 12)
plots
dev.off()
```



#Figures

##Figure 1

###PCA
```{r Figure1_PCA, echo=FALSE, message=FALSE, warning=FALSE}



g <- 
  sample_pca$scores %>%
  as_tibble(rownames = "sample_id") %>%
  separate(sample_id, into = c("tissue_id", "individual"), sep = "_", remove = F) %>%
  
  left_join(tissue_mapping,
            by = c("tissue_id" = "tissue")) %>%
  
  mutate(x = PC1, 
         y = PC2, 
         tissue_color = organ_name) %>%
  mutate(tissue_color = factor(tissue_color, 
                               levels = c('Adipose & soft tissue',
                                          'Brain',
                                          'Bone marrow & immune system',
                                          'Breast and female reproductive system',
                                          'Male reproductive system',
                                          
                                          'Gastrointestinal tract',
                                          'Proximal digestive tract',
                                          'Respiratory system',
                                          'Eye',
                                          'Liver & gallbladder',
                                          'Kidney & urinary bladder',
                                          'Endocrine tissues',
                                          'Muscle tissues',
                                          'Skin'))) %>%
  
  ggplot(aes(x, y, color = tissue_color, fill = tissue_color)) +
  
  
  geom_point(show.legend = T, 
             shape = 21, 
             color = "gray25", 
             size = 2.5, 
             alpha = 0.7) +
  # geom_text(aes(label = group_col ), hjust = 0) +
  
  
  scale_color_manual(values = tissue_colors_palette_full, name = "") +
  scale_fill_manual(values = tissue_colors_palette_full, name = "") +
  # 
  # scale_x_continuous(limits = c(plot_lims_x)) + 
  # scale_y_continuous(limits = c(plot_lims_y)) +
  
  
  # coord_fixed((plot_lims_x[1] - plot_lims_x[2]) /
  #               (plot_lims_y[1] - plot_lims_y[2])) +
  
  facet_zoom(xy = organ_name == "Brain", horizontal = F, zoom.size = 0.5) +
  
  
  xlab("PC1") +
  ylab("PC2") + 
  stripped_theme +
  theme(plot.title = element_text(hjust = 0.5), 
        panel.border = element_blank(), 
        panel.background = element_blank())

ggsave(savepath("Figure1 PCA 1.pdf"), plot = g, width = 8, height = 7, useDingbats = F)
ggsave(savepath("Figure1 PCA 1 no legend.pdf"), plot = g + theme(legend.position = "none"), 
       width = 8, height = 7, useDingbats = F)
ggsave(savepath("Figure1 PCA 1 no legend small.pdf"), plot = g + theme(legend.position = "none"), 
       width = 4, height = 7, useDingbats = F)

g <- 
  sample_pca$scores %>%
  as_tibble(rownames = "sample_id") %>%
  separate(sample_id, into = c("tissue_id", "individual"), sep = "_", remove = F) %>%
  
  left_join(tissue_mapping,
            by = c("tissue_id" = "tissue")) %>%
  
  mutate(x = PC1, 
         y = PC2, 
         tissue_color = ifelse(x > 5, 
                               region_tissue_name,
                               "")) %>%
  mutate(tissue_color = str_to_sentence(tissue_color)) %>%
  
  ggplot(aes(x, y, color = tissue_color, fill = tissue_color)) +
  
  
  geom_point(show.legend = T, 
             shape = 21, 
             color = "gray25", 
             size = 2.5, 
             alpha = 0.7) +
  
  scale_color_manual(values = tissue_colors_palette_full, name = "") +
  scale_fill_manual(values = tissue_colors_palette_full, name = "") +
  facet_zoom(xy = organ_name == "Brain", horizontal = F, zoom.size = 0.5) +
  
  
  xlab("PC1") +
  ylab("PC2") + 
  stripped_theme +
  theme(plot.title = element_text(hjust = 0.5), 
        panel.border = element_blank(), 
        panel.background = element_blank())

ggsave(savepath("Figure1 PCA 2.pdf"), plot = g, width = 8, height = 7, useDingbats = F)
ggsave(savepath("Figure1 PCA 2 no legend.pdf"), plot = g + theme(legend.position = "none"), 
       width = 8, height = 7, useDingbats = F)
ggsave(savepath("Figure1 PCA 2 no legend small.pdf"), plot = g + theme(legend.position = "none"), 
       width = 4, height = 7, useDingbats = F)
```


###Retinagram
```{r Figure1_retina, echo=FALSE, message=FALSE, warning=FALSE}
# 
# 
# tissue_wardclust <- 
#   # tissue_pca$scores[, 1:50] %>% 
#   tissue_pca$scores[, 1:50] %>% 
#   # set_rownames(tissue_colors$tissue_name[match(rownames(.), tissue_colors$tissue)]) %>%
#   dist() %>% 
#   hclust(method = "ward.D2")
# 
# circular_dendrogram_retinastyle_2(clust = tissue_wardclust, 
#                                 color_mapping = tissue_colors, 
#                                 label_col = "tissue_name", 
#                                 color_col = "tissue_color", arc_strength = 0)

tissue_spearclust <-
  pig_atlas_tissue %>% 
  
  select(enssscg_id, tissue_name, tmm) %>% 
  
  spread(enssscg_id, tmm) %>%
  mutate(tissue_name = str_to_sentence(tissue_name)) %>%
  column_to_rownames("tissue_name")  %>%
  t() %>%
  cor(method = "spearman") %>%
  {1 - .} %>%
  as.dist() %>%
  hclust(method = "ward.D2")

circular_dendrogram_retinastyle_2(clust = tissue_spearclust, 
                                color_mapping = tissue_colors %>%
                                  mutate(tissue_name = str_to_sentence(tissue_name) %>%
                                           gsub("lower", "(lower)", .) %>% 
                                           gsub("upper", "(upper)", .) %>%
                                           gsub("Kidney cortex", "Kidney (cortex)", .) %>% 
                                           gsub("Kidney medulla", "Kidney (medulla)", .)), 
                                label_col = "tissue_name", 
                                color_col = "tissue_color", arc_strength = 0, width_range = c(0.75, 7))
ggsave(savepath("Figure1 retinagram.pdf"), width = 6, height = 6)

circular_dendrogram_retinastyle_2(clust = tissue_spearclust, 
                                color_mapping = tissue_colors %>%
                                  mutate(tissue_name = str_to_sentence(tissue_name) %>%
                                           gsub("lower", "(lower)", .) %>% 
                                           gsub("upper", "(upper)", .) %>%
                                           gsub("Kidney cortex", "Kidney (cortex)", .) %>% 
                                           gsub("Kidney medulla", "Kidney (medulla)", .)), 
                                label_col = "tissue_name", 
                                color_col = "tissue_color", arc_strength = 0, width_range = c(0.4, 3.5), 
                                text_size = 2)
ggsave(savepath("Figure1 retinagram 2.pdf"), width = 80, height = 80, units = "mm")

tissue_spearclust %>%
  ggdendrogram(rotate = T)
ggsave(savepath("Figure1 retinagram flat.pdf"), width = 4, height = 10)
```





##Figure 2

###Classification alluvial

```{r Pig_class_alluvial, echo=FALSE, message=FALSE, warning=FALSE}



# pig_gene_classification %>%
#   mutate(distribution_category = str_to_sentence(distribution_category)) %>%
#   multi_alluvial_plot(vars = c("Specificity" = "specificity_category", 
#                                "Distribution" = "distribution_category"), 
#                       chunk_levels = c('Tissue enriched', 'Group enriched', 
#                                        'Tissue enhanced', 'Low tissue specificity', 
#                                        'Detected in single',
#                                        'Detected in some', 
#                                        'Detected in many', 
#                                        'Detected in all',
#                                        'Not detected'), 
#                       pal = c(gene_category_pal, elevation_identity_pal), 
#                       color_by = c(1, 1)) +
#   coord_flip()


width = 0.1

alluv_1 <-
  
  pig_gene_classification %>%
   mutate(Specificity = str_to_sentence(specificity_category),
          Distribution = str_to_sentence(distribution_category)) %>%
  select(Specificity,
         Distribution) %>%
  mutate(row_n = row_number()) %>%
  gather(bar, chunk, -row_n) %>%
  mutate(color_vars = 1) %>%
  group_by(row_n) %>%
  mutate(chunk_color = chunk[match(c("Specificity",
                                     "Distribution")[color_vars], bar)]) %>% 
  ungroup() %>%
  
  
  mutate(chunk = factor(chunk, levels = c('Tissue enriched', 'Group enriched', 
                                          'Tissue enhanced', 'Low tissue specificity', 
                                          'Detected in single',
                                          'Detected in some', 
                                          'Detected in many', 
                                          'Detected in all',
                                          'Not detected')),
         bar = factor(bar, levels = c("Specificity",
                                      "Distribution"))) %>%
  
  
  ggplot(aes(x = bar, stratum = chunk, alluvium = row_n,
             y = 1)) +
  
  geom_flow(aes(fill = chunk_color), 
            show.legend = F, width = width,
            knot.pos = 1/6) +
  geom_stratum(aes(fill = chunk), 
               show.legend = F, color = NA, width = width) +
  
  scale_x_discrete(expand = c(.1, .1), position = "top") +
  scale_fill_manual(values = c(gene_category_pal, elevation_identity_pal)) + 
  
  
  theme(axis.text.y = element_text(size = 18, face = "bold"),
        axis.text.x = element_blank(), 
        axis.ticks = element_blank(), 
        panel.background = element_blank(), 
        axis.title = element_blank()) +
  coord_flip()

# alluv_1
flow_data <-
  ggplot_build(alluv_1)$data[[1]] %>%
  as_tibble() %>%
  {
    if("side" %in% names(.)) {
      .
    } else{
      mutate(.,
             side = case_when(contact == "back" ~ "start",
                              contact == "front" ~ "end"))
    }}

stratum_data <- 
  ggplot_build(alluv_1)$data[[2]]

flow_data_labels <-
  flow_data %>% 
  as_tibble() %>% 
  select(x, stratum, group, side, ymin, ymax) %>% 
  multispread(c(side), c(x, stratum, ymin, ymax)) %>% 
  mutate_at(c("end_x", "end_ymax", "end_ymin", "start_x", "start_ymax", "start_ymin"), as.numeric) %>% 
  group_by(start_stratum, end_stratum, start_x, end_x) %>%
  summarise(end_y = (min(end_ymin) + max(end_ymax)) / 2, 
            start_y = (min(start_ymin) + max(start_ymax)) / 2, 
            size = max(start_ymax) - min(start_ymin))

alluv_2 <- 
  alluv_1 +
  geom_text(data = flow_data_labels,
            aes(x = start_x + width/2,
                y = start_y, 
                label = size), 
            inherit.aes = F, 
            size = 3,
            angle = -90,
            hjust = 1,
            vjust = 0.5) +
  geom_text(data = flow_data_labels,
            aes(x = end_x - width/2,
                y = end_y, 
                label = size), 
            inherit.aes = F, 
            size = 3, 
            angle = -90,
            hjust = 0,
            vjust = 0.5) +
  
  # Stratum label
  
  geom_text(data = stratum_data %>%
              filter(x == 1),
            aes(x = x - width/2,
                y = y,
                label = stratum),
            size = 4, 
            vjust = 1.5,
            inherit.aes = F) + 
  geom_text(data = stratum_data %>%
              filter(x == 2),
            aes(x = x + width/2,
                y = y,
                label = stratum),
            size = 4, 
            vjust = -0.5,
            inherit.aes = F) + 
  
  geom_text(data = stratum_data,
            aes(x = x, 
                y = y,
                label = ymax - ymin), 
            size = 4, 
            fontface = "bold",
            color = "white",
            inherit.aes = F)


alluv_2

ggsave(savepath("Figure2 Alluvial.pdf"), width = 7, height = 2.5, useDingbats = F)

```

###Classification barplot

```{r Figure2_classification_bar, echo=FALSE, message=FALSE, warning=FALSE}

class_table_temp <-
  pig_gene_classification %>%
  select(enssscg_id, specificity_category, enhanced_tissues) %>%
  separate_rows(enhanced_tissues, sep = ";") %>%
  mutate(spec_category = factor(specificity_category, levels = rev(spec_category_levels)),
         enhanced_tissues = str_to_sentence(enhanced_tissues))
  
plot_dendro <-
  pig_atlas_consensus %>% 
  select(enssscg_id, tissue, tpm) %>% 
  mutate(tissue = str_to_sentence(tissue)) %>%
  spread(enssscg_id, tpm) %>%
  column_to_rownames("tissue")  %>%
  t() %>%
  cor(method = "spearman") %>%
  
  {1 - .} %>%
  as.dist() %>% 
  hclust(method = "average") %T>%
  plot %>%
  dendro_data()
  
  
dendro_plot_data <- 
  left_join(plot_dendro$segments, 
            plot_dendro$labels, 
            by = c("x" = "x", "yend" = "y")) 

left_plot <- 
  dendro_plot_data %>%
  ggplot() +
  geom_segment(aes(x=y, y=x, xend=yend, yend=xend, group = label))+
  geom_rect(aes(xmin=0, ymin=x + 0.5, 
                xmax=-0.02, ymax=xend - 0.5, 
                fill = label), 
            show.legend = F) +
  scale_color_manual(values = tissue_colors_palette_full)+
  scale_fill_manual(values = tissue_colors_palette_full)+
  scale_x_reverse(expand = expansion(0), position = "top")+
  scale_y_continuous(expand = expansion()) +
  xlab("1 - Spearman's rho") +
  
  theme(axis.text.y = element_blank(), 
        axis.title.y = element_blank(), 
        axis.ticks.y = element_blank(),
        plot.margin = unit(c(1,1,1,1), units = "mm"), 
        panel.background = element_blank()) 

right_plot <- 
  class_table_temp %>%
  filter(!is.na(enhanced_tissues)) %>%
  group_by(enhanced_tissues, specificity_category) %>%
  summarise(n_genes = n()) %>%
  ungroup() %>%
  mutate(enhanced_tissues = factor(enhanced_tissues, levels = plot_dendro$labels$label),
         specificity_category = factor(specificity_category, rev(spec_category_levels))) %>%
  ggplot(aes(n_genes, enhanced_tissues, fill = specificity_category)) +
  geom_col(width = 0.8, size = 0.1) +
  simple_theme +
  scale_fill_manual(values = gene_category_pal, name = "Specificity") +
  # coord_flip() +
  xlab("Number of genes") +
  
  scale_x_continuous(position = "top", expand = expansion(c(0, 0.1))) + 
  scale_y_discrete(expand = expansion()) +
  
  theme(axis.text.y = element_text(hjust = 0.5), 
        legend.position = c(0.7, 0.5),
        axis.title.y = element_blank(),
        panel.border = element_blank())

left_plot + right_plot +
  plot_layout(widths = c(0.3, 1))

ggsave(savepath("Figure2 barplot.pdf"), width = 5, height = 7)
```

###IHC - example barplots

```{r}


pig_atlas_consensus %>%
  inner_join(gene_mapping %>% 
               filter(display_name %in% c("MOG", "ASGR1", "CRISP2", "TNNT1", "DSC1"))) %>%
  left_join(tissue_mapping %>% 
              select(consensus_tissue_name, organ_name) %>% 
              distinct(),
            by = c("tissue" = "consensus_tissue_name")) %>%
  arrange(organ_name) %>%
  mutate(display_name = factor(display_name, c("MOG", "ASGR1", "CRISP2", "TNNT1", "DSC1")),
         tissue = factor(str_to_sentence(tissue), unique(str_to_sentence(tissue)))) %>%
  
  ggplot(aes(tissue, ptpm, fill = tissue)) +
  geom_col(show.legend = F,
           color = "black",
           size = 0.2) +
  geom_text(data = . %>% 
              filter(display_name == "MOG" & tissue == "Brain" |
                     display_name == "ASGR1" & tissue == "Liver" |
                     display_name == "CRISP2" & tissue == "Testis" |
                     display_name == "TNNT1" & tissue == "Skeletal muscle" |
                     display_name == "DSC1" & tissue == "Skin"),
            aes(label = tissue), 
            vjust = -0.5) +
  facet_wrap(~display_name, ncol = 1, scales = "free_y", strip.position = "left") + 
  scale_fill_manual(values = tissue_colors_palette_full) +
  scale_y_continuous(expand = expansion(c(0, 0.2)), position = "right") +
  stripped_theme_facet + 
  ylab("PTPM") +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.border = element_blank(),
        axis.line = element_blank())
ggsave(savepath("Figure2 IHC example barplots.pdf"), width = 2, height = 5)

```


###Classification network

```{r Figure2_classification_network, echo=FALSE, message=FALSE, warning=FALSE}


# class_network_celltype_enriched <- 
pig_gene_classification %>%
  classification_network_plot_2(gene_col = "enssscg_id",
                                spec_col = "specificity_category",
                                enriched_col = "enhanced_tissues",
                                spec_filter = c("Tissue enriched", 
                                                "Group enriched"),
                                pal = c(tissue_colors_palette_full, gene_category_pal),
                                savename = "Classification network",
                                enriched_sep = ";",
                                node_filter_rank = 2,
                                node_filter_show_cat = "Tissue enriched",
                                node_filter_min = 2,
                                node_filter_n_show = 5, 
                                scale_factor = 1)

ggsave(savepath("Network pig enrichment whole group.pdf"), width = 16, height = 12, useDingbats = F)

# 
# ####
# 
# pig_enrichment_whole_group <- 
#   pig_gene_classification %>% 
#   select(1, spec_category = specificity_category, enriched_tissues = enhanced_tissues) %>% 
#   filter(spec_category %in% c("Tissue enriched", "Group enriched")) %>% 
#   group_by(enriched_tissues, spec_category) %>% 
#   summarise(n_genes = n()) %>%
#   ungroup()
# 
# pig_net_data <-
#   pig_enrichment_whole_group %>% 
#   mutate(all_enriched_tissues = enriched_tissues) %>%
#   separate_rows(enriched_tissues, sep = ",") %>%
#   group_by(enriched_tissues, spec_category) %>% 
#   mutate(rank = rank(-n_genes, ties.method = "min")) %>%
#   group_by(all_enriched_tissues) %>%
#   mutate(any_low_rank = any(rank <= 2)) %>%
#   ungroup() %>%
#   
#   filter((spec_category == "tissue enriched" | any_low_rank | n_genes >= 8) &
#            (spec_category == "tissue enriched" | n_genes >= 2)) %>% 
#   mutate(edge_id = paste("enriched:", all_enriched_tissues)) %>% 
#     arrange(n_genes)
#   
# pig_net_edges <- 
#   pig_net_data %$% 
#   tibble(node1 = enriched_tissues, node2 = edge_id, n = n_genes) %>% 
#   unique()
# 
# g <-
#   pig_net_edges %>%
#   graph_from_data_frame(directed = FALSE) %>%
#   ggraph(layout = "kk") 
# 
# link_map <- 
#   pig_net_edges %>% 
#   gather(node, id, -(3:4)) %>% 
#   mutate(tissue_node = node == "node1", 
#          color_id = ifelse(tissue_node, id, !grepl(",", id)), 
#          label = ifelse(tissue_node, color_id, n)) %>%
#   select(n, node, id, tissue_node, color_id, label) %>%
#   unique()
# 
# 
# edge_data <- get_edges()(g$data)
# node_data <- 
#   get_nodes()(g$data) %>% 
#   as_tibble() %>%
#   left_join(link_map, 
#             by = c("name" = "id")) 
#   
# 
# g + 
#   geom_edge_arc(aes(width = n), 
#                     color = "gray", 
#                 strength = 0, 
#                 show.legend = F) + 
#     scale_edge_alpha_continuous(range = c(0.3, 1)) +
#     scale_edge_width_continuous(range = c(1, 3)) +
#   
#   geom_node_point(data = node_data  %>%
#                     filter(!tissue_node),
#                   aes(size = log(n), 
#                       fill = color_id), 
#                   stroke = 1,
#                   # size = 10,
#                   shape = 21,
#                   show.legend = F)+
#   geom_node_point(data = node_data %>%
#                     filter(tissue_node),
#                   aes(fill = color_id), 
#                   stroke = 1,
#                   size = 20,
#                   shape = 21,
#                   show.legend = F)+
#   geom_node_text(data = node_data,
#                  aes(label = label),
#                  size = 4) +
#   scale_size_continuous(range = c(5, 10)) +
#   scale_fill_manual(values = c(tissue_colors_palette_full, gene_category_pal)) +
#   
#   theme_void()
# 
# ggsave(savepath("Network pig enrichment whole group.pdf"), width = 16, height = 12, useDingbats = F)
# ggsave(savepath("Network pig enrichment whole group.png"), width = 16, height = 12)
# 
# 
# # Save for cytoscape
# pig_net_edges
# pig_net_data
# 
# cyto_summary <- 
#   pig_net_edges %>% 
#   mutate(category = ifelse(!grepl(",", node2), "Tissue enriched", "Group enriched"), 
#          node_id = unclass(factor(node2)),
#          node1 = str_to_sentence(node1), 
#          n_sqrt = sqrt(n), 
#          str_len = str_length(node1)) %>%
#   select(category, node1, node2, node_id, n, n_sqrt, str_len) 
# 
# cyto_summary %T>% 
#   write_delim(savepath("cytoscape nodes summary whole group.txt"), delim = "\t") %>%
#   write_csv(savepath("cytoscape nodes summary whole group.csv"))
# 
# bind_rows(cyto_summary %>% 
#             left_join(tissue_colors_palette_full %>% 
#                         enframe("node1", "color")) %>% 
#             select(node_id = node1, 
#                    color) %>% 
#             unique() %>%
#             mutate(node_type = "Tissue"),
#           cyto_summary %>% 
#             mutate(color = case_when(category == "Tissue enriched" ~ "#e41a1c",
#                                      category == "Group enriched" ~ "#FF9D00"), 
#                    node_id = as.character(node_id)) %>%
#             select(node_id, color) %>% 
#             unique() %>%
#             
#             mutate(node_type = "Enrichment")) %>%
#   mutate(color2 = case_when(node_type == "Enrichment" ~ color,
#                             node_type == "Tissue" ~ "#D3D3D3FF"),
#          color3 = case_when(node_type == "Enrichment" ~ color,
#                             node_type == "Tissue" ~ "#BEBEBEFF")) %T>% 
#   
#   write_delim(savepath("cytoscape nodes color whole group.txt"), delim = "\t") %>%
#   write_csv(savepath("cytoscape nodes color whole group.csv"))
# 
# bind_rows(cyto_summary %>% 
#             select(node_id = node1) %>% 
#             mutate(label = node_id) %>%
#             unique(),
#           cyto_summary %>% 
#             mutate(node_id = as.character(node_id), 
#                    label = as.character(n)) %>%
#             select(node_id, label) %>% 
#             unique()) %T>% 
#   write_delim(savepath("cytoscape nodes label whole group.txt"), delim = "\t") %>%
#   write_csv(savepath("cytoscape nodes label whole group.csv"))
```

##Figure 3

###Barplots

```{r}

plot_data <- 
  intraconsens_gene_enrich_overlap 
# intraconsens_gene_enrich_overlap %>% 
#   group_by(consensus_tissue_name, overlap) %>% 
#   count()
# pig_gene_classification %>% 
#   filter(!is.na(enhanced_tissues)) %>%
#   separate_rows(enhanced_tissues, sep = ";") %>% 
#   filter(enhanced_tissues %in% unique(plot_data$consensus_tissue_name)) %>%
#   rename(consensus_tissue_name = enhanced_tissues) %>%
#   # group_by(consensus_tissue_name) %>% 
#   # count()
#   
#   left_join(plot_data %>% 
#               select(consensus_tissue_name, enssscg_id, overlap)) %>% 
#   mutate(type = case_when(overlap == "variable and elevated" ~ "variable",
#                           is.na(overlap) | overlap == "variable - elevated elsewhere" ~ "not variable")) %>%
#   group_by(consensus_tissue_name, type) %>% 
#   count() %>% 
#   ungroup() %>%
#   mutate(consensus_tissue_name = factor(consensus_tissue_name, levels(plot_data2$consensus_tissue_name)),
#          facet = "Elevated genes")



plot_data2 <- 
  plot_data %>%
  group_by(consensus_tissue_name, overlap) %>% 
  summarise(n_genes = n()) %>%
  group_by(consensus_tissue_name) %>%
  mutate(n_tot = sum(n_genes)) %>%
  ungroup() %>%
  arrange(n_tot) %>%
  mutate(consensus_tissue_name = factor(consensus_tissue_name, unique(consensus_tissue_name)),
         overlap = factor(overlap, variability_levels), 
         facet = "Variable genes") 

plot_data3 <- 
  pig_gene_classification %>% 
  filter(!is.na(enhanced_tissues)) %>%
  separate_rows(enhanced_tissues, sep = ";") %>% 
  filter(enhanced_tissues %in% unique(plot_data$consensus_tissue_name)) %>%
  rename(consensus_tissue_name = enhanced_tissues) %>%
  left_join(plot_data %>% 
              select(consensus_tissue_name, enssscg_id, overlap)) %>% 
  mutate(type = case_when(overlap == "variable and elevated" ~ "variable",
                          is.na(overlap) | overlap == "variable - elevated elsewhere" ~ "not variable")) %>%
  group_by(consensus_tissue_name, type) %>% 
  count() %>% 
  ungroup() %>%
  mutate(consensus_tissue_name = factor(consensus_tissue_name, levels(plot_data2$consensus_tissue_name)),
         facet = "Elevated genes")


xlim <- 
  max(c(plot_data2 %>% 
          group_by(consensus_tissue_name) %>% 
          summarise(n = sum(n_genes)) %>% 
          pull(n) %>% 
          max(),
        plot_data3 %>% 
          group_by(consensus_tissue_name) %>% 
          summarise(n = sum(n)) %>% 
          pull(n) %>% 
          max()))

plot_data2 %>%
  ggplot(aes(n_genes, consensus_tissue_name, fill = overlap)) +
  geom_col() +
  stripped_theme_facet +
  xlab("Number of genes") +
  facet_wrap(~facet) +
  theme(strip.text.y = element_text(angle = 0),
        panel.spacing = unit(0, "mm"),
        axis.title.y = element_blank(),
        legend.position = "top",
        panel.border = element_blank()) +
  scale_fill_manual(values = variability_palette,
                    name = "") +
  scale_x_continuous(expand = expansion(0), limits = c(0, xlim)) +
  
  plot_data3 %>% 
  ggplot(aes(n, consensus_tissue_name, fill = type)) +
  geom_col() +
  stripped_theme_facet +
  xlab("Number of genes") +
  facet_wrap(~facet) +
  theme(strip.text.y = element_text(angle = 0),
        panel.spacing = unit(0, "mm"),
        axis.title.y = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.line.y = element_blank(),
        legend.position = "top",
        panel.border = element_blank()) +
  scale_fill_manual(values = c("not variable" = "darkgray", "variable" = "orangered"),
                    name = "") +
  scale_x_reverse(expand = expansion(0), limits = c(xlim, 0)) 

ggsave(savepath("intraconsensus n var + ele barplot.pdf"), width = 5, height = 3)
```


###Heatmaps

###Kidney

```{r}

plot_data <- 
  pig_atlas_sample_norm %>% 
  inner_join(tissue_mapping %>%
               filter(consensus_tissue_name %in% anova_data$consensus_tissue_name) ,
            by = c("tissue_ID" = "tissue")) %>% 
  
  
  inner_join(pig_gene_classification %>% 
               separate_rows(enhanced_tissues, sep = ",") %>%
               select(enssscg_id, enhanced_tissues) %>%
               bind_rows(intraconsens_relevant_genes %>%
                           select(enssscg_id, 
                                  enhanced_tissues = consensus_tissue_name)) %>%
               distinct(),
             by = c("enssscg_id", "consensus_tissue_name" = "enhanced_tissues")) %>%
  
  group_by(consensus_tissue_name, tissue_name, enssscg_id) %>% 
  summarise(tmm = mean(tmm)) %>%
  ungroup()  %>%
  group_by(enssscg_id, consensus_tissue_name) %>% 
  mutate(log_tmm = log10(tmm + 1),
         tmm2 = tmm / max(tmm, na.rm = T)) %>%
  ungroup() %>%
  filter(is.finite(tmm))

heatmap_data_organ <- 
  plot_data %>%
  filter(consensus_tissue_name == "kidney") %>% 
  select(-log_tmm, -tmm) %>%
  spread(tissue_name, tmm2) %>%
  select(-consensus_tissue_name) %>%
  column_to_rownames("enssscg_id") 

heatmap_meta_organ <- 
  rownames(heatmap_data_organ) %>% 
  enframe("name", "enssscg_id") %>%
  select(-1) %>%
  left_join(intraconsens_gene_enrich_overlap_all %>%
              filter(consensus_tissue_name == "kidney")) %>% 
  
  select(enssscg_id, overlap) %>% 
  mutate(overlap = ifelse(is.na(overlap), "elevated in tissue", overlap)) %>%
  left_join(intraconsens_clusters[["kidney"]]$gene_cluster_id %>%
              select(enssscg_id, 
                     `gene cluster` = cluster_id)) %>%
  
  
  select(1, 3, 2) %>%
  
  column_to_rownames("enssscg_id")


heatmap_settings <- 
  intraconsensus_cluster_settings %>%
  filter(consensus_tissue_name == "kidney")

cluster_pal <- 
  n_distinct(heatmap_meta_organ$`gene cluster`) %>%
  {set_names(viridis(.), LETTERS[1:.])}

heatmap_data_organ %>%
  pheatmap(clustering_method = "ward.D2", 
           color = heatmap_palette,
           border_color = NA,
           show_rownames = F, 
           annotation_row = heatmap_meta_organ,
           annotation_colors = list(overlap = variability_palette, 
                                    "gene cluster" =  cluster_pal),
           cutree_rows = heatmap_settings$cutree_genes,
           cutree_cols = 1,
           filename = savepath("Intraconsensus kidney heatmap.pdf"),
           width = 4.5,
           height = 7)
```


###Lymphoid

```{r}


plot_data <- 
  pig_atlas_sample_norm %>% 
  inner_join(tissue_mapping %>%
               filter(consensus_tissue_name %in% anova_data$consensus_tissue_name) ,
            by = c("tissue_ID" = "tissue")) %>% 
  
  
  inner_join(pig_gene_classification %>% 
               separate_rows(enhanced_tissues, sep = ",") %>%
               select(enssscg_id, enhanced_tissues) %>%
               bind_rows(intraconsens_relevant_genes %>%
                           select(enssscg_id, 
                                  enhanced_tissues = consensus_tissue_name)) %>%
               distinct(),
             by = c("enssscg_id", "consensus_tissue_name" = "enhanced_tissues")) %>%
  
  group_by(consensus_tissue_name, tissue_name, enssscg_id) %>% 
  summarise(tmm = mean(tmm)) %>%
  ungroup()  %>%
  group_by(enssscg_id, consensus_tissue_name) %>% 
  mutate(log_tmm = log10(tmm + 1),
         tmm2 = tmm / max(tmm, na.rm = T)) %>%
  ungroup() %>%
  filter(is.finite(tmm))

heatmap_data_organ <- 
  plot_data %>%
  filter(consensus_tissue_name == "lymphoid tissue") %>% 
  select(-log_tmm, -tmm) %>%
  spread(tissue_name, tmm2) %>%
  select(-consensus_tissue_name) %>%
  column_to_rownames("enssscg_id") 

heatmap_meta_organ <- 
  rownames(heatmap_data_organ) %>% 
  enframe("name", "enssscg_id") %>%
  select(-1) %>%
  left_join(intraconsens_gene_enrich_overlap_all %>%
              filter(consensus_tissue_name == "lymphoid tissue")) %>% 
  
  select(enssscg_id, overlap) %>% 
  mutate(overlap = ifelse(is.na(overlap), "elevated in tissue", overlap)) %>%
  left_join(intraconsens_clusters[["lymphoid tissue"]]$gene_cluster_id %>%
              select(enssscg_id, 
                     `gene cluster` = cluster_id)) %>%
  
  
  select(1, 3, 2) %>%
  
  column_to_rownames("enssscg_id")


heatmap_settings <- 
  intraconsensus_cluster_settings %>%
  filter(consensus_tissue_name == "lymphoid tissue")

cluster_pal <- 
  n_distinct(heatmap_meta_organ$`gene cluster`) %>%
  {set_names(viridis(.), LETTERS[1:.])}

heatmap_data_organ %>%
  pheatmap(clustering_method = "ward.D2", 
           color = heatmap_palette,
           border_color = NA,
           show_rownames = F, 
           annotation_row = heatmap_meta_organ,
           annotation_colors = list(overlap = variability_palette, 
                                    "gene cluster" =  cluster_pal),
           cutree_rows = heatmap_settings$cutree_genes,
           cutree_cols = 1,
           filename = savepath("Intraconsensus Lymphoid tissue heatmap.pdf"),
           width = 5,
           height = 7)



heatmap_meta_organ %>% 
  as_tibble(rownames = "enssscg_id") %>% 
  left_join(gene_orthologs) %>%
  left_join(human_gene_class_bloodlineage %>%
              filter(specificity_category %in% c("Tissue enriched"))) %>%
  
  
  separate_rows(enhanced_tissues, sep = ",") %>%
  filter(!is.na(enhanced_tissues)) %>%
  left_join(plot_data %>%
              filter(consensus_tissue_name == "lymphoid tissue") %>% 
              group_by(enssscg_id) %>% 
              summarise(tmm = sum(tmm))) %>%
  mutate(enhanced_tissues = ifelse(is.na(enhanced_tissues),
                                   "Not blood eriched",
                                   enhanced_tissues)) %>%
  group_by(`gene cluster`, enhanced_tissues) %>% 
  # count() %>%
  # top_n(10, ts_score) %>%
    summarise(n = sum(tmm)) %>%
  group_by(`gene cluster`) %>%
  mutate(fract = n / sum(n)) %>%
  ggplot(aes(1, fract, fill = enhanced_tissues)) +
  geom_col(color = "white",
           size = 0.1) + 
  scale_fill_manual(values = c("Not blood eriched" = "lightgray", tissue_colors_palette_full),
                    name = "Enriched cells") +
  facet_wrap(~`gene cluster`, ncol = 1, strip.position = "right") +
  coord_polar("y") + 
  stripped_theme_facet + 
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank(),
        axis.line = element_blank(),
        panel.border = element_blank())

ggsave(savepath("Intraconsensus lymph cluster pies.pdf"), width = 3, height = 7)


```


##Figure 4

###Gene UMAP

```{r}


if(file.exists("data/processed/gene_umap_res.Rdata")) {
  load("data/processed/gene_umap_res.Rdata")
  
} else {
  source("scripts/gene_umap.R", local = T)
  load("data/processed/gene_umap_res.Rdata")
}

pig_gene_umap %>% 
  select(enssscg_id, V1, V2, merged_cluster) %>% 
  left_join(gene_mapping) %>%
  select(-biotype) %>%
  left_join(human_gene_mapping %>% 
              select(gene_name, gene_description),
            by = c("display_name" = "gene_name")) %>%
  select(1, gene_name = display_name, gene_description, everything()) %>%
  # left_join(pig_atlas_tissue %>% 
  #             select(1, 2, nx) %>% 
  #             spread(tissue_name, nx)) %>%
  write_csv(savepath("UMAP clusters.csv"))


plot_data <- 
  pig_gene_umap %>%
  left_join(pig_gene_classification)

plot_data2 <- 
  pig_gene_umap %>%
  left_join(pig_gene_classification %>% 
              separate_rows(enhanced_tissues, sep = ";"))

####


plot_bins <- 120
plot_width <- 4
plot_height <- 4
plot <- 
  pig_gene_umap %>%
  ggplot(aes(V1, V2)) +
  geom_hex(aes(fill = stat(log10(count))),
           bins = plot_bins,
           show.legend = F) +
  scale_fill_viridis_c() +
  scale_color_viridis_c() +
  stripped_theme_facet 

ggsave(savepath("Gene UMAP hex.pdf"), plot = plot, width = plot_width, height = plot_height)

####
library(ggalt)

plot_data_hulls <- 
  plot_data %>% 
  
  select(merged_cluster, V1, V2) %>% 
  group_by(merged_cluster) %>% 
  do({
    st_as_sf(., coords=c('V1','V2')) %>%
      concaveman(concavity = 2, length_threshold = .1) %$%
      st_coordinates(polygons) %>% 
      as_tibble()
  })

plot <- 
  plot_data %>% 
  group_by(merged_cluster) %>% 
  mutate(mean_V1 = mean(V1),
         mean_V2 = mean(V2),
         dist = sqrt((V1 - mean_V1)^2 + (V2 - mean_V2)^2),
         n = n()) %>% 
  # select(V1, V2, dist, cluster, n) %>% 
  arrange(merged_cluster, dist) %>% 
  # mutate(cum_member = cumsum(1/n)) %>% 
  # filter(cum_member < 0.9) %>%
  # mutate(hull = row_number() %in% chull(V1, V2)) %>% 
  # filter(hull) %>%
  # mutate(hull = order(chull(V1, V2))) %>%
  # arrange(cluster, hull) %>%
  ggplot(aes(V1, V2)) +
  geom_hex(aes(fill = 1),
           data = plot_data,
           fill = "gray90",
           bins = plot_bins) +
  
  geom_polygon(data = plot_data_hulls,
               aes(X, Y, 
                   color = merged_cluster,
                   fill = merged_cluster),
               
               alpha = 0.2,
               size = 0.4, 
               # color = "white",
               show.legend = F) +
  geom_polygon(data = plot_data_hulls,
               aes(X, Y, 
                   color = merged_cluster),
               fill = NA,
               alpha = 0.7,
               size = 0.4, 
               # color = "white",
               show.legend = F) +
  geom_text(data = . %>% 
              select(merged_cluster, mean_V1, mean_V2, n) %>% 
              distinct(),
            aes(mean_V1, mean_V2, label = merged_cluster),
            size = 2,
            alpha = 0.6,
            show.legend = F) +
  stripped_theme

ggsave(savepath("Gene UMAP clusters.pdf"), plot = plot, width = plot_width, height = plot_height)


plot <- 
  pig_gene_umap %>%
  ggplot(aes(V1, V2)) +
  
  geom_hex(data = plot_data2 %>% 
             filter(!is.na(enhanced_tissues)) %>% 
             filter(specificity_category %in% c("Tissue enriched", "Group enriched")), 
           aes(fill = enhanced_tissues),
           alpha = 0.2,
           bins = plot_bins, 
           show.legend = F) +
  geom_text(data = plot_data2 %>% 
              filter(!is.na(enhanced_tissues)) %>% 
              group_by(scale, type, n_neighbors, enhanced_tissues) %>%
              summarise(V1 = median(V1), 
                        V2 = median(V2)),
            aes(label = enhanced_tissues),
            size = 1.5,
            alpha = 0.4) +
  scale_fill_manual(values = tissue_colors_palette_full) +
  # scale_fill_gradient(low = "gray", high = "black") +
  stripped_theme_facet

ggsave(savepath("Gene UMAP tissues.pdf"), plot = plot, width = plot_width, height = plot_height)

# plot <- 
#   plot_data2 %>%
#   filter(!is.na(enhanced_tissues)) %>% 
#   filter(specificity_category %in% c("Tissue enriched", "Group enriched")) %>%
#   group_by(enhanced_tissues) %>% 
#   mutate(mean_V1 = mean(V1),
#          mean_V2 = mean(V2),
#          dist = sqrt((V1 - mean_V1)^2 + (V2 - mean_V2)^2),
#          n = n()) %>% 
#   # select(V1, V2, dist, cluster, n) %>% 
#   arrange(cluster, dist) %>% 
#   mutate(cum_member = cumsum(1/n)) %>%
#   filter(cum_member < 0.3) %>%
#   mutate(hull = row_number() %in% chull(V1, V2)) %>%
#   filter(hull) %>%
#   mutate(hull = order(chull(V1, V2))) %>%
#   arrange(cluster, hull) %>%
#   ggplot(aes(V1, V2)) +
#   
#   # geom_hex(data = plot_data2 %>% 
#   #            filter(!is.na(enhanced_tissues)) %>% 
#   #            filter(specificity_category %in% c("Tissue enriched", "Group enriched")), 
#   #          aes(fill = enhanced_tissues),
#   #          alpha = 0.2,
#   #          bins = plot_bins, 
#   #          show.legend = F) +
#   geom_hex(aes(fill = 1),
#            data = plot_data,
#            fill = "gray90",
#            bins = plot_bins) +
#   geom_encircle(data = plot_data2 , 
#                 aes(fill = enhanced_tissues,
#                     color = enhanced_tissues),
#                 
#                 alpha = 0.1, 
#                 s_shape=1, 
#                 expand=0,
#                 # color = "white",
#                 show.legend = F) +
#   # geom_encircle(aes(color = cluster),
#   #               fill = NA,
#   #               alpha = 0.7, 
#   #               s_shape=1, 
#   #               expand=0,
#   #               # color = "white",
#   #               show.legend = F) +
#   geom_text(data = . %>% 
#               select(enhanced_tissues,
#                      mean_V1, 
#                      mean_V2),
#             aes(mean_V1, mean_V2, label = enhanced_tissues),
#             size = 1.5,
#             alpha = 0.4) +
#   scale_fill_manual(values = tissue_colors_palette_full) +
#   # scale_fill_gradient(low = "gray", high = "black") +
#   stripped_theme_facet
# 
# ggsave(savepath("Gene UMAP tissues 2.pdf"), plot = plot, width = plot_width, height = plot_height)



#####

plot <- 
  plot_data %>%
  mutate(specificity_category = factor(specificity_category, 
                                       c("Tissue enriched", 
                                         "Group enriched", 
                                         "Tissue enhanced", 
                                         "Low tissue specificity", 
                                         "Not detected"))) %>%
  ggplot(aes(V1, V2)) +
  geom_hex(data = . %>% 
             select(-specificity_category),
           fill = "lightgray",
           bins = plot_bins) +
  geom_hex(aes(alpha = stat(log10(count))),
           fill = "red",
           bins = plot_bins,
           show.legend = F) +
  facet_grid(~ specificity_category) +
  stripped_theme_facet

ggsave(savepath("Gene UMAP spec highlighted.pdf"), plot = plot, width = (plot_width * 4)/2, height = plot_height/2)


plot <- 
  plot_data %>%
  mutate(distribution_category = factor(distribution_category, 
                                       c("detected in all", 
                                         "detected in many", 
                                         "detected in some", 
                                         "detected in single", 
                                         "not detected"))) %>%
  ggplot(aes(V1, V2)) +
  geom_hex(data = . %>% 
             select(-distribution_category),
           fill = "lightgray",
           bins = plot_bins) +
  geom_hex(aes(alpha = stat(log10(count))),
           fill = "red",
           bins = plot_bins,
           show.legend = F) +
  facet_grid(~ distribution_category) +
  stripped_theme_facet

ggsave(savepath("Gene UMAP dist highlighted.pdf"), plot = plot, width = (plot_width * 4)/2, height = plot_height/2)

plot <- 
  plot_data %>%
  mutate(specificity_category = factor(specificity_category, 
                                       c("Tissue enriched", 
                                         "Group enriched", 
                                         "Tissue enhanced", 
                                         "Low tissue specificity", 
                                         "Not detected")),
         distribution_category = factor(distribution_category, 
                                       c("detected in single", 
                                         "detected in some", 
                                         "detected in many", 
                                         "detected in all", 
                                         "not detected"))) %>%
  ggplot(aes(V1, V2)) +
  geom_hex(data = . %>% 
             select(-specificity_category, -distribution_category),
           fill = "lightgray",
           bins = plot_bins) +
  geom_hex(aes(alpha = stat(log10(count))),
           fill = "red",
           bins = plot_bins,
           show.legend = F) +
  facet_grid(distribution_category~ specificity_category) +
  stripped_theme_facet + 
  theme(panel.spacing = unit(0, "mm"))

ggsave(savepath("Gene UMAP dist + spec highlighted.pdf"), plot = plot, 
       width = plot_width*1.5, 
       height = plot_height*1.5)


plots <- 
  lapply(sort(unique(pig_atlas_consensus$tissue)),
         function(tis_) {
          plot_data %>%
             
             ggplot(aes(V1, V2)) +
             geom_hex(fill = "lightgray",
                      bins = plot_bins) +
             geom_hex(data = plot_data2 %>%
                        filter(enhanced_tissues == tis_),
                      aes(alpha = stat(log10(count))),
                      fill = "red",
                      bins = plot_bins,
                      show.legend = F) +
             # facet_grid(scale + type ~n_neighbors) +
             ggtitle(tis_) +
             stripped_theme
         })
 
pdf(savepath("Gene UMAP tissue elevated highlighted.pdf"), width = plot_width, height = plot_height)
plots
dev.off()

# plot_data2 %>% 
#   group_by(merged_cluster, enhanced_tissues) %>% 
#   count() %>% 
#   ggplot(aes(n, merged_cluster, fill = enhanced_tissues)) +
#   geom_col(position = "fill") +
#   scale_fill_manual(values = tissue_colors_palette_full)


### Composite plot specificity and examples



plot1 <- 
  plot_data %>%
  filter(specificity_category != "Not detected") %>%
  mutate(specificity_category = factor(specificity_category, 
                                       c("Tissue enriched", 
                                         "Group enriched", 
                                         "Tissue enhanced", 
                                         "Low tissue specificity", 
                                         "Not detected"))) %>%
  ggplot(aes(V1, V2)) +
  geom_hex(data = . %>% 
             select(-specificity_category),
           fill = "lightgray",
           bins = plot_bins) +
  geom_hex(aes(alpha = stat(log10(count))),
           fill = "red",
           bins = plot_bins,
           show.legend = F) +
  facet_wrap(~ specificity_category, ncol = 2) +
  stripped_theme_facet

plot2 <- 
  plot_data %>%
  select(-enhanced_tissues) %>%
  
  
  ggplot(aes(V1, V2)) +
  geom_hex(fill = "lightgray",
           bins = plot_bins) +
  geom_hex(data = plot_data2 %>%
             filter(enhanced_tissues %in% 
                      c("brain",
                        "lung", 
                        "testis", 
                        "lymphoid tissue")),
           aes(alpha = stat(log10(count))),
           fill = "red",
           bins = plot_bins,
           show.legend = F) +
  facet_wrap(~ enhanced_tissues, ncol = 2) +
  stripped_theme_facet


plot <- 
  plot1 + 
  plot2
ggsave(savepath("Gene UMAP examples.pdf"), plot = plot, width = (plot_width * 4)/2, height = plot_height)





```


####hypergeometric test on clusters

```{r}

plot_data <- 
  pig_gene_cluster_enrich_hyper %>%
  mutate(enhanced_tissues = str_to_sentence(enhanced_tissues)) %>%
  filter(enhanced_tissues != "Not enriched")  %>% 
  mutate(log_p = case_when(log_p < 3 ~ 0,
                           log_p > 100 ~ 100,
                           T ~ log_p))

cluster_clus <- 
  plot_data %>%
  select(enhanced_tissues, merged_cluster, log_p) %>%
  spread(enhanced_tissues, log_p) %>% 
  column_to_rownames("merged_cluster") %>% 
  dist(method = "binary") %>% 
  hclust(method = "ward.D2")

tis_clus <- 
  plot_data %>%
  select(enhanced_tissues, merged_cluster, log_p) %>%
  spread(enhanced_tissues, log_p) %>% 
  column_to_rownames("merged_cluster") %>% 
  t() %>%
  dist(method = "binary") %>% 
  hclust(method = "ward.D2")


ggplot() + 
  theme_void() +
  
  #####
  ggdendrogram(cluster_clus, 
               labels = F, 
               leaf_labels = F) + 
  scale_x_continuous(expand = expansion(0, 0.5)) +
  theme_void() +
  theme(panel.spacing = unit(0, "mm"),
        plot.margin = unit(c(0,0,0,0), "mm")) +
  ####
  ggdendrogram(tis_clus, rotate = T) + 
  scale_y_reverse() +
  scale_x_continuous(expand = expansion(0, 0.5)) +
  theme_void() +
  theme(panel.spacing = unit(0, "mm"),
        plot.margin = unit(c(0,0,0,0), "mm")) +
  #####
  plot_data %>% 
  ungroup() %>%
  mutate(enhanced_tissues = factor(enhanced_tissues, labels(tis_clus)),
         merged_cluster = factor(merged_cluster, labels(cluster_clus))) %>%
  ggplot(aes(merged_cluster, enhanced_tissues, fill = log_p, size = log_p)) +
  geom_blank() +
  geom_point(data = . %>% 
               filter(adj_pval < 0.001),
             shape = 21, 
             alpha = 0.8,
             stroke = 0.2,
             color = "black") + 
  # scale_color_viridis(option = "E", direction = 1, 
  #                    name = "Adjusted p-value") + 
  scale_fill_gradientn(colors = ggsci::rgb_material(palette = "deep-orange")) +
  scale_size_continuous(range = c(1, 4)) +
  scale_y_discrete(position = "right") +
  stripped_theme +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.2),
        legend.position = "bottom", 
        panel.spacing = unit(0, "mm"),
        plot.margin = unit(c(0,0,0,0), "mm"),
        panel.grid.major = element_line(color = "gray90", size = 0.2),
        axis.title.y = element_blank()) +
  xlab("Cluster") +
  
  ####
  plot_layout(widths = c(0.09722222, 1), 
              heights = c(0.2, 1))

((0.2/1.2)*7)/12
ggsave(savepath("UMAP hyper heatmap elevated.pdf"), width = 12, height = 7)  

# ####
# cluster_clus <- 
#   plot_data %>%
#   select(enhanced_tissues, merged_cluster, log_p) %>%
#   spread(enhanced_tissues, log_p) %>% 
#   column_to_rownames("merged_cluster") %>% 
#   dist() %>% 
#   hclust(method = "ward.D2")
# 
# tis_clus <- 
#   plot_data %>%
#   select(enhanced_tissues, merged_cluster, log_p) %>%
#   spread(enhanced_tissues, log_p) %>% 
#   column_to_rownames("merged_cluster") %>% 
#   t() %>%
#   dist() %>% 
#   hclust(method = "ward.D2")
# 
# 
# ggplot() + 
#   theme_void() +
#   
#   #####
#   ggdendrogram(cluster_clus, 
#                labels = F, 
#                leaf_labels = F) + 
#   scale_x_continuous(expand = expansion(0, 0.5)) +
#   theme_void() +
#   theme(panel.spacing = unit(0, "mm"),
#         plot.margin = unit(c(0,0,0,0), "mm")) +
#   ####
#   ggdendrogram(tis_clus, rotate = T) + 
#   scale_y_reverse() +
#   scale_x_continuous(expand = expansion(0, 0.5)) +
#   theme_void() +
#   theme(panel.spacing = unit(0, "mm"),
#         plot.margin = unit(c(0,0,0,0), "mm")) +
#   #####
#   plot_data %>% 
#   ungroup() %>%
#   mutate(enhanced_tissues = factor(enhanced_tissues, labels(tis_clus)),
#          merged_cluster = factor(merged_cluster, labels(cluster_clus))) %>%
#   ggplot(aes(merged_cluster, enhanced_tissues, fill = log_p, size = log_p)) +
#   geom_blank() +
#   geom_point(data = . %>% 
#                filter(adj_pval < 0.001),
#              shape = 21, 
#              alpha = 0.8,
#              stroke = 0.2,
#              color = "black") + 
#   # scale_color_viridis(option = "E", direction = 1, 
#   #                    name = "Adjusted p-value") + 
#   scale_fill_gradientn(colors = ggsci::rgb_material(palette = "deep-orange")) +
#   scale_size_continuous(range = c(1, 4)) +
#   scale_y_discrete(position = "right") +
#   stripped_theme +
#   theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.2),
#         legend.position = "bottom", 
#         panel.spacing = unit(0, "mm"),
#         plot.margin = unit(c(0,0,0,0), "mm"),
#         panel.grid.major = element_line(color = "gray90", size = 0.2),
#         axis.title.y = element_blank()) +
#   xlab("Cluster") +
#   
#   ####
#   plot_layout(widths = c(0.09722222, 1), 
#               heights = c(0.2, 1))
# 
# ((0.2/1.2)*7)/12
# ggsave(savepath("UMAP hyper heatmap elevated 2.pdf"), width = 12, height = 7)  
# 
# ####
# pig_gene_cluster_enrich_hyper %>% 
#   filter(enhanced_tissues != "not enriched") %>%
#   select(enhanced_tissues, merged_cluster, adj_pval) %>%
#   filter(adj_pval < 0.000000001) %>% 
#   select(-adj_pval) %>%
#   left_join(., .,
#             by = "merged_cluster",
#             suffix = c("1", "2")) %>% 
#   filter(enhanced_tissues1 < enhanced_tissues2) %>%
#   select(enhanced_tissues1, enhanced_tissues2) %>%
#   distinct() %>%
#   as_tbl_graph() %>% 
#   ggraph(layout = "kk") +
#   geom_node_point() + 
#   geom_node_text(aes(label = name)) +
#   geom_edge_fan() +
#   theme_graph()

```


####Simple cluster metrics

```{r}
pig_gene_umap %>% 
  left_join(pig_gene_classification) %>%
  group_by(merged_cluster, specificity_category) %>% 
  count() %>% 
  ggplot(aes(n, merged_cluster, fill = specificity_category)) +
  geom_col() +
  ylab("Cluster") +
  stripped_theme_facet + 
  scale_x_continuous(expand = expansion(0)) +
  scale_fill_manual(values = gene_category_pal) +
  theme(panel.border = element_blank())
ggsave(savepath("Supplementary Number of genes per UMAP cluster.pdf"), width = 5, height = 8)

plot_data <- 
  pig_atlas_consensus %>% 
  left_join(pig_gene_umap) %>% 
  group_by(enssscg_id) %>%
  mutate(exp = nx / sum(nx)) %>%
  group_by(tissue, merged_cluster) %>% 
  summarise(mean_exp = mean(exp, na.rm = T))

plot_data2 <- 
  plot_data %>%
  group_by(merged_cluster) %>% 
  mutate(rank = rank(-mean_exp)) %>% 
  arrange(merged_cluster, rank) %>%
  mutate(cum_exp = cumsum(mean_exp),
         show = c(0, cum_exp[row_number() - 1]) < 0.5,
         tissue = ifelse(show, tissue, "other"),
         rank = ifelse(show, rank, max(rank))) %>%
  group_by(tissue, merged_cluster, rank) %>% 
  summarise(mean_exp = sum(mean_exp, na.rm = T)) %>% 
  arrange(merged_cluster, rank) %>%
  mutate(plot_id = paste(merged_cluster, rank)) %>%
  
  ungroup()  %>%
  mutate(plot_id = factor(plot_id, plot_id)) 
  
plot_pal <- 
  plot_data2 %>% 
  left_join(tissue_colors,
            by = c("tissue" = "consensus_tissue_name")) %>%
  mutate(consensus_tissue_color = ifelse(is.na(consensus_tissue_color),
                                         "gray", 
                                         consensus_tissue_color)) %$%
  set_names(consensus_tissue_color, plot_id)

plot_data2 %>%
  ggplot(aes(mean_exp, merged_cluster, fill = plot_id)) +
  geom_col(color = "white",
           show.legend = F) +
  geom_text(aes(label = tissue,
                size = mean_exp),
            position = position_stack(vjust = 0.5),
            color = "white",
            fontface = "bold",
            show.legend = F) +
  scale_fill_manual(values = plot_pal) +
  scale_size_continuous(range = c(1, 3)) + 
  scale_x_continuous(expand = expansion(0))+
  stripped_theme +
  xlab("Expression contribution") +
  ylab("Cluster")
ggsave(savepath("UMAP cluster tissue expression contribution.pdf"), width = 8, height = 10)


pig_atlas_consensus %>% 
  group_by(enssscg_id) %>%
  summarise(max_exp = max(ptpm, na.rm = T)) %>%
  left_join(pig_gene_umap) %>% 
  ggplot(aes(log10(max_exp), merged_cluster)) +
  geom_violin(draw_quantiles = 0.5,
              scale = "width",
              fill = "lightgray") + 
  scale_x_continuous(expand = expansion(0))+
  stripped_theme +
  xlab("Expression level (PTPM)") +
  ylab("Cluster")
ggsave(savepath("UMAP cluster expression level violin.pdf"), width = 8, height = 10)

```

#####Combined

```{r}
plot1 <- 
  pig_gene_umap %>% 
  left_join(pig_gene_classification) %>%
  group_by(merged_cluster, specificity_category) %>% 
  count() %>% 
  ungroup() %>%
  mutate(merged_cluster = factor(merged_cluster, labels(cluster_clus)),
         specificity_category = factor(specificity_category, spec_category_levels)) %>%
  ggplot(aes(n, merged_cluster, fill = specificity_category)) +
  geom_col() +
  geom_text(data = . %>% 
              group_by(merged_cluster) %>% 
              summarise(n = sum(n)),
            aes(n, merged_cluster, label = n),
            hjust = 0,
            vjust = 0.5, 
            size = 2,
            inherit.aes = F) +
  ylab("Cluster") +
  xlab("Number of genes") +
  stripped_theme_facet + 
  scale_x_continuous(expand = expansion(c(0, 0.075))) +
  scale_fill_manual(values = gene_category_pal) +
  theme(panel.border = element_blank(),
        legend.position = c(0.8, 0.8),
        panel.grid.major.y = element_line(color = "gray85"))


plot_data <- 
  pig_atlas_consensus %>% 
  left_join(pig_gene_umap) %>% 
  group_by(enssscg_id) %>%
  mutate(exp = nx / sum(nx)) %>%
  group_by(tissue, merged_cluster) %>% 
  summarise(mean_exp = mean(exp, na.rm = T))

plot_data2 <- 
  plot_data %>%
  group_by(merged_cluster) %>% 
  mutate(rank = rank(-mean_exp)) %>% 
  arrange(merged_cluster, rank) %>%
  mutate(cum_exp = cumsum(mean_exp),
         show = c(0, cum_exp[row_number() - 1]) < 0.7,
         tissue = ifelse(show, tissue, "other"),
         rank = ifelse(show, rank, max(rank))) %>%
  group_by(tissue, merged_cluster, rank) %>% 
  summarise(mean_exp = sum(mean_exp, na.rm = T)) %>% 
  arrange(merged_cluster, rank) %>%
  mutate(plot_id = paste(merged_cluster, rank)) %>%
  
  ungroup()  %>%
  mutate(plot_id = factor(plot_id, plot_id)) 
  
plot_pal <- 
  plot_data2 %>% 
  left_join(tissue_colors,
            by = c("tissue" = "consensus_tissue_name")) %>%
  mutate(consensus_tissue_color = ifelse(is.na(consensus_tissue_color),
                                         "gray", 
                                         consensus_tissue_color)) %$%
  set_names(consensus_tissue_color, plot_id)

plot2 <- 
  plot_data2 %>%
  mutate(merged_cluster = factor(merged_cluster, labels(cluster_clus)),
         label = tissue %>% 
           gsub("gland|glands", "gl.", .) %>%
           gsub("upper respiratory system", "upper resp.", .) %>%
           gsub("mesothelial tissue", "mesothelia", .) %>%
           gsub("lymphoid tissue", "lymph. tis.", .) %>%
           gsub("synovial tissue", "synov. tis.", .) %>%
           gsub("skeletal muscle", "sk. muscle", .) %>%
           gsub("small intestine", "sm. intest.", .) %>%
           gsub("large intestine", "lg. intest.", .) %>%
           gsub("adipose tissue", "adipose", .) %>%
           gsub("urinary bladder", "bladder", .),
         label = ifelse(mean_exp < 0.05, NA, label)) %>%
  ggplot(aes(mean_exp, merged_cluster, fill = plot_id)) +
  geom_col(color = "white",
           show.legend = F) +
  geom_text(aes(label = label,
                size = mean_exp),
            position = position_stack(vjust = 0.5),
            color = "white",
            fontface = "bold",
            show.legend = F) +
  scale_fill_manual(values = plot_pal) +
  scale_size_continuous(range = c(0.2, 3)) + 
  scale_x_continuous(expand = expansion(0))+
  stripped_theme +
  xlab("Mean expression contribution") +
  ylab("Cluster")
# plot2
plot3 <- 
  pig_atlas_consensus %>% 
  group_by(enssscg_id) %>%
  summarise(max_exp = max(ptpm, na.rm = T)) %>%
  left_join(pig_gene_umap) %>% 
  mutate(merged_cluster = factor(merged_cluster, labels(cluster_clus))) %>%
  ggplot(aes(log10(max_exp), merged_cluster)) +
  geom_violin(draw_quantiles = 0.5,
              scale = "width",
              fill = "lightgray") + 
  scale_x_continuous(expand = expansion(0))+
  scale_y_discrete(position = "right") +
  stripped_theme +
  xlab("Expression level (PTPM)") +
  ylab("Cluster")

full_plot <- 
  ggdendrogram(cluster_clus, rotate = T) +
  scale_y_reverse() + 
  scale_x_continuous(expand = expansion(0, 0.5)) +
  theme_void() +
  plot1 + 
  theme(plot.margin = unit(c(0,0,0,0), "mm")) +
  plot2 + 
  theme(axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        plot.margin = unit(c(0,0,0,0), "mm")) +
  plot3 +
    theme(plot.margin = unit(c(0,0,0,0), "mm"),
          panel.grid.major.y = element_line(color = "gray85")) +
  
  plot_layout(nrow = 1,
              widths = c(0.1, 0.3, 0.4, 0.2))

ggsave(savepath("Supplementary UMAP clusters.pdf"), plot = full_plot,
       width = 10, height = 10)
```


####Intraconsensus on UMAP

```{r}




# plot_data <- 
#   pig_gene_umap %>%
#   left_join(pig_gene_classification) %>%
#   left_join(intraconsens_classification)

# plot_data2 <- 
#   pig_gene_umap %>%
#   left_join(pig_gene_classification %>% 
#               separate_rows(enhanced_tissues, sep = ","))

####
plot_data <- 
  pig_gene_umap %>% 
  inner_join(pig_gene_classification %>%
               filter(specificity_category == "Low tissue specificity") %>% 
               filter(distribution_category == "detected in all") %>% 
               filter(!enssscg_id %in% intraconsens_classification$enssscg_id))

plot <- 
  pig_gene_umap %>%
  ggplot(aes(V1, V2)) +
  geom_hex(aes(fill = 1),
           fill = "gray90",
           bins = plot_bins) +
  geom_hex(data = plot_data,
           aes(alpha = stat(log10(count))),
           fill = "red",
           bins = plot_bins,
           show.legend = F) +
  scale_fill_viridis_c() +
  scale_color_viridis_c() +
  stripped_theme_facet 

ggsave(savepath("Gene UMAP nonvariable genes.pdf"), plot = plot, width = plot_width, height = plot_height)



plot <- 
  pig_gene_umap %>%
  ggplot(aes(V1, V2)) +
  geom_hex(aes(fill = 1),
           fill = "gray90",
           bins = plot_bins) +
  geom_hex(data = pig_gene_umap %>% 
             inner_join(intraconsens_classification),
           aes(alpha = stat(log10(count))),
           fill = "red",
           bins = plot_bins,
           show.legend = F) +
  scale_fill_viridis_c() +
  scale_color_viridis_c() +
  facet_wrap(~ consensus_tissue_name) +
  stripped_theme_facet 

ggsave(savepath("Gene UMAP variable genes.pdf"), plot = plot, width = plot_width * 3, height = plot_height * 3)

```



##Figure 5
### Pig-Human classification overlap


```{r compare_classification_data_gen, warning=F, message=F, results=F}
# ----- comparison classification -----

if(file.exists("data/processed/gene_overlap_comparison.Rdata")) {
  load("data/processed/gene_overlap_comparison.Rdata")
  
} else {
  
  human_pig_gene_class_comparison_long_all_orth <- 
    full_join(gene_orthologs_all %>% 
                left_join(human_gene_comparison_classification %>%
                            separate_rows(enhanced_tissues, sep = ";") %>% 
                            select(ensg_id,
                                   enriched_tissues = enhanced_tissues, 
                                   spec_score = ts_score, 
                                   spec_category = specificity_category,
                                   dist_category = distribution_category) %>%
                            mutate(enriched_tissues = ifelse(is.na(enriched_tissues), "", 
                                                                   enriched_tissues)), 
                          by = "ensg_id") %>% 
                mutate(comparison_id = enriched_tissues),
              
              
              gene_orthologs_all %>% 
                left_join(pig_gene_comparison_classification %>%
                            separate_rows(enhanced_tissues, sep = ";") %>% 
                            select(enssscg_id,
                                   enriched_tissues = enhanced_tissues, 
                                   spec_score = ts_score, 
                                   spec_category = specificity_category,
                                   dist_category = distribution_category) %>%
                            mutate(enriched_tissues = ifelse(is.na(enriched_tissues), "", 
                                                                   enriched_tissues)), 
                          by = "enssscg_id") %>% 
                mutate(comparison_id = enriched_tissues),
              by = c("enssscg_id", "ensg_id", "ortholog_type", "comparison_id"), 
              suffix = c("_human", "_pig")) %>% 
    select(1, 2, 3, enriched_tissues_human, enriched_tissues_pig, comparison_id, everything()) 
  
  
  human_pig_gene_class_comparison_long <- 
    full_join(gene_orthologs %>% 
                left_join(human_gene_comparison_classification %>%
                            separate_rows(enhanced_tissues, sep = ";") %>% 
                            select(ensg_id,
                                   enriched_tissues = enhanced_tissues, 
                                   spec_score = ts_score, 
                                   spec_category = specificity_category,
                                   dist_category = distribution_category) %>%
                            mutate(enriched_tissues = ifelse(is.na(enriched_tissues), "", 
                                                                   enriched_tissues)), 
                          by = "ensg_id") %>% 
                mutate(comparison_id = enriched_tissues),
              
              
              gene_orthologs %>% 
                left_join(pig_gene_comparison_classification %>%
                            separate_rows(enhanced_tissues, sep = ";") %>% 
                            select(enssscg_id,
                                   enriched_tissues = enhanced_tissues, 
                                   spec_score = ts_score, 
                                   spec_category = specificity_category,
                                   dist_category = distribution_category) %>%
                            mutate(enriched_tissues = ifelse(is.na(enriched_tissues), "", 
                                                                   enriched_tissues)), 
                          by = "enssscg_id") %>% 
                mutate(comparison_id = enriched_tissues),
              by = c("enssscg_id", "ensg_id", "ortholog_type", "comparison_id"), 
              suffix = c("_human", "_pig")) %>% 
    select(1, 2, 3, enriched_tissues_human, enriched_tissues_pig, comparison_id, everything()) 
  
  # human_pig_gene_class_comparison_long %>% 
  #   filter(grepl("heart|intestine", comparison_id)) %>% View
  
  
  
  
  human_pig_gene_class_comparison <- 
    human_pig_gene_class_comparison_long %>% 
    group_by(enssscg_id, ensg_id, ortholog_type) %>% 
    summarise(n_enriched_human = n_distinct(enriched_tissues_human, na.rm = T),
              n_enriched_pig = n_distinct(enriched_tissues_pig, na.rm = T),
              
              
              # Tissues count as overlapping if enriched in both human and pig, and the tissue is common in both datasets
              n_enriched_overlap = length(which((!is.na(enriched_tissues_human) & !is.na(enriched_tissues_pig)))),
              n_enriched_total = n_enriched_human + n_enriched_pig - n_enriched_overlap,
              
              enriched_tissues_overlap = paste(comparison_id[which(!is.na(enriched_tissues_human) & !is.na(enriched_tissues_pig))], 
                                               collapse = ";"),
              enriched_tissues_human = paste(enriched_tissues_human[which(!is.na(enriched_tissues_human))], 
                                             collapse = ";"),
              enriched_tissues_pig = paste(enriched_tissues_pig[which(!is.na(enriched_tissues_pig))], 
                                           collapse = ";"),
              
              enrichment_overlap = case_when(n_enriched_overlap == 0 ~ "no overlap", #n_enriched_total == 0 ~ "no enrichment", 
                                             n_enriched_overlap == n_enriched_total ~ "full overlap",
                                             n_enriched_overlap > 0 ~ "partial overlap")
    ) %>% 
    left_join(human_gene_comparison_classification %>%
                select(ensg_id,
                       dist_category = distribution_category,
                       spec_category = specificity_category,
                       spec_score = ts_score),
              by = "ensg_id") %>% 
    left_join(pig_gene_comparison_classification %>%
                select(enssscg_id,
                       dist_category = distribution_category,
                       spec_category = specificity_category,
                       spec_score = ts_score),
              by = "enssscg_id",
              suffix = c("_human", "_pig")) %>%
    
    mutate(dist_category_human_n = match(dist_category_human, dist_category_levels), 
           spec_category_human_n = match(spec_category_human, spec_category_levels), 
           dist_category_pig_n = match(dist_category_pig, dist_category_levels), 
           spec_category_pig_n = match(spec_category_pig, spec_category_levels), 
           dist_cat_diff = dist_category_human_n - dist_category_pig_n,
           spec_cat_diff = spec_category_human_n - spec_category_pig_n,
           shared_dist_category = case_when(abs(dist_cat_diff) == 0 ~ "shared",
                                            abs(dist_cat_diff) == 1 ~ "minor difference",
                                            abs(dist_cat_diff) == 2 ~ "medium difference",
                                            abs(dist_cat_diff) >= 3 ~ "major difference"),
           shared_spec_category = case_when(abs(spec_cat_diff) == 0 ~ "shared",
                                            abs(spec_cat_diff) == 1 ~ "minor difference",
                                            abs(spec_cat_diff) == 2 ~ "medium difference",
                                            abs(spec_cat_diff) >= 3 ~ "major difference")) %>% 
    ungroup()
  
  # # human_pig_gene_class_comparison_long %>% filter(grepl("heart|intestine", comparison_id))
  # # 
  # # human_pig_gene_class_comparison %>%
  # #   filter(grepl("heart|intestine", enriched_tissues_pig) &
  # #            grepl("heart|intestine", enriched_tissues_human)) %>%
  # #   View
  # # 
  # # human_pig_gene_class_comparison %>%
  # #   filter(enrichment_overlap == "no overlap" &
  # #            enriched_tissues_human == enriched_tissues_pig &
  # #            enriched_tissues_human != "") %>%
  # #   View
  # 
  # # Dist
  # 
  # human_pig_gene_dist_class_comparison_long <- 
  #   full_join(gene_orthologs %>% 
  #               left_join(human_gene_class_comparison %>%
  #                           separate_rows(tissues_detected, sep = ";") %>% 
  #                           select(ensg_id = gene,
  #                                  tissues_detected), 
  #                         by = "ensg_id") %>% 
  #               mutate(comparison_id = tissues_detected),
  #             
  #             
  #             gene_orthologs %>% 
  #               left_join(pig_gene_class_comparison %>%
  #                           separate_rows(tissues_detected, sep = ";") %>% 
  #                           select(enssscg_id = gene,
  #                                  tissues_detected), 
  #                         by = "enssscg_id") %>% 
  #               mutate(comparison_id = tissues_detected),
  #             by = c("enssscg_id", "ensg_id", "ortholog_type", "comparison_id"), 
  #             suffix = c("_human", "_pig")) %>% 
  #   select(1, 2, 3, tissues_detected_human, tissues_detected_pig, comparison_id) 
  # 
  # 
  # 
  # human_pig_gene_dist_class_comparison <- 
  #   human_pig_gene_dist_class_comparison_long %>% 
  #   group_by(enssscg_id, ensg_id, ortholog_type) %>% 
  #   summarise(n_detected_human = n_distinct(tissues_detected_human, na.rm = T),
  #             n_detected_pig = n_distinct(tissues_detected_pig, na.rm = T),
  #             
  #             n_overlap_detected_human = n_distinct(tissues_detected_human[which(comparison_id %in%
  #                                                                                  consensus_tissue_overlap$comparison_id)], na.rm = T),
  #             n_overlap_detected_pig = n_distinct(tissues_detected_pig[which(comparison_id %in%
  #                                                                              consensus_tissue_overlap$comparison_id)], na.rm = T),
  #             
  #             
  #             # Tissues count as overlapping if detected in both human and pig, and the tissue is common in both datasets
  #             n_detected_overlap = length(which((!is.na(tissues_detected_human) & !is.na(tissues_detected_pig)) &
  #                                                 comparison_id %in% consensus_tissue_overlap$comparison_id)),
  #             n_detected_total = n_overlap_detected_human + n_overlap_detected_pig - n_detected_overlap,
  #             
  #             detected_tissues_overlap = paste(comparison_id[which(!is.na(tissues_detected_human) & !is.na(tissues_detected_pig))], 
  #                                              collapse = ";"),
  #             tissues_detected_human = paste(tissues_detected_human[which(!is.na(tissues_detected_human))], 
  #                                            collapse = ";"),
  #             tissues_detected_pig = paste(tissues_detected_pig[which(!is.na(tissues_detected_pig))], 
  #                                          collapse = ";"),
  #             
  #             detection_overlap_n = n_detected_overlap/n_detected_total
  #   ) 
  
  gene_overlap_data <- 
    human_pig_gene_class_comparison_long %>% 
    left_join(human_gene_comparison_classification %>%
                select(1:2),
              by = c("ensg_id")) %>%
    left_join(pig_gene_comparison_classification %>%
                select(1:2),
              by = c("enssscg_id"), 
              suffix = c("_human", "_pig")) %>%
    
    group_by(comparison_id, enssscg_id, ensg_id, 
             spec_score_human, spec_score_pig, 
             spec_category_human, spec_category_pig) %>%
    summarise(type = case_when(any(!is.na(enriched_tissues_human) & 
                                     !is.na(enriched_tissues_pig)) |
                                 all(is.na(enriched_tissues_human) & 
                                     is.na(enriched_tissues_pig)) ~ "Overlap",
                               all(is.na(enriched_tissues_human)) ~ "Pig", 
                               all(is.na(enriched_tissues_pig)) ~ "Human")) %>%
    ungroup() %>%
    arrange(enssscg_id, ensg_id, comparison_id)
  
  gene_overlap_data_summarized <- 
    gene_overlap_data %>%
    left_join(gene_orthologs_all) %>%
    group_by(enssscg_id,
             ensg_id, 
             gene_name = human_gene_name) %>%
    summarise(type = case_when(any(type == "Overlap") ~ "Overlap",
                               any(type == "Pig") & 
                                 any(type == "Human") &
                                 (!any(spec_category_human %in% c("Low tissue specificity", 
                                                                  "Not detected")) &
                                    !any(spec_category_pig %in% c("Low tissue specificity", 
                                                                  "Not detected"))) ~ "Different tissues",
                               any(spec_category_human %in% c("Low tissue specificity", 
                                                              "Not detected")) ~ "Pig",
                               any(spec_category_pig %in% c("Low tissue specificity", 
                                                            "Not detected")) ~ "Human")) %>%
    ungroup()
  
  gene_overlap_data_all_orth <- 
    human_pig_gene_class_comparison_long_all_orth %>% 
    left_join(human_gene_comparison_classification %>%
                select(1:2),
              by = c("ensg_id")) %>%
    left_join(pig_gene_comparison_classification %>%
                select(1:2),
              by = c("enssscg_id"), 
              suffix = c("_human", "_pig")) %>%
    
    group_by(comparison_id, enssscg_id, ensg_id, 
             spec_score_human, spec_score_pig, 
             spec_category_human, spec_category_pig) %>%
    summarise(type = case_when(any(!is.na(enriched_tissues_human) & 
                                     !is.na(enriched_tissues_pig)) |
                                 all(is.na(enriched_tissues_human) & 
                                     is.na(enriched_tissues_pig)) ~ "Overlap",
                               all(is.na(enriched_tissues_human)) ~ "Pig", 
                               all(is.na(enriched_tissues_pig)) ~ "Human")) %>%
    ungroup() %>%
    arrange(enssscg_id, ensg_id, comparison_id)
  
  gene_overlap_data_all_orth_summarized <- 
    gene_overlap_data_all_orth %>%
    left_join(gene_orthologs_all) %>%
    group_by(enssscg_id,
             ensg_id, 
             gene_name = human_gene_name) %>%
    summarise(type = case_when(any(type == "Overlap") ~ "Overlap",
                               any(type == "Pig") & 
                                 any(type == "Human") &
                                 (!any(spec_category_human %in% c("Low tissue specificity", 
                                                                  "Not detected")) &
                                    !any(spec_category_pig %in% c("Low tissue specificity", 
                                                                  "Not detected"))) ~ "Different tissues",
                               any(spec_category_human %in% c("Low tissue specificity", 
                                                              "Not detected")) ~ "Pig",
                               any(spec_category_pig %in% c("Low tissue specificity", 
                                                            "Not detected")) ~ "Human")) %>%
    ungroup()
  
  
  
  
  save(list = c("human_pig_gene_class_comparison_long",
                "human_pig_gene_class_comparison", 
                "gene_overlap_data",
                "gene_overlap_data_summarized",
                "gene_overlap_data_all_orth",
                "gene_overlap_data_all_orth_summarized"),
       file = "data/processed/gene_overlap_comparison.Rdata")
  
  human_pig_gene_class_comparison %>% 
    select(1:3, 
           spec_category_human, spec_category_pig,
           dist_category_human, dist_category_pig,
           contains("enriched_tissues")) %>% 
    left_join(gene_overlap_data_summarized) %>% 
    select(1:3, gene_name, overlap_type = type, everything()) %>%
    write_csv(savepath("Human Pig classification overlap.csv"))
  # write_csv(gene_overlap_data_summarized, savepath("Human Pig classification overlap summary.csv"))
  
  # human_pig_gene_class_comparison %>% 
  #   select(1:3, 
  #          spec_category_human, spec_category_pig,
  #          dist_category_human, dist_category_pig,
  #          contains("enriched_tissues")) %>% 
  #   left_join(gene_overlap_data_summarized) %>% 
  #   select(1:3, gene_name, overlap_type = type, everything()) %>% 
  #   filter(overlap_type == "Different tissues") %>% 
  #   filter(spec_category_human == "Tissue enriched" |
  #            spec_category_pig == "Tissue enriched") %>%
  #   write_csv(savepath("Human Pig classification overlap - enriched different.csv"))
}




```

####Compare overlap rate for ortholog types

```{r}

gene_overlap_data_summarized %>%
  left_join(gene_orthologs) %>%
  group_by(ortholog_type, ortholog_confidence, type) %>% 
  count() %>% 
  ggplot(aes(n, paste(gsub("ortholog_", "", ortholog_type), 
                      c("low conf", "high conf")[ortholog_confidence + 1]), fill = type)) +
  geom_col(position = "fill") +
  geom_text(data = . %>% 
              group_by(ortholog_type, ortholog_confidence) %>%
              summarise(n = sum(n)),
            aes(0.5, paste(gsub("ortholog_", "", ortholog_type), 
                      c("low conf", "high conf")[ortholog_confidence + 1]), 
                label = n),
            color = "white",
            inherit.aes = F) +
  scale_fill_manual(values = overlap_type_pal) +
  stripped_theme +
  theme(axis.title = element_blank())
ggsave(savepath("gene overlap per ortholog 1.pdf"), width = 5, height = 2)

gene_overlap_data_summarized %>%
  left_join(gene_orthologs) %>%
  left_join(pig_gene_comparison_classification) %>%
  group_by(ortholog_type, ortholog_confidence, type, specificity_category) %>% 
  count() %>% 
  ggplot(aes(n, paste(gsub("ortholog_", "", ortholog_type), 
                      c("low conf", "high conf")[ortholog_confidence + 1]), fill = type)) +
  geom_col(position = "fill") +
  geom_text(data = . %>% 
              group_by(ortholog_type, ortholog_confidence, specificity_category) %>%
              summarise(n = sum(n)),
            aes(0.5, paste(gsub("ortholog_", "", ortholog_type), 
                      c("low conf", "high conf")[ortholog_confidence + 1]), 
                label = n),
            color = "white",
            inherit.aes = F) +
  scale_fill_manual(values = overlap_type_pal) +
  facet_wrap(~specificity_category, nrow = 1) +
  stripped_theme_facet +
  theme(axis.title = element_blank())
ggsave(savepath("gene overlap per ortholog 2.pdf"), width = 10, height = 2)

```


#### Venn plot

```{r elevation_overlap_final_venn_plot, echo=FALSE, message=FALSE, warning=FALSE}




plot_data <- 
  gene_overlap_data %>% 
  filter(spec_category_human == "Tissue enriched" |
           spec_category_pig == "Tissue enriched") %>%
  group_by(comparison_id) %>% 
  mutate(n = n()) %>% 
  ungroup() %>%
  mutate(show = n > 30,
         comparison_id = ifelse(show, comparison_id, "Other tissues"))

plot_order <- 
  plot_data %>%
  select(comparison_id, n, show) %>% 
  unique() %>%
  arrange(!show, -n) %$% 
  unique(comparison_id) %>% 
  str_to_sentence() %>% 
  rev()

# Venn

plot_data %>% 
  mutate(comparison_id = factor(str_to_sentence(comparison_id)),
         type = factor(type)) %>%
  group_by(comparison_id, type, .drop = F) %>%
  summarise(n = n()) %>%
  
  ungroup() %>%
  
  mutate(comparison_id = factor(comparison_id, 
                                levels = plot_order)) %>%
  ggplot() + 
  geom_col(mapping = aes(comparison_id, n,fill = type)) + 
  
  geom_text(data = . %>%
              filter(type == "Pig"),
            aes(x = comparison_id, y = 0,
                label = n),
            hjust = 1) +
  geom_text(data = . %>%
              group_by(comparison_id, .drop = F) %>%
              summarise(y = n[which(type == "Overlap")] / 2 +
                          n[which(type == "Pig")],
                        label = n[which(type == "Overlap")]),
            aes(x = comparison_id, y = y,
                label = label),
            hjust = 0.5,
            color = "white") +
  geom_text(data = . %>%
              group_by(comparison_id, .drop = F) %>%
              summarise(y = sum(n),
                        label = n[which(type == "Human")]),
            aes(x = comparison_id, y = y,
                label = label),
            hjust = 0) +
  
  coord_flip() + 
  
  scale_fill_manual(values = c("Human" = "#DB1F48", "Overlap" = "#004369", "Pig" = "#01949A")) +
  scale_y_continuous(expand = expand_scale(0.13)) +
  
  theme(axis.text.y = element_text(face = "bold"), 
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(), 
        axis.title = element_blank(),
        panel.background = element_blank(),
        
        legend.position = c(0.5, 0.5)) + 
  
  ggtitle("Overlap of tissue enriched genes") +
  theme(axis.ticks = element_blank())
  
ggsave(savepath("Overlap enrichment per tissue.pdf"), 
       height = 4, width = 4, useDingbats = F)

#####

plot_data <- 
  gene_overlap_data_summarized %>% 
  left_join(human_pig_gene_class_comparison) 

plot_data <- 
  tibble(spec_cat = c("Tissue enriched","Group enriched","Tissue enhanced",
  "Low tissue specificity","Not detected")) %>% 
  group_by(spec_cat) %>% 
  do({
    spec_cat_ <- .$spec_cat
    plot_data %>%
      filter(spec_category_human == spec_cat_ |
               spec_category_pig == spec_cat_) %>% 
      group_by(type) %>% 
      count()
  }) %>%
  ungroup() 

plot_data %>%
  mutate(spec_cat = factor(spec_cat, rev(spec_category_levels))) %>%
  ggplot(aes(n, spec_cat, fill = type)) +
  geom_col() +
  scale_fill_manual(values = overlap_type_pal) +
  scale_x_continuous(expand = expansion(c(0, 0))) +
  stripped_theme_facet +
  xlab("Number of genes") +
  ggtitle("Overlap of enrichment",
          "Per category in either species") +
  theme(axis.line = element_blank(),
        axis.title.y = element_blank(),
        panel.border = element_blank(),
        legend.position = "top") 
ggsave(savepath("Overlap summary barplot.pdf"), width = 6, height = 3)
  
plot_data%>%
  mutate(spec_cat = factor(spec_cat, (spec_category_levels)))  %>%
  group_by(spec_cat) %>% 
  mutate(percent = 100 * n / sum(n)) %>%
  ggplot(aes(1, percent, fill = type)) +
  geom_col(show.legend = F) +
  geom_text(aes(1.2, label = paste(round(percent, 1), "%")),
            color = "white",
            position = position_stack(vjust = 0.5),
            size = 1) +
  scale_fill_manual(values = overlap_type_pal) +
  scale_x_continuous(expand = expansion(c(0, 0))) +
  facet_wrap(~spec_cat, ncol = 1, strip.position = "right") +
  coord_polar("y") +
  stripped_theme_facet +
  theme(axis.line = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank(),
        axis.text = element_blank(),
        panel.border = element_blank(),
        legend.position = "top") 
ggsave(savepath("Overlap summary pie chart.pdf"), width = 1, height = 3)

  





```

#### Pie summary

```{r}
gene_overlap_data %>%
  filter(spec_category_human == "Tissue enriched" |
           spec_category_pig == "Tissue enriched") %>%
  group_by(enssscg_id,
           ensg_id) %>%
  summarise(type = case_when(any(type == "Overlap") ~ "Overlap",
                             any(type == "Pig") & 
                               any(type == "Human") ~ "Different tissues",
                             any(type == "Pig") ~ "Pig",
                             any(type == "Human") ~ "Human")) %>%
      group_by(type, .drop = F) %>%
      summarise(n = n()) %>%
      
      ungroup() %>%
  mutate(type = factor(type, c("Human", "Pig", "Different tissues", "Overlap"))) %>%
      pie_plot(class_col = "type",
               y_col = "n",
               pal = c("Human" = "#DB1F48", "Overlap" = "#004369", "Pig" = "#01949A", 
                       "Different tissues" = "orangered"),
               dodge = 0.1, 
               prefix = "n = ")
ggsave(savepath("enrichment overlap piechart summary.pdf"), width= 4, height = 4)   
```

####Hypergeometric test

```{r all_elevated}


overlap_hyper_all <- 
  classification_hypergeom_test(class1 = pig_gene_comparison_classification,
                                class2 = human_gene_comparison_classification,
                                gene_col1 = "enssscg_id",
                                gene_col2 = "ensg_id",
                                orthologs = gene_orthologs %>% 
                                  select(1:2)) %>%
  
  mutate(p_value = phyper(q - 1, m, n, k, lower.tail = F)) %>%
  mutate(p_value = ifelse(p_value == 0, .Machine$double.xmin, p_value),
         adj_pval = p.adjust(p_value, method = "BH")) %>%
  rename(pig_id = id1, 
         human_id = id2)

overlap_hyper_enriched <- 
  classification_hypergeom_test(pig_gene_comparison_classification %>%
                                  mutate(enhanced_tissues = ifelse(specificity_category == "Tissue enhanced",
                                                                   NA, 
                                                                   enhanced_tissues)),
                                human_gene_comparison_classification %>%
                                  mutate(enhanced_tissues = ifelse(specificity_category == "Tissue enhanced",
                                                                   NA, 
                                                                   enhanced_tissues)),
                                gene_col1 = "enssscg_id",
                                gene_col2 = "ensg_id",
                                orthologs = gene_orthologs %>% 
                                  select(1:2)) %>%
  
  mutate(p_value = phyper(q - 1, m, n, k, lower.tail = F)) %>%
  mutate(p_value = ifelse(p_value == 0, .Machine$double.xmin, p_value),
         adj_pval = p.adjust(p_value, method = "BH")) %>%
  rename(pig_id = id1, 
         human_id = id2)


plot_order <- 
  overlap_hyper_all %>% 
  ungroup() %>%
  mutate(pig_id = str_to_sentence(pig_id),
         human_id = str_to_sentence(human_id)) %>%
  filter(pig_id == human_id) %>%
  arrange(adj_pval) %>%
  pull(pig_id)

overlap_hyper_all %>% 
  group_by(pig_id, human_id) %>%
  mutate(capped_p = min(c(-log10(adj_pval), 20))) %>%
  ungroup() %>% 
  mutate(pig_id = str_to_sentence(pig_id),
         human_id = str_to_sentence(human_id)) %>%
  mutate(pig_id = factor(pig_id, plot_order),
         human_id = factor(human_id, plot_order)) %>%
  ggplot(aes(human_id, pig_id, fill = capped_p)) +
  geom_tile() + 
  geom_tile(data = . %>% 
              filter(adj_pval >= 0.05),
            fill = "black") + 
  scale_fill_viridis(option = "D", direction = 1, 
                     name = "Adjusted p-value") + 
  stripped_theme +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.2),
        legend.position = "top") +
  xlab("Human") + 
  ylab("Pig")

ggsave(savepath("Overlap hyper heatmap elevated.pdf"), width = 6, height = 6)  

overlap_hyper_all %>% 
  filter(human_id != "not enriched" & 
           pig_id != "not enriched") %>%
  ungroup() %>%
  mutate(pig_id = str_to_sentence(pig_id),
         human_id = str_to_sentence(human_id)) %>%
  filter(adj_pval < 0.05) %>%
  mutate(pig_id = factor(pig_id, plot_order),
         human_id = factor(human_id, plot_order),
         adj_pval = case_when(adj_pval < 1e-100 ~ 1e-100,
                              T ~ adj_pval)) %>%
  ggplot(aes(human_id, pig_id, fill = -log10(adj_pval), size = -log10(adj_pval))) +
  geom_point(shape = 21, 
             alpha = 0.8,
             stroke = 0.2,
             color = "black") + 
  # scale_color_viridis(option = "E", direction = 1, 
  #                    name = "Adjusted p-value") + 
  scale_fill_gradientn(colors = ggsci::rgb_material(palette = "deep-orange")) +
  scale_size_continuous(range = c(1, 6)) +
  stripped_theme +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.2),
        legend.position = "top", 
        panel.grid.major = element_line(color = "gray90", size = 0.2)) +
  xlab("Human") + 
  ylab("Pig")

ggsave(savepath("Overlap hyper heatmap elevated 2.pdf"), width = 5, height = 5.5)  

# 
# plot_order <- 
#   overlap_hyper_enriched %>% 
#   filter(pig_id == human_id) %>%
#   arrange(adj_pval) %>%
#   pull(pig_id)
# 
# overlap_hyper_enriched %>% 
#   group_by(pig_id, human_id) %>%
#   mutate(capped_p = min(c(-log10(adj_pval), 20))) %>%
#   ungroup() %>% 
#   mutate(pig_id = factor(pig_id, plot_order),
#          human_id = factor(human_id, plot_order)) %>%
#   ggplot(aes(human_id, pig_id, fill = capped_p)) +
#   geom_tile() + 
#   geom_tile(data = . %>% 
#               filter(adj_pval >= 0.05),
#             fill = "black") + 
#   scale_fill_viridis(option = "D", direction = 1,
#                      name = "Adjusted p-value") + 
#   stripped_theme +
#   theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust = 0.2),
#         legend.position = "top") +
#   xlab("Human") + 
#   ylab("Pig")
# 
# ggsave(savepath("Overlap hyper heatmap enriched.pdf"), width = 6, height = 6)  
```

#### Alluvial plots




```{r compare_multi_alluvial, echo=F, warning = F, message = F, results=F}

human_pig_gene_class_comparison %>%
  ungroup() %>%
  mutate(spec_category_human = str_to_sentence(paste(spec_category_human,
                                                     case_when(enrichment_overlap %in% 
                                                                 c("full overlap", 
                                                                   "partial overlap") |
                                                                 spec_category_human %in%
                                                                 c("low tissue specificity", 
                                                                   "not detected") ~ 
                                                                 "overlap",
                                                               enrichment_overlap == "no overlap" ~
                                                                 "no overlap"))),
         
         spec_category_pig = str_to_sentence(paste(spec_category_pig,
                                                   case_when(enrichment_overlap %in% 
                                                               c("full overlap", 
                                                                 "partial overlap") |
                                                                 spec_category_pig %in%
                                                                 c("low tissue specificity", 
                                                                   "not detected") ~ 
                                                               "overlap",
                                                             enrichment_overlap == "no overlap" ~
                                                               "no overlap")))) %>%
select(1, 2, spec_category_human, spec_category_pig) %>%
  
  multi_alluvial_plot(vars = c("Pig" = "spec_category_pig", 
                               "Human" = "spec_category_human"), 
                      chunk_levels = c('Tissue enriched overlap', 'Tissue enriched no overlap',
                                       'Group enriched overlap', 'Group enriched no overlap', 
                                       'Tissue enhanced overlap', 'Tissue enhanced no overlap',
                                       'Low tissue specificity overlap', 'Low tissue specificity no overlap', 
                                       'Not detected overlap', 'Not detected no overlap'), 
                      pal = c(set_names(gene_category_pal, 
                                        paste(names(gene_category_pal), "overlap")),
                              set_names(sapply(gene_category_pal,
                                               function(colr) {
                                                 colorRampPalette(c("white", colr))(4)[3]
                                               }), 
                                        paste(names(gene_category_pal), "no overlap"))), 
                      color_by = c(1, 2)) 
ggsave(savepath("pig-human class overlap alluvial.pdf"), width = 5, height = 10, useDingbats = F)

```





#### Detected in single

##### Venn

```{r detected_in_single_venn, echo=FALSE, message=FALSE, warning=FALSE}

# OBS - work in progress


gene_single_overlap_data <- 
  human_pig_gene_class_comparison_long %>% 
  filter(dist_category_human %in% c("detected in single") |
           dist_category_pig %in% c("detected in single")) %>%
  filter(ortholog_type == "ortholog_one2one") %>%
  left_join(human_gene_class_comparison %>%
              select(1, tissues_detected),
            by = c("ensg_id" = "gene")) %>%
  left_join(pig_gene_class_comparison %>%
              select(1, tissues_detected),
            by = c("enssscg_id" = "gene"), 
            suffix = c("_human", "_pig")) %>%
  filter(!is.na(comparison_id)) %>%
  
  group_by(comparison_id, enssscg_id, ensg_id, 
           spec_score_human, spec_score_pig, 
           spec_category_human, spec_category_pig,
           dist_category_human, dist_category_pig) %>%
  summarise(type = case_when(any(!is.na(enriched_tissues_human) & 
                                   !is.na(enriched_tissues_pig)) ~ "Overlap",
                             all(is.na(enriched_tissues_human)) ~ "Pig", 
                             all(is.na(enriched_tissues_pig)) ~ "Human")) %>%
  ungroup()


# 
# plot_data <- 
#   gene_overlap_data %>% 
#   filter(spec_category_human == "tissue enriched" |
#            spec_category_pig == "tissue enriched") %>%
#   group_by(comparison_id) %>% 
#   mutate(n = n()) %>% 
#   ungroup() %>%
#   mutate(show = n > 30,
#          comparison_id = ifelse(show, comparison_id, "Other tissues"))
# 
# plot_order <- 
#   plot_data %>%
#   select(comparison_id, n, show) %>% 
#   unique() %>%
#   arrange(!show, -n) %$% 
#   unique(comparison_id) %>% 
#   str_to_sentence() %>% 
#   rev()
# 
# # Venn
# 
# overlap_plot_only_enriched <-
#   plot_data %>% 
#   mutate(comparison_id = factor(str_to_sentence(comparison_id)),
#          type = factor(type)) %>%
#   group_by(comparison_id, type, .drop = F) %>%
#   summarise(n = n()) %>%
#   
#   ungroup() %>%
#   
#   mutate(comparison_id = factor(comparison_id, 
#                                 levels = plot_order)) %>%
#   ggplot() + 
#   geom_col(mapping = aes(comparison_id, n,fill = type)) + 
#   
#   geom_text(data = . %>%
#               filter(type == "Pig"),
#             aes(x = comparison_id, y = 0,
#                 label = n),
#             hjust = 1) +
#   geom_text(data = . %>%
#               group_by(comparison_id, .drop = F) %>%
#               summarise(y = n[which(type == "Overlap")] / 2 +
#                           n[which(type == "Pig")],
#                         label = n[which(type == "Overlap")]),
#             aes(x = comparison_id, y = y,
#                 label = label),
#             hjust = 0.5,
#             color = "white") +
#   geom_text(data = . %>%
#               group_by(comparison_id, .drop = F) %>%
#               summarise(y = sum(n),
#                         label = n[which(type == "Human")]),
#             aes(x = comparison_id, y = y,
#                 label = label),
#             hjust = 0) +
#   
#   coord_flip() + 
#   
#   scale_fill_manual(values = c("Human" = "#DB1F48", "Overlap" = "#004369", "Pig" = "#01949A")) +
#   scale_y_continuous(expand = expand_scale(0.13)) +
#   
#   theme(axis.text.y = element_text(face = "bold"), 
#         axis.text.x = element_blank(),
#         axis.ticks.x = element_blank(), 
#         axis.title = element_blank(),
#         panel.background = element_blank(),
#         
#         legend.position = c(0.5, 0.5)) + 
#   
#   ggtitle("Overlap of tissue enriched genes") +
#   theme(axis.ticks = element_blank())
#   
# # Summary pie
# 
# overlap_summary_pie_plot <- 
#   gene_overlap_data %>%
#   filter(spec_category_human == "tissue enriched" |
#            spec_category_pig == "tissue enriched") %>%
#   
#   left_join(joined_atlas_comparison_exp_ranks, 
#             by = c("comparison_id" = "comparison_tissue",
#                    "enssscg_id", 
#                    "ensg_id")) %>% 
#   
#   mutate(corresponding_rank = case_when(type == "Pig" ~ rank_human,
#                                         type == "Human" ~ rank_pig, 
#                                         type == "Overlap" ~ (rank_pig + rank_human) / 2),
#          
#          comparison_id = factor(str_to_sentence(comparison_id)),
#          type = factor(type))  %>% 
#   # filter(any(type == "Pig") & 
#   #                              any(type == "Human")) %>% View
#   group_by(enssscg_id,
#            ensg_id) %>%
#   summarise(type = case_when(any(type == "Overlap") ~ "Overlap",
#                              any(type == "Pig") & 
#                                any(type == "Human") ~ "Different tissues",
#                              any(type == "Pig") ~ "Pig",
#                              any(type == "Human") ~ "Human"), 
#             rank_group = case_when(max(corresponding_rank) == 1 ~ "First",
#                                    max(corresponding_rank) <= 2 ~ "Second",
#                                    max(corresponding_rank) <= 3 ~ "Third",
#                                    max(corresponding_rank) > 3 ~ ">Third")) %>%
#   group_by(type, rank_group, .drop = F) %>%
#       summarise(n = n()) %>%
#       
#       ungroup() %>%
#   mutate(type_rank = paste(type, rank_group),
#          level = 2) %>%
#   select(group = type, 
#          label = type_rank, 
#          n, 
#          level) %>%
#   bind_rows(group_by(., group) %>% 
#               summarise(label = unique(group), 
#                         n = sum(n)) %>% 
#               ungroup() %>% 
#               mutate(level = 1)) %>%
#   mutate(label = factor(label,
#                         levels = c(c("Human >Third","Human Third", 
#                                          "Human Second", "Human First"),
#                                    c("Pig >Third", "Pig Third", 
#                                      "Pig Second", "Pig First"),
#                                    c("Different tissues >Third", "Different tissues Second"),
#                                    c("Overlap First", "Overlap Second", "Overlap Third"),
#                                    c("Human", "Pig", "Different tissues", "Overlap")))) %>%
#   donut_plot(class_col = "label", 
#              group_col = "group", 
#              level_col = "level",
#              y_col = "n", 
#              pal = c(set_names(colorRampPalette(c("#DB1F48", "white"))(6)[1:4],
#                                paste("Human", c("First", "Second","Third",">Third"))),
#                      set_names(colorRampPalette(c("#01949A", "white"))(6)[1:4],
#                                paste("Pig", c("First", "Second","Third",">Third"))),
#                      c("Human" = "#DB1F48", "Overlap" = "#004369", 
#                        "Pig" = "#01949A", "Different tissues" = "orangered"), 
#                      "NA" = NA, 
#                      "Different tissues >Third" = NA,
#                      "Different tissues Second" = NA,
#                      "Overlap First" = NA,
#                      "Overlap Second" = NA,
#                      "Overlap Third" = NA), 
#              dodge = 0.1, 
#              widths = c(1, 0.5), 
#              inner_label = c(T, F), 
#              outer_label = c(F, F), 
#              curved_label = c(F, T), 
#              text_color = c("white", "black"), 
#              label_format = c("lnp", "n"))
# 
# 
#   
# # Enrichement score distributions
# 
# overlap_enrichment_score_distr_plot <- 
#   plot_data %>% 
#   mutate(comparison_id = str_to_sentence(comparison_id)) %>%
#   compare_class_enrichment_score_lineplot("Enrichment score of tissue enriched genes", man_order = T, plot_order = plot_order) +
#   coord_flip() + 
#   # scale_y_reverse() + 
#   theme(axis.text.y = element_blank(), 
#         axis.ticks.x = element_line(), 
#         axis.text.x = element_text(angle = 0, 
#                                    hjust = 0.5))
# 
# overlap_enrichment_score_distr_plot_all_elevated <- 
#   gene_overlap_data %>% 
#   mutate(comparison_id = ifelse(str_to_sentence(comparison_id) %in% plot_order, 
#                                 comparison_id, "Other tissues")) %>%
#   mutate(comparison_id = str_to_sentence(comparison_id)) %>%
#   compare_class_enrichment_score_lineplot("Enrichment score of all elevated genes", man_order = T, plot_order = plot_order) +
#   coord_flip() + 
#   # scale_y_reverse() + 
#   theme(axis.text.y = element_blank(), 
#         axis.ticks.x = element_line(), 
#         axis.text.x = element_text(angle = 0, 
#                                    hjust = 0.5))
# 
# 
# overlap_enrichment_score_distr_plot + 
#   overlap_enrichment_score_distr_plot_all_elevated + 
#   
#   overlap_plot_only_enriched + 
#   # overlap_summary_pie_plot +
#   plot_layout(nrow = 1, widths = c(1, 1, 2, 3))
# ggsave(savepath("Final overlap Venn.pdf"), height = 3, width = 20, useDingbats = F)
# 
# gene_overlap_data %>% 
#   mutate(enrichment_score = case_when(type == "Overlap" ~ (spec_score_human + spec_score_pig) / 2,
#                                       type == "Human" ~ spec_score_human,
#                                       type == "Pig" ~ spec_score_pig)) %>%
#   ggplot(aes(type, enrichment_score, fill = type)) + 
#   geom_violin(draw_quantiles = 0.5,
#               color = "white", 
#               show.legend = F) +
#   
#   # coord_flip() + 
#   
#   scale_y_log10() +
#   scale_fill_manual(values = c("Human" = "#DB1F48", "Overlap" = "#004369", "Pig" = "#01949A")) +
#   
#   ylab("Enrichment score") +
#   ggtitle("Enrichment score of all elevated genes") +
#   
#   stripped_theme +
#   theme(axis.title.x = element_blank(),
#         axis.line = element_blank())
#   
# ggsave(savepath("Overlap enrichment scora all elevated violin.pdf"), 
#        height = 4, width = 4, useDingbats = F)
```





###Human-pig alluvial

```{r Figure4_humanpig_alluvial, echo=FALSE, message=FALSE, warning=FALSE}

width = 0.1

alluv_1 <-
  
  human_pig_gene_class_comparison %>%
   mutate(Pig = str_to_sentence(spec_category_pig),
          Human = str_to_sentence(spec_category_human)) %>%
  select(Pig,
         Human) %>%
  mutate(row_n = row_number()) %>%
  gather(bar, chunk, -row_n) %>%
  mutate(color_vars = 1) %>%
  group_by(row_n) %>%
  mutate(chunk_color = chunk[match(c("Pig",
                                     "Human")[color_vars], bar)]) %>% 
  ungroup() %>%
  
  
  mutate(chunk = factor(chunk, levels = c('Tissue enriched', 'Group enriched', 
                                          'Tissue enhanced', 'Low tissue specificity', 
                                          'Not detected')),
         bar = factor(bar, levels = c("Human",
                                      "Pig"))) %>%
  
  
  ggplot(aes(x = bar, stratum = chunk, alluvium = row_n,
             y = 1)) +
  
  geom_flow(aes(fill = chunk_color), 
            show.legend = F, width = width,
            knot.pos = 1/6) +
  geom_stratum(aes(fill = chunk), 
               show.legend = F, color = NA, width = width) +
  
  scale_x_discrete(expand = c(.1, .1), position = "top") +
  scale_fill_manual(values = c(gene_category_pal, elevation_identity_pal)) + 
  
  
  theme(axis.text.y = element_text(size = 18, face = "bold"),
        axis.text.x = element_blank(), 
        axis.ticks = element_blank(), 
        panel.background = element_blank(), 
        axis.title = element_blank()) +
  coord_flip()

# alluv_1
flow_data <-
  ggplot_build(alluv_1)$data[[1]] %>%
  as_tibble() %>%
  {
    if("side" %in% names(.)) {
      .
    } else{
      mutate(.,
             side = case_when(contact == "back" ~ "start",
                              contact == "front" ~ "end"))
    }}

stratum_data <- 
  ggplot_build(alluv_1)$data[[2]]

flow_data_labels <-
  flow_data %>% 
  as_tibble() %>% 
  select(x, stratum, group, side, ymin, ymax) %>% 
  multispread(c(side), c(x, stratum, ymin, ymax)) %>% 
  mutate_at(c("end_x", "end_ymax", "end_ymin", "start_x", "start_ymax", "start_ymin"), as.numeric) %>% 
  group_by(start_stratum, end_stratum, start_x, end_x) %>%
  summarise(end_y = (min(end_ymin) + max(end_ymax)) / 2, 
            start_y = (min(start_ymin) + max(start_ymax)) / 2, 
            size = max(start_ymax) - min(start_ymin))

alluv_2 <- 
  alluv_1 +
  geom_text(data = flow_data_labels,
            aes(x = start_x + width/2,
                y = start_y, 
                label = size), 
            inherit.aes = F, 
            size = 3,
            angle = -90,
            hjust = 1,
            vjust = 0.5) +
  geom_text(data = flow_data_labels,
            aes(x = end_x - width/2,
                y = end_y, 
                label = size), 
            inherit.aes = F, 
            size = 3, 
            angle = -90,
            hjust = 0,
            vjust = 0.5) +
  
  # Stratum label
  
  geom_text(data = stratum_data %>%
              filter(x == 1),
            aes(x = x - width/2,
                y = y,
                label = stratum),
            size = 4, 
            vjust = 1.5,
            inherit.aes = F) + 
  geom_text(data = stratum_data %>%
              filter(x == 2),
            aes(x = x + width/2,
                y = y,
                label = stratum),
            size = 4, 
            vjust = -0.5,
            inherit.aes = F) + 
  
  geom_text(data = stratum_data,
            aes(x = x, 
                y = y,
                label = ymax - ymin), 
            size = 4, 
            fontface = "bold",
            color = "white",
            inherit.aes = F)


# alluv_2

ggsave(savepath("Figure4 Alluvial.pdf"), plot = alluv_2, width = 7, height = 2.5, useDingbats = F)


```

###Scatter plots

```{r}
plot_data <- 
  joined_atlas_comparison %>% 
  filter(mutual_id %in% with(gene_orthologs, paste(enssscg_id, ensg_id, sep = "_"))) %>%
  select(mutual_id, comparison_tissue, species, ptpm) %>%
  spread(species, ptpm) %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id")) %>% 
  left_join(gene_overlap_data,
            by = c("enssscg_id", "ensg_id", "comparison_tissue" = "comparison_id")) %>%
  left_join(human_gene_mapping) %>%
  mutate(x = log10(human + 1), 
         y = log10(pig + 1))
  
  
plot_cor_data <- 
  plot_data %>%
  group_by(comparison_tissue) %>% 
  summarise(cor = cor(human, pig, method = "spearman", use = "pairwise.complete.obs")) %>%
  ungroup() %>%
  arrange(cor) %>%
  mutate(comparison_tissue = factor(str_to_sentence(comparison_tissue),
                                    str_to_sentence(comparison_tissue)))
  
plot_ellipse <- 
  plot_data %>%
  mutate(comparison_tissue = str_to_sentence(comparison_tissue)) %>%
  group_by(comparison_tissue) %>%
  do({
    g_data <- .
    g_data %$%
      ellipse::ellipse(cor(x, y, use = "pairwise.complete.obs"),
                       scale=c(sd(1.5 * x, na.rm = T), 
                               sd(1.5 * y, na.rm = T)),
                       centre=c(median(x, na.rm = T), 
                                median(y, na.rm = T))) %>% 
      
      as_tibble() %>% 
      mutate(x = ifelse(x < 0, 0, x),
             y = ifelse(y < 0, 0, y))
  }) %>%
  ungroup()


plot_data_2 <-
  plot_data %>%
  mutate(comparison_tissue = str_to_sentence(comparison_tissue)) %>%
  group_by(comparison_tissue) %>%
  mutate(label = ifelse(sp::point.in.polygon(x, y, 
                                             plot_ellipse$x[which(plot_ellipse$comparison_tissue == unique(comparison_tissue))], 
                                             plot_ellipse$y[which(plot_ellipse$comparison_tissue == unique(comparison_tissue))]),
                        NA, 
                        gene_name),
         distance = sqrt(log10(human + 1)^2 + log10(pig + 1)^2)) %>% 
  ungroup() %>% 
  mutate(comparison_tissue = factor(comparison_tissue, plot_cor_data$comparison_tissue)) 

plot <- 
  plot_data_2 %>% 
  # filter(comparison_tissue == "liver") %>%
  ggplot(aes(x, y)) + 
  geom_abline(slope = 1) + 
  geom_hex(aes(fill = stat(log10(count))),
           bins = 100) +
  
  geom_point(data = . %>%
             filter(!is.na(type)),
             aes(color = type),
             size = 0.4,
             alpha = 0.5) +
  geom_text(data = . %>%
              filter(!is.na(type)),
            aes(label = label, size = distance, color = type),
             hjust = 0.5,
             alpha = 0.8) +
  geom_text(data = plot_cor_data,
            aes(x = 1, y = 4.5, label = round(cor, 2))) +
  
  facet_wrap(~ comparison_tissue) +
  
  coord_fixed() +
  xlab("Human expression") +
  ylab("Pig expression") +
  scale_fill_gradientn(colors = c("gray90", "gray10")) +
  scale_color_manual(values = overlap_type_pal) +
  scale_size_continuous(range = c(0.8, 2)) +
  stripped_theme_facet 

ggsave(savepath("Figure4 scatter overlap.pdf"), plot = plot, width = 10, height = 8)

plot <- 
  plot_data_2 %>% 
  filter(comparison_tissue %in% c("Brain",
                                  "Esophagus")) %>%
  ggplot(aes(x, y)) + 
  geom_abline(slope = 1) + 
  geom_hex(aes(fill = stat(log10(count))),
           bins = 100,
           show.legend = F) +
  
  geom_point(data = . %>%
             filter(!is.na(type)),
             aes(color = type),
             size = 0.4,
             alpha = 0.5) +
  geom_text(data = . %>%
              filter(!is.na(type)),
            aes(label = label, size = distance, color = type),
             hjust = 0.5,
             alpha = 0.8,
            show.legend = F) +
  geom_text(data = plot_cor_data %>%
              filter(comparison_tissue %in% c("Brain",
                                              "Esophagus")),
            aes(x = 1, y = 4.5, label = round(cor, 2)),
            size = 8,
            show.legend = F) +
  
  facet_wrap(~ comparison_tissue, nrow = 1) +
  
  coord_fixed() +
  xlab("Human expression") +
  ylab("Pig expression") +
  scale_fill_gradientn(colors = c("gray90", "gray10")) +
  scale_color_manual(values = overlap_type_pal, 
                     name = "Specificity overlap") +
  scale_size_continuous(range = c(0.8, 2), guide = NA) +
  stripped_theme_facet +
  theme(legend.position = "top") +
  guides(size = F,
         fill = F)

ggsave(savepath("Figure4 scatter overlap examples.pdf"), plot = plot, width = 4, height = 3)

```

### IHC example Barplots

```{r}



joined_atlas_comparison %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>%
  inner_join(gene_mapping %>% 
               filter(display_name %in% c("PLN", "SATB2", "PDHB"))) %>%
  left_join(tissue_mapping %>% 
              select(comparison_tissue, organ_name) %>% 
              distinct(),
            by = c("comparison_tissue")) %>%
  arrange(organ_name) %>%
  mutate(display_name = factor(display_name, c("PLN", "SATB2", "PDHB")),
         comparison_tissue = factor(str_to_sentence(comparison_tissue), 
                                    unique(str_to_sentence(comparison_tissue))),
         species = factor(species, c("pig", "human"))) %>%
  
  ggplot(aes(comparison_tissue, ptpm, fill = organ_name)) +
  geom_col(show.legend = F,
           color = "black",
           size = 0.2) +
  geom_text(data = . %>% 
              filter(display_name == "PLN" & comparison_tissue == "Heart muscle" |
                     display_name == "SATB2" & comparison_tissue %in% c("Intestine", "Brain")),
            aes(label = comparison_tissue), 
            vjust = -0.5) +
  facet_wrap(~display_name + species, ncol = 1, scales = "free_y", strip.position = "left") + 
  scale_fill_manual(values = tissue_colors_palette_full) +
  scale_y_continuous(expand = expansion(c(0, 0.2)), position = "right") +
  stripped_theme_facet + 
  ylab("PTPM") +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.border = element_blank(),
        axis.line = element_blank())
ggsave(savepath("Figure5 IHC example barplots.pdf"), width = 2, height = 6)

```


### UMAP

```{r}
plot_data1 <- 
  joined_atlas_comparison %>% 
  select(mutual_id, comparison_tissue, species, mutual_limma) %>%
  unite(id, comparison_tissue, species) %>% 
  spread(id, mutual_limma) %>% 
  filter(complete.cases(.)) %>%
  column_to_rownames("mutual_id") %>% 
  t() %>%
  {log10(. + 1)}

set.seed(44)
plot_data <- 
  plot_data1 %>%
  uwot::umap(n_neighbors = 5, n_epochs = 1000)

plot_data %>% 
  as.tibble() %>% 
  mutate(id = rownames(plot_data1)) %>% 
  separate(id, into = c("comparison_tissue", "species"), sep = "_") %>% 
  left_join(tissue_mapping %>% 
              select(comparison_tissue, 
                     organ_name) %>% 
              distinct()) %>%
     
  
  ggplot(aes(V1, V2, fill = organ_name, shape = species)) +
  geom_path(aes(group = comparison_tissue),
            size = 0.3,
            color = "gray60") +
  geom_point(stroke = 0.3,
             size = 3) +
  geom_text(data = . %>% 
              group_by(comparison_tissue) %>% 
              summarise(angle = atan((V2[2] - V2[1]) /
                          (V1[2] - V1[1])) * 180 / pi,
                        V1 = mean(V1),
                        V2 = mean(V2)),
            aes(V1, V2, label = comparison_tissue, angle = angle),
            inherit.aes = F,
            vjust = -1,
            size = 2.5) +
  scale_fill_manual(values = tissue_colors_palette_organs) +
  scale_shape_manual(values = c("pig" = 21, "human" = 22)) +
  stripped_theme +
  coord_fixed() +
  guides(fill = F)
ggsave(savepath("Fig5 Human-pig UMAP.pdf"), width = 6, height = 6)



```

### Distance barplot 

```{r}
plot_data1 <- 
  joined_atlas_comparison %>% 
  select(mutual_id, comparison_tissue, species, mutual_limma) %>%
  unite(id, comparison_tissue, species) %>% 
  spread(id, mutual_limma) %>% 
  filter(complete.cases(.)) %>%
  column_to_rownames("mutual_id") %>% 
  t() %>%
  {log10(. + 1)}

plot_data1 %>%
  dist() %>%
  as.matrix() %>% 
  as_tibble(rownames = "id1") %>%
  gather(id2, dist, -1) %>%
  separate(id1, into = c("human", "species1"), sep = "_") %>%
  separate(id2, into = c("pig", "species2"), sep = "_") %>%
  filter(human == pig,
         species1 == "human", 
         species2 == "pig") %>% 
  arrange(dist) %>%
  mutate(tissue = factor(human, human)) %>%
  ggplot(aes(dist, tissue)) +
  geom_col()



```

###Sequence identity comparisons

```{r}
gene_overlap_data_all_orth_summarized %>%
  left_join(gene_orthologs_all) %>%
  ggplot(aes(type, sequence_identity)) +
  geom_violin(draw_quantiles = 0.5,
              fill = "gray90") +
  ggpubr::stat_compare_means(comparisons = list(c('Different tissues', 'Overlap'),
                                                c('Human', 'Overlap'),
                                                c('Pig', 'Overlap')),
                             method = "wilcox.test",
                             size=3, paired = F) +
  geom_text(data = . %>% 
              group_by(type) %>% 
              summarise(mean = median(sequence_identity)),
            aes(y = mean, label = paste(round(mean, 1), "%")),
            vjust = 0) +
  stripped_theme
  
ggsave(savepath("sequence identity per overlap all orth.pdf"),
       width = 4, height = 4)

gene_overlap_data_all_orth_summarized %>%
  left_join(human_gene_comparison_classification %>% 
              select(ensg_id, specificity_category),
            by = "ensg_id") %>%
  left_join(pig_gene_comparison_classification %>% 
              select(enssscg_id, specificity_category),
            by = "enssscg_id",
            suffix = c("_human", "_pig")) %>%
  mutate(elevated = specificity_category_human %in% c("Tissue enhanced", 
                                                      "Tissue enriched",
                                                      "Group enriched") |
           specificity_category_pig %in% c("Tissue enhanced", 
                                           "Tissue enriched",
                                           "Group enriched"),
         enriched = specificity_category_human %in% c("Tissue enriched",
                                                      "Group enriched") |
           specificity_category_pig %in% c("Tissue enriched",
                                           "Group enriched"),
         type2 = case_when(type == "Overlap" & enriched ~ "Overlap\nenriched",
                           type == "Overlap" & elevated ~ "Overlap\nelevated",
                           type == "Overlap" & !elevated ~ "Overlap\nlow specificity",
                           T ~ type)) %>%
  left_join(gene_orthologs_all) %>%


  ggplot(aes(type2, sequence_identity)) +
  geom_violin(draw_quantiles = 0.5,
              fill = "gray90") +
  geom_text(data = . %>% 
              group_by(type2) %>% 
              summarise(mean = median(sequence_identity)),
            aes(y = mean, label = paste(round(mean, 1), "%")),
            vjust = 0) +
  stripped_theme
  
ggsave(savepath("sequence identity per overlap all orth 2.pdf"),
       width = 6, height = 4)

gene_overlap_data_all_orth_summarized %>%
  left_join(human_gene_comparison_classification %>% 
              select(ensg_id, specificity_category),
            by = "ensg_id") %>%
  left_join(pig_gene_comparison_classification %>% 
              select(enssscg_id, specificity_category),
            by = "enssscg_id",
            suffix = c("_human", "_pig")) %>%
  
  left_join(gene_orthologs_all) %>%


  ggplot(aes(specificity_category_pig, sequence_identity)) +
  geom_violin(draw_quantiles = 0.5,
              fill = "gray90") +
  geom_text(data = . %>% 
              group_by(specificity_category_pig) %>% 
              summarise(mean = median(sequence_identity)),
            aes(y = mean, label = paste(round(mean, 1), "%")),
            vjust = 0) +
  stripped_theme
  
ggsave(savepath("sequence identity per pig class all orth.pdf"),
       width = 6, height = 4)

pig_gene_umap %>%
  select(enssscg_id, merged_cluster) %>%
  left_join(gene_orthologs_all) %>%
  filter(!is.na(ensg_id)) %>%
  ggplot(aes(sequence_identity, merged_cluster)) +
  geom_boxplot(fill = "gray90", 
               outlier.colour = NA) +
  geom_text(data = . %>% 
              group_by(merged_cluster) %>% 
              summarise(mean = median(sequence_identity)),
            aes(x = mean, y = merged_cluster, label = paste(round(mean, 1), "%")),
            hjust = 0) +
  stripped_theme

#### Abundance

joined_atlas_comparison %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>%
  group_by(enssscg_id, ensg_id, species) %>% 
  summarise(mutual_tmm = max(mutual_tmm, na.rm = T)) %>% 
  
  left_join(gene_orthologs_all) %>%
  group_by(species) %>% 
  mutate(cor = cor(log10(mutual_tmm + 1), sequence_identity, method = "spearman", use = "pairwise.complete.obs")) %>%
  ggplot(aes(log10(mutual_tmm + 1), sequence_identity)) +
  geom_hex(aes(fill = stat(log10(count))), 
           bins = 100) +
  facet_wrap(~species + round(cor, 3)) +
  scale_fill_viridis() +
  stripped_theme_facet
ggsave(savepath("sequence identity vs tmm hex.pdf"),
       width = 8, height = 4)


# linear model

seq_id_anova <- 
  joined_atlas_comparison %>%
  # head(100) %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>%
  group_by(enssscg_id, ensg_id) %>% 
  summarise(mutual_tmm = max(mutual_tmm, na.rm = T)) %>%
  ungroup() %>% 
  mutate(exp_quartile = as.factor(unclass(cut(mutual_tmm, breaks = quantile(mutual_tmm), include.lowest = T)))) %>%
  left_join(gene_orthologs_all) %>%
  
  left_join(human_gene_comparison_classification %>% 
              select(ensg_id, specificity_category),
            by = "ensg_id") %>%
  left_join(pig_gene_comparison_classification %>% 
              select(enssscg_id, specificity_category),
            by = "enssscg_id",
            suffix = c("_human", "_pig")) %>% 
  left_join(gene_overlap_data_all_orth_summarized) %>%
  mutate(elevated = ifelse(specificity_category_human %in% c("Tissue enhanced", 
                                                             "Tissue enriched",
                                                             "Group enriched") |
                             specificity_category_pig %in% c("Tissue enhanced", 
                                                             "Tissue enriched",
                                                             "Group enriched"),
                           "yes",
                           "no"),
         enriched = ifelse(specificity_category_human %in% c("Tissue enriched",
                                                             "Group enriched") |
                             specificity_category_pig %in% c("Tissue enriched",
                                                             "Group enriched"),
                           "yes",
                           "no"),
         overlap = ifelse(type == "Overlap",
                          "yes", 
                          "no")) %>%
  aov(sequence_identity ~ 
        exp_quartile + 
        overlap + 
        elevated,
      data = .)

seq_id_anova$coefficients %>%
  enframe("coefficient", "estimate") %>%
  ggplot(aes(estimate, coefficient)) +
  geom_col()
omega_sq(seq_id_anova) %>%
  enframe("coefficient", "var_exp") %>%
  ggplot(aes(var_exp, coefficient)) +
  geom_col() +
  geom_text(aes(label = paste0(100 * round(var_exp, 4), "%")),
            hjust = 0) +
  scale_x_continuous(expand = expansion(0),
                     limits = c(0,1)) +
  stripped_theme
  





cor_orth_anova <- 
  joined_atlas_comparison %>%
  # head(100) %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>%
  select(enssscg_id, ensg_id, comparison_tissue, species, ptpm) %>%
  spread(species, ptpm) %>%
  group_by(enssscg_id, ensg_id) %>% 
  summarise(cor = cor(human, pig, use = "pairwise.complete.obs", method = "spearman")) %>%
  ungroup() %>% 
  left_join(joined_atlas_comparison %>%
              # head(100) %>%
              separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>%
              group_by(enssscg_id, ensg_id) %>% 
              summarise(mutual_tmm = max(mutual_tmm, na.rm = T)) %>%
              ungroup() %>% 
              mutate(exp_quartile = as.factor(unclass(cut(mutual_tmm, 
                                                          breaks = quantile(mutual_tmm), 
                                                          include.lowest = T))))) %>%
  
  left_join(gene_orthologs_all) %>%
  
  left_join(human_gene_comparison_classification %>% 
              select(ensg_id, specificity_category),
            by = "ensg_id") %>%
  left_join(pig_gene_comparison_classification %>% 
              select(enssscg_id, specificity_category),
            by = "enssscg_id",
            suffix = c("_human", "_pig")) %>% 
  left_join(gene_overlap_data_all_orth_summarized) %>%
  mutate(elevated = ifelse(specificity_category_human %in% c("Tissue enhanced", 
                                                             "Tissue enriched",
                                                             "Group enriched") |
                             specificity_category_pig %in% c("Tissue enhanced", 
                                                             "Tissue enriched",
                                                             "Group enriched"),
                           "yes",
                           "no"),
         enriched = ifelse(specificity_category_human %in% c("Tissue enriched",
                                                             "Group enriched") |
                             specificity_category_pig %in% c("Tissue enriched",
                                                             "Group enriched"),
                           "yes",
                           "no"),
         overlap = ifelse(type == "Overlap",
                          "yes", 
                          "no")) %>%
  mutate(exp_quartile = factor(exp_quartile, 1:4),
         overlap = factor(overlap),
         elevated = factor(elevated)) %>%
  aov(cor ~ 
        sequence_identity +
        gene_order_conservation_score +
        whole_genome_alignment_coverage +
        exp_quartile + 
        overlap:elevated,
      data = .)

cor_orth_anova$coefficients %>%
  enframe("coefficient", "estimate") %>%
  ggplot(aes(estimate, coefficient)) +
  geom_col()
omega_sq(cor_orth_anova) %>%
  enframe("coefficient", "var_exp") %>%
  ggplot(aes(var_exp, coefficient)) +
  geom_col() +
  geom_text(aes(label = paste0(100 * round(var_exp, 4), "%")),
            hjust = 0) +
  scale_x_continuous(expand = expansion(0),
                     limits = c(0,1)) +
  stripped_theme

```



##Figure 6

##Figure 7


###Selected genes 3

```{r}


plot_genes <- 
  disease_gene_classes %>%
  filter(gene_name %in% c("PLP1",
                          "MBP",
                          "ABDB11",
                          "PDX1",
                          "GCGR",
                          "TNNT2",
                          "NPPA",
                          "LEPR",
                          "GRXCR2")) %>%
  mutate(disease = disease %>% 
           gsub("^H\\d{5}  ", "", .) %>% 
           gsub(" \\[.*\\]$", "", .)) %>%
  group_by(ensg_id, gene_name) %>% 
  summarise(disease = paste(sort(unique(disease)), collapse = ";"),
            disease_type = paste(sort(unique(disease_type)), collapse = ";"),
            disease_system = paste(sort(unique(disease_system)), collapse = ";"))


plot_data <- 
  joined_atlas_comparison %>% 
  separate(mutual_id, into = c("enssscg_id", "ensg_id")) %>% 
  inner_join(plot_genes) %>%
  group_by(enssscg_id, species) %>% 
  mutate(max_tmm = max(mutual_tmm, na.rm = T),
         mutual_tmm1 = mutual_tmm,
         mutual_tmm = ifelse(max_tmm == 0, 0, mutual_tmm / max_tmm)) %>%
  left_join(human_gene_mapping) 


plot_data_frac <- 
  plot_data %>%
  select(comparison_tissue, species, ptpm, gene_name, enssscg_id) %>%
  group_by(comparison_tissue, species) %>%
  mutate(frac = 100 * ptpm / sum(ptpm, na.rm = T),
         label = ifelse(frac >= 10, paste0(round(frac, 1), "%"), NA)) %>%
  
  left_join(select(., -label, -ptpm) %>% 
              spread(species, frac) %>%
              group_by(comparison_tissue) %>%
              summarise(cor = cor(human, pig, method = "spearman")) %>%
              select(comparison_tissue, cor)) %>%
  ungroup() %>%
  arrange(cor) %>% 
  mutate(comparison_tissue = factor(comparison_tissue, unique(comparison_tissue)),
         species = factor(species, c("human", "pig"))) 

plot_data2 <-
  gene_overlap_data_summarized %>%
  inner_join(plot_genes)

plot_data3 <-
  human_pig_gene_class_comparison %>% 
  inner_join(plot_data %>% 
               select(enssscg_id, ensg_id) %>% 
               distinct()) %>%
  select(enssscg_id, ensg_id, spec_category_pig, spec_category_human) %>%
  left_join(human_gene_mapping) %>%
  gather(species, spec_category, spec_category_pig, spec_category_human) %>%
  mutate(species = str_to_sentence(gsub("spec_category_", "", species))) %>%
  inner_join(plot_genes)


#####


cluster_data <-
  plot_data %>%
  ungroup() %>%
  select(comparison_tissue, species, mutual_tmm, gene_name) %>%
  group_by(gene_name, comparison_tissue) %>%
  summarise(mutual_tmm = mean(mutual_tmm)) %>%
  spread(gene_name, mutual_tmm) %>%
  column_to_rownames("comparison_tissue")


tis_cluster <- 
  cluster_data %>%
  dist() %>% 
  hclust(method = "ward.D2")

tis_order <- 
  tis_cluster$labels[tis_cluster$order]


gene_order <- 
  plot_data %>% 
  ungroup() %>%
  arrange(disease_type) %>% 
  pull(gene_name) %>% 
  unique()
  

disease_order <- 
  plot_data %>%
  group_by(disease) %>%
  top_n(1, ptpm) %>%
  mutate(tissue_n = match(comparison_tissue, tis_order)) %>%
  arrange(tissue_n, ptpm) %>%
  pull(disease)


# Plot:

plot_data %>% 
  ungroup() %>%
  select(gene_name, species, disease, disease) %>% 
  distinct() %>%
  mutate(species = toupper(substr(species, 1, 1))) %>%
  mutate(gene_name = factor(gene_name, gene_order),
         disease = factor(disease, disease_order)) %>%
  
  
  ggplot(aes(1, gene_name, group = species, fill = species, label = species)) +
  geom_tile(show.legend = F, 
            position = "dodge", 
            height = 0.9) +
  geom_text(position = position_dodge(width = 1), 
            size = 2, 
            fontface = "bold",
            color = "white") +
  scale_fill_manual(values = c("H" = "#DB1F48",
                               "P" = "#01949A")) +
  # scale_x_discrete(limits = "Species") +
  facet_grid(disease ~ ., scales = "free", space = "free") +
  theme_void() +
  theme(plot.margin = unit(c(0,0,0,0), "mm"),
        strip.background = element_blank(),
        strip.text = element_blank(),
        panel.spacing = unit(0, "mm")) +
  
  # Specificity
  plot_data3 %>%
  mutate(gene_name = factor(gene_name, gene_order),
         disease = factor(disease, disease_order)) %>%
  
  ggplot(aes(1, gene_name, group = species, fill = spec_category, label = species)) +
  geom_tile(show.legend = F,
            position = "dodge", 
            height = 0.9, width = 1) +
  scale_fill_manual(values = gene_category_pal) +
  # scale_x_discrete(limits = "Specificity") +
  facet_grid(disease ~ ., scales = "free", space = "free") +
  theme_void() +
  theme(plot.margin = unit(c(0,0,0,0), "mm"),
        strip.background = element_blank(),
        strip.text = element_blank(),
        panel.spacing = unit(0, "mm")) +
  
  
  # Overlap
  plot_data2 %>%
  mutate(gene_name = factor(gene_name, gene_order),
         disease = factor(disease, disease_order)) %>%
  
  ggplot(aes(1, gene_name, fill = type)) +
  geom_tile(show.legend = F,
            height = 0.9) +
  scale_fill_manual(values = overlap_type_pal) +
  # scale_x_discrete(limits = "Overlap") +
  facet_grid(disease ~ ., scales = "free", space = "free") +
  theme_void() +
  theme(strip.background = element_blank(),
        strip.text = element_blank(),
        panel.spacing = unit(0, "mm")) + 
  
  # Heatmap
  plot_data %>% 
  ungroup() %>%
  mutate(gene_name = factor(gene_name, gene_order),
         comparison_tissue = factor(comparison_tissue, tis_order),
         disease = factor(disease, disease_order)) %>%
  
  ggplot(aes(gene_name, comparison_tissue, group = species, fill = mutual_tmm)) +
  geom_tile(position= "dodge", width = 0.9) +
  coord_flip() +
  scale_fill_viridis_c(option = "B", direction = -1, name = "Expression") +
  facet_grid(disease ~ ., scales = "free", space = "free", labeller = label_wrap_gen()) +
  # facet_wrap(~disease, ncol = 1, strip.position = "right") +
  stripped_theme_facet +
  theme(axis.text.x = element_text(angle = 60, hjust = 1),
        # strip.placement = "outside",
        strip.text.y = element_text(angle = 0),
        strip.background = element_rect(color = "white"),
        panel.spacing = unit(0, "mm"),
        legend.position = "right",
        panel.border = element_rect(fill = NA, color = "white"),
        axis.title = element_blank(), 
        axis.line = element_blank(),
        plot.margin = unit(c(0,0,0,0), "mm")) +
  
  
  
  plot_layout(widths = c(0.025, 0.025, 0.025, 1), nrow = 1)

ggsave(savepath("Fig7 selected genes heatmap 4.pdf"),
       width = 8, height = 4)


```


###All disease genes

```{r}


disease_genes_full %>% 
  filter(!is.na(disease)) %>% 
  group_by(disease) %>% 
  summarise(n = n(), 
            genes = paste(sort(gene_name), collapse = ", ")) %>% 
  arrange(-n) %T>%
  write_csv(savepath("Disease genes.csv")) %>%
  mutate(disease = factor(disease, unique(disease))) %>% 
  ggplot(aes(n, disease, label = n)) + 
  geom_col() +
  geom_text(hjust = 0, 
            size = 2)

ggsave(savepath("Disease genes.pdf"), width = 5, height = 20)
  
disease_genes %>% 
  filter(!is.na(disease)) %>%
  mutate(connection = 1) %>% 
  select(-gene_name) %>%
  spread(disease, connection, fill = 0) %>% 
  
  column_to_rownames("ensg_id") %>% 
  cor(method = "spearman") %>% 
  {
    filtr <- apply(., 
                   MARGIN = 1, 
                   function(x) any(x[x < 1] > 0.5))
    
    .[filtr, filtr]
  } %>%
  
  {
    clu <- 
      dist(1 - .) %>% 
      hclust(method = "average")
    pheatmap(., 
             border_color = NA, 
             cluster_rows = clu,
             cluster_cols = clu,
             filename = savepath("Disease genes overlap heatmap.pdf"), 
             width = 8, 
             height = 8)
  }




pig_atlas_consensus %>% 
  inner_join(disease_genes_full) %>% 
  select(gene_name, tissue, nx) %>%
  distinct() %>%
  spread(tissue, nx) %>% 
  column_to_rownames("gene_name") %>% 
  {log10(. + 1)} %>% 
  pheatmap(filename = savepath("A.pdf"),
           width = 5,
           height = 50,
           fontsize_row = 2,
           clustering_method = "ward.D2")


```

### Heatmaps per disease

```{r}

plot_data <-
  joined_atlas_comparison %>%
  select(mutual_id, comparison_tissue, species, mutual_tmm) %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id")) %>%
  inner_join(gene_orthologs) %>%
  filter(enssscg_id %in% disease_genes_full$enssscg_id) %>%
  
  
  # mutate(mutual_tmm = ifelse(mutual_tmm < 1, 0, mutual_tmm)) %>%
  
  group_by(enssscg_id, species) %>% 
  mutate(max_tmm = max(mutual_tmm, na.rm = T),
         mutual_tmm = ifelse(max_tmm == 0, 0, mutual_tmm / max_tmm)) %>%
  left_join(human_gene_mapping) 

plot_data2 <-
  gene_overlap_data_summarized 

plot_data3 <-
  human_pig_gene_class_comparison %>% 
  inner_join(plot_data %>% 
               select(enssscg_id, ensg_id) %>% 
               distinct()) %>%
  select(enssscg_id, ensg_id, spec_category_pig, spec_category_human) %>%
  left_join(human_gene_mapping) %>%
  gather(species, spec_category, spec_category_pig, spec_category_human) %>%
  mutate(species = str_to_sentence(gsub("spec_category_", "", species)))


#####

plots <- 
  lapply(sort(unique(disease_genes_full$disease)),
         function(pw) {
           cat(paste(pw, "\n"))
           filter_ensg <- 
             disease_genes_full %>% 
             filter(disease == pw) %>% 
             pull(enssscg_id)
           
           cluster_data <-
             plot_data %>%
             ungroup() %>%
             filter(enssscg_id %in% filter_ensg) %>%
             select(comparison_tissue, species, mutual_tmm, gene_name) %>%
             group_by(gene_name, comparison_tissue) %>%
             summarise(mutual_tmm = mean(mutual_tmm, na.rm = T)) %>%
             spread(gene_name, mutual_tmm) %>%
             column_to_rownames("comparison_tissue")
           
           # cluster_data <-
           #   plot_data %>% 
           #   ungroup() %>%
           #   filter(ensg_id %in% filter_ensg) %>%
           #   select(comparison_tissue, species, mutual_tmm, gene_name) %>%
           #     spread(species, mutual_tmm) %>%
           #   mutate(diff = human - pig) %>% 
           #     select(-human, -pig) %>% 
           #     spread(gene_name, diff) %>%
           #   column_to_rownames("comparison_tissue") 
           
           tis_cluster <- 
             cluster_data %>%
             dist() %>% 
             hclust(method = "ward.D2")
           
           tis_order <- 
             tis_cluster$labels[tis_cluster$order]
           
           if(dim(cluster_data)[2] >= 2) {
             # gene_cluster <- 
             #   cluster_data %>%
             #   t() %>%
             #   dist() %>% 
             #   hclust(method = "ward.D2")
             
             gene_order <- 
               plot_data %>% 
               ungroup() %>%
               filter(enssscg_id %in% filter_ensg) %>%
               select(comparison_tissue, species, mutual_tmm, gene_name) %>%
               spread(species, mutual_tmm) %>%
               group_by(gene_name) %>% 
               summarise(dist = sqrt(sum((human - pig) ^ 2))) %>% 
               arrange(dist) %>% 
               pull(gene_name)
                 
             
           } else {
             gene_cluster <- 
               tibble(labels = colnames(cluster_data),
                      order = 1)
             
             gene_order <- 
               gene_cluster$labels[gene_cluster$order]
             
           }
           
           
           
           # Plot:
           
           plot_data %>% 
             ungroup() %>%
             filter(enssscg_id %in% filter_ensg) %>%
             select(gene_name, species) %>% 
             distinct() %>%
             mutate(species = toupper(substr(species, 1, 1))) %>%
             mutate(gene_name = factor(gene_name, gene_order)) %>%
             
             
             ggplot(aes(1, gene_name, group = species, fill = species, label = species)) +
             geom_tile(show.legend = F, 
                       position = "dodge", 
                       height = 0.9) +
             geom_text(position = position_dodge(width = 1), 
                       size = 2, 
                       fontface = "bold",
                       color = "white") +
             scale_fill_manual(values = c("H" = "#DB1F48",
                                          "P" = "#01949A")) +
             scale_x_discrete(limits = "Species") +
             theme_void() +
             theme(plot.margin = unit(c(0,0,0,0), "mm"),
                   axis.text.x = element_text(angle = 90, hjust = 0, size = 6, vjust = 0.2)) +
             
             # Specificity
             plot_data3 %>%
             filter(enssscg_id %in% filter_ensg) %>%
             mutate(gene_name = factor(gene_name, gene_order)) %>%
             
             ggplot(aes(1, gene_name, group = species, fill = spec_category, label = species)) +
             geom_tile(show.legend = F,
                       position = "dodge", 
                       height = 0.9, width = 1) +
             # geom_text(position = position_dodge(width = 1), 
             #           size = 2, 
             #           fontface = "bold",
             #           color = "white") +
             scale_fill_manual(values = gene_category_pal) +
             scale_x_discrete(limits = "Specificity") +
             theme_void() +
             theme(plot.margin = unit(c(0,0,0,0), "mm"),
                   axis.text.x = element_text(angle = 90, hjust = 0, size = 6, vjust = 0.2)) +

             
             # Overlap
             plot_data2 %>%
             filter(enssscg_id %in% filter_ensg) %>%
             mutate(gene_name = factor(gene_name, gene_order)) %>%
             
             ggplot(aes(1, gene_name, fill = type)) +
             geom_tile(show.legend = F,
                       height = 0.9) +
             scale_fill_manual(values = overlap_type_pal) +
             scale_x_discrete(limits = "Overlap") +
             theme_void() +
             theme(axis.text.x = element_text(angle = 90, hjust = 0, size = 6, vjust = 0.2)) + 
             
             # Heatmap
             plot_data %>% 
             ungroup() %>%
             filter(enssscg_id %in% filter_ensg) %>%
             mutate(gene_name = factor(gene_name, gene_order),
                    comparison_tissue = factor(comparison_tissue, tis_order)) %>%
             
             ggplot(aes(gene_name, comparison_tissue, group = species, fill = mutual_tmm)) +
             geom_tile(position= "dodge", width = 0.9) +
             coord_flip() +
             scale_fill_viridis_c(option = "B", direction = -1, name = "Expression") +
             # facet_grid(enssscg_id ~ .) + 
             # facet_wrap(~gene_name, ncol = 1, strip.position = "left") + 
             stripped_theme_facet +
             theme(axis.text.x = element_text(angle = 60, hjust = 1),
                   strip.placement = "outside",
                   strip.text.y.left = element_text(angle = 0),
                   panel.spacing = unit(0, "mm"),
                   legend.position = "right",
                   panel.border = element_rect(fill = NA, color = "white"),
                   axis.title = element_blank(), 
                   axis.line = element_blank()) +
             theme(plot.margin = unit(c(0,0,0,0), "mm")) +
             ggtitle(pw) +
             
             
             plot_layout(widths = c(0.025, 0.025, 0.025, 1), nrow = 1)
         })
  
plot_heights <- 
  plot_data %>% 
  ungroup() %>% 
  select(enssscg_id) %>% 
  distinct() %>% 
  inner_join(disease_genes_full)  %>%
  group_by(disease) %>%
  summarise(n = n()) %>%
  arrange(disease) %>% 
  mutate(height = 2 + 0.3*n / 2)  
  

pdf_multiple(plots, 
             filepath = savepath("Disease expression heatmaps.pdf"), 
             widths = rep(6, nrow(plot_heights)), 
             heights = plot_heights$height)

```

### Correlation

```{r}


cor_data <- 
  joined_atlas_comparison %>% 
  select(mutual_id, comparison_tissue, species, mutual_tmm) %>% 
  separate(mutual_id, into = c("enssscg_id", "ensg_id")) %>%
  
  inner_join(disease_gene_classes %>% 
               distinct) %>% 
  inner_join(gene_orthologs %>% 
               select(1:6)) %>%
  mutate(mutual_tmm = log10(mutual_tmm + 1)) %>%
  spread(species, mutual_tmm)

disease_gene_cor <-
  cor_data %>% 
  group_by(disease_system, disease_type, disease, 
           enssscg_id, ensg_id, gene_name, 
           ortholog_type, ortholog_confidence) %>% 
  do({
    cor.test(.$human, .$pig, method = "pearson") %>% 
      tidy()
  }) %>%
  ungroup



disease_gene_cor %>% 
  select(-disease_type, -disease) %>% 
  distinct() %>% 
  left_join(gene_overlap_data_summarized) %>%
  group_by(disease_system) %>%
  mutate(median = median(estimate, na.rm = T)) %>% 
  ungroup() %>% 
  arrange(median) %>%
  mutate(disease_system = factor(disease_system, unique(disease_system))) %>%
  ggplot(aes(disease_system, estimate)) +
  geom_boxplot(fill = "lightgray") + 
  # geom_violin(draw_quantiles = 0.5, alpha = 0.5) +
  geom_beeswarm(aes(color = type),
                cex = 0.5, 
                alpha = 0.5,
                size = 1) +
  geom_text(data = . %>% 
              arrange(estimate) %>%
              group_by(disease_system) %>% 
              do({bind_rows(head(., 3),
                            tail(., 3))}),
            aes(label = gene_name)) +
  scale_color_manual(values = overlap_type_pal) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))
 
ggsave(savepath("Pearson per disease.pdf"), width = 16, height = 10)


disease_tis_cor <-
  cor_data %>% 
  filter(!is.na(human) & !is.na(pig)) %>%
  group_by(disease) %>% 
  mutate(n = n_distinct(ensg_id)) %>% 
  ungroup() %>% 
  filter(n >= 5) %>%
  group_by(disease, comparison_tissue) %>% 
  do({
    cor.test(.$human, .$pig, method = "pearson") %>% 
      tidy()
  }) %>%
  ungroup %>% 
  mutate(adj_p = p.adjust(p.value, method = "BH"))

disease_tis_cor %>% 
  mutate(estimate = ifelse(adj_p < 0.05, estimate, 0)) %>%
  select(disease, comparison_tissue, estimate) %>% 
  
  spread(comparison_tissue, estimate) %>% 
  column_to_rownames("disease") %>%
  pheatmap(clustering_method = "ward.D2",
           border_color = NA,
           main = "Pearson correlation of disease in tissues",
           filename = savepath("Pearson per disease and tissue.pdf"),
           width = 9, 
           height = 16
           )


```





## Figure ? - Immunology

###correlation between lymphoid tissues

```{r}

plot_data <- 
  pig_atlas_tissue %>%
  filter(tissue_name %in% c("lymph node", "spleen", "thymus", "tonsil")) %>%
  
  inner_join(gene_orthologs %>% 
               select(1:2)) %>%
  select(enssscg_id, ensg_id, tissue_name, ptpm) %>% 
  left_join(human_atlas_tissue %>%
              filter(tissue_name %in% c("lymph node", "spleen", "thymus", "tonsil")) %>%
              select(1, 2, ptpm),
            by = c("ensg_id", "tissue_name"),
            suffix = c("_pig", "_human"))

plot_data %>%
  gather(species, ptpm, ptpm_pig, ptpm_human) %>%
  mutate(id = paste(gsub("^ptpm_", "", species), tissue_name)) %>%
  select(1, 2, id, ptpm) %>% 
  spread(id, ptpm) %>%
  select(-1, -2) %>%
  cor(use = "pairwise.complete.obs", method = "spearman") %>% 
  {.[which(grepl("human", rownames(.))), which(grepl("pig", rownames(.)))]} %>%
  
  
  pheatmap(clustering_method = "average",
           # cluster_rows = {dist(1 - .)} %>%
           #   hclust(method = "average"),
           # cluster_cols = t(.) %>%
           #   {dist(1 - .)} %>%
           #   hclust(method = "average"),
           cluster_rows = F,
           cluster_cols = F, 
           border_color = "white", 
           display_numbers = T,
           legend = F,
           filename = savepath("Lymphoid pig human spearman heatmap.pdf"),
           width = 3.2, height = 3)


lymph_panther <- 
  read.delim("data/meta/PANTHER15.0_HMM_classifications", header = F) %>%
  as_tibble() %>%
  set_names(c("panther_id",
              "name",
              "molecular_function",
              "biological_process",
              "cellular_components",
              "protein_class",
              "Pathway")) %>%
  filter(grepl("CATHELICIDIN|DEFENSIN|HEPCIDIN", name)) 

# LEAP-2 = LIVER EXPRESSED ANTIMICROBIAL PEPTIDE 2 (LEAP2)
# PGRP-S = PGLYRP1
# NKL - only pig? 
# gene_mapping %>%
#   filter(display_name %in% c("LEAP2", "PGLYRP1", "NKL"))

DEF_human <- 
  human_gene_mapping %>%
  select(ensg_id, gene_name) %>%
  full_join(read_delim("data/meta/ensembl100_panther_families_human.txt", delim = "\t") %>%
              set_colnames(c("ensg_id", "panther_id", "gene_name")) %>%
              inner_join(lymph_panther[, 1:2]) ,
            by = "gene_name",
            suffix = c("","100")) %>%
  filter(!is.na(ensg_id100)) %>%
  filter(!grepl("SF", panther_id)) %>%
  distinct() %>%
  select(ensg_id, gene_name, panther_id, name) %>%
  bind_rows(human_gene_mapping %>%
              filter(gene_name %in% c("LEAP2", "PGLYRP1", "NKL")) %>%
              mutate(panther_id = NA, 
                     name = gene_name) %>%
              select(ensg_id, gene_name, panther_id, name)) %>% 
  distinct()

DEF_pig <- 
  gene_mapping %>%
  select(enssscg_id, gene_name = display_name) %>%
  full_join(read_delim("data/meta/ensembl100_panther_families_pig.txt", delim = "\t") %>%
              set_colnames(c("enssscg_id", "gene_name", "panther_id")) %>%
              inner_join(lymph_panther[, 1:2]) ,
            by = c("enssscg_id"),
            suffix = c("","100")) %>%
  filter(!is.na(panther_id)) %>% 
  
  filter(!grepl("SF", panther_id)) %>%
  distinct() %>%
  select(enssscg_id, gene_name, panther_id, name) %>% 
  bind_rows(gene_mapping %>%
              filter(display_name %in% c("LEAP2", "PGLYRP1", "NKL")) %>%
              mutate(panther_id = NA, 
                     name = display_name) %>%
              select(enssscg_id, gene_name = display_name, panther_id, name)) %>% 
  distinct()


DEF_joined <- 
  full_join(right_join(gene_orthologs %>%
                         select(1:2),
                       DEF_human),
            
            right_join(gene_orthologs %>%
                         select(1:2),
                       DEF_pig),
            
            by = c("enssscg_id", "ensg_id", "panther_id", "name"),
            suffix = c("_human", "_pig")) %>%
  select(enssscg_id, ensg_id, gene_name_pig, gene_name_human, panther_id, name)

plot_levels <- 
  bind_rows(pig_atlas_comparison %>%
              inner_join(DEF_joined) %>%
              select(gene = enssscg_id, comparison_tissue, nx),
            
            human_atlas_comparison %>%
              inner_join(DEF_joined) %>%
              select(gene = ensg_id, comparison_tissue, nx)) %>%
  spread(comparison_tissue, nx) %>% 
  column_to_rownames("gene") %>% 
  {log10(. + 1)} %>% 
  t() %>%
  dist() %>% 
  hclust(method = "ward.D2") %>%
  labels()



pig_atlas_comparison %>%
  inner_join(DEF_joined) %>%
  mutate(comparison_tissue = factor(comparison_tissue, 
                                    levels = plot_levels)) %>%
  
  mutate(gene_name = case_when((gene_name_pig == "NULL" | is.na(gene_name_pig) & 
                                  is.na(gene_name_human)) ~ enssscg_id,
                               !(gene_name_pig == "NULL" | is.na(gene_name_pig)) ~ gene_name_pig,
                               T ~ gene_name_human)) %>%
  
  ggplot(aes(gene_name, comparison_tissue, fill = log10(nx + 1))) +
  geom_tile() +
  facet_grid(~name, scales = "free_x", space = "free",
             labeller = label_wrap_gen()) +
  scale_fill_viridis(option = "B", direction = -1) +
  scale_x_discrete(expand = expansion(0)) +
  stripped_theme_facet + 
  theme(panel.spacing = unit(0.5, "mm"),
        axis.text.x = element_text(angle = 60, hjust = 1),
        axis.title = element_blank(),
        strip.text.x = element_text(angle = 90), 
        legend.position = "top") + 
  ggtitle("Pig") +
  
  human_atlas_comparison %>%
  inner_join(DEF_joined) %>%
  mutate(comparison_tissue = factor(comparison_tissue, 
                                    levels = plot_levels)) %>%
  
  mutate(gene_name = human_gene_mapping$gene_name[match(ensg_id, 
                                                        human_gene_mapping$ensg_id)]) %>%
  
  ggplot(aes(gene_name, comparison_tissue, fill = log10(nx + 1))) +
  geom_tile() +
  facet_grid(~name, scales = "free_x", space = "free",
             labeller = label_wrap_gen()) +
  scale_fill_viridis(option = "B", direction = -1) +
  scale_x_discrete(expand = expansion(0)) +
  stripped_theme_facet + 
  theme(panel.spacing = unit(0.5, "mm"),
        axis.text.x = element_text(angle = 60, hjust = 1),
        axis.title = element_blank(),
        strip.text.x = element_text(angle = 90), 
        legend.position = "top",
        axis.text.y = element_blank()) + 
  ggtitle("Human")
  
ggsave(savepath("HDP human pig expression heatmap.pdf"),
       width = 10, height = 8)


DEF_joined %>%
  select(1:2) %>%
  mutate(type = case_when(!is.na(enssscg_id) & !is.na(ensg_id) ~ "Overlap",
                          !is.na(enssscg_id) ~ "Pig",
                          !is.na(ensg_id) ~ "Human")) %>%
  group_by(type) %>% 
  count() %>%
  ungroup() %>%
  pie_plot(class_col = "type", y_col = "n", pal = overlap_type_pal)
  
ggsave(savepath("HDP overlap pie.pdf"),width = 4, height = 4)

```


####Overlap bipartite network

```{r}

n_conn <- 2

plot_graph <- 
  overlap_hyper_all %>% 
  filter(adj_pval < 0.05) %>%
  group_by(pig_id) %>% 
  mutate(rank1 = rank(adj_pval)) %>%
  group_by(human_id) %>% 
  mutate(rank2 = rank(adj_pval)) %>% 
  ungroup() %>% 
  mutate(keep = rank1 <= n_conn |
           rank2 <= n_conn) %>%
  filter(keep) %>% 
  mutate(weight = -log10(adj_pval) / m) %>%
  arrange(-weight) %>%
  mutate(pig_id = paste("pig", pig_id),
         human_id = paste("human", human_id)) %>%
  select(pig_id, human_id, weight) %>%
  as_tbl_graph() %>%
  mutate(species = str_to_sentence(str_extract(name, "pig|human")),
         comparison_tissue = gsub("pig |human ", "", name)) %>%
  left_join(tissue_mapping %>% 
              select(comparison_tissue,
                     organ_name) %>% 
              distinct())

plot_layout <- 
  plot_graph %>% 
  ggraph::create_layout(layout = "kk", 
                        weight = 1 - .$weight) %>%
  as_tibble()

plot_graph %>% 
  
  ggraph(layout = "manual", x = plot_layout$x, y = plot_layout$y) +

  geom_point(data = tibble(x = rep(range(plot_layout$x) * 1.2, 2),
                           y = rep(range(plot_layout$y) * 1.2, each = 2)),
             aes(x, y),
             color = NA) +
  
  stat_density_2d(geom = "polygon", 
                  aes(x, y, 
                      fill = organ_name),
                  inherit.aes = F,
                  alpha = 0.5,
                  bins = 3, 
                  show.legend = F, 
                  h = rep(1, 2)) +
  
  geom_edge_link(aes(alpha =  weight,
                     width =  weight)) +
  geom_node_point(aes(color = species),
                  size = 5) +
  
  geom_node_text(aes(label = comparison_tissue),
                 size = 2) +
  
  scale_edge_width_continuous(range = c(1, 3)) +
  scale_color_manual(values = overlap_type_pal) +
  scale_fill_manual(values = tissue_colors_palette_full) +
  scale_x_continuous(expand = expansion(0)) +
  scale_y_continuous(expand = expansion(0)) +
  coord_flip() +
  theme_stripped +
  theme(axis.line = element_blank())

ggsave(savepath("Classification overlap network.pdf"), width = 8, height = 8)

```


###TLR & IL

```{r}

TLR_genes <- 
  read_csv("data/meta/pig_immunology_TLR_list.csv")  %>%
  select(2) %>%
  left_join(gene_orthologs,
            by = c("Approved symbol" = "human_gene_name")) %>%
  filter(!is.na(enssscg_id))

IL_genes <-
  read_csv("data/meta/pig_immunology_IL_list.csv") %>%
  select(2) %>%
  left_join(gene_orthologs,
            by = c("Approved symbol" = "human_gene_name")) %>%
  filter(!is.na(enssscg_id))



# TLR

plot_data <- 
  joined_atlas_comparison %>% 
  separate(mutual_id, into = c("enssscg_id", "ensg_id")) %>% 
  inner_join(TLR_genes) %>%
  group_by(enssscg_id, species) %>% 
  mutate(max_tmm = max(mutual_tmm, na.rm = T),
         mutual_tmm1 = mutual_tmm,
         mutual_tmm = ifelse(max_tmm == 0, 0, mutual_tmm / max_tmm)) %>%
  left_join(human_gene_mapping) 

plot_data_frac <- 
  plot_data %>%
  select(comparison_tissue, species, ptpm, gene_name, enssscg_id) %>%
  group_by(comparison_tissue, species) %>%
  mutate(frac = 100 * ptpm / sum(ptpm),
         label = ifelse(frac >= 10, paste0(round(frac, 1), "%"), NA)) %>%
  
  left_join(select(., -label, -ptpm) %>% 
              spread(species, frac) %>%
              group_by(comparison_tissue) %>%
              summarise(cor = cor(human, pig, method = "spearman")) %>%
              select(comparison_tissue, cor)) %>%
  ungroup() %>%
  arrange(cor) %>% 
  mutate(comparison_tissue = factor(comparison_tissue, unique(comparison_tissue)),
         species = factor(species, c("human", "pig"))) 

plot_data_frac %>%
  
  
  {ggplot(., aes(as.numeric(species), frac, fill = gene_name, label = label)) +
      geom_col(color = "white",
               size = 0.3,
               width = 0.55) +
      geom_area(aes(scales::rescale(as.numeric(species), c(1.25, 1.75))),
                size = 0.3,
                color = "white",
                alpha = 1) +
      geom_text(size = 2,
                position = position_stack(vjust = 0.5), color = "white") +
      facet_wrap(~comparison_tissue, nrow = 4, scales = "free_x") +
      scale_x_discrete(labels = levels(.$species),
                       limits=c("1","2"),
                       expand = expansion(0)) +
      scale_y_continuous(expand = expansion(c(0, 0.05))) +
      scale_fill_manual(values = colorRampPalette(ggthemes::tableau_color_pal(palette = "Hue Circle", 
                                                                              type = c("regular"), 
                                                                              direction = 1)(19))(n_distinct(plot_data$gene_name))) + 
      # directlabels::geom_dl(aes(label = label), list("top.points", cex = .6)) +
      stripped_theme_facet +
      ylab("Fraction of expression") +
      theme(panel.border = element_blank(),
            axis.title.x = element_blank())}
ggsave(savepath("TLR species fraction bar.pdf"), width = 10, height = 7)

plot_data_frac %>%
  unite(id, comparison_tissue, species) %>%
  select(id, frac, gene_name) %>% 
  spread(id, frac) %>% 
  column_to_rownames("gene_name") %>% 
  pheatmap(clustering_method = "ward.D2",
           border_color = NA,
           annotation_col = colnames(.) %>% 
             enframe() %>%
             separate(value, into = c("Tissue", "Species"), sep = "_", remove = F) %>% 
             mutate(Species = str_to_sentence(Species)) %>%
             select(-name, -Tissue) %>% 
             column_to_rownames("value"),
           annotation_colors = list(Species = overlap_type_pal[c(1,3)]),
           filename = savepath("TLR frac heatmap.pdf"),
           width = 10, 
           height = 4)

plot_data %>%
  ungroup() %>%
  filter(!comparison_tissue %in% c( "bone marrow", "retina")) %>%
  unite(id, comparison_tissue, species) %>%
  select(id, mutual_tmm1, gene_name) %>% 
  mutate(mutual_tmm1 = ifelse(mutual_tmm1 < 1, 0 , mutual_tmm1)) %>%
  spread(id, mutual_tmm1) %>% 
  column_to_rownames("gene_name") %>% 
  {log10(. + 1)} %>%
  pheatmap(clustering_method = "ward.D2",
           border_color = NA,
           annotation_col = colnames(.) %>% 
             enframe() %>%
             separate(value, into = c("Tissue", "Species"), sep = "_", remove = F) %>% 
             mutate(Species = str_to_sentence(Species)) %>%
             select(-name, -Tissue) %>% 
             column_to_rownames("value"),
           annotation_colors = list(Species = overlap_type_pal[c(1,3)]),
           filename = savepath("TLR log tmm heatmap.pdf"),
           width = 10, 
           height = 8)

plot_data %>% 
  group_by(comparison_tissue, species) %>% 
  summarise(expression = sum(mutual_tmm1)) %>%
  {ggplot(., aes(species, expression)) +
      geom_violin(draw_quantiles = 0.5) +
      geom_path(aes(group = comparison_tissue), 
                alpha = 0.2) +
      geom_point(aes(color = comparison_tissue),
                 show.legend = F) +
      group_by(., comparison_tissue) %>%
      top_n(1, expression) %>%
      geom_text(data = ., 
                aes(label = comparison_tissue),
                vjust = 0,
                hjust = ifelse(.$species == "human",
                               1, 0),
                size = 2) +
      scale_color_manual(values = tissue_colors_palette_comparison) + 
      stripped_theme}
ggsave(savepath("TLR violin.pdf"), 
       width = 4, height = 5)


plot_data2 <-
  gene_overlap_data_summarized %>%
  inner_join(TLR_genes) 

plot_data3 <-
  human_pig_gene_class_comparison %>% 
  inner_join(plot_data %>% 
               select(enssscg_id, ensg_id) %>% 
               distinct()) %>%
  select(enssscg_id, ensg_id, spec_category_pig, spec_category_human) %>%
  left_join(human_gene_mapping) %>%
  gather(species, spec_category, spec_category_pig, spec_category_human) %>%
  mutate(species = str_to_sentence(gsub("spec_category_", "", species))) %>%
  left_join(TLR_genes)


#####


cluster_data <-
  plot_data %>%
  ungroup() %>%
  select(comparison_tissue, species, mutual_tmm, gene_name) %>%
  group_by(gene_name, comparison_tissue) %>%
  summarise(mutual_tmm = mean(mutual_tmm)) %>%
  spread(gene_name, mutual_tmm) %>%
  column_to_rownames("comparison_tissue")


tis_cluster <- 
  cluster_data %>%
  dist() %>% 
  hclust(method = "ward.D2")

tis_order <- 
  tis_cluster$labels[tis_cluster$order]

gene_order <- 
  plot_data %>% 
  ungroup() %>%
  select(comparison_tissue, species, mutual_tmm, gene_name) %>%
  spread(species, mutual_tmm) %>%
  group_by(gene_name) %>% 
  summarise(dist = sqrt(sum((human - pig) ^ 2))) %>% 
  arrange(dist) %>% 
  pull(gene_name)
  
  


# Plot:


plot_data %>% 
  ungroup() %>%
  select(gene_name, species) %>% 
  distinct() %>%
  mutate(species = toupper(substr(species, 1, 1))) %>%
  mutate(gene_name = factor(gene_name, gene_order)) %>%
  
  
  ggplot(aes(1, gene_name, group = species, fill = species, label = species)) +
  geom_tile(show.legend = F, 
            position = "dodge", 
            height = 0.9) +
  geom_text(position = position_dodge(width = 1), 
            size = 2, 
            fontface = "bold",
            color = "white") +
  scale_fill_manual(values = c("H" = "#DB1F48",
                               "P" = "#01949A")) +
  # scale_x_discrete(limits = "Species") +
  theme_void() +
  theme(plot.margin = unit(c(0,0,0,0), "mm"),
        axis.text.x = element_blank()) +
  
  # Specificity
  plot_data3 %>%
  mutate(gene_name = factor(gene_name, gene_order)) %>%
  
  ggplot(aes(1, gene_name, group = species, fill = spec_category, label = species)) +
  geom_tile(show.legend = F,
            position = "dodge", 
            height = 0.9, width = 1) +
  # geom_text(position = position_dodge(width = 1), 
  #           size = 2, 
  #           fontface = "bold",
  #           color = "white") +
  scale_fill_manual(values = gene_category_pal) +
  # scale_x_discrete(limits = "Specificity") +
  theme_void() +
  theme(plot.margin = unit(c(0,0,0,0), "mm"),
        axis.text.y = element_blank()) +
  
  
  # Overlap
  plot_data2 %>%
  mutate(gene_name = factor(gene_name, gene_order)) %>%
  
  ggplot(aes(1, gene_name, fill = type)) +
  geom_tile(show.legend = F,
            height = 0.9) +
  scale_fill_manual(values = overlap_type_pal) +
  # scale_x_discrete(limits = "Overlap") +
  theme_void() +
  theme(axis.text.x = element_blank()) + 
  
  # Heatmap
  plot_data %>% 
  ungroup() %>%
  mutate(gene_name = factor(gene_name, gene_order),
         comparison_tissue = factor(comparison_tissue, tis_order)) %>%
  
  ggplot(aes(gene_name, comparison_tissue, group = species, fill = mutual_tmm)) +
  geom_tile(position= "dodge", width = 0.9) +
  coord_flip() +
  scale_fill_viridis_c(option = "B", direction = -1, name = "Expression") +
  # facet_grid(ensg_id ~ .) + 
  # facet_wrap(~gene_name, ncol = 1, strip.position = "left") + 
  stripped_theme_facet +
  theme(axis.text.x = element_text(angle = 60, hjust = 1),
        strip.placement = "outside",
        strip.text.y.left = element_text(angle = 0),
        panel.spacing = unit(0, "mm"),
        legend.position = "right",
        panel.border = element_rect(fill = NA, color = "white"),
        axis.title = element_blank(), 
        axis.line = element_blank()) +
  theme(plot.margin = unit(c(0,0,0,0), "mm")) +
  
  
  
  plot_layout(widths = c(0.025, 0.025, 0.025, 1), nrow = 1)

ggsave(savepath("Fig7 TLR heatmap.pdf"),
       width = 6, height = 2 + 0.15 * n_distinct(plot_data$ensg_id))







# IL

plot_data <- 
  joined_atlas_comparison %>% 
  separate(mutual_id, into = c("enssscg_id", "ensg_id")) %>% 
  inner_join(IL_genes) %>%
  group_by(enssscg_id, species) %>% 
  mutate(max_tmm = max(mutual_tmm, na.rm = T),
         mutual_tmm1 = mutual_tmm,
         mutual_tmm = ifelse(max_tmm == 0, 0, mutual_tmm / max_tmm)) %>%
  left_join(human_gene_mapping) 

plot_data_frac <- 
  plot_data %>%
  select(comparison_tissue, species, ptpm, gene_name, enssscg_id) %>%
  group_by(comparison_tissue, species) %>%
  mutate(frac = 100 * ptpm / sum(ptpm, na.rm = T),
         label = ifelse(frac >= 10, paste0(gene_name, "\n", round(frac, 1), "%"), NA)) %>%
  
  left_join(select(., -label, -ptpm) %>% 
              spread(species, frac) %>%
              group_by(comparison_tissue) %>%
              summarise(cor = cor(human, pig, method = "spearman")) %>%
              select(comparison_tissue, cor)) %>%
  ungroup() %>%
  arrange(cor) %>% 
  mutate(comparison_tissue = factor(comparison_tissue, unique(comparison_tissue)),
         species = factor(species, c("human", "pig"))) 

plot_data_frac %>%
  
  
  {ggplot(., aes(as.numeric(species), frac, fill = gene_name, label = label)) +
      geom_col(color = "white",
               size = 0.3,
               width = 0.55) +
      geom_area(aes(scales::rescale(as.numeric(species), c(1.25, 1.75))),
                size = 0.3,
                color = "white",
                alpha = 1) +
      geom_text(size = 2,
                position = position_stack(vjust = 0.5), color = "white") +
      facet_wrap(~comparison_tissue, nrow = 4, scales = "free_x") +
      scale_x_discrete(labels = levels(.$species),
                       limits=c("1","2"),
                       expand = expansion(0)) +
      scale_y_continuous(expand = expansion(c(0, 0.05))) +
      scale_fill_manual(values = colorRampPalette(ggthemes::tableau_color_pal(palette = "Hue Circle", 
                                                                              type = c("regular"), 
                                                                              direction = 1)(19))(n_distinct(plot_data$gene_name))) + 
      # directlabels::geom_dl(aes(label = label), list("top.points", cex = .6)) +
      stripped_theme_facet +
      ylab("Fraction of expression") +
      theme(panel.border = element_blank(),
            axis.title.x = element_blank())}
ggsave(savepath("IL species fraction bar.pdf"), width = 10, height = 7)

plot_data_frac %>%
  unite(id, comparison_tissue, species) %>%
  select(id, frac, gene_name) %>% 
  spread(id, frac) %>% 
  column_to_rownames("gene_name") %>% 
  pheatmap(clustering_method = "ward.D2",
           border_color = NA,
           annotation_col = colnames(.) %>% 
             enframe() %>%
             separate(value, into = c("Tissue", "Species"), sep = "_", remove = F) %>% 
             mutate(Species = str_to_sentence(Species)) %>%
             select(-name, -Tissue) %>% 
             column_to_rownames("value"),
           annotation_colors = list(Species = overlap_type_pal[c(1,3)]),
           filename = savepath("IL frac heatmap.pdf"),
           width = 10, 
           height = 8)

plot_data %>%
  ungroup() %>%
  filter(!comparison_tissue %in% c( "bone marrow", "retina")) %>%
  unite(id, comparison_tissue, species) %>%
  select(id, mutual_tmm1, gene_name) %>% 
  mutate(mutual_tmm1 = ifelse(mutual_tmm1 < 1, 0 , mutual_tmm1)) %>%
  spread(id, mutual_tmm1) %>% 
  column_to_rownames("gene_name") %>% 
  {log10(. + 1)} %>%
  pheatmap(clustering_method = "ward.D2",
           border_color = NA,
           annotation_col = colnames(.) %>% 
             enframe() %>%
             separate(value, into = c("Tissue", "Species"), sep = "_", remove = F) %>% 
             mutate(Species = str_to_sentence(Species)) %>%
             select(-name, -Tissue) %>% 
             column_to_rownames("value"),
           annotation_colors = list(Species = overlap_type_pal[c(1,3)]),
           filename = savepath("IL log tmm heatmap.pdf"),
           width = 10, 
           height = 8)

plot_data %>% 
  group_by(comparison_tissue, species) %>% 
  summarise(expression = sum(mutual_tmm1)) %>%
  {ggplot(., aes(species, expression)) +
      geom_violin(draw_quantiles = 0.5) +
      geom_path(aes(group = comparison_tissue), 
                alpha = 0.2) +
      geom_point(aes(color = comparison_tissue),
                 show.legend = F) +
      group_by(., comparison_tissue) %>%
      top_n(1, expression) %>%
      geom_text(data = ., 
                aes(label = comparison_tissue),
                vjust = 0,
                hjust = ifelse(.$species == "human",
                               1, 0),
                size = 2) +
      scale_color_manual(values = tissue_colors_palette_comparison) + 
      stripped_theme}
ggsave(savepath("IL violin.pdf"), 
       width = 4, height = 10)

plot_data2 <-
  gene_overlap_data_summarized %>%
  inner_join(IL_genes) 

plot_data3 <-
  human_pig_gene_class_comparison %>% 
  inner_join(plot_data %>% 
               select(enssscg_id, ensg_id) %>% 
               distinct()) %>%
  select(enssscg_id, ensg_id, spec_category_pig, spec_category_human) %>%
  left_join(human_gene_mapping) %>%
  gather(species, spec_category, spec_category_pig, spec_category_human) %>%
  mutate(species = str_to_sentence(gsub("spec_category_", "", species))) %>%
  left_join(IL_genes)


#####


cluster_data <-
  plot_data %>%
  ungroup() %>%
  select(comparison_tissue, species, mutual_tmm, gene_name) %>%
  group_by(gene_name, comparison_tissue) %>%
  summarise(mutual_tmm = mean(mutual_tmm)) %>%
  spread(gene_name, mutual_tmm) %>%
  column_to_rownames("comparison_tissue")


tis_cluster <- 
  cluster_data %>%
  dist() %>% 
  hclust(method = "ward.D2")

tis_order <- 
  tis_cluster$labels[tis_cluster$order]

gene_order <- 
  plot_data %>% 
  ungroup() %>%
  select(comparison_tissue, species, mutual_tmm, gene_name) %>%
  spread(species, mutual_tmm) %>%
  group_by(gene_name) %>% 
  summarise(dist = sqrt(sum((human - pig) ^ 2))) %>% 
  arrange(dist) %>% 
  pull(gene_name)
  
  


# Plot:


plot_data %>% 
  ungroup() %>%
  select(gene_name, species) %>% 
  distinct() %>%
  mutate(species = toupper(substr(species, 1, 1))) %>%
  mutate(gene_name = factor(gene_name, gene_order)) %>%
  
  
  ggplot(aes(1, gene_name, group = species, fill = species, label = species)) +
  geom_tile(show.legend = F, 
            position = "dodge", 
            height = 0.9) +
  geom_text(position = position_dodge(width = 1), 
            size = 2, 
            fontface = "bold",
            color = "white") +
  scale_fill_manual(values = c("H" = "#DB1F48",
                               "P" = "#01949A")) +
  # scale_x_discrete(limits = "Species") +
  theme_void() +
  theme(plot.margin = unit(c(0,0,0,0), "mm"),
        axis.text.x = element_blank()) +
  
  # Specificity
  plot_data3 %>%
  mutate(gene_name = factor(gene_name, gene_order)) %>%
  
  ggplot(aes(1, gene_name, group = species, fill = spec_category, label = species)) +
  geom_tile(show.legend = F,
            position = "dodge", 
            height = 0.9, width = 1) +
  # geom_text(position = position_dodge(width = 1), 
  #           size = 2, 
  #           fontface = "bold",
  #           color = "white") +
  scale_fill_manual(values = gene_category_pal) +
  # scale_x_discrete(limits = "Specificity") +
  theme_void() +
  theme(plot.margin = unit(c(0,0,0,0), "mm"),
        axis.text.y = element_blank()) +
  
  
  # Overlap
  plot_data2 %>%
  mutate(gene_name = factor(gene_name, gene_order)) %>%
  
  ggplot(aes(1, gene_name, fill = type)) +
  geom_tile(show.legend = F,
            height = 0.9) +
  scale_fill_manual(values = overlap_type_pal) +
  # scale_x_discrete(limits = "Overlap") +
  theme_void() +
  theme(axis.text.x = element_blank()) + 
  
  # Heatmap
  plot_data %>% 
  ungroup() %>%
  mutate(gene_name = factor(gene_name, gene_order),
         comparison_tissue = factor(comparison_tissue, tis_order)) %>%
  
  ggplot(aes(gene_name, comparison_tissue, group = species, fill = mutual_tmm)) +
  geom_tile(position= "dodge", width = 0.9) +
  coord_flip() +
  scale_fill_viridis_c(option = "B", direction = -1, name = "Expression") +
  # facet_grid(ensg_id ~ .) + 
  # facet_wrap(~gene_name, ncol = 1, strip.position = "left") + 
  stripped_theme_facet +
  theme(axis.text.x = element_text(angle = 60, hjust = 1),
        strip.placement = "outside",
        strip.text.y.left = element_text(angle = 0),
        panel.spacing = unit(0, "mm"),
        legend.position = "right",
        panel.border = element_rect(fill = NA, color = "white"),
        axis.title = element_blank(), 
        axis.line = element_blank()) +
  theme(plot.margin = unit(c(0,0,0,0), "mm")) +
  
  
  
  plot_layout(widths = c(0.025, 0.025, 0.025, 1), nrow = 1)

ggsave(savepath("Fig7 IL heatmap.pdf"),
       width = 6, height = 2 + 0.15 * n_distinct(plot_data$ensg_id))



```

### Blood cell composition

```{r}



# selected markers

selected_blood_markers <- 
  readxl::read_excel("data/meta/High sensitivity immune cell markers.xlsx") %>%
  select(gene_name = 2, 
         cell_type = 3) %>%
  inner_join(human_gene_mapping %>% 
               select(gene_name, ensg_id)) 
  

plot_data <- 
  selected_blood_markers %>%
  inner_join(gene_orthologs) %>%
  inner_join(joined_atlas_comparison %>% 
               filter(!comparison_tissue %in% c("bone marrow", "retina")) %>%
               select(mutual_id, comparison_tissue, species, mutual_tmm) %>%
               separate(mutual_id, into = c("enssscg_id", "ensg_id"))) %>%
  group_by(ensg_id, species, cell_type) %>% 
  mutate(frac = mutual_tmm / sum(mutual_tmm, na.rm = T)) 

plot_data %>%
  ggplot(aes(frac, comparison_tissue, fill = species)) +
  geom_boxplot() +
  facet_wrap(~cell_type, scales = "free")
  

plot_data %>%
  group_by(cell_type, comparison_tissue, species) %>% 
  summarise(frac = mean(frac, na.rm = T)) %>%
  ungroup() %>%
  left_join(tissue_mapping %>% 
              select(comparison_tissue, organ_name) %>% 
              distinct()) %>%
  arrange(organ_name) %>%
  mutate(frac = frac * 100,
         label = ifelse(frac >= 10, 
                        paste0(comparison_tissue, "\n", 
                               round(frac, 1), "%"),
                        NA),
         species = factor(species, c("human", "pig")),
         comparison_tissue = factor(comparison_tissue, unique(comparison_tissue))) %>% 
  
  
  {ggplot(., aes(as.numeric(species), frac, fill = comparison_tissue, label = label)) +
      geom_col(color = "white",
               size = 0.3,
               width = 0.55) +
      geom_area(aes(scales::rescale(as.numeric(species), c(1.25, 1.75))),
                size = 0.3,
                color = "white",
                alpha = 1) +
      geom_text(size = 2,
                position = position_stack(vjust = 0.5), color = "white") +
      facet_wrap(~cell_type, nrow = 4, scales = "free_x") +
      scale_x_discrete(labels = levels(.$species),
                       limits=c("1","2"),
                       expand = expansion(0)) +
      scale_y_continuous(expand = expansion(c(0, 0.05))) +
      # scale_fill_manual(values = colorRampPalette(ggthemes::tableau_color_pal(palette = "Hue Circle", 
      #                                                                         type = c("regular"), 
      #                                                                         direction = 1)(19))(n_distinct(plot_data$comparison_tissue))) + 
      scale_fill_manual(values = tissue_colors_palette_comparison) +
      # directlabels::geom_dl(aes(label = label), list("top.points", cex = .6)) +
      stripped_theme_facet +
      ylab("Fraction of expression") +
      theme(panel.border = element_blank(),
            axis.title.x = element_blank())}
ggsave(savepath("blood decomposition selected markers bar.pdf"),
       width = 12, height = 10)




tibble(gene = rep(c("GENE1", "GENE2"), 2),
       species = rep(c("human", "pig"), each = 2),
       frac = c(0.7, 0.3, 0.1, 0.9)) %>%
  mutate(species = factor(species)) %>%
   
                  
  {ggplot(., aes(as.numeric(species), frac, fill = gene, label = gene)) +
      geom_col(color = "white",
               size = 0.3,
               width = 0.55,
               show.legend = F) +
      geom_area(aes(scales::rescale(as.numeric(species), c(1.25, 1.75))),
                size = 0.3,
                color = "white",
                alpha = 1,
                show.legend = F) +
      geom_text(data = . %>%
                  mutate(y = scales::rescale(unclass(factor(gene)), c(1,0))) %>%
                  select(gene, y) %>% 
                  distinct() %>% 
                  arrange(y), 
                aes(x = 1.5, y = y),
                size = 3,
                vjust = c(-1, 2),
                # position = position_stack(vjust = 0.5), 
                color = "white") +
      scale_x_discrete(labels = levels(.$species),
                       limits=c("1","2"),
                       expand = expansion(0)) +
      scale_y_continuous(expand = expansion(c(0, 0.05))) +
      scale_fill_manual(values = c("gray50", "gray20")) +
      ylab("Fraction of expression") +
      theme(panel.border = element_blank(),
            axis.title.x = element_blank(),
            panel.grid = element_blank(),
            panel.background = element_blank())}
ggsave(savepath("Gen ratio example plot.pdf"), width = 2, height = 3)

# Lymphoid tissues - selected markers
plot_data1 <- 
  gene_orthologs %>%
  select(1:2, 4) %>% 
  left_join(pig_atlas_tissue %>% 
              filter(tissue_name %in% c("lymph node",
                                        "spleen",
                                        "thymus",
                                        "tonsil")) %>% 
              select(enssscg_id, tissue_name, pig = ptpm)) %>% 
  left_join(human_atlas_tissue %>% 
              filter(tissue_name %in% c("lymph node",
                                        "spleen",
                                        "thymus",
                                        "tonsil")) %>% 
              select(ensg_id, tissue_name, human = ptpm), 
            by = c("ensg_id", "tissue_name")) %>%
  gather(species, ptpm, human, pig) 
            

plot_data <- 
  selected_blood_markers %>%
  inner_join(gene_orthologs) %>%
  inner_join(plot_data1) %>%
  group_by(ensg_id, species, cell_type) %>% 
  mutate(frac = ptpm / sum(ptpm, na.rm = T)) 

plot_data %>%
  ggplot(aes(frac, tissue_name, fill = species)) +
  geom_boxplot() +
  facet_wrap(~cell_type, scales = "free")
  

plot_data %>%
  group_by(cell_type, tissue_name, species) %>% 
  summarise(frac = mean(frac, na.rm = T)) %>%
  ungroup() %>%
  left_join(tissue_mapping %>% 
              select(tissue_name, organ_name) %>% 
              distinct()) %>%
  arrange(organ_name) %>%
  mutate(frac = frac * 100,
         label = ifelse(frac >= 10, 
                        paste0(tissue_name, "\n", 
                               round(frac, 1), "%"),
                        NA),
         species = factor(species, c("human", "pig")),
         tissue_name = factor(tissue_name, unique(tissue_name))) %>% 
  
  
  {ggplot(., aes(as.numeric(species), frac, fill = tissue_name, label = label)) +
      geom_col(color = "white",
               size = 0.3,
               width = 0.55) +
      geom_area(aes(scales::rescale(as.numeric(species), c(1.25, 1.75))),
                size = 0.3,
                color = "white",
                alpha = 1) +
      geom_text(size = 2,
                position = position_stack(vjust = 0.5), color = "white") +
      facet_wrap(~cell_type, nrow = 4, scales = "free_x") +
      scale_x_discrete(labels = levels(.$species),
                       limits=c("1","2"),
                       expand = expansion(0)) +
      scale_y_continuous(expand = expansion(c(0, 0.05))) +
      # scale_fill_manual(values = colorRampPalette(ggthemes::tableau_color_pal(palette = "Hue Circle",
      #                                                                         type = c("regular"),
      #                                                                         direction = 1)(19))(n_distinct(plot_data$tissue_name))) +
      scale_fill_manual(values = prismatic::color(colorRampPalette(c("black", "#A1A8AA", "white"))(9)[2:6])) +
      # directlabels::geom_dl(aes(label = label), list("top.points", cex = .6)) +
      stripped_theme_facet +
      ylab("Fraction of expression") +
      theme(panel.border = element_blank(),
            axis.title.x = element_blank())}
ggsave(savepath("blood decomposition selected markers lymphoid bar.pdf"),
       width = 12, height = 10)



# cell enriched - blood cell
plot_data <- 
  human_gene_class_blood %>%
  filter(specificity_category %in% c("Tissue enriched")) %>%
  select(ensg_id, enhanced_tissues, ts_score) %>% 
  left_join(human_gene_class %>%
              select(ensg_id, enhanced_tissues),
            by = "ensg_id",
            suffix = c("", "_tissue")) %>% 
  separate_rows(enhanced_tissues_tissue, sep = ";") %>%
  mutate(blood = enhanced_tissues_tissue %in% c("blood", "lymphoid tissue", "bone marrow")) %>%
  group_by(ensg_id) %>% 
  mutate(only_blood = all(blood)) %>%
  ungroup() %>% 
  filter(only_blood) %>%
  select(1:3) %>%
  distinct() %>%
  inner_join(gene_orthologs) %>%
  inner_join(joined_atlas_comparison %>% 
               filter(!comparison_tissue %in% c("bone marrow", "retina")) %>%
               select(mutual_id, comparison_tissue, species, mutual_tmm) %>%
               separate(mutual_id, into = c("enssscg_id", "ensg_id"))) %>%
  group_by(ensg_id, species, enhanced_tissues) %>% 
  mutate(frac = mutual_tmm / sum(mutual_tmm, na.rm = T)) 

plot_data %>%
  ggplot(aes(frac, comparison_tissue, fill = species)) +
  geom_boxplot() +
  facet_wrap(~enhanced_tissues, scales = "free")
  

plot_data %>%
  group_by(enhanced_tissues, comparison_tissue, species) %>% 
  summarise(frac = mean(frac, na.rm = T)) %>%
  ungroup() %>%
  left_join(tissue_mapping %>% 
              select(comparison_tissue, organ_name) %>% 
              distinct()) %>%
  arrange(organ_name) %>%
  mutate(frac = frac * 100,
         label = ifelse(frac >= 10, 
                        paste0(comparison_tissue, "\n", 
                               round(frac, 1), "%"),
                        NA),
         species = factor(species, c("human", "pig")),
         comparison_tissue = factor(comparison_tissue, unique(comparison_tissue))) %>% 
  
  
  {ggplot(., aes(as.numeric(species), frac, fill = comparison_tissue, label = label)) +
      geom_col(color = "white",
               size = 0.3,
               width = 0.55) +
      geom_area(aes(scales::rescale(as.numeric(species), c(1.25, 1.75))),
                size = 0.3,
                color = "white",
                alpha = 1) +
      geom_text(size = 2,
                position = position_stack(vjust = 0.5), color = "white") +
      facet_wrap(~enhanced_tissues, nrow = 4, scales = "free_x") +
      scale_x_discrete(labels = levels(.$species),
                       limits=c("1","2"),
                       expand = expansion(0)) +
      scale_y_continuous(expand = expansion(c(0, 0.05))) +
      # scale_fill_manual(values = colorRampPalette(ggthemes::tableau_color_pal(palette = "Hue Circle", 
      #                                                                         type = c("regular"), 
      #                                                                         direction = 1)(19))(n_distinct(plot_data$comparison_tissue))) + 
      scale_fill_manual(values = tissue_colors_palette_comparison) +
      # directlabels::geom_dl(aes(label = label), list("top.points", cex = .6)) +
      stripped_theme_facet +
      ylab("Fraction of expression") +
      theme(panel.border = element_blank(),
            axis.title.x = element_blank())}
ggsave(savepath("blood decomposition cell enriched comparison bar.pdf"),
       width = 12, height = 10)


# cell or group enriched - blood cell
plot_data <- 
  human_gene_class_blood %>%
  filter(specificity_category %in% c("Tissue enriched", "Group enriched")) %>%
  select(ensg_id, enhanced_tissues, ts_score) %>% 
  separate_rows(enhanced_tissues, sep = ",") %>%
  left_join(human_gene_class %>%
              select(ensg_id, enhanced_tissues),
            by = "ensg_id",
            suffix = c("", "_tissue")) %>% 
  separate_rows(enhanced_tissues_tissue, sep = ";") %>%
  mutate(blood = enhanced_tissues_tissue %in% c("blood", "lymphoid tissue", "bone marrow")) %>%
  group_by(ensg_id) %>% 
  mutate(only_blood = all(blood)) %>%
  ungroup() %>% 
  filter(only_blood) %>%
  select(1:3) %>%
  distinct() %>%
  inner_join(gene_orthologs) %>%
  inner_join(joined_atlas_comparison %>% 
               filter(!comparison_tissue %in% c("bone marrow", "retina")) %>%
               select(mutual_id, comparison_tissue, species, mutual_tmm) %>%
               separate(mutual_id, into = c("enssscg_id", "ensg_id"))) %>%
  group_by(ensg_id, species, enhanced_tissues) %>% 
  mutate(frac = mutual_tmm / sum(mutual_tmm, na.rm = T)) 

plot_data %>%
  ggplot(aes(frac, comparison_tissue, fill = species)) +
  geom_boxplot() +
  facet_wrap(~enhanced_tissues, scales = "free")
  

plot_data %>%
  group_by(enhanced_tissues, comparison_tissue, species) %>% 
  summarise(frac = mean(frac, na.rm = T)) %>%
  ungroup() %>%
  left_join(tissue_mapping %>% 
              select(comparison_tissue, organ_name) %>% 
              distinct()) %>%
  arrange(organ_name) %>%
  mutate(frac = frac * 100,
         label = ifelse(frac >= 10, 
                        paste0(comparison_tissue, "\n", 
                               round(frac, 1), "%"),
                        NA),
         species = factor(species, c("human", "pig")),
         comparison_tissue = factor(comparison_tissue, unique(comparison_tissue))) %>% 
  
  
  {ggplot(., aes(as.numeric(species), frac, fill = comparison_tissue, label = label)) +
      geom_col(color = "white",
               size = 0.3,
               width = 0.55) +
      geom_area(aes(scales::rescale(as.numeric(species), c(1.25, 1.75))),
                size = 0.3,
                color = "white",
                alpha = 1) +
      geom_text(size = 2,
                position = position_stack(vjust = 0.5), color = "white") +
      facet_wrap(~enhanced_tissues, nrow = 4, scales = "free_x") +
      scale_x_discrete(labels = levels(.$species),
                       limits=c("1","2"),
                       expand = expansion(0)) +
      scale_y_continuous(expand = expansion(c(0, 0.05))) +
      # scale_fill_manual(values = colorRampPalette(ggthemes::tableau_color_pal(palette = "Hue Circle", 
      #                                                                         type = c("regular"), 
      #                                                                         direction = 1)(19))(n_distinct(plot_data$comparison_tissue))) + 
      scale_fill_manual(values = tissue_colors_palette_comparison) +
      # directlabels::geom_dl(aes(label = label), list("top.points", cex = .6)) +
      stripped_theme_facet +
      ylab("Fraction of expression") +
      theme(panel.border = element_blank(),
            axis.title.x = element_blank())}
ggsave(savepath("blood decomposition cell or group enriched comparison bar.pdf"),
       width = 12, height = 10)

# cell enriched - blood lineage
plot_data <- 
  human_gene_class_bloodlineage %>%
  filter(specificity_category %in% c("Tissue enriched")) %>%
  select(ensg_id, enhanced_tissues, ts_score) %>% 
  left_join(human_gene_class %>%
              select(ensg_id, enhanced_tissues),
            by = "ensg_id",
            suffix = c("", "_tissue")) %>% 
  separate_rows(enhanced_tissues_tissue, sep = ";") %>%
  mutate(blood = enhanced_tissues_tissue %in% c("blood", "lymphoid tissue", "bone marrow")) %>%
  group_by(ensg_id) %>% 
  mutate(only_blood = all(blood)) %>%
  ungroup() %>% 
  filter(only_blood) %>%
  select(1:3) %>%
  distinct() %>%
  inner_join(gene_orthologs) %>%
  inner_join(joined_atlas_comparison %>% 
               filter(!comparison_tissue %in% c("bone marrow", "retina")) %>%
               select(mutual_id, comparison_tissue, species, mutual_tmm) %>%
               separate(mutual_id, into = c("enssscg_id", "ensg_id"))) %>%
  group_by(ensg_id, species, enhanced_tissues) %>% 
  mutate(frac = mutual_tmm / sum(mutual_tmm, na.rm = T)) 

plot_data %>%
  ggplot(aes(frac, comparison_tissue, fill = species)) +
  geom_boxplot() +
  facet_wrap(~enhanced_tissues, scales = "free")
  

plot_data %>%
  group_by(enhanced_tissues, comparison_tissue, species) %>% 
  summarise(frac = mean(frac, na.rm = T)) %>%
  ungroup() %>%
  left_join(tissue_mapping %>% 
              select(comparison_tissue, organ_name) %>% 
              distinct()) %>%
  arrange(organ_name) %>%
  mutate(frac = frac * 100,
         label = ifelse(frac >= 10, 
                        paste0(comparison_tissue, "\n", 
                               round(frac, 1), "%"),
                        NA),
         species = factor(species, c("human", "pig")),
         comparison_tissue = factor(comparison_tissue, unique(comparison_tissue))) %>% 
  
  
  {ggplot(., aes(as.numeric(species), frac, fill = comparison_tissue, label = label)) +
      geom_col(color = "white",
               size = 0.3,
               width = 0.55) +
      geom_area(aes(scales::rescale(as.numeric(species), c(1.25, 1.75))),
                size = 0.3,
                color = "white",
                alpha = 1) +
      geom_text(size = 2,
                position = position_stack(vjust = 0.5), color = "white") +
      facet_wrap(~enhanced_tissues, nrow = 4, scales = "free_x") +
      scale_x_discrete(labels = levels(.$species),
                       limits=c("1","2"),
                       expand = expansion(0)) +
      scale_y_continuous(expand = expansion(c(0, 0.05))) +
      # scale_fill_manual(values = colorRampPalette(ggthemes::tableau_color_pal(palette = "Hue Circle", 
      #                                                                         type = c("regular"), 
      #                                                                         direction = 1)(19))(n_distinct(plot_data$comparison_tissue))) + 
      scale_fill_manual(values = tissue_colors_palette_comparison) +
      # directlabels::geom_dl(aes(label = label), list("top.points", cex = .6)) +
      stripped_theme_facet +
      ylab("Fraction of expression") +
      theme(panel.border = element_blank(),
            axis.title.x = element_blank())}
ggsave(savepath("blood decomposition lineage enriched comparison bar.pdf"),
       width = 12, height = 10)


# Lymphoid tissues
plot_data1 <- 
  gene_orthologs %>%
  select(1:2, 4) %>% 
  left_join(pig_atlas_tissue %>% 
              filter(tissue_name %in% c("lymph node",
                                        "spleen",
                                        "thymus",
                                        "tonsil")) %>% 
              select(enssscg_id, tissue_name, pig = ptpm)) %>% 
  left_join(human_atlas_tissue %>% 
              filter(tissue_name %in% c("lymph node",
                                        "spleen",
                                        "thymus",
                                        "tonsil")) %>% 
              select(ensg_id, tissue_name, human = ptpm), 
            by = c("ensg_id", "tissue_name")) %>%
  gather(species, ptpm, human, pig) 
            

plot_data <- 
  human_gene_class_bloodlineage %>%
  filter(specificity_category %in% c("Tissue enriched")) %>%
  select(ensg_id, enhanced_tissues, ts_score) %>% 
  left_join(human_gene_class %>%
              select(ensg_id, enhanced_tissues),
            by = "ensg_id",
            suffix = c("", "_tissue")) %>% 
  separate_rows(enhanced_tissues_tissue, sep = ";") %>%
  mutate(blood = enhanced_tissues_tissue %in% c("blood", "lymphoid tissue", "bone marrow")) %>%
  group_by(ensg_id) %>% 
  mutate(only_blood = all(blood)) %>%
  ungroup() %>% 
  filter(only_blood) %>%
  select(1:3) %>%
  distinct() %>%
  inner_join(gene_orthologs) %>%
  inner_join(plot_data1) %>%
  group_by(ensg_id, species, enhanced_tissues) %>% 
  mutate(frac = ptpm / sum(ptpm, na.rm = T)) 

plot_data %>%
  ggplot(aes(frac, tissue_name, fill = species)) +
  geom_boxplot() +
  facet_wrap(~enhanced_tissues, scales = "free")
  

plot_data %>%
  group_by(enhanced_tissues, tissue_name, species) %>% 
  summarise(frac = mean(frac, na.rm = T)) %>%
  ungroup() %>%
  left_join(tissue_mapping %>% 
              select(tissue_name, organ_name) %>% 
              distinct()) %>%
  arrange(organ_name) %>%
  mutate(frac = frac * 100,
         label = ifelse(frac >= 10, 
                        paste0(tissue_name, "\n", 
                               round(frac, 1), "%"),
                        NA),
         species = factor(species, c("human", "pig")),
         tissue_name = factor(tissue_name, unique(tissue_name))) %>% 
  
  
  {ggplot(., aes(as.numeric(species), frac, fill = tissue_name, label = label)) +
      geom_col(color = "white",
               size = 0.3,
               width = 0.55) +
      geom_area(aes(scales::rescale(as.numeric(species), c(1.25, 1.75))),
                size = 0.3,
                color = "white",
                alpha = 1) +
      geom_text(size = 2,
                position = position_stack(vjust = 0.5), color = "white") +
      facet_wrap(~enhanced_tissues, nrow = 4, scales = "free_x") +
      scale_x_discrete(labels = levels(.$species),
                       limits=c("1","2"),
                       expand = expansion(0)) +
      scale_y_continuous(expand = expansion(c(0, 0.05))) +
      # scale_fill_manual(values = colorRampPalette(ggthemes::tableau_color_pal(palette = "Hue Circle",
      #                                                                         type = c("regular"),
      #                                                                         direction = 1)(19))(n_distinct(plot_data$tissue_name))) +
      scale_fill_manual(values = prismatic::color(colorRampPalette(c("black", "#A1A8AA", "white"))(9)[2:6])) +
      # directlabels::geom_dl(aes(label = label), list("top.points", cex = .6)) +
      stripped_theme_facet +
      ylab("Fraction of expression") +
      theme(panel.border = element_blank(),
            axis.title.x = element_blank())}
ggsave(savepath("blood decomposition lineage enriched lymphoid bar.pdf"),
       width = 12, height = 10)




```

###Disease genes overlap comparison

```{r}

disease_gene_classes <- 
  # read_delim("data/meta/HumanGeneticDisease_Genes.tsv",
  #          delim = "\t", na = "-") %>%
  readxl::read_excel("data/meta/HumanGeneticDisease_Genes V2.xlsx") %>%
  select(disease_system = 1, 
         disease_type = 2, 
         disease = 3, 
         ensg_id = EnsemblID) %>%
  left_join(human_gene_mapping %>% 
              select(1:2),
            by = "ensg_id") %>%
  filter(!is.na(gene_name)) 


joined_comparison_maxtissue <- 
  joined_atlas_comparison %>% 
  arrange(-mutual_tmm) %>%
  group_by(mutual_id, species) %>%
  top_n(1, mutual_tmm) %>%
  do({head(., 1)}) %>%
  ungroup() %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>% 
  select(1, 2, comparison_tissue, species, mutual_tmm)

human_comparison_maxtissue <- 
  human_atlas_comparison %>% 
  arrange(-nx) %>%
  group_by(ensg_id) %>%
  top_n(1, nx) %>%
  do({head(., 1)}) %>%
  ungroup() %>%
  select(1, 2, comparison_tissue, nx)


joined_comparison_maxexp_classification <- 
  joined_atlas_comparison %>% 
  select(mutual_id, comparison_tissue, species, ptpm) %>% 
  arrange(-ptpm) %>% 
  group_by(mutual_id, species) %>% 
  do({head(., 1)}) %>%
  ungroup() %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id")) %>%
  mutate(gene_id = ifelse(species == "human",
                          ensg_id, 
                          enssscg_id)) %>%
  left_join(bind_rows(pig_gene_comparison_classification %>% 
                         filter(!is.na(enhanced_tissues)) %>%
                         # separate_rows(enhanced_tissues, sep = ";") %>% 
                         select(gene_id = 1, 2, enhanced_tissues),
                       human_gene_comparison_classification %>% 
                         filter(!is.na(enhanced_tissues)) %>%
                         # separate_rows(enhanced_tissues, sep = ";") %>% 
                         select(gene_id = 1, 2, enhanced_tissues)),
             by = c("gene_id", "comparison_tissue" = "enhanced_tissues")) %>%
  select(gene_id, enssscg_id, ensg_id, species, comparison_tissue, specificity_category, ptpm) %>%
  mutate(specificity_category = ifelse(is.na(specificity_category),
                                       "not elevated", 
                                       specificity_category))


# 
# disease_entrez_ensembl_genes <- 
#   getBM(attributes=c('hgnc_symbol', 'entrezgene_id', 'ensembl_gene_id'),
#       filters = 'entrezgene_id', values = unique(disease_gene_classes$entrez_id), mart = ensembl) %>%
#   as_tibble()
# 
# disease_entrez_ensembl_genes %>%
#   write_delim("data/meta/disease_entrez_ensembl_genes.txt", delim = "\t")
# 
# 
# a %>%
#   select(gene_name, ensg_id, entrez_id) %>% 
#   distinct %>%
#   left_join(b,
#             by = c("entrez_id" = "entrezgene")) %>%
#   filter(ensembl_gene_id != ensg_id)
#  
# b


####  
plot_data <- 
  disease_gene_classes %>%
  left_join(gene_orthologs %>% 
              select(1:2)) %>%
  left_join(gene_overlap_data_summarized) %>% 
  mutate(type = case_when(type == "Overlap" ~ "Overlap",
                          is.na(enssscg_id) ~ "No ortholog",
                          T ~ "No overlap")) %>%
  left_join(human_gene_comparison_classification %>% 
              select(1:2)) %>%
  left_join(human_comparison_maxtissue %>%
              select(2, enhanced_tissue_human = comparison_tissue) %>%
              distinct()) %>% 
  select(1, ensg_id, type, specificity_category, enhanced_tissue_human) %>%
  distinct() %>%
  mutate(tissue_category = ifelse(specificity_category %in% c("Tissue enhanced", 
                                                              "Tissue enriched", 
                                                              "Group enriched"),
                                  enhanced_tissue_human,
                                  "Not elevated")) %>%
  left_join(tissue_mapping %>%
              select(comparison_tissue, organ_name) %>% 
              distinct(),
            by = c("tissue_category" = "comparison_tissue")) %>%
  mutate(organ_name = ifelse(is.na(organ_name), "Not elevated", organ_name)) 

  

plot_data %>%
  group_by(disease_system, type) %>% 
  count() %>%
  group_by(disease_system) %>%
  mutate(frac = n / sum(n)) %>%
  ungroup %>%
  arrange(frac) %>%
  mutate(disease_system = factor(disease_system, filter(., type == "Overlap") %$% 
                                   unique(disease_system))) %>%
  
                                   {ggplot(., aes(frac * 100, disease_system,
                                               fill = type, label = paste0(round(frac * 100, 1), "%"))) +
                                       geom_col() +
                                       geom_text(position = position_stack(vjust = 0.5),
                                                 color = "white",
                                                 fontface = "bold", 
                                                 size = 2) + 
                                       scale_x_continuous(expand = expansion(0)) +
                                       scale_fill_manual(values = c("Overlap" = "#004369",
                                                                    "No overlap" = "gray",
                                                                    "No ortholog" = "darkgray")) +
                                       stripped_theme +
                                       theme(legend.position = "top") +
                                       ggplot(., aes(n, disease_system,
                                                     label = n)) +
                                       geom_col(aes(fill = type),
                                                show.legend = F) +
                                       geom_text(data = . %>% 
                                                   group_by(disease_system) %>% 
                                                   summarise(n = sum(n)),
                                                 color = "white",
                                                 fontface = "bold", 
                                                 hjust = 1,
                                                 size = 2) +
                                       scale_x_continuous(expand = expansion(0)) +
                                       scale_fill_manual(values = c("Overlap" = "#004369",
                                                                    "No overlap" = "gray",
                                                                    "No ortholog" = "darkgray")) +
                                       stripped_theme +
                                       theme(axis.text.y = element_blank(),
                                             axis.ticks.y = element_blank(),
                                             axis.title = element_blank())}
ggsave(savepath("Disease system overlap barplot.pdf"), width = 7, height = 4)  
  
plot_order <-
  plot_data %>%
  filter(organ_name != "Not elevated") %>%
  filter(type == "Overlap") %>%
  mutate(id = paste(disease_system, type)) %>% 
  group_by(id, organ_name) %>% 
  count() %>% 
  spread(organ_name, n, fill = 0) %>% 
  column_to_rownames("id") %>% 
  {. / rowSums(.)} %>%
  t() %>% 
  dist() %>% 
  hclust(method = "ward.D2") %>% 
  labels

plot_data %>%
  
  gather(x, stratum, type, organ_name) %>%
  mutate(stratum = factor(stratum, c(plot_order,
                                     "Not elevated",
                                     "No ortholog",
                                     "No overlap",
                                     "Overlap"))) %>%
  ggplot(aes(x = x, stratum = stratum, alluvium = paste(ensg_id, disease_system),
             fill = stratum, label = stratum)) +
  geom_flow(show.legend = F) +
  geom_stratum(show.legend = T,
               color = NA) +
  # geom_alluvium(aes(fill = human),
  #               width = 1/8, knot.pos = 0.25, reverse = FALSE,
  #               show.legend = F) +
  # geom_stratum(width = 1/8, reverse = FALSE) +
  # geom_text(stat = "stratum", aes(label = ifelse(ymax-ymin >= 5, after_stat(as.character(stratum)), "")),
  #           reverse = T, 
  #           size = 1.5,
  #           angle = 90) +
  stripped_theme_facet +
  scale_fill_manual(values = c(tissue_colors_palette_organs, 
                               c("Overlap" = "#004369",
                                 "No overlap" = "gray",
                                 "No ortholog" = "darkgray"), 
                               "not elevated" = "darkgray",
                               "Not elevated" = "darkgray")) +
  scale_x_discrete(expand = expansion(1/6)) +
  scale_y_continuous(expand = expansion(0)) +
  facet_wrap(~disease_system, scales = "free", ncol = 1, strip.position = "right",
             labeller = label_wrap_gen()) +
  coord_flip() +
  theme(axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank(),
        panel.spacing = unit(0, "mm"))
ggsave(savepath("Disease system overlap tissue alluvialt.pdf"), width = 6, height = 16)  



#####
cor_data <- 
  disease_gene_classes %>%
  select(disease_system, ensg_id, gene_name) %>%
  distinct() %>%
  left_join(gene_orthologs %>% 
              select(1:2)) %>%
  filter(!is.na(enssscg_id)) %>%
  distinct() %>% 
  inner_join(joined_atlas_comparison %>% 
               separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>% 
               select(1:2, comparison_tissue, species, ptpm)) %>% 
  spread(species, ptpm) %>% 
  group_by(ensg_id) %>% 
  summarise(pearson = cor(log10(human + 1), log10(pig + 1), method = "pearson", use = "pairwise.complete.obs"))

disease_gene_classes %>%
  left_join(gene_orthologs %>% 
              select(1:2)) %>%
  left_join(gene_overlap_data_summarized) %>% 
  mutate(type = case_when(type == "Overlap" ~ "Overlap",
                          is.na(enssscg_id) ~ "No ortholog",
                          T ~ "No overlap")) %>%
  left_join(human_gene_comparison_classification %>% 
              select(1:2)) %>%
  left_join(human_comparison_maxtissue %>%
              select(2, enhanced_tissue_human = comparison_tissue) %>%
              distinct()) %>% 
  select(ensg_id, enssscg_id, gene_name, disease_system, disease_type, disease, 
         overlap_type = type, 
         specificity_category_human = specificity_category, 
         max_tissue_human = enhanced_tissue_human) %>%
  left_join(cor_data) %>%
  write_csv(savepath("Disease genes overlap.csv"))



########


plot_data <- 
  disease_gene_classes %>%
  left_join(gene_orthologs %>% 
              select(1:2)) %>%
  filter(!is.na(enssscg_id)) %>%
  left_join(joined_comparison_maxexp_classification %>% 
              mutate(tissue_category = ifelse(specificity_category == "not elevated",
                                              "not elevated",
                                              comparison_tissue)) %>%
              select(enssscg_id, ensg_id, species, tissue_category) %>%  
              inner_join(gene_orthologs %>% 
                           select(1:2))) %>%
  # group_by(species, disease_system) %>% 
  # count() %>%
  # ungroup %>%
  select(disease_system, ensg_id, enssscg_id, gene_name, species, tissue_category) %>%
  distinct() 

plot_order <- 
  plot_data %>%
  mutate(id = paste(disease_system, species)) %>% 
  group_by(id, tissue_category) %>% 
  count() %>% 
  spread(tissue_category, n, fill = 0) %>% 
  column_to_rownames("id") %>% 
  t() %>% 
  {log10(. + 1)} %>%
  dist() %>% 
  hclust(method = "ward.D2") %>% 
  labels

plot_data %>%
  mutate(tissue_category = factor(tissue_category, plot_order)) %>%
  ggplot(aes(x = species, stratum = tissue_category, alluvium = paste(ensg_id, disease_system),
             fill = tissue_category, label = tissue_category)) +
  geom_flow(show.legend = F) +
  geom_stratum(show.legend = F,
               color = NA) +
  # geom_alluvium(aes(fill = human),
  #               width = 1/8, knot.pos = 0.25, reverse = FALSE,
  #               show.legend = F) +
  # geom_stratum(width = 1/8, reverse = FALSE) +
  geom_text(stat = "stratum", aes(label = ifelse(ymax-ymin >= 5, after_stat(as.character(stratum)), "")),
            reverse = T, 
            size = 1.5) +
  stripped_theme_facet +
  scale_fill_manual(values = c(tissue_colors_palette_comparison, "not elevated" = "darkgray")) +
  scale_x_discrete(expand = expansion(1/6)) +
  scale_y_continuous(expand = expansion(0)) +
  facet_wrap(~disease_system, scales = "free") 
ggsave(savepath("Disease system overlap alluvial.pdf"), width = 10, height = 10)


# plot_data <- 
#   disease_gene_classes %>%
#   select(disease_system, ensg_id, gene_name) %>%
#   distinct() %>%
#   left_join(gene_orthologs %>% 
#               select(1:2)) %>%
#   filter(!is.na(enssscg_id)) %>%
#   distinct() %>% 
#   inner_join(joined_atlas_comparison %>% 
#                separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>% 
#                select(1:2, comparison_tissue, species, ptpm)) %>% 
#   spread(species, ptpm) %>% 
#   group_by(disease_system, comparison_tissue) %>% 
#   summarise(pearson = cor(log10(human + 1), log10(pig + 1), method = "pearson", use = "pairwise.complete.obs"))
# write_csv(plot_data, savepath("Disease system tissue pearson.csv"))
# 
# plot_data %>% 
#   ggplot(aes(disease_system, pearson)) +
#   geom_violin(draw_quantiles = 0.5) +
#   geom_text(data = . %>% 
#               group_by(disease_system) %>%
#               top_n(3, pearson),
#             aes(label = comparison_tissue, 
#                 color = comparison_tissue),
#             show.legend = F) +
#   scale_color_manual(values = tissue_colors_palette_comparison) +
#   theme(axis.text.x = element_text(angle = 60, hjust = 1))
  

# plot_data %>% 
#   spread(comparison_tissue, pearson) %>% 
#   column_to_rownames("disease_system") %>% 
#   pheatmap(clustering_method = "ward.D2", 
#            display_numbers = T, 
#            number_color = alpha("white", 0.5), 
#            color = heatmap_palette,
#            border_color = NA,
#            filename = savepath("Disease system human pig correlation per tissue.pdf"),
#            width = 12, height = 6)
# 
# 
# plot_data <- 
#   disease_gene_classes %>%
#   select(disease_type, ensg_id, gene_name) %>%
#   distinct() %>%
#   left_join(gene_orthologs %>% 
#               select(1:2)) %>%
#   filter(!is.na(enssscg_id)) %>%
#   distinct() %>% 
#   inner_join(joined_atlas_comparison %>% 
#                separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>% 
#                select(1:2, comparison_tissue, species, ptpm)) %>% 
#   spread(species, ptpm) %>% 
#   group_by(disease_type, comparison_tissue) %>% 
#   summarise(pearson = cor(log10(human + 1), log10(pig + 1), method = "pearson", use = "pairwise.complete.obs"))
# write_csv(plot_data, savepath("Disease type tissue pearson.csv"))
# 
# plot_data %>% 
#   spread(comparison_tissue, pearson) %>% 
#   column_to_rownames("disease_type") %>% 
#   pheatmap(clustering_method = "ward.D2", 
#            display_numbers = T, 
#            number_color = alpha("white", 0.5), 
#            color = heatmap_palette,
#            border_color = NA,
#            filename = savepath("Disease type human pig correlation per tissue.pdf"),
#            width = 18, height = 12)


# plot_data <- 
#   disease_gene_classes %>%
#   select(disease_system, ensg_id, gene_name) %>%
#   distinct() %>%
#   left_join(gene_orthologs %>% 
#               select(1:2)) %>%
#   filter(!is.na(enssscg_id)) %>%
#   distinct() %>% 
#   inner_join(joined_atlas_comparison %>% 
#                separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>% 
#                select(1:2, comparison_tissue, species, ptpm)) %>% 
#   spread(species, ptpm) %>% 
#   group_by(disease_system, ensg_id) %>% 
#   summarise(spearman = cor(human, pig, method = "spearman", use = "pairwise.complete.obs"),
#             pearson = cor(log10(human + 1), log10(pig + 1), method = "pearson", use = "pairwise.complete.obs"))
# 
# plot_order <- 
#   plot_data %>% 
#   group_by(disease_system) %>% 
#   summarise(median = median(pearson)) %>% 
#   arrange(median) %>% 
#   pull(disease_system)
# 
# plot_data %>% 
#   ungroup() %>%
#   mutate(disease_system = factor(disease_system, plot_order)) %>%
#   ggplot(aes(pearson, disease_system)) +
#   geom_boxplot() +
#   plot_data %>% 
#   ungroup() %>%
#   mutate(disease_system = factor(disease_system, plot_order)) %>%
#   ggplot(aes(spearman, disease_system)) +
#   geom_boxplot()


########

g <-
  disease_gene_classes %>%
  # filter(disease_system == "Cancers") %>%
  select(disease_system, ensg_id, gene_name) %>%
  distinct() %>%
  left_join(gene_orthologs %>% 
              select(1:2)) %>%
  filter(!is.na(enssscg_id)) %>%
  distinct() %>% 
  inner_join(joined_atlas_comparison %>% 
               separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>% 
               select(1:2, comparison_tissue, species, ptpm)) %>% 
  spread(species, ptpm) %>%
  left_join(gene_overlap_data,
             by = c("ensg_id", "enssscg_id", "comparison_tissue" = "comparison_id")) %>% 
  mutate(type = ifelse(is.na(type), "Not elevated", type)) %>%
  ggplot(aes(log10(human + 1), log10(pig + 1), color = type)) +
  geom_point(data = . %>% 
               filter(type == "Not elevated")) +
  geom_point(data = . %>% 
               filter(type != "Not elevated")) +
  facet_grid(disease_system ~ comparison_tissue) +
  scale_color_manual(values = c(overlap_type_pal, "Not elevated" = "darkgray")) +
  stripped_theme_facet +
  theme(panel.spacing = unit(0, "mm"),
        strip.text.y = element_text(angle = 0))

ggsave(savepath("Disease system tissue scatter.pdf"), plot = g,
       width = 50, height = 20, limitsize = F)

######

human_gene_class_bloodlineage %>% 
  filter(!is.na(ts_score)) %>% 
  filter(!grepl(",", enhanced_tissues)) %>% 
  group_by(enhanced_tissues) %>% 
  top_n(50, ts_score) %>%
  left_join(human_gene_mapping) %>%
  select(ensg_id, gene_name, ts_score, cell_type = enhanced_tissues) %>%
  arrange(cell_type, -ts_score) %>%
  write_csv(savepath("Blood lineage best markers.csv"))



#####

plot_data <- 
  disease_gene_classes %>%
  left_join(gene_orthologs %>% 
              select(1:2)) %>%
  left_join(human_gene_comparison_classification %>% 
              select(1:2))  %>%
  left_join(pig_gene_comparison_classification %>% 
              select(1:2), 
            by = "enssscg_id",
            suffix = c("_human", "_pig"))  %>%
  mutate(specificity_category_pig = ifelse(is.na(specificity_category_pig),
                                           "No ortholog",
                                           specificity_category_pig)) %>%
  select(ensg_id, enssscg_id, specificity_category_human, specificity_category_pig) %>%
  distinct()

plot_data %>%
  gather(x, stratum, specificity_category_human, specificity_category_pig) %>%
  mutate(stratum = factor(stratum, c(spec_category_levels,
                                     "No ortholog"))) %>%
  ggplot(aes(x = x, stratum = stratum, alluvium = paste(ensg_id, enssscg_id),
             fill = stratum, label = stratum)) +
  geom_flow(show.legend = F) +
  geom_stratum(show.legend = F,
               color = NA) +
  geom_text(stat = "stratum", aes(label = ifelse(ymax-ymin >= 5, after_stat(as.character(stratum)), "")),
            reverse = T, 
            size = 1.5) +
  stripped_theme_facet +
  scale_fill_manual(values = c(gene_category_pal, 
                               c("Overlap" = "#004369",
                                 "No overlap" = "gray",
                                 "No ortholog" = "darkgray"), 
                               "not elevated" = "darkgray",
                               "Not elevated" = "darkgray")) +
  scale_x_discrete(expand = expansion(1/6)) +
  scale_y_continuous(expand = expansion(0))
ggsave(savepath("Disease genes specificity alluvial.pdf"), width = 5, height = 5)  



```

####Cancer genes

```{r}

disease_gene_classes %>%
  filter(disease_system == "Cancers") %>%
  left_join(gene_overlap_data_summarized) %>% 
  mutate(type = case_when(type == "Overlap" ~ "Overlap",
                          is.na(type) ~ "No ortholog",
                          T ~ "No overlap")) %>%
  group_by(disease_type, type) %>% 
  count() %>%
  group_by(disease_type) %>%
  mutate(frac = n / sum(n)) %>%
  ungroup %>%
  arrange(frac) %>%
  mutate(disease_type = factor(disease_type, filter(., type == "Overlap") %$% 
                                 unique(disease_type))) %>%
  
                                 {ggplot(., aes(frac * 100, disease_type,
                                                fill = type, label = paste0(round(frac * 100, 1), "%"))) +
                                     geom_col() +
                                     geom_text(position = position_stack(vjust = 0.5),
                                               color = "white",
                                               fontface = "bold", 
                                               size = 2) + 
                                     scale_x_continuous(expand = expansion(0)) +
                                     scale_fill_manual(values = c("Overlap" = "#004369",
                                                                  "No overlap" = "gray",
                                                                  "No ortholog" = "darkgray")) +
                                     stripped_theme +
                                     theme(legend.position = "top") +
                                     ggplot(., aes(n, disease_type,
                                                   label = n)) +
                                     geom_col(aes(fill = type),
                                              show.legend = F) +
                                     geom_text(data = . %>% 
                                                 group_by(disease_type) %>% 
                                                 summarise(n = sum(n)),
                                               color = "white",
                                               fontface = "bold", 
                                               hjust = 1,
                                               size = 2) +
                                     scale_x_continuous(expand = expansion(0)) +
                                     scale_fill_manual(values = c("Overlap" = "#004369",
                                                                  "No overlap" = "gray",
                                                                  "No ortholog" = "darkgray")) +
                                     stripped_theme +
                                     theme(axis.text.y = element_blank(),
                                           axis.ticks.y = element_blank(),
                                           axis.title = element_blank())}
ggsave(savepath("Disease system overlap barplot cancers.pdf"), width = 7, height = 4) 

```

###MHC genes

```{r}

MHC_genes <- 
  readxl::read_excel("data/meta/NCBI_MHC_gene.xlsx")

library(biomaRt)
select <- dplyr::select
ensembl <- useMart("ensembl",
                   dataset="sscrofa_gene_ensembl") 

listDatasets(ensembl) %>% 
  as_tibble %>% 
  filter(grepl("pig", description)) %>% 
  pull(1)

a <- getBM(attributes=c('hgnc_symbol', 'entrezgene_id', 'ensembl_gene_id'),mart = ensembl) %>% 
  as_tibble()

MHC_ensembl_genes <- 
  getBM(attributes=c('hgnc_symbol', 'entrezgene_id', 'ensembl_gene_id'),
        filters = 'entrezgene_id', values = unique(MHC_genes$GeneID), mart = ensembl) %>%
  as_tibble()

MHC_genes %>% 
  select(GeneID, Symbol) %>% 
  left_join(MHC_ensembl_genes,
            by = c("GeneID" = "entrezgene_id")) 


a %>%
  select(gene_name, ensg_id, entrez_id) %>%
  distinct %>%
  left_join(b,
            by = c("entrez_id" = "entrezgene")) %>%
  filter(ensembl_gene_id != ensg_id)





####



MHC_genes %>% 
  select(Symbol) %>% 
  left_join(gene_mapping,
            by = c("Symbol" = "display_name")) %>%
  filter(!is.na(enssscg_id))


MHC_genes %>% 
  select(GeneID, Symbol) %>% 
  left_join(MHC_ensembl_genes,
            by = c("GeneID" = "entrezgene_id")) %>%
  select(enssscg_id = ensembl_gene_id) %>% 
  filter(!is.na(enssscg_id)) %>% 
  distinct() %>%
  left_join(pig_atlas_tissue) %>%
  select(1, 2, nx) %>% 
  spread(tissue_name, nx) %>% 
  column_to_rownames("enssscg_id") %>% 
  {log10(. + 1)} %>%
  pheatmap(color = heatmap_palette,
           clustering_method = "ward.D2",
           border_color = NA)
  

```



## Supplementary figures

### Figure 1 supporting

```{r Figure1_spearman, echo=FALSE, message=FALSE, warning=FALSE}

pig_atlas_consensus %>% 
  select(enssscg_id, tissue, tpm) %>% 
  mutate(tissue = str_to_sentence(tissue)) %>%
  spread(enssscg_id, tpm) %>%
  column_to_rownames("tissue")  %>%
  t() %>%
  cor(method = "spearman") %T>%
  {g_data <- .
  cluster_ <- g_data %>% 
    {1 - .} %>%
    as.dist() %>% 
    hclust(method = "average")
  pheatmap(g_data, 
           color = heatmap_palette,
           border_color = NA,
           cluster_cols = cluster_,
           cluster_rows = cluster_,
           
           filename = savepath("Figure1 spearman heatmap average spearman.pdf"),
           width = 7.9, height = 7.4)
  } %>%
  pheatmap(color = heatmap_palette,
           border_color = NA,
           clustering_method = "ward.D2",
           
           filename = savepath("Figure1 spearman heatmap ward spearman.pdf"),
           width = 7.9, height = 7.4)

pig_atlas_tissue %>% 
  
  select(enssscg_id, tissue_name, tpm) %>% 
  mutate(tissue_name = str_to_sentence(tissue_name)) %>%
  spread(enssscg_id, tpm) %>%
  column_to_rownames("tissue_name")  %>%
  t() %>%
  cor(method = "spearman") %T>%
  {g_data <- .
  cluster_ <- g_data %>% 
  {1 - .} %>%
    as.dist() %>% 
    hclust(method = "average")
  pheatmap(g_data, 
           color = heatmap_palette,
           border_color = NA,
           cluster_cols = cluster_,
           cluster_rows = cluster_,
           
           filename = savepath("Figure1 tissue spearman heatmap average spearman.pdf"),
           width = 14.5, height = 14)
  } %>%
  pheatmap(color = heatmap_palette,
           border_color = NA,
           clustering_method = "ward.D2",
           
           filename = savepath("Figure1 tissue spearman heatmap ward spearman.pdf"),
           width = 14.5, height = 14)

```

```{r Figure1_big_PCA_UMAP}

set.seed(42)
g <- 
  sample_pca$scores %>%
  uwot::umap(n_neighbors = 10) %>%
  as_tibble() %>%
  mutate(sample_id = rownames(sample_pca$scores)) %>%
  separate(sample_id, into = c("tissue_id", "individual"), sep = "_", remove = F) %>%
  
  left_join(tissue_mapping,
            by = c("tissue_id" = "tissue")) %>%
  group_by(tissue_name) %>%
  mutate(V1_mean = mean(V1), 
         V2_mean = mean(V2)) %>%
  
  ggplot(aes(V1, V2, color = tissue_name, fill = tissue_name)) +
  
  
  geom_point(shape = 21, 
             color = "gray25", 
             size = 2.5, 
             alpha = 0.7, 
             show.legend = F) +
  
  geom_segment(aes(xend = V1_mean, yend = V2_mean),
               alpha = 0.4,
               show.legend = F) +
  geom_text(data = . %>% 
              select(V1_mean, V2_mean, tissue_name) %>% 
              distinct(), 
            aes(V1_mean, V2_mean, label = tissue_name),
            show.legend = F) +
  
  
  scale_color_manual(values = tissue_colors_palette_full, name = "") +
  scale_fill_manual(values = tissue_colors_palette_full, name = "") +
  
  stripped_theme +
  theme(plot.title = element_text(hjust = 0.5), 
        panel.border = element_blank(), 
        panel.background = element_blank())

ggsave(savepath("Supplementary sample UMAP.pdf"), plot = g, width = 25, height = 25, useDingbats = F)

g <- 
  sample_pca$scores %>%
  as_tibble(rownames = "sample_id") %>%
  separate(sample_id, into = c("tissue_id", "individual"), sep = "_", remove = F) %>%
  
  left_join(tissue_mapping,
            by = c("tissue_id" = "tissue")) %>%
  group_by(tissue_name) %>%
  mutate(PC1_mean = mean(PC1), 
         PC2_mean = mean(PC2)) %>%
  
  ggplot(aes(PC1, PC2, color = tissue_name, fill = tissue_name)) +
  
  
  geom_point(shape = 21, 
             color = "gray25", 
             size = 2.5, 
             alpha = 0.7, 
             show.legend = F) +
  
  geom_segment(aes(xend = PC1_mean, yend = PC2_mean),
               alpha = 0.4,
               show.legend = F) +
  geom_text(data = . %>% 
              select(PC1_mean, PC2_mean, tissue_name) %>% 
              distinct(), 
            aes(PC1_mean, PC2_mean, label = tissue_name),
            show.legend = F) +
  
  
  scale_color_manual(values = tissue_colors_palette_full, name = "") +
  scale_fill_manual(values = tissue_colors_palette_full, name = "") +
  
  stripped_theme +
  theme(plot.title = element_text(hjust = 0.5), 
        panel.border = element_blank(), 
        panel.background = element_blank())

ggsave(savepath("Supplementary sample PCA.pdf"), plot = g, width = 25, height = 25, useDingbats = F)


```


#### Barplot # detected

```{r figure_n_detected}



plot_data <- 
  pig_atlas_tissue %>%
  filter(nx >= 1) %>%
  select(1:2, nx) %>% 
  select(tissue_name, enssscg_id) %>% 
  distinct() %>% 
  left_join(pig_gene_classification) %>%
  group_by(tissue_name, distribution_category) %>%
  summarise(n_detected = n_distinct(enssscg_id)) %>%
  mutate(sum_n = sum(n_detected)) %>%
  ungroup() %>% 
  
  arrange(sum_n) %>%
  mutate(distribution_category = str_to_sentence(distribution_category),
         distribution_category = factor(distribution_category, rev(dist_category_levels)),
         tissue_name = str_to_sentence(tissue_name),
         tissue_name = factor(tissue_name, unique(tissue_name))) 

plot_data %>%
  pull(sum_n) %>%
  min()

plot_data %>%
  pull(sum_n) %>%
  max()
  

plot_data %>%
  ggplot(aes(n_detected, tissue_name, fill = distribution_category)) +
  geom_col(width = 0.8) +
  geom_text(data = . %>% 
              select(tissue_name, sum_n) %>% 
              distinct(), 
            aes(sum_n, tissue_name, label = sum_n), 
            hjust = 0,
            size = 3,
            inherit.aes = F) + 
  xlab("Number of detected genes") +
  ggtitle("Number of detected genes per tissue type") +
  scale_x_continuous(expand = expansion(c(0, 0.15))) +
  scale_fill_manual(values = gene_category_pal, name = "") +
  theme(axis.title.y = element_blank(),
        panel.background = element_blank(),
        panel.grid = element_blank(), 
        legend.position = "bottom") +
  guides(fill = guide_legend(nrow=2,byrow=TRUE))
ggsave(savepath("Supplementary detected genes bar.pdf"), width = 5, height = 12)  

```

#### Sequence quality

```{r}
sample_quality <-
  read_delim("data/meta/Sequencing QC data/QC-tissue.tsv", delim = "\t") %>% 
  separate(Sample, into = c("individual_n", "tissue_ID"), sep = "-") %>% 
  mutate(individual = letters[as.integer(individual_n)],
         sample_ID = paste(tissue_ID, individual, sep = "_")) %>%
  select(sample_ID, tissue_ID, individual, everything()) %>%
  select(-individual_n) %>%
  filter(sample_ID %in% pig_atlas_sample$sample_ID)

sample_quality %>%
  pull(`Clean Reads(M)`) %>%
  mean() 
sample_quality %>%
  pull(`Clean Reads(M)`) %>%
  min() 
sample_quality %>%
  pull(`Clean Reads(M)`) %>%
  max() 

sample_quality %>% 
  left_join(tissue_mapping,
            by = c("tissue_ID" = "tissue")) %>% 
  select(1, tissue_name, 3:7) %>%
  write_csv(savepath("Supplementary table seq QC.csv"))
```

###Figuere 5 supporting

####Venn complete

```{r}

plot_data <- 
  gene_overlap_data %>% 
  filter(spec_category_human == "Tissue enriched" |
           spec_category_pig == "Tissue enriched") %>%
  group_by(comparison_id) %>% 
  mutate(n = n()) %>% 
  ungroup() %>%
  mutate(show = T,
         comparison_id = ifelse(show, comparison_id, "Other tissues"))

plot_order <- 
  plot_data %>%
  select(comparison_id, n, show) %>% 
  unique() %>%
  arrange(!show, -n) %$% 
  unique(comparison_id) %>% 
  str_to_sentence() %>% 
  rev()

# Venn

plot_data %>% 
  mutate(comparison_id = factor(str_to_sentence(comparison_id)),
         type = factor(type)) %>%
  group_by(comparison_id, type, .drop = F) %>%
  summarise(n = n()) %>%
  
  ungroup() %>%
  
  mutate(comparison_id = factor(comparison_id, 
                                levels = plot_order)) %>%
  ggplot() + 
  geom_col(mapping = aes(comparison_id, n,fill = type)) + 
  
  geom_text(data = . %>%
              filter(type == "Pig"),
            aes(x = comparison_id, y = 0,
                label = n),
            hjust = 1) +
  geom_text(data = . %>%
              group_by(comparison_id, .drop = F) %>%
              summarise(y = n[which(type == "Overlap")] / 2 +
                          n[which(type == "Pig")],
                        label = n[which(type == "Overlap")]),
            aes(x = comparison_id, y = y,
                label = label),
            hjust = 0.5,
            color = "white") +
  geom_text(data = . %>%
              group_by(comparison_id, .drop = F) %>%
              summarise(y = sum(n),
                        label = n[which(type == "Human")]),
            aes(x = comparison_id, y = y,
                label = label),
            hjust = 0) +
  
  coord_flip() + 
  
  scale_fill_manual(values = c("Human" = "#DB1F48", "Overlap" = "#004369", "Pig" = "#01949A")) +
  scale_y_continuous(expand = expand_scale(0.05)) +
  
  theme(axis.text.y = element_text(face = "bold"), 
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(), 
        axis.title = element_blank(),
        panel.background = element_blank(),
        
        legend.position = c(0.5, 0.5)) + 
  
  ggtitle("Overlap of tissue enriched genes") +
  theme(axis.ticks = element_blank())
  
ggsave(savepath("Supplementary Overlap enrichment per tissue full.pdf"), 
       height = 6, width = 12, useDingbats = F)
```

#### Fract overlap bar

```{r}
human_pig_gene_class_comparison_long %>% 
  select(1, 2, enriched_tissues_human, enriched_tissues_pig, comparison_id) %>%
  mutate(type = case_when(!is.na(enriched_tissues_pig) & !is.na(enriched_tissues_human) ~ "Overlap",
                          !is.na(enriched_tissues_pig) ~ "Pig",
                          !is.na(enriched_tissues_human) ~ "Human")) %>%
  group_by(comparison_id, type) %>% 
  # group_by(enriched_tissues_human, enriched_tissues_pig) %>%
  count() %>% 
  group_by(comparison_id) %>% 
  mutate(fract = n / sum(n)) %>%
  arrange(fract) %>% 
  ungroup %>% 
  filter(type == "Overlap") %>%
  mutate(comparison_id = factor(str_to_sentence(comparison_id), str_to_sentence(comparison_id))) %>%
  ggplot(aes(fract * 100, comparison_id, label =  paste(round(100*fract, 1), "%"))) + 
  geom_col() +
  geom_text(hjust = 0) +
  scale_x_continuous(expand = expansion(c(0, 0.10)), limits = c(0,100)) +
  stripped_theme +
  ylab("") +
  xlab("% overlap") +
  ggtitle("% total elevated overlap",
          "overlap vs number of genes elevated in either pig or human") +
  
  human_pig_gene_class_comparison_long %>% 
  filter(spec_category_pig %in% c("Tissue enriched", "Group enriched") |
           spec_category_human %in% c("Tissue enriched", "Group enriched")) %>%
  select(1, 2, enriched_tissues_human, enriched_tissues_pig, comparison_id) %>%
  mutate(type = case_when(!is.na(enriched_tissues_pig) & !is.na(enriched_tissues_human) ~ "Overlap",
                          !is.na(enriched_tissues_pig) ~ "Pig",
                          !is.na(enriched_tissues_human) ~ "Human")) %>%
  group_by(comparison_id, type) %>% 
  # group_by(enriched_tissues_human, enriched_tissues_pig) %>%
  count() %>% 
  group_by(comparison_id) %>% 
  mutate(fract = n / sum(n)) %>%
  arrange(fract) %>% 
  ungroup %>% 
  filter(type == "Overlap") %>%
  mutate(comparison_id = factor(str_to_sentence(comparison_id), str_to_sentence(comparison_id))) %>%
  ggplot(aes(fract * 100, comparison_id, label =  paste(round(100*fract, 1), "%"))) + 
  geom_col() +
  geom_text(hjust = 0) +
  scale_x_continuous(expand = expansion(c(0, 0.10)), limits = c(0,100)) +
  stripped_theme +
  ylab("") +
  xlab("% overlap") +
  ggtitle("% total enriched overlap",
          "overlap vs number of genes enriched in either pig or human") + 
  
  
  
  human_pig_gene_class_comparison_long %>% 
  select(1, 2, enriched_tissues_human, enriched_tissues_pig, comparison_id) %>%
  mutate(type = case_when(!is.na(enriched_tissues_pig) & !is.na(enriched_tissues_human) ~ "Overlap",
                          !is.na(enriched_tissues_pig) ~ "Pig",
                          !is.na(enriched_tissues_human) ~ "Human")) %>%
  group_by(comparison_id, type) %>% 
  # group_by(enriched_tissues_human, enriched_tissues_pig) %>%
  count() %>% 
  group_by(comparison_id) %>% 
  mutate(fract = n / (min(n[which(type != "Overlap")]) + n[which(type == "Overlap")])) %>%
  arrange(fract) %>% 
  ungroup %>% 
  filter(type == "Overlap") %>%
  mutate(comparison_id = factor(str_to_sentence(comparison_id), str_to_sentence(comparison_id))) %>%
  ggplot(aes(fract * 100, comparison_id, label =  paste(round(100*fract, 1), "%"))) + 
  geom_col() +
  geom_text(hjust = 0) +
  scale_x_continuous(expand = expansion(c(0, 0.10)), limits = c(0,100)) +
  stripped_theme +
  ylab("") +
  xlab("% overlap") +
  ggtitle("% possible elevated overlap",
          "overlap vs smallest number of genes elevated in either pig or human") +
  
  human_pig_gene_class_comparison_long %>% 
  filter(spec_category_pig %in% c("Tissue enriched", "Group enriched") |
           spec_category_human %in% c("Tissue enriched", "Group enriched")) %>%
  select(1, 2, enriched_tissues_human, enriched_tissues_pig, comparison_id) %>%
  mutate(type = case_when(!is.na(enriched_tissues_pig) & !is.na(enriched_tissues_human) ~ "Overlap",
                          !is.na(enriched_tissues_pig) ~ "Pig",
                          !is.na(enriched_tissues_human) ~ "Human")) %>%
  group_by(comparison_id, type) %>% 
  # group_by(enriched_tissues_human, enriched_tissues_pig) %>%
  count() %>% 
  group_by(comparison_id) %>% 
  mutate(fract = n / (min(n[which(type != "Overlap")]) + n[which(type == "Overlap")])) %>%
  arrange(fract) %>% 
  ungroup %>% 
  filter(type == "Overlap") %>%
  mutate(comparison_id = factor(str_to_sentence(comparison_id), str_to_sentence(comparison_id))) %>%
  ggplot(aes(fract * 100, comparison_id, label =  paste(round(100*fract, 1), "%"))) + 
  geom_col() +
  geom_text(hjust = 0) +
  scale_x_continuous(expand = expansion(c(0, 0.10)), limits = c(0,100)) +
  stripped_theme +
  ylab("") +
  xlab("% overlap") +
  ggtitle("% possible enriched overlap", 
          "overlap vs smallest number of genes enriched in either pig or human") 

ggsave(savepath("Supplementary 5 fraction of total overlap alternatives.pdf"), width = 12, height = 10)

overlap_hyper_all %>% 
  filter(pig_id == human_id) %>% 
  # arrange(-adj_pval)
  mutate(fract = q/m) %>% 
  arrange(fract) %>% 
  ungroup %>% 
  mutate(pig_id = factor(str_to_sentence(pig_id), str_to_sentence(pig_id))) %>%
  ggplot(aes(fract * 100, pig_id, label =  paste(round(100*fract, 1), "%"))) + 
  geom_col() +
  geom_text(hjust = 0) +
  scale_x_continuous(expand = expansion(c(0, 0.10))) +
  stripped_theme +
  ylab("") +
  xlab("% overlap")

ggsave(savepath("Supplementary 5 fraction of total overlap.pdf"), width = 6, height = 6)

```

#### Fract overlap density?
```{r}




plot_data <- 
  human_pig_gene_class_comparison_long %>% 
  filter(spec_category_pig %in% c("Tissue enriched", "Group enriched") |
           spec_category_human %in% c("Tissue enriched", "Group enriched")) %>%
  select(1, 2, enriched_tissues_human, enriched_tissues_pig, comparison_id) %>%
  mutate(type = case_when(!is.na(enriched_tissues_pig) & !is.na(enriched_tissues_human) ~ "Overlap",
                          !is.na(enriched_tissues_pig) ~ "Pig",
                          !is.na(enriched_tissues_human) ~ "Human")) %>%
  
  left_join(joined_atlas_comparison %>% 
              group_by(mutual_id, comparison_tissue) %>% 
              summarise(tmm = mean(mutual_tmm, na.rm = T)) %>%
              separate(mutual_id, into = c("enssscg_id", "ensg_id")),
            by = c("enssscg_id", "ensg_id", "comparison_id" = "comparison_tissue")) %>%
  arrange(comparison_id, tmm) %>% 
  
  group_by(comparison_id, type) %>% 
  mutate(cum_n = row_number()) %>%
  group_by(comparison_id) %>% 
  mutate(cum_fract = cum_n / n_distinct(ensg_id)) %>%
  ungroup() %>%
  
  mutate(comparison_id = str_to_sentence(comparison_id))

 

plot_data %>%
  filter(type == "Overlap") %>%
  ggplot(aes(log10(tmm + 1), cum_fract, color = comparison_id)) +
  geom_line() +
  scale_color_manual(values = tissue_colors_palette_comparison) +
  stripped_theme
ggsave(savepath("sup5 cum fract overlap.pdf"), width = 7, height = 5)
  
plot_data %>%
  ggplot(aes(log10(tmm + 1), type)) +
  geom_violin(draw_quantiles = 0.5) +
  scale_fill_manual(values = tissue_colors_palette_comparison) +
  stripped_theme
ggsave(savepath("sup5 overlap tmm distribution.pdf"), width = 3, height = 3)




###

plot_data2 <- 
  plot_data %>%
  group_by(comparison_id, type) %>% 
  
  count %>%
  group_by(comparison_id, .drop = F) %>%
  mutate(fract = n / sum(n)) %>%
  arrange(fract) %>% 
  ungroup %>% 
  filter(type == "Overlap") %>%
  mutate(comparison_id = factor(str_to_sentence(comparison_id), str_to_sentence(comparison_id))) 


plot_data2 %>%
  ggplot(aes(fract * 100, comparison_id, label =  paste(round(100*fract, 1), "%"))) + 
  geom_col() +
  geom_text(hjust = 0) +
  scale_x_continuous(expand = expansion(c(0)), limits = c(0,100)) +
  stripped_theme +
  ylab("") +
  xlab("% overlap") +
  ggtitle("% total enriched overlap",
          "overlap vs number of genes enriched in either pig or human") +
  
  plot_data %>%
  group_by(comparison_id) %>% 
  mutate(quartile = as.factor(unclass(cut(tmm, breaks = quantile(tmm), include.lowest = T)))) %>%
  ungroup(comparison_id = factor(comparison_id), 
          type = factor(type)) %>%
  group_by(comparison_id, type, quartile, .drop = F) %>% 
  count %>%
  group_by(comparison_id, quartile, .drop = F) %>%
  mutate(fract = n / sum(n)) %>%
  arrange(fract) %>% 
  ungroup %>% 
  filter(type == "Overlap") %>%
  mutate(comparison_id = factor(comparison_id, levels(plot_data2$comparison_id))) %>%
  ggplot(aes(fract * 100, comparison_id, label =  paste(round(100*fract, 1), "%"))) + 
  geom_col() +
  geom_text(hjust = 0) +
  scale_x_continuous(expand = expansion(c(0)), limits = c(0,100)) +
  facet_wrap(~quartile, nrow = 1) +
  stripped_theme_facet +
  theme(panel.grid.major.y = element_line(color = "gray85"), 
        panel.spacing = unit(4, "mm"),
        axis.text.x = element_blank()) +
  ylab("") +
  xlab("% overlap") +
  ggtitle("% Total enriched overlap", 
          "Overlap vs number of genes enriched in either pig or human") +
  
  plot_layout(widths = c(0.5, 1))
ggsave(savepath("sup5 overlap per quartile.pdf"), width = 12, height = 5)

 


```



#Side stories

##Calculations of numbers for manuscript

```{r}
gene_overlap_data_summarized %>%
  inner_join(gene_overlap_data %>% 
               # filter(spec_category_human %in% c("Tissue enriched", "Group enriched") |
               #          spec_category_pig %in% c("Tissue enriched", "Group enriched"))
               filter(spec_category_pig %in% c("Tissue enriched", "Group enriched")) %>% 
               select(enssscg_id) %>% 
               distinct()) %>% 
  group_by(type) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(fract = n / sum(n))

gene_overlap_data_summarized %>%
  left_join(pig_gene_classification) %>%
  group_by(specificity_category, type) %>% 
  count() %>% 
  ungroup() %>% 
  group_by(specificity_category) %>%
  mutate(fract = n / sum(n))


pig_gene_classification %>% 
  group_by(specificity_category) %>%
  count
pig_gene_classification %>% 
  group_by(distribution_category) %>%
  count

pig_gene_classification %>% 
  group_by(specificity_category, distribution_category) %>%
  count

pig_gene_classification %>% 
  filter(specificity_category == "Tissue enriched", 
         distribution_category == "detected in single") %>%
  group_by(enhanced_tissues) %>%
  count %>% 
  arrange(-n)


pig_gene_classification %>% 
  filter(!is.na(enhanced_tissues)) %>%
  separate_rows(enhanced_tissues, sep = ";") %>%
  left_join(tissue_mapping %>% 
              select(consensus_tissue_name, organ_name) %>% 
              distinct(),
            by = c("enhanced_tissues" = "consensus_tissue_name")) %>%
  group_by(enhanced_tissues, organ_name, specificity_category) %>% 
  count() -> a

a %>% 
  arrange(-n)

a %>%
  group_by(enhanced_tissues, organ_name) %>% 
  summarise(n = sum(n)) %>% 
  arrange(-n)

a %>% 
  filter(organ_name == "Adipose & soft tissue") %>%
  group_by(enhanced_tissues, organ_name) %>% 
  summarise(n = sum(n)) %>% 
  arrange(-n)


intraconsens_classification %>% 
  group_by(enssscg_id) %>% 
  count %>%
  mutate(a = n > 1) %>% 
  group_by(a) %>% 
  count

intraconsens_classification %>% 
  group_by(consensus_tissue_name) %>% 
  count%>%
  arrange(-n)


# Variable / elevated overlap

intraconsens_classification %>% 
  select(1, 2) %>%
  left_join(pig_gene_classification %>% 
              filter(!is.na(enhanced_tissues)) %>%
              separate_rows(enhanced_tissues, sep = ";") ,
            by = c("enssscg_id", "consensus_tissue_name" = "enhanced_tissues")) %>% 
  mutate(elevated = specificity_category %in% c("Tissue enhanced", "Tissue enriched", "Group enriched")) -> b

b %>% 
  group_by(consensus_tissue_name, elevated) %>% 
  count() %>% View

b %>% 
  group_by(consensus_tissue_name, elevated) %>% 
  count() %>%
  group_by(consensus_tissue_name) %>% 
  mutate(fract = n / sum(n)) %>% 
  filter(elevated) %>% 
  arrange(-fract)

intraconsens_classification %>% 
  select(-contains("cluster")) %>%
  left_join(pig_gene_classification %>% 
              filter(is.na(enhanced_tissues)) %>%
              select(-ts_score, -enhanced_score, -enhanced_tissues),
            by = c("enssscg_id")) %>% 
  filter(consensus_tissue_name == "brain",
         specificity_category == "Low tissue specificity") %T>%
  write_csv(savepath("variable brain low spec.csv")) %>%
  select(enssscg_id, high_tissues) %>% 
  separate_rows(high_tissues, sep = ";") %>%
  mutate(y = 1) %>%
  spread(high_tissues, y, fill = 0) %>% 
  left_join(gene_mapping %>% 
              select(1:2)) %>%
  unite(id, enssscg_id, display_name) %>%
  column_to_rownames("id") %>%
  pheatmap(clustering_method = "ward.D2",
           filename = savepath("Brain variable low tis spec.pdf"), 
           main = "Brain variable and low tissue spec",
           fontsize_row = 1,
           width = 6, 
           height = 12)
  

intraconsens_gene_enrich_overlap_all %>% 
  filter(consensus_tissue_name == "lymphoid tissue") %>%
  group_by(overlap) %>% 
  count

intraconsens_gene_enrich_overlap_all %>% 
  group_by(consensus_tissue_name, overlap) %>%
  count() %>%
  filter(overlap != "not variable - elevated") %>%
  group_by(consensus_tissue_name) %>% 
  mutate(fract = n / sum(n)) %>% 
  filter(overlap == "variable and elevated") %>% 
  arrange(-fract)

```

## Supplementary tables

```{r}

pig_gene_classification %>% 
  mutate(distribution_category = str_to_sentence(distribution_category)) %>% 
  select(Gene = 1, 
         Specificity = 2, 
         Distribution = 3, 
         `Elevated tissues` = enhanced_tissues) %>% 
  write_csv(savepath("supplementary pig gene classification.csv"))

```

##Metabolic example gene scatter

```{r}

joined_atlas_comparison %>% 
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>%
  inner_join(human_gene_mapping %>% 
               filter(gene_name %in% c("ABAT", "AGXT2")),
               # filter(gene_name %in% c("DBH", "MOXD1")),
             by = "ensg_id") %>%
  select(enssscg_id, ensg_id, comparison_tissue, species, mutual_tmm, gene_name) %>% 
  spread(species, mutual_tmm) %>% 
  ggplot(aes(pig, human, color = comparison_tissue, label = comparison_tissue)) +
  geom_text(show.legend = F) +
  facet_wrap(~gene_name) +
  scale_fill_manual(values = tissue_colors_palette_comparison) +
  stripped_theme_facet
  

```


##Ortholog alluvial

```{r}

gene_orthologs_pig_summary <-
  gene_orthologs_all %>% 
  select(enssscg_id, ortholog_type, ortholog_confidence) %>% 
  distinct()

gene_orthologs_pig_summary_n <- 
  gene_mapping %>%
  left_join(gene_orthologs_pig_summary) %>% 
  mutate(type = ifelse(is.na(ortholog_type), "no ortholog",
                       gsub("^ortholog_", "", ortholog_type))) %>% 
  group_by(type, ortholog_confidence) %>% 
  summarise(n = n()) %>% 
  ungroup()

bind_rows(gene_orthologs_pig_summary_n %>% 
            group_by(type) %>% 
            summarise(n = sum(n)) %>% 
            mutate(bar = "All genes"),
          gene_orthologs_pig_summary_n %>% 
            filter(type != "no ortholog") %>%
            group_by(type) %>% 
            summarise(n = sum(n)) %>% 
            mutate(bar = "Orthologs"),
          gene_orthologs_pig_summary_n %>% 
            filter(type != "no ortholog") %>%
            filter(ortholog_confidence == 1) %>%
            group_by(type) %>% 
            summarise(n = sum(n)) %>% 
            mutate(bar = "High confidence\northologs")) %>% 
  mutate(bar = factor(bar, unique(bar)),
         type = factor(type, c("no ortholog", "many2many", "one2many", "one2one"))) %>%
  ggplot(aes(bar, n, fill = type)) +
  geom_col(color = "black") +
  geom_text(data = . %>%
              group_by(bar) %>% 
              summarise(n = sum(n)),
            aes(bar, n, label = n),
            inherit.aes = F,
            hjust = 0) +
  scale_y_continuous(expand = expansion(c(0, 0.05))) +
  scale_fill_brewer(type = "seq", palette = 4) +
  stripped_theme +
  ggtitle("Number of pig genes with human orthologs") +
  coord_flip() +
  theme(legend.position = "bottom")

ggsave(savepath("Barplot number pig-human orthologs.pdf"), width = 4, height = 3)


```


##SLA genes

```{r}

pig_atlas_tissue %>% 
  inner_join(SLA_genes) %>% 
  select(gene_name, tissue_name, nx) %>%
  spread(gene_name, nx) %>%
  column_to_rownames("tissue_name") %>%
  {log10(. + 1)} %>% 
  # scale(center = T, scale = F) %>%
  pheatmap(border_color = NA, 
           color = heatmap_palette,
           clustering_method = "ward.D2", 
           cutree_rows = 5, 
           cutree_cols = 3,
           filename = savepath("SLA heatmap.pdf"),
           width = 5, height = 10)

pig_atlas_tissue %>% 
  inner_join(SLA_genes) %>% 
  left_join(tissue_mapping,
            by = "tissue_name") %>% 
  filter(consensus_tissue_name %in% c("heart",
                                      "liver",
                                      "kidney",
                                      "lung",
                                      "cornea",
                                      "skin")) %>%
  select(gene_name, tissue_name, nx) %>%
  spread(gene_name, nx) %>%
  column_to_rownames("tissue_name") %>%
  {log10(. + 1)} %>% 
  # scale(center = T, scale = F) %>%
  pheatmap(border_color = NA, 
           color = heatmap_palette,
           clustering_method = "ward.D2", 
           cutree_rows = 2, 
           cutree_cols = 3,
           filename = savepath("SLA heatmap specific.pdf"),
           width = 5, height = 4)
```

##GO on classification

###Main

```{r}
# library(topGO)




pig_gene_classification_GO_res <-
  pig_gene_classification %>%
  separate_rows(enhanced_tissues, sep = ";") %>%
  mutate(enhanced_tissues = ifelse(is.na(enhanced_tissues),
                                   "not elevated",
                                   enhanced_tissues)) %>%
  left_join(gene_orthologs) %>% 
  select(enhanced_tissues,
         human_gene_name) %>% 
  filter(!is.na(human_gene_name)) %>%
  distinct() %>%
  group_by(enhanced_tissues) %>% 
  
  do({
    enrichr(.$human_gene_name,
            databases = c("GO_Biological_Process_2018")) %>%
      bind_rows(.id = "database")
  }) %>%
  as_tibble()

plot_data <- 
  pig_gene_classification_GO_res %>%
  filter(Adjusted.P.value < 0.05) %>%
  # select(xeno_id, Term, Adjusted.P.value, Combined.Score) %>%
  mutate(Term = gsub(" \\(.*\\)$", "", Term)) %>%
  mutate(Term = gsub("B cell", "B-cell", Term)) %>%
  mutate(Term = gsub("T cell", "T-cell", Term)) %>%
  mutate(Term = gsub("\\ acid", "\\-acid", Term)) %>%
  mutate(Term = tolower(Term)) %>% 
  separate_rows(Term, sep = " ") %>% 
  filter(!Term %in% c(stopwords("english"), 
                      c("regulation", "positive", "cell", "response", "via",
                        "negative", "signaling", "cellular", "pathway", "process"))) %>% 
  mutate(score = -log10(Adjusted.P.value)) %>%
  group_by(enhanced_tissues, Term) %>% 
  summarise(n = n(),
            score = sum(score),
            mean_score = sum(score) / n) %>%
  arrange(-n) %>% 
  ungroup()
  
plot_data %>% 
  filter(n > 2) %>%
  ggplot(aes(log10(n))) +
  geom_density() +
  plot_data %>% 
  ggplot(aes(log10(score))) +
  geom_density() +
  plot_data %>% 
  ggplot(aes(log10(mean_score))) +
  geom_density()


plot <- 
  plot_data %>%
  group_by(enhanced_tissues) %>%
  top_n(20, score) %>%
  mutate(score = score / max(score)) %>%
  
  ggplot(aes(label = Term, size = score)) +
  geom_text_wordcloud(eccentricity = 0.8) +
  theme_minimal() +
  facet_wrap(~enhanced_tissues) +
  theme(strip.text = element_text(face = "bold")) +
  scale_size_continuous(range = c(1, 3))
ggsave(savepath("Pig classification GO mean score wordcloud.pdf"), plot = plot, width = 16, height = 12)


#   
# a <- 
#   pig_gene_classification %>% 
#   filter(enhanced_tissues == "liver")
# 
# topGOdata <- new( # New creates an object of a defined class
#   'topGOdata', # The class of object to create
#   ontology = 'BP', # We can choose from Molecular Function (MF), 
#   # Biological Process (BP), 
#   # and Cellular Component (CC)
#   allGenes = geneList, # The p-values of all genes in your analysis
#   geneSel = function(), # A function for defining the subset you'd like to analyze
#   annot = annotFUN(), # A function defining your annotation
#   # and some other arguments
# )
# 
# GOdata <- new('topGOdata',
#                ontology='BP', # We'll use Biological Process for this example
#                allGenes=geneList,
#                geneSel = selFun,
#                annot=annFUN.org,
#                mapping = 'org.Hs.eg.db', # The annotation package for the human genome
#                ID = 'symbol' # We're using gene symbols
# )
# 
#  annot = annFUN.gene2GO, gene2GO = geneID2GO


human_gene_classification_GO_res <-
  human_gene_classification %>%
  separate_rows(enhanced_tissues, sep = ";") %>%
  mutate(enhanced_tissues = ifelse(is.na(enhanced_tissues),
                                   "not elevated",
                                   enhanced_tissues)) %>%
  left_join(gene_orthologs) %>% 
  select(enhanced_tissues,
         human_gene_name) %>% 
  filter(!is.na(human_gene_name)) %>%
  distinct() %>%
  group_by(enhanced_tissues) %>% 
  
  do({
    enrichr(.$human_gene_name,
            databases = c("GO_Biological_Process_2018")) %>%
      bind_rows(.id = "database")
  }) %>%
  as_tibble()

```

###Variable lymph


```{r}
# library(topGO)




pig_gene_variability_GO_res <-
  intraconsens_classification %>%
  select(1:3) %>%
  separate_rows(high_tissues, sep = ";") %>%
  left_join(gene_orthologs) %>% 
  select(high_tissues,
         human_gene_name) %>% 
  filter(!is.na(human_gene_name)) %>%
  distinct() %>%
  group_by(high_tissues) %>% 
  
  do({
    enrichr(.$human_gene_name,
            databases = c("GO_Biological_Process_2018")) %>%
      bind_rows(.id = "database")
  }) %>%
  as_tibble()


plot_data <- 
  pig_gene_variability_GO_res %>%
  filter(Adjusted.P.value < 0.05) %>%
  # select(xeno_id, Term, Adjusted.P.value, Combined.Score) %>%
  mutate(Term = gsub(" \\(.*\\)$", "", Term)) %>%
  mutate(Term = gsub("B cell", "B-cell", Term)) %>%
  mutate(Term = gsub("T cell", "T-cell", Term)) %>%
  mutate(Term = gsub("\\ acid", "\\-acid", Term)) %>%
  mutate(Term = tolower(Term)) %>% 
  separate_rows(Term, sep = " ") %>% 
  filter(!Term %in% c(stopwords("english"), 
                      c("regulation", "positive", "cell", "response", "via",
                        "negative", "signaling", "cellular", "pathway", "process"))) %>% 
  mutate(score = -log10(Adjusted.P.value)) %>%
  group_by(high_tissues, Term) %>% 
  summarise(n = n(),
            score = sum(score),
            mean_score = sum(score) / n) %>%
  arrange(-n) %>% 
  ungroup()
  
plot_data %>% 
  filter(n > 2) %>%
  ggplot(aes(log10(n))) +
  geom_density() +
  plot_data %>% 
  ggplot(aes(log10(score))) +
  geom_density() +
  plot_data %>% 
  ggplot(aes(log10(mean_score))) +
  geom_density()


plot <- 
  plot_data %>%
  group_by(high_tissues) %>%
  top_n(20, score) %>%
  mutate(score = score / max(score)) %>%
  
  ggplot(aes(label = Term, size = score)) +
  geom_text_wordcloud(eccentricity = 0.8) +
  theme_minimal() +
  facet_wrap(~high_tissues) +
  theme(strip.text = element_text(face = "bold")) +
  scale_size_continuous(range = c(1, 3))
ggsave(savepath("Pig variability GO mean score wordcloud.pdf"), plot = plot, width = 16, height = 12)


#   
# a <- 
#   pig_gene_classification %>% 
#   filter(enhanced_tissues == "liver")
# 
# topGOdata <- new( # New creates an object of a defined class
#   'topGOdata', # The class of object to create
#   ontology = 'BP', # We can choose from Molecular Function (MF), 
#   # Biological Process (BP), 
#   # and Cellular Component (CC)
#   allGenes = geneList, # The p-values of all genes in your analysis
#   geneSel = function(), # A function for defining the subset you'd like to analyze
#   annot = annotFUN(), # A function defining your annotation
#   # and some other arguments
# )
# 
# GOdata <- new('topGOdata',
#                ontology='BP', # We'll use Biological Process for this example
#                allGenes=geneList,
#                geneSel = selFun,
#                annot=annFUN.org,
#                mapping = 'org.Hs.eg.db', # The annotation package for the human genome
#                ID = 'symbol' # We're using gene symbols
# )
# 
#  annot = annFUN.gene2GO, gene2GO = geneID2GO
```

###UMAP class


```{r}
# library(topGO)




pig_gene_UMAP_GO_res <-
  pig_gene_umap %>%
  select(enssscg_id, merged_cluster) %>%
  left_join(gene_orthologs) %>% 
  
  filter(!is.na(human_gene_name)) %>%
  distinct() %>%
  group_by(merged_cluster) %>% 
  
  do({
    enrichr(.$human_gene_name,
            databases = c("GO_Biological_Process_2018",
                          "GO_Cellular_Component_2018",
                          "GO_Molecular_Function_2018")) %>%
      bind_rows(.id = "database")
  }) %>%
  as_tibble()


pig_gene_UMAP_GO_res %>%
  filter(Adjusted.P.value < 0.05) %>%
  filter(merged_cluster == 23) %>%
  select(Term, Genes) %>%
  pull(1)

pig_gene_UMAP_GO_res %>%
  filter(Adjusted.P.value < 0.05) %>%
  filter(merged_cluster == 1) %>%
  select(Term, Genes) 

pig_gene_UMAP_GO_res %>%
  filter(Adjusted.P.value < 0.05) %>%
  filter(merged_cluster == 80) %>%
  select(Term, Genes) 

pig_gene_UMAP_GO_res %>%
  ggplot(aes(merged_cluster, -log10(Adjusted.P.value))) +
  geom_violin(scale = "width") +
  geom_hline(yintercept = -log10(0.05))
  
pig_gene_UMAP_GO_res %>%
  group_by(Term) %>% 
  summarise(n_sign = length(which(Adjusted.P.value < 0.05))) %>%
  group_by(n_sign) %>% 
  count() %>%
  ggplot(aes(as_factor(n_sign), n)) +
  geom_col()
# pig_gene_UMAP_GO_res %>%
#   filter(Adjusted.P.value < 0.05) %>%
#   filter(merged_cluster == 23) %>%
#   select(Term, Genes) %>%
#   separate_rows(Genes, sep = ";") %>%
#   mutate(a = 1) %>% 
#   spread(Genes, a, fill = 0) %>% 
#   column_to_rownames("Term") %>%
#   uwot::umap()
# 
# pig_gene_UMAP_GO_res %>%
#   filter(Adjusted.P.value < 0.05) %>%
#   filter(merged_cluster == 23)%>%
#   select(Term, Genes, Adjusted.P.value) %>%
#   bind_cols(as_tibble(a)) %>%
#   mutate(logp = -log10(Adjusted.P.value)) %>%
#   ggplot(aes(V1, V2, label = Term, size = logp))+
#   geom_point() +
#   geom_text()

plot_data1 <- 
  pig_gene_UMAP_GO_res %>%
  filter(Adjusted.P.value < 0.05)

plot_data2 <- 
  pig_gene_cluster_enrich_hyper %>%
  arrange(adj_pval) %>% 
  filter(adj_pval < 0.05)

plot_data3 <- 
  pig_atlas_tissue %>%
  left_join(pig_gene_umap %>% 
              select(enssscg_id, merged_cluster)) %>%
  
  mutate(log_nx = log10(nx + 1))


plots <- 
  pblapply(1:max(as.numeric(pig_gene_umap$merged_cluster)),
           function(i) {
             plot_data1 %>%
               filter(merged_cluster == i) %>%
               mutate(Term = factor(Term, rev(Term))) %>%
               head(20) %>%
               ggplot(aes(-log10(Adjusted.P.value), Term)) +
               geom_col() +
               ggtitle(i,
                       paste(pig_gene_umap %>% filter(merged_cluster == i) %>% nrow(),
                             "genes")) +
               
               plot_data2 %>% 
               filter(merged_cluster == i) %>%
               mutate(enhanced_tissues = factor(enhanced_tissues, rev(enhanced_tissues))) %>%
               ggplot(aes(-log10(adj_pval), enhanced_tissues)) +
               geom_col() +
               
               plot_data3 %>%
               filter(merged_cluster == i) %>%
               {
                 clus_data <- 
                   plot_data3 %>%
                   filter(merged_cluster == i) %>%
                   select(1, 2, log_nx) %>% 
                   spread(tissue_name, log_nx) %>% 
                   column_to_rownames("enssscg_id")
                 
                 mutate(., 
                        enssscg_id = factor(enssscg_id,
                                            clus_data %>%
                                              dist() %>% 
                                              hclust(method = "ward.D2") %>%
                                              labels()),
                        tissue_name = factor(tissue_name,
                                             clus_data %>%
                                               t() %>%
                                               dist() %>% 
                                               hclust(method = "ward.D2") %>%
                                               labels()))
               } %>%
               ggplot(aes(enssscg_id, tissue_name, fill = log_nx)) +
               geom_tile() +
               scale_fill_viridis(option = "B", direction = -1) +
               theme(axis.text.x = element_blank()) +
               
               plot_layout(widths = c(0.25, 0.25, 0.5))
           })

pdf(savepath("UMAP cluster detailed profile.pdf"),
    width = 16,
    height = 10)
plots
dev.off()

plot_data <- 
  pig_gene_UMAP_GO_res %>%
  filter(Adjusted.P.value < 0.05) %>%
  # select(xeno_id, Term, Adjusted.P.value, Combined.Score) %>%
  mutate(Term = gsub(" \\(.*\\)$", "", Term)) %>%
  mutate(Term = gsub("B cell", "B-cell", Term)) %>%
  mutate(Term = gsub("T cell", "T-cell", Term)) %>%
  mutate(Term = gsub("\\ acid", "\\-acid", Term)) %>%
  mutate(Term = tolower(Term)) %>% 
  separate_rows(Term, sep = " ") %>% 
  filter(!Term %in% c(stopwords("english"), 
                      c("regulation", "positive", "cell", "response", "via",
                        "negative", "signaling", "cellular", "pathway", "process"))) %>% 
  mutate(score = -log10(Adjusted.P.value)) %>%
  group_by(merged_cluster, Term) %>% 
  summarise(n = n(),
            score = sum(score),
            mean_score = sum(score) / n) %>%
  arrange(-n) %>% 
  ungroup()
  
plot_data %>% 
  filter(n > 2) %>%
  ggplot(aes(log10(n))) +
  geom_density() +
  plot_data %>% 
  ggplot(aes(log10(score))) +
  geom_density() +
  plot_data %>% 
  ggplot(aes(log10(mean_score))) +
  geom_density()


plot <- 
  plot_data %>%
  group_by(merged_cluster) %>%
  top_n(20, score) %>%
  mutate(score = score / max(score)) %>%
  
  ggplot(aes(label = Term, size = score)) +
  geom_text_wordcloud(eccentricity = 0.8) +
  theme_minimal() +
  facet_wrap(~merged_cluster) +
  theme(strip.text = element_text(face = "bold")) +
  scale_size_continuous(range = c(1, 3))
ggsave(savepath("Pig umap cluster GO mean score wordcloud.pdf"), plot = plot, width = 20, height = 20)


pig_gene_UMAP_GO_res %>%
  filter(Adjusted.P.value < 0.05) %>%
  select(1, Term, Overlap, Adjusted.P.value, Genes, Combined.Score) %>%
  write_csv(savepath("UMAP GO res.csv"))

```

####ReviGO

```{r}

library(rrvgo)
library(treemapify)

plot_data_all <- 
  pig_gene_UMAP_GO_res %>%
  filter(Adjusted.P.value < 0.05) %>%
  mutate(GO = str_extract(Term, "\\(.*\\)") %>%
           gsub("\\(|\\)", "", .)) %>%
  select(merged_cluster, GO, P = Adjusted.P.value) 

plots <- 
  pblapply(1:max(as.numeric(pig_gene_umap$merged_cluster)),
           function(i) {
             cat(paste(i, "\n"))
             plot_data <- 
               plot_data_all %>% 
               filter(merged_cluster == i)
             
             if(nrow(plot_data) > 1) {
               plot_mat <- 
                 calculateSimMatrix(plot_data$GO, 
                                    orgdb="org.Hs.eg.db",
                                    ont="BP",
                                    method="Rel") 
               plot_mat2 <- 
                 reduceSimMatrix(plot_mat, 
                                 setNames(-log10(plot_data$P), plot_data$GO),
                                 threshold=0.7,
                                 orgdb="org.Hs.eg.db") 
               
               plot_mat2 %>% 
                 as_tibble() %>%
                 group_by(parentTerm) %>%
                 mutate(n = row_number()) %>%
                 ggplot(aes(area = size, subgroup = parentTerm)) +
                 
                 geom_treemap(aes(fill = parentTerm,
                                  alpha = n), 
                              color = "black",
                              show.legend = F) +
                 geom_treemap_subgroup_border(color = "black") +
                 
                 geom_treemap_text(aes(label = term),
                                   colour = "black", 
                                   place = "centre",
                                   alpha = 0.4,
                                   grow = TRUE) +
                 
                 geom_treemap_subgroup_text(place = "centre", 
                                            grow = T, 
                                            reflow = T,
                                            alpha = 1, 
                                            colour = "white", 
                                            fontface = "bold",
                                            min.size = 0) +
                 scale_alpha_continuous(range = c(0.8, 1)) +
                 scale_fill_manual(values = rep(ggthemes::tableau_color_pal(palette = "Hue Circle", 
                                                                            type = c("regular"), 
                                                                            direction = 1)(19),
                                                10)) +
                 theme_void() +
                 ggtitle(i)
             } else {
               ggplot() +
                 theme_void() +
                 ggtitle(i)
             }
             
           })


pdf(savepath("UMAP cluster GSEA treemaps.pdf"), width = 8, height = 5)
plots
dev.off()

g_gene_UMAP_GO_res %>%
  filter(merged_cluster == 66,
         Adjusted.P.value < 0.05) %>%
  mutate(GO = str_extract(Term, "\\(.*\\)") %>%
           gsub("\\(|\\)", "", .)) %>%
  select(GO, P = Adjusted.P.value) %>%
  write.csv(savepath("cluster 66.csv"))

```

####UMAP cluster scatter

```{r}
plots <- 
  pblapply(unique(pig_gene_umap$merged_cluster),
           function(i) {
             pig_gene_umap %>%
               filter(merged_cluster == i) %>%
               left_join(gene_orthologs) %>%
               mutate(id = case_when(is.na(human_gene_name) & 
                                       is.na(pig_gene_name) ~ gsub("^ENSSSCG0*", "", enssscg_id), 
                                     is.na(human_gene_name) ~ pig_gene_name,
                                     T ~ human_gene_name)) %>% 
               # select(human_gene_name, pig_gene_name, id)
               ggplot(aes(V1, V2, label = id)) +
               geom_text(alpha = 0.5, 
                         size = 2.5) +
               stripped_theme +
               ggtitle(i)
           })
  
pdf(savepath("UMAP cluster scatter.pdf"), width = 8, height = 8)
plots
dev.off()

```

####UMAP cluster prop

```{r}


plot_data <- 
  pig_atlas_tissue %>%
  left_join(pig_gene_umap %>% 
              select(enssscg_id, merged_cluster)) %>%
  select(1, 2, ptpm, merged_cluster) %>%
  spread(tissue_name, ptpm)

# 
# plot_data %>% 
#   # group_by(merged_cluster) %>%
#   filter(merged_cluster == 66) %>%
#   
#   select(., -merged_cluster) %>%
#   column_to_rownames("enssscg_id") %>%
#   # t() %>%
#   # {
#   #              names <- colnames(.)
#   #              
#   #              propr(.) %>%
#   #                {.@results} %>%
#   #                as_tibble() %>%
#   #                mutate(id1 = names[Partner],
#   #                       id2 = names[Pair]) %>% 
#   #                select(id1, id2, propr)
#   #            }
#   {. + 1} %>%
#   t() %>%
#   
#   {left_join(cor(log10(.), method = "pearson") %>%
#                as_tibble(rownames = "id1") %>%
#                gather(id2, cor, -1) %>%
#                filter(id1 > id2),
#              {
#                names <- colnames(.)
#                
#                propr(., metric = "rho") %>%
#                  {.@results} %>%
#                  as_tibble() %>%
#                  mutate(id1 = names[Partner],
#                         id2 = names[Pair]) %>% 
#                  select(id1, id2, propr)
#              })} %>%
#   ggplot(aes(cor, propr)) +
#   geom_hex(bins = 50)
#    
# 
# 
pig_umap_cluster_tissue_cor <- 
  plot_data %>% 
  group_by(merged_cluster) %>%
  # filter(merged_cluster == 66) %>%
  do({
    select(., -merged_cluster) %>%
      column_to_rownames("enssscg_id") %>%
      {log10(. + 1)} %>%
      # t() %>%
      cor(method = "pearson") %>%
      as_tibble(rownames = "id1") %>%
      gather(id2, cor, -1) %>%
      filter(id1 < id2) 
  }) 

pig_umap_cluster_gene_cor <- 
  plot_data %>% 
  group_by(merged_cluster) %>%
  # filter(merged_cluster == 66) %>%
  do({
    select(., -merged_cluster) %>%
      column_to_rownames("enssscg_id") %>%
      {log10(. + 1)} %>%
      t() %>%
      cor(method = "pearson") %>%
      as_tibble(rownames = "id1") %>%
      gather(id2, cor, -1) %>%
      filter(id1 < id2) 
  }) 
  
left_join(pig_umap_cluster_tissue_cor %>%
            summarise(cor = median(cor)),
          pig_umap_cluster_gene_cor %>%
            summarise(cor = median(cor)),
          by = c("merged_cluster"),
          suffix = c("_tissue", "_gene")) %>%
  ggplot(aes(cor_tissue, cor_gene, label = merged_cluster)) +
  geom_text()

pig_umap_cluster_tissue_cor %>%
  ggplot(aes(merged_cluster, cor)) +
  geom_violin(draw_quantiles = 0.5, scale = "width") +
  stripped_theme +
  theme(panel.grid.major = element_line(color = "gray80"))


```






###XGSA

```{r}

library(xgsa)

cardiac.perturbation.data <- read.table("http://cardiaccode.victorchang.edu.au/data/Cardiaccode_in_vivo_evidence_2014_06_10.txt", sep="\t", header=TRUE, quote="\"")
mouse.cardiac.genes <- unique(unlist(cardiac.perturbation.data[cardiac.perturbation.data$Species == "Mus musculus", c("Regulator","Target")]))

mouse.ensembl.symbol.map <- get_ENSEMBL_symbol_map(species = 'mmusculus')
mouse.cardiac.ensembl.symbols <- mouse.ensembl.symbol.map$ensembl_gene_id[mouse.ensembl.symbol.map$external_gene_name %in% mouse.cardiac.genes]

mouse.data <- new_XGSA_dataset(species = 'mmusculus', data = list(mouseCardiacGenes = mouse.cardiac.ensembl.symbols), type = 'genesetlist', name = 'MouseCardiacGenes', universe = unique(mouse.ensembl.symbol.map$ensembl_gene_id))
#

zebrafish.GO <- get_GO('drerio', ontologies = "biological_process")
#> [1] "retrieved GO"
zebrafish.GO <- zebrafish.GO[lapply(zebrafish.GO, length) > 10 & lapply(zebrafish.GO, length) < 500]
zebrafish.GO.data <- new_XGSA_dataset(species = "drerio", data = zebrafish.GO, type = 'genesetlist', name = "ZebrafishGO", universe = unique(unlist(zebrafish.GO)))

# Now we can compare the mouse cardiac genes to the zebrafish gene ontology.
mouse.cardiac.vs.zebrafish.GO.results <- run_XGSA_test(mouse.data, zebrafish.GO.data)
#
# We need to separate the pvalues and the overlapping gene IDs, because XGSA returns both.
# The p.values are stored in the first element of each result, and the overlapping genes are stored in the second element.
resulting.pvals <- lapply(mouse.cardiac.vs.zebrafish.GO.results, function(X){ X[["pvals"]] })
resulting.overlap.genes <- lapply(mouse.cardiac.vs.zebrafish.GO.results, function(X){ X[["genes"]] })
#
# Now we perform Benjamini Hochberg multiple hypothesis testing correction to the pvalues.
adjusted.pvals <- p.adjust(unlist(resulting.pvals), method = "BH")
#
# We need to make the names of our results interpretable for humans, so we extract the GO Term IDs
names(adjusted.pvals) <- unlist(lapply(strsplit(names(adjusted.pvals) ,"\\."), function(X){return(X[[2]])}))
# We can use another XGSA helper function to find out the GO term names.
zebrafish.GO.names <- get_GO_names('drerio')
# And finally we get interpretable names
names(adjusted.pvals) <- zebrafish.GO.names[match( names(adjusted.pvals), zebrafish.GO.names$go_id),"name_1006"]
#
Now let's look at the 10 most significant GO term results

# Now let's look at the 10 most significant GO term results
significant.GO.Terms <- adjusted.pvals[which(adjusted.pvals < 0.05)]
print(head(sort(significant.GO.Terms),10))
#>                             cell fate commitment 
#>                                     8.363504e-09 
#>                             mesoderm development 
#>                                     2.305281e-07 
#>                          cell fate specification 
#>                                     2.734734e-07 
#>                 embryonic heart tube development 
#>                                     9.498289e-07 
#>                       cardiocyte differentiation 
#>                                     4.236265e-06 
#>                         diencephalon development 
#>                                     4.506317e-06 
#>                              heart morphogenesis 
#>                                     9.151835e-06 
#>                             endoderm development 
#>                                     3.238483e-05 
#>                                heart development 
#>                                     3.238483e-05 
#> regulation of endodermal cell fate specification 
#>                                     3.238483e-05
par(mar=c(10,5,4,2))
barplot(-log10(head(sort(significant.GO.Terms),10)), ylab = "- log 10 p-value", las=2)

```


##GSVA

```{r}



pig_gene_comparison_classification_GO_res <-
  pig_gene_comparison_classification %>%
  separate_rows(enhanced_tissues, sep = ";") %>%
  mutate(enhanced_tissues = ifelse(is.na(enhanced_tissues),
                                   "not elevated",
                                   enhanced_tissues)) %>%
  left_join(gene_orthologs) %>% 
  select(enhanced_tissues,
         human_gene_name) %>% 
  filter(!is.na(human_gene_name)) %>%
  distinct() %>%
  group_by(enhanced_tissues) %>% 
  
  do({
    enrichr(.$human_gene_name,
            databases = c("GO_Biological_Process_2018")) %>%
      bind_rows(.id = "database")
  }) %>%
  as_tibble()

human_gene_comparison_classification_GO_res <-
  human_gene_comparison_classification %>%
  separate_rows(enhanced_tissues, sep = ";") %>%
  mutate(enhanced_tissues = ifelse(is.na(enhanced_tissues),
                                   "not elevated",
                                   enhanced_tissues)) %>%
  left_join(gene_orthologs) %>% 
  select(enhanced_tissues,
         human_gene_name) %>% 
  filter(!is.na(human_gene_name)) %>%
  distinct() %>%
  group_by(enhanced_tissues) %>% 
  
  do({
    enrichr(.$human_gene_name,
            databases = c("GO_Biological_Process_2018")) %>%
      bind_rows(.id = "database")
  }) %>%
  as_tibble()

#####

left_join(pig_gene_comparison_classification_GO_res %>%
            select(enhanced_tissues, Term, adj_p = Adjusted.P.value),
          human_gene_comparison_classification_GO_res %>%
            select(enhanced_tissues, Term, adj_p = Adjusted.P.value),
          by = c("enhanced_tissues", "Term"),
          suffix = c("_pig", "_human")) %>% 
  mutate(log_p_pig = -log10(adj_p_pig),
         log_p_human = -log10(adj_p_human)) %>%
  group_by(enhanced_tissues) %>% 
  mutate(spearman = cor(log_p_pig, log_p_human, method = "spearman", use = "pairwise.complete.obs")) %>%
  ungroup() %>%
  arrange(spearman) %>% 
  mutate(enhanced_tissues = factor(enhanced_tissues, unique(enhanced_tissues))) %>%
  ggplot(aes(log_p_pig, log_p_human)) +
  geom_point() +
  geom_hline(yintercept = -log10(0.05)) +
  geom_vline(xintercept = -log10(0.05)) +
  facet_wrap(~enhanced_tissues + round(spearman, 3), scales = "free") +
  stripped_theme_facet
ggsave(savepath("pig human GSEA comparison scatter.pdf"), width = 10, height = 10)

```

##UMAP Cluster zoom

```{r}

read_delim("data/meta/ensembl92 pig gene loci.txt", delim = "\t") %>%
  filter(str_length(`Chromosome/scaffold name`) < 5) %>%
  mutate(selected = `Gene stable ID` %in% c("ENSSSCG00000014730",
                                            "ENSSSCG00000014740",
                                            "ENSSSCG00000014745",
                                            "ENSSSCG00000014746",
                                            "ENSSSCG00000014747",
                                            "ENSSSCG00000014748",
                                            "ENSSSCG00000014749",
                                            "ENSSSCG00000014752",
                                            "ENSSSCG00000014754",
                                            "ENSSSCG00000014755",
                                            "ENSSSCG00000014756",
                                            "ENSSSCG00000014757",
                                            "ENSSSCG00000014758",
                                            "ENSSSCG00000014759",
                                            "ENSSSCG00000014760",
                                            "ENSSSCG00000014761",
                                            "ENSSSCG00000014762")) %>%
  filter(`Chromosome/scaffold name` == 9) %>%
  ggplot(aes(x = Strand, xend = Strand,
             y = `Gene start (bp)`, yend = `Gene end (bp)`,
             color = selected,
             size = selected)) +
  geom_segment() +
 
  # facet_wrap(~`Chromosome/scaffold name`, nrow = 1) +
    facet_zoom(y = selected) +
  # scale_x_continuous(expand = expansion(2)) +
  stripped_theme


read_delim("data/meta/ensembl92 pig gene loci.txt", delim = "\t") %>%
  filter(str_length(`Chromosome/scaffold name`) < 5) %>%
  mutate(selected = `Gene stable ID` %in% c("ENSSSCG00000031149",
                                            "ENSSSCG00000031850",
                                            "ENSSSCG00000032099",
                                            "ENSSSCG00000032368",
                                            "ENSSSCG00000032868",
                                            "ENSSSCG00000033254",
                                            "ENSSSCG00000033297",
                                            "ENSSSCG00000033330",
                                            "ENSSSCG00000033601",
                                            "ENSSSCG00000033603",
                                            "ENSSSCG00000033722",
                                            "ENSSSCG00000033910",
                                            "ENSSSCG00000034088",
                                            "ENSSSCG00000034800",
                                            "ENSSSCG00000034940",
                                            "ENSSSCG00000035188",
                                            "ENSSSCG00000035199",
                                            "ENSSSCG00000035294",
                                            "ENSSSCG00000036134",
                                            "ENSSSCG00000036510",
                                            "ENSSSCG00000036535",
                                            "ENSSSCG00000036584",
                                            "ENSSSCG00000036586",
                                            "ENSSSCG00000037603",
                                            "ENSSSCG00000037624",
                                            "ENSSSCG00000037708",
                                            "ENSSSCG00000037750",
                                            "ENSSSCG00000037797",
                                            "ENSSSCG00000037813",
                                            "ENSSSCG00000037818",
                                            "ENSSSCG00000038373",
                                            "ENSSSCG00000038727",
                                            "ENSSSCG00000038862",
                                            "ENSSSCG00000039065",
                                            "ENSSSCG00000039164",
                                            "ENSSSCG00000039434",
                                            "ENSSSCG00000039623",
                                            "ENSSSCG00000039987",
                                            "ENSSSCG00000040054",
                                            "ENSSSCG00000040246",
                                            "ENSSSCG00000040321",
                                            "ENSSSCG00000040403",
                                            "ENSSSCG00000040502")) %>%
  filter(selected) %>% 
  group_by(`Chromosome/scaffold name`) %>%
  count
  filter(`Chromosome/scaffold name` == 1) %>%
  ggplot(aes(x = Strand, xend = Strand,
             y = `Gene start (bp)`, yend = `Gene end (bp)`,
             color = selected,
             size = selected)) +
  geom_segment() +
 
  # facet_wrap(~`Chromosome/scaffold name`, nrow = 1) +
    facet_zoom(y = selected) +
  # scale_x_continuous(expand = expansion(2)) +
  stripped_theme_facet



```

##Ortholog correlation vs homology

```{r}

comparison_ortholog_cor <- 
  joined_atlas_comparison %>%
  select(mutual_id, comparison_tissue, species, ptpm) %>% 
  mutate(ptpm = log10(ptpm + 1)) %>%
  spread(species, ptpm) %>% 
  separate(mutual_id, into = c("enssscg_id", "ensg_id"), sep = "_") %>%
  group_by(enssscg_id, ensg_id) %>%
  summarise(spearman = cor(human, pig, use = "pairwise.complete.obs", method = "spearman"),
            pearson = cor(human, pig, use = "pairwise.complete.obs", method = "pearson"),
            prop = 1 - (var(human - pig)/(var(human) + var(pig)))) %>%
  left_join(gene_orthologs_all)

comparison_ortholog_cor %>% 
  ungroup %>% 
  filter(!is.na(sequence_identity)) %>%
  select(1:5, sequence_identity) %>% 
  gather(cor_type, cor, spearman, pearson, prop) %>%
  ggplot(aes(cor, sequence_identity)) +
  geom_hex(aes(fill = stat(log10(count))),
           bins = 100) +
  geom_text(data = . %>%
              group_by(cor_type) %>%
              summarise(spear = cor(cor, sequence_identity, use = "pairwise.complete.obs", method = "spearman")),
            aes(x = 0, y = 0, label = round(spear, 3))) +
  facet_wrap(~cor_type) +
  scale_fill_viridis() + 
  stripped_theme_facet
```

##Filter metabolic list

```{r}

read_delim("data/processed/TMM_negtive_spearman_reaction_mapped_diff.txt",
           delim = "\t") %>%
  arrange(react_id) %>%
  filter(pig_TMM > 1 |
           human_TMM > 1) %>%
  group_by(react_id, tissue, mutual_id) %>%
  mutate(direction = sign(pig_TMM - human_TMM),
         ratio = max(c(pig_TMM, 1)) / max(c(1, human_TMM)),
         ratio = ifelse(ratio < 1, 1 / ratio, ratio)) %>%
  ungroup() %>%
  filter(ratio > 4) %>%
  inner_join(group_by(., tissue, react_id, direction) %>%
               count %>%
               spread(direction, n, fill = 0) %>%
               filter(`-1` > 0,
                      `1` > 0) %>%
               select(1, 2)) %>%
  group_by(react_id, tissue) %>%
  mutate(mean_ratio = psych::geometric.mean(ratio)) %>%
    
  arrange(-mean_ratio, react_id, tissue, mutual_id) %>%
  write_csv(savepath("most_extreme_flipdiff_reactions.csv")) 
  
  

```


# Xenotransplantation knockout genes

## Heatmaps

```{r}



# pig_gene_classification %>%
#   inner_join(xeno_ko_genes) %>% View

pig_atlas_tissue %>% 
  inner_join(xeno_ko_genes) %>%
  # filter(enssscg_id %in% c(xeno_ko_genes$enssscg_id, SLA_genes$enssscg_id)) %>%
  left_join(tissue_mapping,
            by = c("tissue_name")) %>% 
   filter(consensus_tissue_name %in% c("heart",
                                      "liver",
                                      "kidney",
                                      "lung",
                                      "cornea",
                                      "skin")) %>%
  select(name_id, tissue_name, nx) %>%
  spread(name_id, nx) %>%
  column_to_rownames("tissue_name") %>%
  {log10(. + 1)} %>%
  # scale(center = F, scale = T) %>%
  pheatmap(border_color = NA, 
           color = heatmap_palette,
           clustering_method = "ward.D2", 
           cutree_rows = 3, 
           cutree_cols = 3,
           filename = savepath("Xeno KO heatmap.pdf"),
           width = 5, height = 5)


pig_atlas_tissue %>% 
  inner_join(xeno_ko_genes) %>%
  # filter(enssscg_id %in% c(xeno_ko_genes$enssscg_id, SLA_genes$enssscg_id)) %>%
  
  ggplot(aes(tmm, tissue_name)) + 
  geom_col() +
  facet_wrap(~enssscg_id + gene_name)
```

## Co-expression network

```{r}


xeno_ko_prop_res <- 
  pig_gene_propr_res %>%
  {.[, xeno_ko_genes$enssscg_id]} %>% 
  as_tibble(rownames = "enssscg_id1") %>% 
  gather(enssscg_id2, prop, -1) %>%
  left_join(gene_mapping,
            by = c("enssscg_id1" = "enssscg_id")) %>%
  left_join(xeno_ko_genes,
            by = c("enssscg_id2" = "enssscg_id"),
            suffix = c("1", "2")) %>%
  select(-biotype) %>%
  rename(gene_name1 = display_name, 
         gene_name2 = gene_name, 
         gene_id2 = name_id) %>%
  mutate(gene_id1 = ifelse(gene_name1 == "NULL", enssscg_id1, paste(gene_name1, enssscg_id1))) %>%
  select(enssscg_id1, enssscg_id2, gene_name1, gene_name2, gene_id2, gene_id1, prop)



net_data <- 
  xeno_ko_prop_res %>%
  filter(abs(prop) > 0.65)


####

net_edges <- 
  net_data %>%
  select(gene_id1, gene_id2, prop)

xeno_ko_prop_res %>% 
  filter(enssscg_id1 %in% net_data$enssscg_id1) %>%
  select(gene_id1, gene_id2, prop) %>%
  spread(gene_id2, prop, fill = 0) %>% 
  column_to_rownames("gene_id1") %>% 
  pheatmap(clustering_method = "ward.D2", 
           cutree_rows = 5,
           border_color = NA,
           filename = savepath("Xeno KO genes propr network heatmap.pdf"),
           width = 6, height = 100)

g <-
  net_edges %>%
  graph_from_data_frame(directed = FALSE) %>%
  ggraph(layout = "kk") 



edge_data <- get_edges()(g$data)
node_data <- 
  get_nodes()(g$data) %>% 
  as_tibble()
  

g + 
  geom_edge_arc(aes(color = prop,
                    width = abs(prop),
                    alpha = abs(prop)),
                strength = 0, 
                show.legend = F) + 
    scale_edge_alpha_continuous(range = c(0.3, 1)) +
    scale_edge_width_continuous(range = c(1, 3)) +
  
  geom_node_point(data = node_data, 
                  stroke = 1,
                  # size = 10,
                  shape = 21,
                  show.legend = F)+
  # geom_node_text(data = node_data,
  #                aes(label = name),
  #                size = 4) +
  scale_size_continuous(range = c(5, 10)) +
  scale_edge_color_gradient2(low = "blue", high = "red", mid = "white", midpoint = 0) +
  scale_fill_manual(values = c(tissue_colors_palette_full, gene_category_pal)) +
  
  theme_void()

ggsave(savepath("Network pig enrichment whole group.pdf"), width = 16, height = 12, useDingbats = F)
ggsave(savepath("Network pig enrichment whole group.png"), width = 16, height = 12)


```

## Whole genome network

```{r}



pig_atlas_sample %>%
  filter(enssscg_id %in% c("ENSSSCG00000003444", "ENSSSCG00000004183")) %>%
  select(1, 2, ptpm) %>%
  spread(enssscg_id, ptpm) %>% 
  ggplot(aes(ENSSSCG00000003444, ENSSSCG00000004183)) +
  geom_point()
    

pig_gene_spearman_propr %>% 
  head(100000) %>%
  ggplot(aes(prop, cor)) +
  geom_hex(aes(fill = log10(stat(count))),
           bins = 100) +
  scale_fill_viridis()




gene_coexpression_network %>%
  as_tibble() %>%
  select(contains("group")) %>%
  gather(type, community) %>%
  group_by(type, community) %>% 
  summarise(n = n()) %>% 
  ggplot(aes(n)) +
  geom_histogram() +
  facet_wrap(~type)



gene_coexpression_network %>%
  activate(edges) %>% 
  as_tibble() %>%
  filter(from == 1)

?tidygraph::crystallise()

#######
graph_plot <- 
  gene_coexpression_network %>% 
  filter(group_louvain == 4) %>%
  ggraph(layout = 'lgl') +
  geom_edge_fan(alpha = 0.1) +
  theme_graph()

ggsave(savepath("graph plot.pdf"), plot = graph_plot, width = 10, height = 10)

graph_plot <- 
  net_graph %>%
  ggraph(layout = "kk") +
  geom_edge_arc(strength = 0)

ggsave(savepath("graph plot.pdf"), plot = graph_plot, width = 10, height = 10)


prop_gene_graph_plot <- 
  prop_gene_graph %>%
  ggraph(layout = "kk") +
  geom_edge_arc(aes(color = sign(prop)),
                strength = 0, 
                show.legend = F) + 
  scale_color_manual(values = c("-1" = "skyblue", "1" = "orangered"))

ggsave(savepath("Global gene prop network.pdf"), plot = prop_gene_graph_plot, width = 10, height = 10)

```




#MHC genes

```{r}

MHC_GO <- 
  read_delim("data/meta/MHC GO terms.txt", delim = "\t", col_names = c("GO", "term"))

library(biomaRt)
select <- dplyr::select
ensembl <- useMart("ensembl",
                   dataset="hsapiens_gene_ensembl",
                   host = "http://apr2018.archive.ensembl.org") #uses human ensembl annotations

MHC_genes <- 
  getBM(attributes=c('hgnc_symbol', 'ensembl_gene_id', "go_id"),
        filters = 'go', values = MHC_GO$GO, mart = ensembl) %>%
  as_tibble() %>%
  inner_join(MHC_GO,
             by = c("go_id" = "GO"))

plot_data <- 
  joined_atlas_comparison %>% 
  select(mutual_id, comparison_tissue, species, mutual_tmm) %>%
  separate(mutual_id, into = c("enssscg_id", "ensg_id")) %>%
  filter(ensg_id %in% MHC_genes$ensembl_gene_id) %>%
  inner_join(gene_orthologs) %>%
  group_by(enssscg_id, ensg_id, species) %>% 
  mutate(mutual_tmm = mutual_tmm / max(mutual_tmm))


cluster_data <-
  plot_data %>%
  ungroup() %>%
  select(comparison_tissue, species, mutual_tmm, human_gene_name) %>%
  group_by(human_gene_name, comparison_tissue) %>%
  summarise(mutual_tmm = mean(mutual_tmm)) %>%
  spread(human_gene_name, mutual_tmm) %>%
  column_to_rownames("comparison_tissue")

tis_cluster <- 
  cluster_data %>%
  dist() %>% 
  hclust(method = "ward.D2")

tis_order <- 
  tis_cluster$labels[tis_cluster$order]


gene_order <- 
  plot_data %>% 
  ungroup() %>%
  select(comparison_tissue, species, mutual_tmm, gene_name = human_gene_name) %>%
  spread(species, mutual_tmm) %>%
  group_by(gene_name) %>% 
  summarise(dist = sqrt(sum((human - pig) ^ 2))) %>% 
  arrange(dist) %>% 
  pull(gene_name)
  
# Plot:
plot_data %>% 
  ungroup() %>%
  select(gene_name = human_gene_name, species) %>% 
  distinct() %>%
  mutate(species = toupper(substr(species, 1, 1))) %>%
  mutate(gene_name = factor(gene_name, gene_order)) %>%
  
  
  ggplot(aes(1, gene_name, group = species, fill = species, label = species)) +
  geom_tile(show.legend = F, 
            position = "dodge", 
            height = 0.9) +
  geom_text(position = position_dodge(width = 1), 
            size = 2, 
            fontface = "bold",
            color = "white") +
  scale_fill_manual(values = c("H" = "#DB1F48",
                               "P" = "#01949A")) +
  scale_x_discrete(limits = "Species") +
  theme_void() +
  theme(plot.margin = unit(c(0,0,0,0), "mm"),
        axis.text.x = element_text(angle = 90, hjust = 0, size = 6, vjust = 0.2)) +
  
  plot_data %>% 
  ungroup() %>%
  mutate(gene_name = factor(human_gene_name, gene_order),
         comparison_tissue = factor(comparison_tissue, tis_order)) %>%
  
  ggplot(aes(gene_name, comparison_tissue, group = species, fill = mutual_tmm)) +
  geom_tile(position= "dodge", width = 0.9) +
  coord_flip() +
  scale_fill_viridis_c(option = "B", direction = -1, name = "Expression") +
  # facet_grid(ensg_id ~ .) + 
  # facet_wrap(~gene_name, ncol = 1, strip.position = "left") + 
  stripped_theme_facet +
  theme(axis.text.x = element_text(angle = 60, hjust = 1),
        strip.placement = "outside",
        strip.text.y.left = element_text(angle = 0),
        panel.spacing = unit(0, "mm"),
        legend.position = "right",
        panel.border = element_rect(fill = NA, color = "white"),
        axis.title = element_blank(), 
        axis.line = element_blank()) +
  theme(plot.margin = unit(c(0,0,0,0), "mm")) +
  
  plot_layout(widths = c(0.025, 1), nrow = 1)

ggsave(savepath("MHC genes heatmap.pdf"), width = 7, height = 30)

```

